// @pittorica/hover-card-react@0.26.3/dist/index.js
import { jsxs as E, jsx as H } from "react/jsx-runtime";
import { useState as R, useRef as l, useCallback as f, useLayoutEffect as L } from "react";
import { createPortal as O } from "react-dom";
const W = ({
  renderTrigger: T,
  children: C,
  openDelay: a = 500,
  closeDelay: d = 300
}) => {
  const [s, m] = R(!1), [p, M] = R({ top: 0, left: 0 }), o = l(null), n = l(null), e = l(null), h = f(() => {
    const t = o.current, b = n.current;
    if (!b || !t) return;
    const c = t.getBoundingClientRect(), { innerWidth: g, innerHeight: y } = globalThis, { offsetWidth: i, offsetHeight: w } = b;
    let r = c.left + c.width / 2 - i / 2, u = c.bottom + 8;
    r < 8 && (r = 8), r + i > g - 8 && (r = g - i - 8), u + w > y - 8 && (u = c.top - w - 8), M({ left: r, top: u });
  }, []), x = f(() => {
    e.current && clearTimeout(e.current), e.current = setTimeout(() => {
      m(!0);
    }, a);
  }, [a]), v = f(() => {
    e.current && clearTimeout(e.current), e.current = setTimeout(() => {
      m(!1);
    }, d);
  }, [d]);
  return L(() => {
    if (!s || !o.current || !n.current) return;
    const t = new ResizeObserver(() => h());
    return t.observe(o.current), t.observe(n.current), () => t.disconnect();
  }, [s, h]), /* @__PURE__ */ E(
    "div",
    {
      style: { display: "inline-block" },
      onMouseEnter: x,
      onMouseLeave: v,
      children: [
        T({ ref: o }),
        s && O(
          /* @__PURE__ */ H(
            "div",
            {
              ref: n,
              className: "pittorica-hover-card-content",
              style: { top: p.top, left: p.left },
              onMouseEnter: () => {
                e.current && clearTimeout(e.current);
              },
              onMouseLeave: v,
              children: C
            }
          ),
          document.body
        )
      ]
    }
  );
};
export {
  W as HoverCard
};
