// @pittorica/card-react@0.26.3/dist/index.js
import { jsx as x } from "react/jsx-runtime";
function T(r) {
  var p, i, a = "";
  if (typeof r == "string" || typeof r == "number") a += r;
  else if (typeof r == "object") if (Array.isArray(r)) {
    var o = r.length;
    for (p = 0; p < o; p++) r[p] && (i = T(r[p])) && (a && (a += " "), a += i);
  } else for (i in r) r[i] && (a && (a += " "), a += i);
  return a;
}
function k() {
  for (var r, p, i = 0, a = "", o = arguments.length; i < o; i++) (r = arguments[i]) && (p = T(r)) && (a && (a += " "), a += p);
  return a;
}
function N(r) {
  var p, i, a = "";
  if (typeof r == "string" || typeof r == "number") a += r;
  else if (typeof r == "object") if (Array.isArray(r)) {
    var o = r.length;
    for (p = 0; p < o; p++) r[p] && (i = N(r[p])) && (a && (a += " "), a += i);
  } else for (i in r) r[i] && (a && (a += " "), a += i);
  return a;
}
function S() {
  for (var r, p, i = 0, a = "", o = arguments.length; i < o; i++) (r = arguments[i]) && (p = N(r)) && (a && (a += " "), a += p);
  return a;
}
const A = ({
  as: r,
  children: p,
  display: i,
  m: a,
  mt: o,
  mr: e,
  mb: c,
  ml: s,
  mx: n,
  my: d,
  p: f,
  pt: v,
  pr: l,
  pb: u,
  pl: $,
  px: g,
  py: m,
  width: y,
  height: h,
  position: b,
  style: B,
  className: L,
  disabled: R,
  required: j,
  ...q
}) => {
  const w = r || "div", t = {};
  i && (t.display = i), y && (t.width = y), h && (t.height = h), b && (t.position = b), a && (t.margin = `var(--pittorica-space-${a})`), o && (t.marginTop = `var(--pittorica-space-${o})`), e && (t.marginRight = `var(--pittorica-space-${e})`), c && (t.marginBottom = `var(--pittorica-space-${c})`), s && (t.marginLeft = `var(--pittorica-space-${s})`), n && (t.marginLeft = `var(--pittorica-space-${n})`, t.marginRight = `var(--pittorica-space-${n})`), d && (t.marginTop = `var(--pittorica-space-${d})`, t.marginBottom = `var(--pittorica-space-${d})`), f && (t.padding = `var(--pittorica-space-${f})`), v && (t.paddingTop = `var(--pittorica-space-${v})`), l && (t.paddingRight = `var(--pittorica-space-${l})`), u && (t.paddingBottom = `var(--pittorica-space-${u})`), $ && (t.paddingLeft = `var(--pittorica-space-${$})`), g && (t.paddingLeft = `var(--pittorica-space-${g})`, t.paddingRight = `var(--pittorica-space-${g})`), m && (t.paddingTop = `var(--pittorica-space-${m})`, t.paddingBottom = `var(--pittorica-space-${m})`);
  const C = {
    ...B,
    ...t
  };
  return /* @__PURE__ */ x(
    w,
    {
      className: S("pittorica-box", L),
      style: C,
      disabled: R,
      required: j,
      ...q,
      children: p
    }
  );
};
A.displayName = "Box";
const z = ({
  variant: r = "surface",
  translucent: p = !1,
  children: i,
  className: a,
  style: o,
  as: e,
  ...c
}) => /* @__PURE__ */ x(
  A,
  {
    as: e || "div",
    className: k(
      "pittorica-card",
      `pittorica-card--variant-${r}`,
      { "pittorica-card--translucent": p },
      a
    ),
    style: o,
    ...c,
    children: i
  }
);
z.displayName = "Card";
export {
  z as Card
};
