// @pittorica/aspect-ratio-react@0.26.3/dist/index.js
import { jsx as x } from "react/jsx-runtime";
function R(i) {
  var p, t, a = "";
  if (typeof i == "string" || typeof i == "number") a += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (p = 0; p < o; p++) i[p] && (t = R(i[p])) && (a && (a += " "), a += t);
  } else for (t in i) i[t] && (a && (a += " "), a += t);
  return a;
}
function S() {
  for (var i, p, t = 0, a = "", o = arguments.length; t < o; t++) (i = arguments[t]) && (p = R(i)) && (a && (a += " "), a += p);
  return a;
}
function A(i) {
  var p, t, a = "";
  if (typeof i == "string" || typeof i == "number") a += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (p = 0; p < o; p++) i[p] && (t = A(i[p])) && (a && (a += " "), a += t);
  } else for (t in i) i[t] && (a && (a += " "), a += t);
  return a;
}
function z() {
  for (var i, p, t = 0, a = "", o = arguments.length; t < o; t++) (i = arguments[t]) && (p = A(i)) && (a && (a += " "), a += p);
  return a;
}
const N = ({
  as: i,
  children: p,
  display: t,
  m: a,
  mt: o,
  mr: e,
  mb: g,
  ml: m,
  mx: s,
  my: c,
  p: f,
  pt: v,
  pr: l,
  pb: y,
  pl: $,
  px: n,
  py: d,
  width: u,
  height: h,
  position: b,
  style: T,
  className: B,
  disabled: L,
  required: j,
  ...q
}) => {
  const w = i || "div", r = {};
  t && (r.display = t), u && (r.width = u), h && (r.height = h), b && (r.position = b), a && (r.margin = `var(--pittorica-space-${a})`), o && (r.marginTop = `var(--pittorica-space-${o})`), e && (r.marginRight = `var(--pittorica-space-${e})`), g && (r.marginBottom = `var(--pittorica-space-${g})`), m && (r.marginLeft = `var(--pittorica-space-${m})`), s && (r.marginLeft = `var(--pittorica-space-${s})`, r.marginRight = `var(--pittorica-space-${s})`), c && (r.marginTop = `var(--pittorica-space-${c})`, r.marginBottom = `var(--pittorica-space-${c})`), f && (r.padding = `var(--pittorica-space-${f})`), v && (r.paddingTop = `var(--pittorica-space-${v})`), l && (r.paddingRight = `var(--pittorica-space-${l})`), y && (r.paddingBottom = `var(--pittorica-space-${y})`), $ && (r.paddingLeft = `var(--pittorica-space-${$})`), n && (r.paddingLeft = `var(--pittorica-space-${n})`, r.paddingRight = `var(--pittorica-space-${n})`), d && (r.paddingTop = `var(--pittorica-space-${d})`, r.paddingBottom = `var(--pittorica-space-${d})`);
  const k = {
    ...T,
    ...r
  };
  return /* @__PURE__ */ x(
    w,
    {
      className: z("pittorica-box", B),
      style: k,
      disabled: L,
      required: j,
      ...q,
      children: p
    }
  );
};
N.displayName = "Box";
const C = ({
  ratio: i = 1 / 1,
  children: p,
  className: t,
  style: a,
  as: o,
  ...e
}) => /* @__PURE__ */ x(
  N,
  {
    as: o || "div",
    className: S("pittorica-aspect-ratio", t),
    style: {
      aspectRatio: `${i}`,
      ...a
    },
    ...e,
    children: p
  }
);
C.displayName = "AspectRatio";
export {
  C as AspectRatio
};
