// @pittorica/skeleton-react@0.26.3/dist/index.js
import { jsx as l, Fragment as S } from "react/jsx-runtime";
function k(t) {
  var e, a, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (e = 0; e < o; e++) t[e] && (a = k(t[e])) && (i && (i += " "), i += a);
  } else for (a in t) t[a] && (i && (i += " "), i += a);
  return i;
}
function F() {
  for (var t, e, a = 0, i = "", o = arguments.length; a < o; a++) (t = arguments[a]) && (e = k(t)) && (i && (i += " "), i += e);
  return i;
}
function T(t) {
  var e, a, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (e = 0; e < o; e++) t[e] && (a = T(t[e])) && (i && (i += " "), i += a);
  } else for (a in t) t[a] && (i && (i += " "), i += a);
  return i;
}
function z() {
  for (var t, e, a = 0, i = "", o = arguments.length; a < o; a++) (t = arguments[a]) && (e = T(t)) && (i && (i += " "), i += e);
  return i;
}
const N = ({
  as: t,
  children: e,
  display: a,
  m: i,
  mt: o,
  mr: c,
  mb: p,
  ml: n,
  mx: s,
  my: g,
  p: f,
  pt: h,
  pr: v,
  pb: u,
  pl: y,
  px: d,
  py: m,
  width: $,
  height: x,
  position: b,
  style: w,
  className: A,
  disabled: B,
  required: L,
  ...R
}) => {
  const j = t || "div", r = {};
  a && (r.display = a), $ && (r.width = $), x && (r.height = x), b && (r.position = b), i && (r.margin = `var(--pittorica-space-${i})`), o && (r.marginTop = `var(--pittorica-space-${o})`), c && (r.marginRight = `var(--pittorica-space-${c})`), p && (r.marginBottom = `var(--pittorica-space-${p})`), n && (r.marginLeft = `var(--pittorica-space-${n})`), s && (r.marginLeft = `var(--pittorica-space-${s})`, r.marginRight = `var(--pittorica-space-${s})`), g && (r.marginTop = `var(--pittorica-space-${g})`, r.marginBottom = `var(--pittorica-space-${g})`), f && (r.padding = `var(--pittorica-space-${f})`), h && (r.paddingTop = `var(--pittorica-space-${h})`), v && (r.paddingRight = `var(--pittorica-space-${v})`), u && (r.paddingBottom = `var(--pittorica-space-${u})`), y && (r.paddingLeft = `var(--pittorica-space-${y})`), d && (r.paddingLeft = `var(--pittorica-space-${d})`, r.paddingRight = `var(--pittorica-space-${d})`), m && (r.paddingTop = `var(--pittorica-space-${m})`, r.paddingBottom = `var(--pittorica-space-${m})`);
  const q = {
    ...w,
    ...r
  };
  return /* @__PURE__ */ l(
    j,
    {
      className: z("pittorica-box", A),
      style: q,
      disabled: B,
      required: L,
      ...R,
      children: e
    }
  );
};
N.displayName = "Box";
const C = ({
  children: t,
  loading: e = !0,
  variant: a = "rect",
  className: i,
  style: o,
  as: c,
  ...p
}) => e ? /* @__PURE__ */ l(
  N,
  {
    as: c || "div",
    className: F(
      "pittorica-skeleton",
      "pittorica-skeleton--loading",
      {
        "pittorica-skeleton--hiding-children": !!t,
        "pittorica-skeleton--circle": a === "circle",
        "pittorica-skeleton--text": a === "text"
      },
      i
    ),
    style: {
      height: a === "text" && !p.height ? "0.8em" : p.height,
      width: a === "text" && !p.width ? "100%" : p.width,
      ...o
    },
    ...p,
    children: t
  }
) : /* @__PURE__ */ l(S, { children: t });
C.displayName = "Skeleton";
export {
  C as Skeleton
};
