// @pittorica/text-area-react@0.26.3/dist/index.js
import { jsx as x, jsxs as j } from "react/jsx-runtime";
import { createContext as G, useRef as J, useEffect as K, useId as S, use as M } from "react";
function q(t) {
  var a, i, r = "";
  if (typeof t == "string" || typeof t == "number") r += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var n = t.length;
    for (a = 0; a < n; a++) t[a] && (i = q(t[a])) && (r && (r += " "), r += i);
  } else for (i in t) t[i] && (r && (r += " "), r += i);
  return r;
}
function z() {
  for (var t, a, i = 0, r = "", n = arguments.length; i < n; i++) (t = arguments[i]) && (a = q(t)) && (r && (r += " "), r += a);
  return r;
}
function C(t) {
  var a, i, r = "";
  if (typeof t == "string" || typeof t == "number") r += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var n = t.length;
    for (a = 0; a < n; a++) t[a] && (i = C(t[a])) && (r && (r += " "), r += i);
  } else for (i in t) t[i] && (r && (r += " "), r += i);
  return r;
}
function O() {
  for (var t, a, i = 0, r = "", n = arguments.length; i < n; i++) (t = arguments[i]) && (a = C(t)) && (r && (r += " "), r += a);
  return r;
}
const I = ({
  as: t,
  children: a,
  display: i,
  m: r,
  mt: n,
  mr: m,
  mb: o,
  ml: c,
  mx: p,
  my: d,
  p: f,
  pt: h,
  pr: g,
  pb: l,
  pl: v,
  px: s,
  py: u,
  width: $,
  height: y,
  position: b,
  style: T,
  className: A,
  disabled: w,
  required: N,
  ...B
}) => {
  const L = t || "div", e = {};
  i && (e.display = i), $ && (e.width = $), y && (e.height = y), b && (e.position = b), r && (e.margin = `var(--pittorica-space-${r})`), n && (e.marginTop = `var(--pittorica-space-${n})`), m && (e.marginRight = `var(--pittorica-space-${m})`), o && (e.marginBottom = `var(--pittorica-space-${o})`), c && (e.marginLeft = `var(--pittorica-space-${c})`), p && (e.marginLeft = `var(--pittorica-space-${p})`, e.marginRight = `var(--pittorica-space-${p})`), d && (e.marginTop = `var(--pittorica-space-${d})`, e.marginBottom = `var(--pittorica-space-${d})`), f && (e.padding = `var(--pittorica-space-${f})`), h && (e.paddingTop = `var(--pittorica-space-${h})`), g && (e.paddingRight = `var(--pittorica-space-${g})`), l && (e.paddingBottom = `var(--pittorica-space-${l})`), v && (e.paddingLeft = `var(--pittorica-space-${v})`), s && (e.paddingLeft = `var(--pittorica-space-${s})`, e.paddingRight = `var(--pittorica-space-${s})`), u && (e.paddingTop = `var(--pittorica-space-${u})`, e.paddingBottom = `var(--pittorica-space-${u})`);
  const R = {
    ...T,
    ...e
  };
  return /* @__PURE__ */ x(
    L,
    {
      className: O("pittorica-box", A),
      style: R,
      disabled: w,
      required: N,
      ...B,
      children: a
    }
  );
};
I.displayName = "Box";
function k(t) {
  var a, i, r = "";
  if (typeof t == "string" || typeof t == "number") r += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var n = t.length;
    for (a = 0; a < n; a++) t[a] && (i = k(t[a])) && (r && (r += " "), r += i);
  } else for (i in t) t[i] && (r && (r += " "), r += i);
  return r;
}
function W() {
  for (var t, a, i = 0, r = "", n = arguments.length; i < n; i++) (t = arguments[i]) && (a = k(t)) && (r && (r += " "), r += a);
  return r;
}
function E(t) {
  var a, i, r = "";
  if (typeof t == "string" || typeof t == "number") r += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var n = t.length;
    for (a = 0; a < n; a++) t[a] && (i = E(t[a])) && (r && (r += " "), r += i);
  } else for (i in t) t[i] && (r && (r += " "), r += i);
  return r;
}
function P() {
  for (var t, a, i = 0, r = "", n = arguments.length; i < n; i++) (t = arguments[i]) && (a = E(t)) && (r && (r += " "), r += a);
  return r;
}
const H = ({
  as: t,
  children: a,
  display: i,
  m: r,
  mt: n,
  mr: m,
  mb: o,
  ml: c,
  mx: p,
  my: d,
  p: f,
  pt: h,
  pr: g,
  pb: l,
  pl: v,
  px: s,
  py: u,
  width: $,
  height: y,
  position: b,
  style: T,
  className: A,
  disabled: w,
  required: N,
  ...B
}) => {
  const L = t || "div", e = {};
  i && (e.display = i), $ && (e.width = $), y && (e.height = y), b && (e.position = b), r && (e.margin = `var(--pittorica-space-${r})`), n && (e.marginTop = `var(--pittorica-space-${n})`), m && (e.marginRight = `var(--pittorica-space-${m})`), o && (e.marginBottom = `var(--pittorica-space-${o})`), c && (e.marginLeft = `var(--pittorica-space-${c})`), p && (e.marginLeft = `var(--pittorica-space-${p})`, e.marginRight = `var(--pittorica-space-${p})`), d && (e.marginTop = `var(--pittorica-space-${d})`, e.marginBottom = `var(--pittorica-space-${d})`), f && (e.padding = `var(--pittorica-space-${f})`), h && (e.paddingTop = `var(--pittorica-space-${h})`), g && (e.paddingRight = `var(--pittorica-space-${g})`), l && (e.paddingBottom = `var(--pittorica-space-${l})`), v && (e.paddingLeft = `var(--pittorica-space-${v})`), s && (e.paddingLeft = `var(--pittorica-space-${s})`, e.paddingRight = `var(--pittorica-space-${s})`), u && (e.paddingTop = `var(--pittorica-space-${u})`, e.paddingBottom = `var(--pittorica-space-${u})`);
  const R = {
    ...T,
    ...e
  };
  return /* @__PURE__ */ x(
    L,
    {
      className: P("pittorica-box", A),
      style: R,
      disabled: w,
      required: N,
      ...B,
      children: a
    }
  );
};
H.displayName = "Box";
const F = ({
  children: t,
  as: a,
  size: i,
  weight: r = "regular",
  align: n,
  truncate: m = !1,
  color: o,
  className: c,
  style: p,
  ...d
}) => {
  const f = a || "span", h = (o == null ? void 0 : o.startsWith("#")) || (o == null ? void 0 : o.startsWith("rgb")) || (o == null ? void 0 : o.startsWith("hsl")), g = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), l = (() => {
    if (o)
      return o === "inherit" ? "inherit" : h ? o : g.has(o) ? `var(--pittorica-${o}-9)` : o;
  })(), v = i ? typeof i == "string" ? `pittorica-text--size-${i}` : W({
    [`pittorica-text--size-${i.initial}`]: i.initial,
    [`pittorica-text--xs-size-${i.xs}`]: i.xs,
    [`pittorica-text--sm-size-${i.sm}`]: i.sm,
    [`pittorica-text--md-size-${i.md}`]: i.md,
    [`pittorica-text--lg-size-${i.lg}`]: i.lg,
    [`pittorica-text--xl-size-${i.xl}`]: i.xl
  }) : null, s = {
    ...p,
    textAlign: n,
    color: l
  };
  return /* @__PURE__ */ x(
    H,
    {
      as: f,
      className: W(
        "pittorica-text",
        v,
        { "pittorica-text--truncate": m },
        c
      ),
      "data-weight": r,
      style: s,
      ...d,
      children: t
    }
  );
};
F.displayName = "Text";
const D = G(null), Q = () => {
  const t = M(D);
  if (!t)
    throw new Error("TextArea components must be wrapped in <TextArea.Root />");
  return t;
}, U = ({
  children: t,
  label: a,
  helperText: i,
  error: r,
  color: n = "source",
  disabled: m,
  name: o,
  required: c,
  size: p = "sm",
  className: d,
  style: f,
  as: h,
  ...g
}) => {
  const l = S(), v = S(), u = n !== "inherit" && !(n != null && n.startsWith("#")) && !(n != null && n.startsWith("rgb")) ? `var(--pittorica-${n}-9)` : n;
  return /* @__PURE__ */ x(
    D,
    {
      value: { inputId: l, helperId: v, disabled: m, size: p, name: o, required: c },
      children: /* @__PURE__ */ j(
        I,
        {
          as: h || "div",
          className: z(
            "pittorica-textarea-root",
            `pittorica-textarea--${p}`,
            d
          ),
          "data-error": r,
          ...g,
          children: [
            a && /* @__PURE__ */ j(
              F,
              {
                as: "label",
                htmlFor: l,
                weight: "medium",
                style: {
                  paddingLeft: "4px",
                  fontSize: "var(--pittorica-font-size-1)",
                  marginBottom: "4px",
                  display: "inline-block"
                },
                children: [
                  a,
                  " ",
                  c && /* @__PURE__ */ x("span", { "aria-hidden": "true", children: "*" })
                ]
              }
            ),
            /* @__PURE__ */ x(
              "div",
              {
                className: "pittorica-textarea-wrapper",
                "data-disabled": m,
                style: {
                  "--pittorica-source-color": u,
                  ...f
                },
                children: t
              }
            ),
            i && /* @__PURE__ */ x("div", { id: v, className: "pittorica-textarea-helper", children: i })
          ]
        }
      )
    }
  );
}, X = ({
  className: t,
  autoResize: a = !1,
  onChange: i,
  value: r,
  defaultValue: n,
  name: m,
  required: o,
  // Add required prop here
  ...c
}) => {
  const {
    inputId: p,
    helperId: d,
    disabled: f,
    name: h
  } = Q(), g = J(null), l = () => {
    if (globalThis.window === void 0) return;
    const s = g.current;
    !s || !a || (s.style.height = "auto", s.style.height = `${s.scrollHeight}px`);
  };
  return K(() => {
    a && l();
  }, [r, n, a]), /* @__PURE__ */ x(
    "textarea",
    {
      name: m ?? h,
      ...c,
      id: p,
      ref: g,
      value: r,
      defaultValue: n,
      disabled: f,
      required: o,
      "aria-describedby": d,
      onChange: (s) => {
        a && l(), i == null || i(s);
      },
      className: z("pittorica-textarea-input", t),
      style: {
        overflow: a ? "hidden" : void 0,
        ...c.style
      }
    }
  );
}, _ = {
  Root: U,
  Content: X
};
export {
  _ as TextArea,
  X as TextAreaContent,
  U as TextAreaRoot
};
