// eslint-plugin-vuoto@0.1.2/dist/index.js
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

//#region src/configs/recommended.ts
const baseRules = {
	"vuoto/byte-order-mark": "error",
	"vuoto/em-space": "error",
	"vuoto/en-space": "error",
	"vuoto/form-feed": "error",
	"vuoto/ideographic-space": "error",
	"vuoto/invisible-separators": "error",
	"vuoto/line-separator": "error",
	"vuoto/narrow-no-break-space": "error",
	"vuoto/non-breaking-space": "error",
	"vuoto/paragraph-separator": "error",
	"vuoto/vertical-tab": "error",
	"vuoto/visible-misc-spaces": "error",
	"vuoto/zero-width": "error"
};
const configRecommended = [{
	name: "vuoto/recommended",
	rules: baseRules
}];

//#endregion
//#region src/configs/all.ts
const configAll = [{
	name: "vuoto/all",
	rules: baseRules
}];

//#endregion
//#region src/configs/off.ts
const configOff = [{
	name: "vuoto/off",
	rules: Object.fromEntries(Object.keys(baseRules).map((rule) => [rule, "off"]))
}];

//#endregion
//#region src/configs/strict.ts
const configStrict = [{
	name: "vuoto/strict",
	rules: baseRules
}];

//#endregion
//#region src/helpers.ts
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const packageJsonPath = path.resolve(__dirname, "../package.json");
const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf8"));

//#endregion
//#region ../vuoto/dist/consts.js
/**
* Byte order mark regex pattern.
*/
const BYTE_ORDER_MARK = new RegExp(String.raw`\uFEFF`, "g");
/**
* Em space regex pattern.
*/
const EM_SPACE = new RegExp(String.raw`\u2003`, "g");
/**
* En space regex pattern.
*/
const EN_SPACE = new RegExp(String.raw`\u2002`, "g");
/**
* Form feed regex pattern.
*/
const FORM_FEED = new RegExp(String.raw`\f`, "g");
/**
* Ideographic space regex pattern.
*/
const IDEOGRAPHIC_SPACE = new RegExp(String.raw`\u3000`, "g");
/**
* Invisible separators regex pattern.
*/
const INVISIBLE_SEPARATORS = new RegExp(String.raw`[\u2060\uFEFF]`, "g");
/**
* Line separator regex pattern.
*/
const LINE_SEPARATOR = new RegExp(String.raw`\u2028`, "g");
/**
* Narrow no-break space regex pattern.
*/
const NARROW_NO_BREAK_SPACE = new RegExp(String.raw`\u202F`, "g");
/**
* Non-breaking space regex pattern.
*/
const NON_BREAKING_SPACE = new RegExp(String.raw`\u00A0`, "g");
/**
* Paragraph separator regex pattern.
*/
const PARAGRAPH_SEPARATOR = new RegExp(String.raw`\u2029`, "g");
/**
* Vertical tab regex pattern.
*/
const VERTICAL_TAB = new RegExp(String.raw`\x0B`, "g");
/**
* Visible miscellaneous spaces regex pattern.
*/
const VISIBLE_MISC_SPACES = new RegExp(String.raw`[\u1680\u180E\u2000-\u200A\u205F\u3000]`, "g");
/**
* Zero-width characters regex pattern.
*/
const ZERO_WIDTH = new RegExp(String.raw`\u200B|\u200C|\u200D|\uFEFF|\u2060`, "g");
/**
* Map of Unicode characters to their human-readable names.
*/
const UNICODE_NAME_MAP = {
	"\xA0": "NON-BREAKING SPACE",
	" ": "OGHAM SPACE MARK",
	"᠎": "MONGOLIAN VOWEL SEPARATOR",
	" ": "EN QUAD",
	" ": "EM QUAD",
	" ": "EN SPACE",
	" ": "EM SPACE",
	" ": "THREE-PER-EM SPACE",
	" ": "FOUR-PER-EM SPACE",
	" ": "SIX-PER-EM SPACE",
	" ": "FIGURE SPACE",
	" ": "PUNCTUATION SPACE",
	" ": "THIN SPACE",
	" ": "HAIR SPACE",
	"​": "ZERO WIDTH SPACE",
	"‌": "ZERO WIDTH NON-JOINER",
	"‍": "ZERO WIDTH JOINER",
	" ": "NARROW NO-BREAK SPACE",
	" ": "MEDIUM MATHEMATICAL SPACE",
	"⁠": "WORD JOINER",
	"　": "IDEOGRAPHIC SPACE",
	"﻿": "ZERO WIDTH NO-BREAK SPACE (BOM)",
	"\u2028": "LINE SEPARATOR",
	"\u2029": "PARAGRAPH_SEPARATOR",
	"\f": "FORM FEED",
	"\v": "VERTICAL TAB"
};

//#endregion
//#region src/rules/factory.ts
/**
* Create a rule that disallows a set of characters.
* @param regex The regex pattern to match.
* @param description The rule description.
* @param replacement The string to replace the match with.
* @param messageId The message ID.
* @returns The rule module.
*/
function createWhitespaceRule(regex, description, replacement = "", messageId = "unexpected") {
	return {
		meta: {
			type: "problem",
			docs: {
				description,
				recommended: true
			},
			fixable: "code",
			schema: [],
			messages: { [messageId]: `Unexpected {{name}} character {{code}}.` }
		},
		create(context) {
			const sourceCode = context.sourceCode;
			const text = sourceCode.text;
			return { Program() {
				let match;
				const globalRegex = new RegExp(regex.source, regex.flags.includes("g") ? regex.flags : regex.flags + "g");
				while ((match = globalRegex.exec(text)) !== null) {
					const index = match.index;
					const char = match[0];
					context.report({
						loc: sourceCode.getLocFromIndex(index),
						messageId,
						data: {
							code: `U+${char.codePointAt(0).toString(16).toUpperCase().padStart(4, "0")}`,
							name: UNICODE_NAME_MAP[char] || "Unknown whitespace"
						},
						fix(fixer) {
							return fixer.replaceTextRange([index, index + char.length], replacement);
						}
					});
				}
			} };
		}
	};
}

//#endregion
//#region src/rules/index.ts
const rules = {
	"byte-order-mark": createWhitespaceRule(BYTE_ORDER_MARK, "disallow byte order mark (BOM) (U+FEFF)"),
	"em-space": createWhitespaceRule(EM_SPACE, "disallow em space (U+2003)", " "),
	"en-space": createWhitespaceRule(EN_SPACE, "disallow en space (U+2002)", " "),
	"form-feed": createWhitespaceRule(FORM_FEED, "disallow form feed (U+000C)", "\n"),
	"ideographic-space": createWhitespaceRule(IDEOGRAPHIC_SPACE, "disallow ideographic space (U+3000)", " "),
	"invisible-separators": createWhitespaceRule(INVISIBLE_SEPARATORS, "disallow invisible separators (U+2060, U+FEFF)"),
	"line-separator": createWhitespaceRule(LINE_SEPARATOR, "disallow line separator (U+2028)", "\n"),
	"narrow-no-break-space": createWhitespaceRule(NARROW_NO_BREAK_SPACE, "disallow narrow no-break space (U+202F)", " "),
	"non-breaking-space": createWhitespaceRule(NON_BREAKING_SPACE, "disallow non-breaking space (U+00A0)", " "),
	"paragraph-separator": createWhitespaceRule(PARAGRAPH_SEPARATOR, "disallow paragraph separator (U+2029)", "\n\n"),
	"vertical-tab": createWhitespaceRule(VERTICAL_TAB, "disallow vertical tab (U+000B)", "\n"),
	"visible-misc-spaces": createWhitespaceRule(VISIBLE_MISC_SPACES, "disallow miscellaneous visible spaces", " "),
	"zero-width": createWhitespaceRule(ZERO_WIDTH, "disallow zero-width characters (U+200B, U+200C, U+200D, U+FEFF, U+2060)")
};

//#endregion
//#region src/index.ts
const configs = {
	recommended: configRecommended,
	strict: configStrict,
	all: configAll,
	off: configOff
};
const plugin = {
	meta: {
		name: packageJson.name,
		version: packageJson.version
	},
	configs,
	rules,
	processors: {}
};

//#endregion
export { plugin as default };
