// @pittorica/radio-group-react@0.26.3/dist/index.js
import { jsx as b } from "react/jsx-runtime";
import { forwardRef as H, createElement as q, useState as G, createContext as O, useMemo as U, use as J } from "react";
function L(a) {
  var e, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var o = a.length;
    for (e = 0; e < o; e++) a[e] && (r = L(a[e])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function K() {
  for (var a, e, r = 0, t = "", o = arguments.length; r < o; r++) (a = arguments[r]) && (e = L(a)) && (t && (t += " "), t += e);
  return t;
}
function j(a) {
  var e, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var o = a.length;
    for (e = 0; e < o; e++) a[e] && (r = j(a[e])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function P() {
  for (var a, e, r = 0, t = "", o = arguments.length; r < o; r++) (a = arguments[r]) && (e = j(a)) && (t && (t += " "), t += e);
  return t;
}
const z = ({
  as: a,
  children: e,
  display: r,
  m: t,
  mt: o,
  mr: l,
  mb: s,
  ml: g,
  mx: n,
  my: d,
  p: c,
  pt: m,
  pr: f,
  pb: u,
  pl: p,
  px: h,
  py: v,
  width: y,
  height: $,
  position: k,
  style: R,
  className: x,
  disabled: w,
  required: C,
  ...N
}) => {
  const T = a || "div", i = {};
  r && (i.display = r), y && (i.width = y), $ && (i.height = $), k && (i.position = k), t && (i.margin = `var(--pittorica-space-${t})`), o && (i.marginTop = `var(--pittorica-space-${o})`), l && (i.marginRight = `var(--pittorica-space-${l})`), s && (i.marginBottom = `var(--pittorica-space-${s})`), g && (i.marginLeft = `var(--pittorica-space-${g})`), n && (i.marginLeft = `var(--pittorica-space-${n})`, i.marginRight = `var(--pittorica-space-${n})`), d && (i.marginTop = `var(--pittorica-space-${d})`, i.marginBottom = `var(--pittorica-space-${d})`), c && (i.padding = `var(--pittorica-space-${c})`), m && (i.paddingTop = `var(--pittorica-space-${m})`), f && (i.paddingRight = `var(--pittorica-space-${f})`), u && (i.paddingBottom = `var(--pittorica-space-${u})`), p && (i.paddingLeft = `var(--pittorica-space-${p})`), h && (i.paddingLeft = `var(--pittorica-space-${h})`, i.paddingRight = `var(--pittorica-space-${h})`), v && (i.paddingTop = `var(--pittorica-space-${v})`, i.paddingBottom = `var(--pittorica-space-${v})`);
  const B = {
    ...R,
    ...i
  };
  return /* @__PURE__ */ b(
    T,
    {
      className: P("pittorica-box", x),
      style: B,
      disabled: w,
      required: C,
      ...N,
      children: e
    }
  );
};
z.displayName = "Box";
function W(a) {
  var e, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var o = a.length;
    for (e = 0; e < o; e++) a[e] && (r = W(a[e])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function Q() {
  for (var a, e, r = 0, t = "", o = arguments.length; r < o; r++) (a = arguments[r]) && (e = W(a)) && (t && (t += " "), t += e);
  return t;
}
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var X = {
  outline: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  },
  filled: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "currentColor",
    stroke: "none"
  }
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const D = (a, e, r, t) => {
  const o = H(
    ({ color: l = "currentColor", size: s = 24, stroke: g = 2, title: n, className: d, children: c, ...m }, f) => q(
      "svg",
      {
        ref: f,
        ...X[a],
        width: s,
        height: s,
        className: ["tabler-icon", `tabler-icon-${e}`, d].join(" "),
        ...a === "filled" ? {
          fill: l
        } : {
          strokeWidth: g,
          stroke: l
        },
        ...m
      },
      [
        n && q("title", { key: "svg-title" }, n),
        ...t.map(([u, p]) => q(u, p)),
        ...Array.isArray(c) ? c : [c]
      ]
    )
  );
  return o.displayName = `${r}`, o;
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const Y = [["path", { d: "M3 12a9 9 0 1 0 18 0a9 9 0 1 0 -18 0", key: "svg-0" }]], Z = D("outline", "circle", "Circle", Y);
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const _ = [["path", { d: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-1.293 5.953a1 1 0 0 0 -1.32 -.083l-.094 .083l-3.293 3.292l-1.293 -1.292l-.094 -.083a1 1 0 0 0 -1.403 1.403l.083 .094l2 2l.094 .083a1 1 0 0 0 1.226 0l.094 -.083l4 -4l.083 -.094a1 1 0 0 0 -.083 -1.32z", key: "svg-0" }]], V = D("filled", "circle-check-filled", "CircleCheckFilled", _);
function I(a) {
  var e, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var o = a.length;
    for (e = 0; e < o; e++) a[e] && (r = I(a[e])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function tt() {
  for (var a, e, r = 0, t = "", o = arguments.length; r < o; r++) (a = arguments[r]) && (e = I(a)) && (t && (t += " "), t += e);
  return t;
}
const M = ({
  as: a,
  children: e,
  display: r,
  m: t,
  mt: o,
  mr: l,
  mb: s,
  ml: g,
  mx: n,
  my: d,
  p: c,
  pt: m,
  pr: f,
  pb: u,
  pl: p,
  px: h,
  py: v,
  width: y,
  height: $,
  position: k,
  style: R,
  className: x,
  disabled: w,
  required: C,
  ...N
}) => {
  const T = a || "div", i = {};
  r && (i.display = r), y && (i.width = y), $ && (i.height = $), k && (i.position = k), t && (i.margin = `var(--pittorica-space-${t})`), o && (i.marginTop = `var(--pittorica-space-${o})`), l && (i.marginRight = `var(--pittorica-space-${l})`), s && (i.marginBottom = `var(--pittorica-space-${s})`), g && (i.marginLeft = `var(--pittorica-space-${g})`), n && (i.marginLeft = `var(--pittorica-space-${n})`, i.marginRight = `var(--pittorica-space-${n})`), d && (i.marginTop = `var(--pittorica-space-${d})`, i.marginBottom = `var(--pittorica-space-${d})`), c && (i.padding = `var(--pittorica-space-${c})`), m && (i.paddingTop = `var(--pittorica-space-${m})`), f && (i.paddingRight = `var(--pittorica-space-${f})`), u && (i.paddingBottom = `var(--pittorica-space-${u})`), p && (i.paddingLeft = `var(--pittorica-space-${p})`), h && (i.paddingLeft = `var(--pittorica-space-${h})`, i.paddingRight = `var(--pittorica-space-${h})`), v && (i.paddingTop = `var(--pittorica-space-${v})`, i.paddingBottom = `var(--pittorica-space-${v})`);
  const B = {
    ...R,
    ...i
  };
  return /* @__PURE__ */ b(
    T,
    {
      className: tt("pittorica-box", x),
      style: B,
      disabled: w,
      required: C,
      ...N,
      children: e
    }
  );
};
M.displayName = "Box";
const E = ({
  checked: a,
  defaultChecked: e,
  disabled: r,
  color: t = "source",
  className: o,
  style: l,
  onCheckedChange: s,
  name: g,
  required: n = !1,
  as: d,
  ...c
}) => {
  const [m, f] = G(
    e ?? !1
  ), u = a !== void 0, p = u ? a : m, h = t !== "inherit" && !(t != null && t.startsWith("#")) && !(t != null && t.startsWith("rgb")) ? `var(--pittorica-${t}-9)` : t, v = d || "button";
  return /* @__PURE__ */ b(
    M,
    {
      as: v,
      type: v === "button" ? "button" : void 0,
      role: "radio",
      name: g,
      "aria-checked": p,
      "aria-required": n,
      "data-state": p ? "checked" : "unchecked",
      disabled: r,
      onClick: (y) => {
        r || (u || f(!p), s == null || s(!p), typeof c.onClick == "function" && c.onClick(y));
      },
      className: Q("pittorica-radio-root", o),
      style: {
        "--pittorica-source-color": h,
        ...l
      },
      ...c,
      children: p ? /* @__PURE__ */ b(V, { size: 14 }) : /* @__PURE__ */ b(Z, { size: 14, opacity: 0.5 })
    }
  );
};
E.displayName = "Radio";
const S = O(null), at = () => {
  const a = J(S);
  if (!a)
    throw new Error("RadioGroup components must be wrapped in <RadioGroup />");
  return a;
}, A = ({
  children: a,
  value: e,
  defaultValue: r,
  onValueChange: t,
  color: o = "indigo",
  disabled: l,
  name: s,
  required: g,
  className: n,
  as: d,
  ...c
}) => {
  const [m, f] = G(r), u = e !== void 0, p = u ? e : m, h = ($) => {
    u || f($), t == null || t($);
  }, v = U(
    () => ({
      value: p,
      onValueChange: h,
      color: o,
      disabled: l,
      name: s,
      required: g
    }),
    [p, o, l, s, g]
  );
  return /* @__PURE__ */ b(S, { value: v, children: /* @__PURE__ */ b(
    z,
    {
      as: d || "div",
      role: "radiogroup",
      className: K("pittorica-radio-group-root", n),
      ...c,
      children: a
    }
  ) });
}, F = ({
  value: a,
  disabled: e,
  as: r,
  ...t
}) => {
  const {
    value: o,
    onValueChange: l,
    color: s,
    disabled: g,
    name: n,
    required: d
  } = at();
  return /* @__PURE__ */ b(
    E,
    {
      as: r || "button",
      name: n,
      value: a,
      color: s,
      checked: o === a,
      disabled: g || e,
      required: t.required || d,
      onCheckedChange: () => l(a),
      ...t
    }
  );
}, it = Object.assign(A, {
  Root: A,
  Item: F,
  displayName: "RadioGroup"
});
F.displayName = "RadioGroup.Item";
export {
  it as RadioGroup,
  F as RadioGroupItem
};
