// @pittorica/blockquote-react@0.26.3/dist/index.js
import { jsx as h } from "react/jsx-runtime";
function N(r) {
  var a, i, t = "";
  if (typeof r == "string" || typeof r == "number") t += r;
  else if (typeof r == "object") if (Array.isArray(r)) {
    var s = r.length;
    for (a = 0; a < s; a++) r[a] && (i = N(r[a])) && (t && (t += " "), t += i);
  } else for (i in r) r[i] && (t && (t += " "), t += i);
  return t;
}
function V() {
  for (var r, a, i = 0, t = "", s = arguments.length; i < s; i++) (r = arguments[i]) && (a = N(r)) && (t && (t += " "), t += a);
  return t;
}
function z(r) {
  var a, i, t = "";
  if (typeof r == "string" || typeof r == "number") t += r;
  else if (typeof r == "object") if (Array.isArray(r)) {
    var s = r.length;
    for (a = 0; a < s; a++) r[a] && (i = z(r[a])) && (t && (t += " "), t += i);
  } else for (i in r) r[i] && (t && (t += " "), t += i);
  return t;
}
function q() {
  for (var r, a, i = 0, t = "", s = arguments.length; i < s; i++) (r = arguments[i]) && (a = z(r)) && (t && (t += " "), t += a);
  return t;
}
function A(r) {
  var a, i, t = "";
  if (typeof r == "string" || typeof r == "number") t += r;
  else if (typeof r == "object") if (Array.isArray(r)) {
    var s = r.length;
    for (a = 0; a < s; a++) r[a] && (i = A(r[a])) && (t && (t += " "), t += i);
  } else for (i in r) r[i] && (t && (t += " "), t += i);
  return t;
}
function _() {
  for (var r, a, i = 0, t = "", s = arguments.length; i < s; i++) (r = arguments[i]) && (a = A(r)) && (t && (t += " "), t += a);
  return t;
}
const B = ({
  as: r,
  children: a,
  display: i,
  m: t,
  mt: s,
  mr: p,
  mb: o,
  ml: g,
  mx: c,
  my: n,
  p: d,
  pt: u,
  pr: m,
  pb: f,
  pl: v,
  px: l,
  py: $,
  width: y,
  height: x,
  position: b,
  style: W,
  className: k,
  disabled: T,
  required: j,
  ...L
}) => {
  const R = r || "div", e = {};
  i && (e.display = i), y && (e.width = y), x && (e.height = x), b && (e.position = b), t && (e.margin = `var(--pittorica-space-${t})`), s && (e.marginTop = `var(--pittorica-space-${s})`), p && (e.marginRight = `var(--pittorica-space-${p})`), o && (e.marginBottom = `var(--pittorica-space-${o})`), g && (e.marginLeft = `var(--pittorica-space-${g})`), c && (e.marginLeft = `var(--pittorica-space-${c})`, e.marginRight = `var(--pittorica-space-${c})`), n && (e.marginTop = `var(--pittorica-space-${n})`, e.marginBottom = `var(--pittorica-space-${n})`), d && (e.padding = `var(--pittorica-space-${d})`), u && (e.paddingTop = `var(--pittorica-space-${u})`), m && (e.paddingRight = `var(--pittorica-space-${m})`), f && (e.paddingBottom = `var(--pittorica-space-${f})`), v && (e.paddingLeft = `var(--pittorica-space-${v})`), l && (e.paddingLeft = `var(--pittorica-space-${l})`, e.paddingRight = `var(--pittorica-space-${l})`), $ && (e.paddingTop = `var(--pittorica-space-${$})`, e.paddingBottom = `var(--pittorica-space-${$})`);
  const S = {
    ...W,
    ...e
  };
  return /* @__PURE__ */ h(
    R,
    {
      className: _("pittorica-box", k),
      style: S,
      disabled: T,
      required: j,
      ...L,
      children: a
    }
  );
};
B.displayName = "Box";
const w = ({
  children: r,
  as: a,
  size: i,
  weight: t = "regular",
  align: s,
  truncate: p = !1,
  color: o,
  className: g,
  style: c,
  ...n
}) => {
  const d = a || "span", u = (o == null ? void 0 : o.startsWith("#")) || (o == null ? void 0 : o.startsWith("rgb")) || (o == null ? void 0 : o.startsWith("hsl")), m = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), f = (() => {
    if (o)
      return o === "inherit" ? "inherit" : u ? o : m.has(o) ? `var(--pittorica-${o}-9)` : o;
  })(), v = i ? typeof i == "string" ? `pittorica-text--size-${i}` : q({
    [`pittorica-text--size-${i.initial}`]: i.initial,
    [`pittorica-text--xs-size-${i.xs}`]: i.xs,
    [`pittorica-text--sm-size-${i.sm}`]: i.sm,
    [`pittorica-text--md-size-${i.md}`]: i.md,
    [`pittorica-text--lg-size-${i.lg}`]: i.lg,
    [`pittorica-text--xl-size-${i.xl}`]: i.xl
  }) : null, l = {
    ...c,
    textAlign: s,
    color: f
  };
  return /* @__PURE__ */ h(
    B,
    {
      as: d,
      className: q(
        "pittorica-text",
        v,
        { "pittorica-text--truncate": p },
        g
      ),
      "data-weight": t,
      style: l,
      ...n,
      children: r
    }
  );
};
w.displayName = "Text";
const C = ({
  className: r,
  style: a,
  variant: i = "classic",
  color: t,
  ...s
}) => {
  const p = t && t !== "inherit" && !t.startsWith("#") && !t.startsWith("rgb"), o = {
    "--_bq-accent": p ? `var(--pittorica-${t}-9)` : t,
    ...i === "soft" && p && {
      "--pittorica-source-3": `var(--pittorica-${t}-3)`,
      "--pittorica-source-9": `var(--pittorica-${t}-9)`,
      "--pittorica-source-11": `var(--pittorica-${t}-11)`
    },
    ...i === "solid" && p && {
      "--pittorica-source-9": `var(--pittorica-${t}-9)`,
      "--pittorica-on-source-9": `var(--pittorica-on-${t}-9)`
    },
    ...i === "outline" && p && {
      "--pittorica-source-6": `var(--pittorica-${t}-6)`,
      "--pittorica-source-9": `var(--pittorica-${t}-9)`
    },
    ...a
  };
  return /* @__PURE__ */ h(
    w,
    {
      as: "blockquote",
      className: V("pittorica-blockquote", r),
      "data-variant": i,
      style: o,
      ...s
    }
  );
};
C.displayName = "Blockquote";
export {
  C as Blockquote
};
