// vuoto@0.1.2/dist/cli.js
#!/usr/bin/env node
import { t as runNormalize, x as expandPatterns } from "./chunks/actions-CAIbGFBy.js";
import path from "node:path";
import chalk from "chalk";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { Command } from "commander";

//#region src/cli.ts
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const pkg = JSON.parse(readFileSync(path.resolve(__dirname, "../package.json"), "utf8"));
const program = new Command();
program.name("vuoto").description(chalk.cyanBright.bold("✨ Modern whitespace normalizer for cleaner codebases")).version(pkg.version, "-v, --version", "output the current version").helpOption("-h, --help", "display help for command").argument("[patterns...]", "files or globs to normalize (default: .)").option("--fix", "apply fixes to the files").option("--exclude <globs...>", "glob patterns to exclude (like in .gitignore)").option("--verbose", "show detailed output").option("--silent", "suppress all output except errors").addHelpText("after", `
${chalk.bold("Examples:")}
  ${chalk.gray("# Check all files in current directory")}
  ${chalk.cyanBright("$ vuoto")}

  ${chalk.gray("# Fix all TypeScript files")}
  ${chalk.cyanBright("$ vuoto src/**/*.ts --fix")}

  ${chalk.gray("# Check specific files excluding tests")}
  ${chalk.cyanBright("$ vuoto src/ --exclude \"**/*.test.ts\"")}

${chalk.bold("Configuration:")}
  Create a ${chalk.cyanBright(".vuotoignore")} file with patterns to exclude (one per line)
  ${chalk.dim("# Example .vuotoignore")}
  ${chalk.dim("dist/**")}
  ${chalk.dim("build/**")}
  ${chalk.dim("*.min.js")}

${chalk.bold("Learn more:")} ${chalk.blueBright("https://github.com/dcdavidev/vuoto")}
`).action(async (patterns, options) => {
	try {
		if (!patterns || patterns.length === 0) patterns = ["."];
		if (options.verbose && !options.silent) console.log(chalk.dim(`Expanding patterns: ${patterns.join(", ")}`));
		const expanded = await expandPatterns(patterns);
		if (options.verbose && !options.silent) console.log(chalk.dim(`Resolved to: ${expanded.join(", ")}`));
		await runNormalize({
			patterns: expanded,
			fix: options.fix,
			exclude: options.exclude,
			verbose: options.verbose,
			silent: options.silent
		});
	} catch (error) {
		console.error(chalk.red("✗ Error:"), error instanceof Error ? error.message : String(error));
		process.exitCode = 1;
	}
});
program.exitOverride((err) => {
	throw err;
});
/**
* Run the vuoto CLI program.
* @example
*
* ```ts
* run();
* ```
*/
function run() {
	program.parse();
}
try {
	run();
} catch (error_) {
	const error = error_;
	if (error.code === "commander.help" || error.code === "commander.version") {} else process.exitCode = 1;
}

//#endregion
export { run };
