// vuoto@0.1.2/dist/consts.js
//#region src/consts.ts
/**
* Default ignore patterns for vuoto
* (these files/folders are skipped unless explicitly included).
*/
const DEFAULT_IGNORES = [
	"**/node_modules/**",
	".git/**",
	"**/*.png",
	"**/*.jpg",
	"**/*.jpeg",
	"**/*.gif",
	"**/*.webp",
	"**/*.avif",
	"**/*.svg",
	"**/*.ico",
	"**/*.mp3",
	"**/*.wav",
	"**/*.ogg",
	"**/*.mp4",
	"**/*.mkv",
	"**/*.avi",
	"**/*.mov",
	"**/*.pdf",
	"**/*.zip",
	"**/*.tar",
	"**/*.gz",
	"**/*.7z",
	"**/*.exe",
	"**/*.dll",
	"**/*.bin"
];
/**
* Name of the vuoto ignore file.
*/
const VUOTO_IGNORE_FILENAME = ".vuotoignore";
/**
* Byte order mark regex pattern.
*/
const BYTE_ORDER_MARK = new RegExp(String.raw`\uFEFF`, "g");
/**
* Em space regex pattern.
*/
const EM_SPACE = new RegExp(String.raw`\u2003`, "g");
/**
* En space regex pattern.
*/
const EN_SPACE = new RegExp(String.raw`\u2002`, "g");
/**
* Form feed regex pattern.
*/
const FORM_FEED = new RegExp(String.raw`\f`, "g");
/**
* Ideographic space regex pattern.
*/
const IDEOGRAPHIC_SPACE = new RegExp(String.raw`\u3000`, "g");
/**
* Invisible separators regex pattern.
*/
const INVISIBLE_SEPARATORS = new RegExp(String.raw`[\u2060\uFEFF]`, "g");
/**
* Line separator regex pattern.
*/
const LINE_SEPARATOR = new RegExp(String.raw`\u2028`, "g");
/**
* Narrow no-break space regex pattern.
*/
const NARROW_NO_BREAK_SPACE = new RegExp(String.raw`\u202F`, "g");
/**
* Non-breaking space regex pattern.
*/
const NON_BREAKING_SPACE = new RegExp(String.raw`\u00A0`, "g");
/**
* Paragraph separator regex pattern.
*/
const PARAGRAPH_SEPARATOR = new RegExp(String.raw`\u2029`, "g");
/**
* Vertical tab regex pattern.
*/
const VERTICAL_TAB = new RegExp(String.raw`\x0B`, "g");
/**
* Visible miscellaneous spaces regex pattern.
*/
const VISIBLE_MISC_SPACES = new RegExp(String.raw`[\u1680\u180E\u2000-\u200A\u205F\u3000]`, "g");
/**
* Zero-width characters regex pattern.
*/
const ZERO_WIDTH = new RegExp(String.raw`\u200B|\u200C|\u200D|\uFEFF|\u2060`, "g");
/**
* Map of Unicode characters to their human-readable names.
*/
const UNICODE_NAME_MAP = {
	"\xA0": "NON-BREAKING SPACE",
	" ": "OGHAM SPACE MARK",
	"᠎": "MONGOLIAN VOWEL SEPARATOR",
	" ": "EN QUAD",
	" ": "EM QUAD",
	" ": "EN SPACE",
	" ": "EM SPACE",
	" ": "THREE-PER-EM SPACE",
	" ": "FOUR-PER-EM SPACE",
	" ": "SIX-PER-EM SPACE",
	" ": "FIGURE SPACE",
	" ": "PUNCTUATION SPACE",
	" ": "THIN SPACE",
	" ": "HAIR SPACE",
	"​": "ZERO WIDTH SPACE",
	"‌": "ZERO WIDTH NON-JOINER",
	"‍": "ZERO WIDTH JOINER",
	" ": "NARROW NO-BREAK SPACE",
	" ": "MEDIUM MATHEMATICAL SPACE",
	"⁠": "WORD JOINER",
	"　": "IDEOGRAPHIC SPACE",
	"﻿": "ZERO WIDTH NO-BREAK SPACE (BOM)",
	"\u2028": "LINE SEPARATOR",
	"\u2029": "PARAGRAPH_SEPARATOR",
	"\f": "FORM FEED",
	"\v": "VERTICAL TAB"
};

//#endregion
export { BYTE_ORDER_MARK, DEFAULT_IGNORES, EM_SPACE, EN_SPACE, FORM_FEED, IDEOGRAPHIC_SPACE, INVISIBLE_SEPARATORS, LINE_SEPARATOR, NARROW_NO_BREAK_SPACE, NON_BREAKING_SPACE, PARAGRAPH_SEPARATOR, UNICODE_NAME_MAP, VERTICAL_TAB, VISIBLE_MISC_SPACES, VUOTO_IGNORE_FILENAME, ZERO_WIDTH };
