// @pittorica/heading-react@0.26.3/dist/index.js
import { jsx as v } from "react/jsx-runtime";
function A(r) {
  var a, t, i = "";
  if (typeof r == "string" || typeof r == "number") i += r;
  else if (typeof r == "object") if (Array.isArray(r)) {
    var s = r.length;
    for (a = 0; a < s; a++) r[a] && (t = A(r[a])) && (i && (i += " "), i += t);
  } else for (t in r) r[t] && (i && (i += " "), i += t);
  return i;
}
function H() {
  for (var r, a, t = 0, i = "", s = arguments.length; t < s; t++) (r = arguments[t]) && (a = A(r)) && (i && (i += " "), i += a);
  return i;
}
function N(r) {
  var a, t, i = "";
  if (typeof r == "string" || typeof r == "number") i += r;
  else if (typeof r == "object") if (Array.isArray(r)) {
    var s = r.length;
    for (a = 0; a < s; a++) r[a] && (t = N(r[a])) && (i && (i += " "), i += t);
  } else for (t in r) r[t] && (i && (i += " "), i += t);
  return i;
}
function z() {
  for (var r, a, t = 0, i = "", s = arguments.length; t < s; t++) (r = arguments[t]) && (a = N(r)) && (i && (i += " "), i += a);
  return i;
}
function w(r) {
  var a, t, i = "";
  if (typeof r == "string" || typeof r == "number") i += r;
  else if (typeof r == "object") if (Array.isArray(r)) {
    var s = r.length;
    for (a = 0; a < s; a++) r[a] && (t = w(r[a])) && (i && (i += " "), i += t);
  } else for (t in r) r[t] && (i && (i += " "), i += t);
  return i;
}
function k() {
  for (var r, a, t = 0, i = "", s = arguments.length; t < s; t++) (r = arguments[t]) && (a = w(r)) && (i && (i += " "), i += a);
  return i;
}
const T = ({
  as: r,
  children: a,
  display: t,
  m: i,
  mt: s,
  mr: o,
  mb: n,
  ml: l,
  mx: p,
  my: c,
  p: d,
  pt: g,
  pr: f,
  pb: u,
  pl: h,
  px: m,
  py: $,
  width: y,
  height: x,
  position: b,
  style: j,
  className: B,
  disabled: R,
  required: L,
  ...q
}) => {
  const S = r || "div", e = {};
  t && (e.display = t), y && (e.width = y), x && (e.height = x), b && (e.position = b), i && (e.margin = `var(--pittorica-space-${i})`), s && (e.marginTop = `var(--pittorica-space-${s})`), o && (e.marginRight = `var(--pittorica-space-${o})`), n && (e.marginBottom = `var(--pittorica-space-${n})`), l && (e.marginLeft = `var(--pittorica-space-${l})`), p && (e.marginLeft = `var(--pittorica-space-${p})`, e.marginRight = `var(--pittorica-space-${p})`), c && (e.marginTop = `var(--pittorica-space-${c})`, e.marginBottom = `var(--pittorica-space-${c})`), d && (e.padding = `var(--pittorica-space-${d})`), g && (e.paddingTop = `var(--pittorica-space-${g})`), f && (e.paddingRight = `var(--pittorica-space-${f})`), u && (e.paddingBottom = `var(--pittorica-space-${u})`), h && (e.paddingLeft = `var(--pittorica-space-${h})`), m && (e.paddingLeft = `var(--pittorica-space-${m})`, e.paddingRight = `var(--pittorica-space-${m})`), $ && (e.paddingTop = `var(--pittorica-space-${$})`, e.paddingBottom = `var(--pittorica-space-${$})`);
  const C = {
    ...j,
    ...e
  };
  return /* @__PURE__ */ v(
    S,
    {
      className: k("pittorica-box", B),
      style: C,
      disabled: R,
      required: L,
      ...q,
      children: a
    }
  );
};
T.displayName = "Box";
const W = ({
  children: r,
  as: a,
  size: t,
  weight: i = "regular",
  align: s,
  truncate: o = !1,
  color: n,
  className: l,
  style: p,
  ...c
}) => {
  const d = a || "span", g = (n == null ? void 0 : n.startsWith("#")) || (n == null ? void 0 : n.startsWith("rgb")) || (n == null ? void 0 : n.startsWith("hsl")), f = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), u = (() => {
    if (n)
      return n === "inherit" ? "inherit" : g ? n : f.has(n) ? `var(--pittorica-${n}-9)` : n;
  })(), h = t ? typeof t == "string" ? `pittorica-text--size-${t}` : z({
    [`pittorica-text--size-${t.initial}`]: t.initial,
    [`pittorica-text--xs-size-${t.xs}`]: t.xs,
    [`pittorica-text--sm-size-${t.sm}`]: t.sm,
    [`pittorica-text--md-size-${t.md}`]: t.md,
    [`pittorica-text--lg-size-${t.lg}`]: t.lg,
    [`pittorica-text--xl-size-${t.xl}`]: t.xl
  }) : null, m = {
    ...p,
    textAlign: s,
    color: u
  };
  return /* @__PURE__ */ v(
    T,
    {
      as: d,
      className: z(
        "pittorica-text",
        h,
        { "pittorica-text--truncate": o },
        l
      ),
      "data-weight": i,
      style: m,
      ...c,
      children: r
    }
  );
};
W.displayName = "Text";
const O = (r, a) => a ? typeof a == "string" ? [`pittorica-heading--${r}-${a}`] : Object.entries(a).filter(([t, i]) => i !== void 0).map(
  ([t, i]) => t === "initial" ? `pittorica-heading--${r}-${i}` : `pittorica-heading--${t}-${r}-${i}`
) : [], V = ({
  children: r,
  as: a,
  size: t = "6",
  weight: i = "bold",
  variant: s = "classic",
  color: o,
  className: n,
  style: l,
  ...p
}) => {
  const c = O("size", t), d = a || "h1", g = o && o !== "inherit" && !o.startsWith("#") && !o.startsWith("rgb"), f = {
    ...s === "soft" && g && {
      "--pittorica-source-3": `var(--pittorica-${o}-3)`,
      "--pittorica-source-11": `var(--pittorica-${o}-11)`
    },
    ...s === "outline" && g && {
      "--pittorica-source-9": `var(--pittorica-${o}-9)`
    },
    ...l
  };
  return /* @__PURE__ */ v(
    W,
    {
      as: d,
      weight: i,
      className: H("pittorica-heading", c, n),
      "data-variant": s,
      style: f,
      color: o,
      ...p,
      children: r
    }
  );
};
V.displayName = "Heading";
export {
  V as Heading
};
