// @pittorica/callout-react@0.26.3/dist/index.js
import { jsx as u } from "react/jsx-runtime";
function C(i) {
  var t, r, a = "";
  if (typeof i == "string" || typeof i == "number") a += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var p = i.length;
    for (t = 0; t < p; t++) i[t] && (r = C(i[t])) && (a && (a += " "), a += r);
  } else for (r in i) i[r] && (a && (a += " "), a += r);
  return a;
}
function b() {
  for (var i, t, r = 0, a = "", p = arguments.length; r < p; r++) (i = arguments[r]) && (t = C(i)) && (a && (a += " "), a += t);
  return a;
}
function B(i) {
  var t, r, a = "";
  if (typeof i == "string" || typeof i == "number") a += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var p = i.length;
    for (t = 0; t < p; t++) i[t] && (r = B(i[t])) && (a && (a += " "), a += r);
  } else for (r in i) i[r] && (a && (a += " "), a += r);
  return a;
}
function W() {
  for (var i, t, r = 0, a = "", p = arguments.length; r < p; r++) (i = arguments[r]) && (t = B(i)) && (a && (a += " "), a += t);
  return a;
}
const $ = ({
  as: i,
  children: t,
  display: r,
  m: a,
  mt: p,
  mr: n,
  mb: l,
  ml: d,
  mx: e,
  my: c,
  p: s,
  pt: g,
  pr: m,
  pb: v,
  pl: f,
  px: y,
  py: h,
  width: x,
  height: T,
  position: N,
  style: A,
  className: L,
  disabled: q,
  required: I,
  ..._
}) => {
  const k = i || "div", o = {};
  r && (o.display = r), x && (o.width = x), T && (o.height = T), N && (o.position = N), a && (o.margin = `var(--pittorica-space-${a})`), p && (o.marginTop = `var(--pittorica-space-${p})`), n && (o.marginRight = `var(--pittorica-space-${n})`), l && (o.marginBottom = `var(--pittorica-space-${l})`), d && (o.marginLeft = `var(--pittorica-space-${d})`), e && (o.marginLeft = `var(--pittorica-space-${e})`, o.marginRight = `var(--pittorica-space-${e})`), c && (o.marginTop = `var(--pittorica-space-${c})`, o.marginBottom = `var(--pittorica-space-${c})`), s && (o.padding = `var(--pittorica-space-${s})`), g && (o.paddingTop = `var(--pittorica-space-${g})`), m && (o.paddingRight = `var(--pittorica-space-${m})`), v && (o.paddingBottom = `var(--pittorica-space-${v})`), f && (o.paddingLeft = `var(--pittorica-space-${f})`), y && (o.paddingLeft = `var(--pittorica-space-${y})`, o.paddingRight = `var(--pittorica-space-${y})`), h && (o.paddingTop = `var(--pittorica-space-${h})`, o.paddingBottom = `var(--pittorica-space-${h})`);
  const S = {
    ...A,
    ...o
  };
  return /* @__PURE__ */ u(
    k,
    {
      className: W("pittorica-box", L),
      style: S,
      disabled: q,
      required: I,
      ..._,
      children: t
    }
  );
};
$.displayName = "Box";
const R = ({
  variant: i = "soft",
  color: t = "indigo",
  size: r = "md",
  direction: a = "row",
  align: p = "start",
  children: n,
  className: l,
  style: d,
  as: e,
  ...c
}) => {
  const s = t !== "inherit" && !(t != null && t.startsWith("#")) && !(t != null && t.startsWith("rgb")), g = s ? `var(--pittorica-${t}-3)` : `color-mix(in srgb, ${t} 12%, var(--pittorica-white))`, m = s ? `var(--pittorica-${t}-11)` : `color-mix(in srgb, ${t} 80%, var(--pittorica-black))`, v = s ? `var(--pittorica-${t}-6)` : `color-mix(in srgb, ${t} 20%, transparent)`;
  return /* @__PURE__ */ u(
    $,
    {
      as: e || "div",
      className: b(
        "pittorica-callout",
        `pittorica-callout--${i}`,
        `pittorica-callout--${r}`,
        `pittorica-callout--dir-${a}`,
        `pittorica-callout--align-${p}`,
        l
      ),
      style: {
        "--_callout-bg": g,
        "--_callout-text": m,
        "--_callout-border": v,
        ...d
      },
      ...c,
      children: n
    }
  );
}, j = ({
  children: i,
  className: t,
  as: r,
  ...a
}) => /* @__PURE__ */ u(
  $,
  {
    as: r || "div",
    className: b("pittorica-callout-icon", t),
    ...a,
    children: i
  }
), w = ({
  children: i,
  className: t,
  as: r,
  ...a
}) => /* @__PURE__ */ u(
  $,
  {
    as: r || "div",
    className: b("pittorica-callout-content", t),
    ...a,
    children: i
  }
), z = Object.assign(R, {
  Icon: j,
  Text: w
});
R.displayName = "Callout.Root";
j.displayName = "Callout.Icon";
w.displayName = "Callout.Text";
export {
  z as Callout
};
