// @pittorica/chip-react@0.26.3/dist/index.js
import { jsx as f, jsxs as q } from "react/jsx-runtime";
import { forwardRef as M, createElement as $ } from "react";
function x(i) {
  var r, a, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var p = i.length;
    for (r = 0; r < p; r++) i[r] && (a = x(i[r])) && (t && (t += " "), t += a);
  } else for (a in i) i[a] && (t && (t += " "), t += a);
  return t;
}
function S() {
  for (var i, r, a = 0, t = "", p = arguments.length; a < p; a++) (i = arguments[a]) && (r = x(i)) && (t && (t += " "), t += r);
  return t;
}
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var X = {
  outline: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  },
  filled: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "currentColor",
    stroke: "none"
  }
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const _ = (i, r, a, t) => {
  const p = M(
    ({ color: h = "currentColor", size: n = 24, stroke: m = 2, title: o, className: g, children: c, ...l }, d) => $(
      "svg",
      {
        ref: d,
        ...X[i],
        width: n,
        height: n,
        className: ["tabler-icon", `tabler-icon-${r}`, g].join(" "),
        strokeWidth: m,
        stroke: h,
        ...l
      },
      [
        o && $("title", { key: "svg-title" }, o),
        ...t.map(([s, v]) => $(s, v)),
        ...Array.isArray(c) ? c : [c]
      ]
    )
  );
  return p.displayName = `${a}`, p;
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const z = [["path", { d: "M18 6l-12 12", key: "svg-0" }], ["path", { d: "M6 6l12 12", key: "svg-1" }]], E = _("outline", "x", "X", z);
function N(i) {
  var r, a, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var p = i.length;
    for (r = 0; r < p; r++) i[r] && (a = N(i[r])) && (t && (t += " "), t += a);
  } else for (a in i) i[a] && (t && (t += " "), t += a);
  return t;
}
function I() {
  for (var i, r, a = 0, t = "", p = arguments.length; a < p; a++) (i = arguments[a]) && (r = N(i)) && (t && (t += " "), t += r);
  return t;
}
const C = ({
  as: i,
  children: r,
  display: a,
  m: t,
  mt: p,
  mr: h,
  mb: n,
  ml: m,
  mx: o,
  my: g,
  p: c,
  pt: l,
  pr: d,
  pb: s,
  pl: v,
  px: u,
  py: y,
  width: b,
  height: w,
  position: k,
  style: A,
  className: j,
  disabled: B,
  required: L,
  ...R
}) => {
  const T = i || "div", e = {};
  a && (e.display = a), b && (e.width = b), w && (e.height = w), k && (e.position = k), t && (e.margin = `var(--pittorica-space-${t})`), p && (e.marginTop = `var(--pittorica-space-${p})`), h && (e.marginRight = `var(--pittorica-space-${h})`), n && (e.marginBottom = `var(--pittorica-space-${n})`), m && (e.marginLeft = `var(--pittorica-space-${m})`), o && (e.marginLeft = `var(--pittorica-space-${o})`, e.marginRight = `var(--pittorica-space-${o})`), g && (e.marginTop = `var(--pittorica-space-${g})`, e.marginBottom = `var(--pittorica-space-${g})`), c && (e.padding = `var(--pittorica-space-${c})`), l && (e.paddingTop = `var(--pittorica-space-${l})`), d && (e.paddingRight = `var(--pittorica-space-${d})`), s && (e.paddingBottom = `var(--pittorica-space-${s})`), v && (e.paddingLeft = `var(--pittorica-space-${v})`), u && (e.paddingLeft = `var(--pittorica-space-${u})`, e.paddingRight = `var(--pittorica-space-${u})`), y && (e.paddingTop = `var(--pittorica-space-${y})`, e.paddingBottom = `var(--pittorica-space-${y})`);
  const W = {
    ...A,
    ...e
  };
  return /* @__PURE__ */ f(
    T,
    {
      className: I("pittorica-box", j),
      style: W,
      disabled: B,
      required: L,
      ...R,
      children: r
    }
  );
};
C.displayName = "Box";
const P = ({
  children: i,
  variant: r = "soft",
  size: a = "2",
  color: t = "indigo",
  startDecorator: p,
  endDecorator: h,
  onDelete: n,
  deleteIcon: m,
  appearance: o,
  className: g,
  style: c,
  as: l,
  ...d
}) => {
  const s = t !== "inherit" && !(t != null && t.startsWith("#")) && !(t != null && t.startsWith("rgb")), v = {
    "--chip-base": s ? `var(--pittorica-${t}-9)` : t,
    "--chip-soft-bg": s ? `var(--pittorica-${t}-3)` : "transparent",
    "--chip-soft-text": s ? `var(--pittorica-${t}-11)` : "inherit",
    "--chip-on-base": s ? `var(--pittorica-on-${t}-9)` : "var(--pittorica-white)",
    ...c
  }, u = l || (d.onClick && !n ? "button" : "div");
  return /* @__PURE__ */ q(
    C,
    {
      as: u,
      className: S(
        "pittorica-chip",
        `pittorica-chip--size-${a}`,
        {
          "pittorica-chip--clickable": !!d.onClick
        },
        g
      ),
      "data-variant": r,
      "data-appearance": o,
      style: v,
      ...d,
      children: [
        p && /* @__PURE__ */ f("span", { className: "pittorica-chip-decorator", children: p }),
        /* @__PURE__ */ f("span", { className: "pittorica-chip-content", children: i }),
        h && /* @__PURE__ */ f("span", { className: "pittorica-chip-decorator", children: h }),
        n && /* @__PURE__ */ f(
          "button",
          {
            type: "button",
            "aria-label": "Delete",
            className: "pittorica-chip-delete",
            onClick: (y) => {
              y.stopPropagation(), n();
            },
            children: m || /* @__PURE__ */ f(E, { size: 14 })
          }
        )
      ]
    }
  );
};
P.displayName = "Chip";
export {
  P as Chip
};
