// @pittorica/link-react@0.26.3/dist/index.js
import { jsx as y } from "react/jsx-runtime";
function z(r) {
  var a, t, i = "";
  if (typeof r == "string" || typeof r == "number") i += r;
  else if (typeof r == "object") if (Array.isArray(r)) {
    var o = r.length;
    for (a = 0; a < o; a++) r[a] && (t = z(r[a])) && (i && (i += " "), i += t);
  } else for (t in r) r[t] && (i && (i += " "), i += t);
  return i;
}
function C() {
  for (var r, a, t = 0, i = "", o = arguments.length; t < o; t++) (r = arguments[t]) && (a = z(r)) && (i && (i += " "), i += a);
  return i;
}
function A(r) {
  var a, t, i = "";
  if (typeof r == "string" || typeof r == "number") i += r;
  else if (typeof r == "object") if (Array.isArray(r)) {
    var o = r.length;
    for (a = 0; a < o; a++) r[a] && (t = A(r[a])) && (i && (i += " "), i += t);
  } else for (t in r) r[t] && (i && (i += " "), i += t);
  return i;
}
function N() {
  for (var r, a, t = 0, i = "", o = arguments.length; t < o; t++) (r = arguments[t]) && (a = A(r)) && (i && (i += " "), i += a);
  return i;
}
function T(r) {
  var a, t, i = "";
  if (typeof r == "string" || typeof r == "number") i += r;
  else if (typeof r == "object") if (Array.isArray(r)) {
    var o = r.length;
    for (a = 0; a < o; a++) r[a] && (t = T(r[a])) && (i && (i += " "), i += t);
  } else for (t in r) r[t] && (i && (i += " "), i += t);
  return i;
}
function D() {
  for (var r, a, t = 0, i = "", o = arguments.length; t < o; t++) (r = arguments[t]) && (a = T(r)) && (i && (i += " "), i += a);
  return i;
}
const w = ({
  as: r,
  children: a,
  display: t,
  m: i,
  mt: o,
  mr: s,
  mb: n,
  ml: g,
  mx: p,
  my: c,
  p: d,
  pt: m,
  pr: f,
  pb: u,
  pl: v,
  px: l,
  py: h,
  width: $,
  height: x,
  position: b,
  style: k,
  className: B,
  disabled: j,
  required: R,
  ...W
}) => {
  const q = r || "div", e = {};
  t && (e.display = t), $ && (e.width = $), x && (e.height = x), b && (e.position = b), i && (e.margin = `var(--pittorica-space-${i})`), o && (e.marginTop = `var(--pittorica-space-${o})`), s && (e.marginRight = `var(--pittorica-space-${s})`), n && (e.marginBottom = `var(--pittorica-space-${n})`), g && (e.marginLeft = `var(--pittorica-space-${g})`), p && (e.marginLeft = `var(--pittorica-space-${p})`, e.marginRight = `var(--pittorica-space-${p})`), c && (e.marginTop = `var(--pittorica-space-${c})`, e.marginBottom = `var(--pittorica-space-${c})`), d && (e.padding = `var(--pittorica-space-${d})`), m && (e.paddingTop = `var(--pittorica-space-${m})`), f && (e.paddingRight = `var(--pittorica-space-${f})`), u && (e.paddingBottom = `var(--pittorica-space-${u})`), v && (e.paddingLeft = `var(--pittorica-space-${v})`), l && (e.paddingLeft = `var(--pittorica-space-${l})`, e.paddingRight = `var(--pittorica-space-${l})`), h && (e.paddingTop = `var(--pittorica-space-${h})`, e.paddingBottom = `var(--pittorica-space-${h})`);
  const S = {
    ...k,
    ...e
  };
  return /* @__PURE__ */ y(
    q,
    {
      className: D("pittorica-box", B),
      style: S,
      disabled: j,
      required: R,
      ...W,
      children: a
    }
  );
};
w.displayName = "Box";
const L = ({
  children: r,
  as: a,
  size: t,
  weight: i = "regular",
  align: o,
  truncate: s = !1,
  color: n,
  className: g,
  style: p,
  ...c
}) => {
  const d = a || "span", m = (n == null ? void 0 : n.startsWith("#")) || (n == null ? void 0 : n.startsWith("rgb")) || (n == null ? void 0 : n.startsWith("hsl")), f = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), u = (() => {
    if (n)
      return n === "inherit" ? "inherit" : m ? n : f.has(n) ? `var(--pittorica-${n}-9)` : n;
  })(), v = t ? typeof t == "string" ? `pittorica-text--size-${t}` : N({
    [`pittorica-text--size-${t.initial}`]: t.initial,
    [`pittorica-text--xs-size-${t.xs}`]: t.xs,
    [`pittorica-text--sm-size-${t.sm}`]: t.sm,
    [`pittorica-text--md-size-${t.md}`]: t.md,
    [`pittorica-text--lg-size-${t.lg}`]: t.lg,
    [`pittorica-text--xl-size-${t.xl}`]: t.xl
  }) : null, l = {
    ...p,
    textAlign: o,
    color: u
  };
  return /* @__PURE__ */ y(
    w,
    {
      as: d,
      className: N(
        "pittorica-text",
        v,
        { "pittorica-text--truncate": s },
        g
      ),
      "data-weight": i,
      style: l,
      ...c,
      children: r
    }
  );
};
L.displayName = "Text";
const E = ({
  underline: r = "hover",
  className: a,
  style: t,
  color: i,
  // Default handled by CSS or inherited if not provided
  as: o,
  ...s
}) => /* @__PURE__ */ y(
  L,
  {
    as: o || "a",
    className: C(
      "pittorica-link",
      `pittorica-link--underline-${r}`,
      a
    ),
    color: i,
    style: t,
    ...s
  }
);
E.displayName = "Link";
export {
  E as Link
};
