// @pittorica/data-list-react@0.26.3/dist/index.js
import { jsx as p } from "react/jsx-runtime";
function D(t) {
  var r, i, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var s = t.length;
    for (r = 0; r < s; r++) t[r] && (i = D(t[r])) && (a && (a += " "), a += i);
  } else for (i in t) t[i] && (a && (a += " "), a += i);
  return a;
}
function c() {
  for (var t, r, i = 0, a = "", s = arguments.length; i < s; i++) (t = arguments[i]) && (r = D(t)) && (a && (a += " "), a += r);
  return a;
}
function R(t) {
  var r, i, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var s = t.length;
    for (r = 0; r < s; r++) t[r] && (i = R(t[r])) && (a && (a += " "), a += i);
  } else for (i in t) t[i] && (a && (a += " "), a += i);
  return a;
}
function V() {
  for (var t, r, i = 0, a = "", s = arguments.length; i < s; i++) (t = arguments[i]) && (r = R(t)) && (a && (a += " "), a += r);
  return a;
}
const e = ({
  as: t,
  children: r,
  display: i,
  m: a,
  mt: s,
  mr: g,
  mb: v,
  ml: f,
  mx: n,
  my: d,
  p: u,
  pt: $,
  pr: h,
  pb: y,
  pl: b,
  px: l,
  py: m,
  width: L,
  height: N,
  position: x,
  style: T,
  className: j,
  disabled: A,
  required: B,
  ...q
}) => {
  const w = t || "div", o = {};
  i && (o.display = i), L && (o.width = L), N && (o.height = N), x && (o.position = x), a && (o.margin = `var(--pittorica-space-${a})`), s && (o.marginTop = `var(--pittorica-space-${s})`), g && (o.marginRight = `var(--pittorica-space-${g})`), v && (o.marginBottom = `var(--pittorica-space-${v})`), f && (o.marginLeft = `var(--pittorica-space-${f})`), n && (o.marginLeft = `var(--pittorica-space-${n})`, o.marginRight = `var(--pittorica-space-${n})`), d && (o.marginTop = `var(--pittorica-space-${d})`, o.marginBottom = `var(--pittorica-space-${d})`), u && (o.padding = `var(--pittorica-space-${u})`), $ && (o.paddingTop = `var(--pittorica-space-${$})`), h && (o.paddingRight = `var(--pittorica-space-${h})`), y && (o.paddingBottom = `var(--pittorica-space-${y})`), b && (o.paddingLeft = `var(--pittorica-space-${b})`), l && (o.paddingLeft = `var(--pittorica-space-${l})`, o.paddingRight = `var(--pittorica-space-${l})`), m && (o.paddingTop = `var(--pittorica-space-${m})`, o.paddingBottom = `var(--pittorica-space-${m})`);
  const I = {
    ...T,
    ...o
  };
  return /* @__PURE__ */ p(
    w,
    {
      className: V("pittorica-box", j),
      style: I,
      disabled: A,
      required: B,
      ...q,
      children: r
    }
  );
};
e.displayName = "Box";
const k = ({
  orientation: t = "horizontal",
  children: r,
  className: i,
  as: a,
  ...s
}) => /* @__PURE__ */ p(
  e,
  {
    as: a || "div",
    className: c(
      "pittorica-data-list-root",
      `pittorica-data-list-root--orientation-${t}`,
      i
    ),
    ...s,
    children: r
  }
), z = ({
  children: t,
  className: r,
  as: i,
  ...a
}) => /* @__PURE__ */ p(
  e,
  {
    as: i || "div",
    className: c("pittorica-data-list-item", r),
    ...a,
    children: t
  }
), O = ({
  children: t,
  className: r,
  as: i,
  ...a
}) => /* @__PURE__ */ p(
  e,
  {
    as: i || "div",
    className: c("pittorica-data-list-label", r),
    ...a,
    children: t
  }
), S = ({
  children: t,
  className: r,
  as: i,
  ...a
}) => /* @__PURE__ */ p(
  e,
  {
    as: i || "div",
    className: c("pittorica-data-list-value", r),
    ...a,
    children: t
  }
), E = Object.assign(k, {
  Item: z,
  Label: O,
  Value: S
});
export {
  E as DataList
};
