// @pittorica/divider-react@0.26.3/dist/index.js
import { jsx as p, jsxs as C } from "react/jsx-runtime";
import { useId as E } from "react";
function H(i) {
  var r, a, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var s = i.length;
    for (r = 0; r < s; r++) i[r] && (a = H(i[r])) && (t && (t += " "), t += a);
  } else for (a in i) i[a] && (t && (t += " "), t += a);
  return t;
}
function M() {
  for (var i, r, a = 0, t = "", s = arguments.length; a < s; a++) (i = arguments[a]) && (r = H(i)) && (t && (t += " "), t += r);
  return t;
}
function S(i) {
  var r, a, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var s = i.length;
    for (r = 0; r < s; r++) i[r] && (a = S(i[r])) && (t && (t += " "), t += a);
  } else for (a in i) i[a] && (t && (t += " "), t += a);
  return t;
}
function I() {
  for (var i, r, a = 0, t = "", s = arguments.length; a < s; a++) (i = arguments[a]) && (r = S(i)) && (t && (t += " "), t += r);
  return t;
}
const b = ({
  as: i,
  children: r,
  display: a,
  m: t,
  mt: s,
  mr: n,
  mb: d,
  ml: c,
  mx: l,
  my: g,
  p: y,
  pt: m,
  pr: w,
  pb: v,
  pl: $,
  px: h,
  py: f,
  width: u,
  height: x,
  position: L,
  style: q,
  className: R,
  disabled: A,
  required: V,
  ...j
}) => {
  const k = i || "div", e = {};
  a && (e.display = a), u && (e.width = u), x && (e.height = x), L && (e.position = L), t && (e.margin = `var(--pittorica-space-${t})`), s && (e.marginTop = `var(--pittorica-space-${s})`), n && (e.marginRight = `var(--pittorica-space-${n})`), d && (e.marginBottom = `var(--pittorica-space-${d})`), c && (e.marginLeft = `var(--pittorica-space-${c})`), l && (e.marginLeft = `var(--pittorica-space-${l})`, e.marginRight = `var(--pittorica-space-${l})`), g && (e.marginTop = `var(--pittorica-space-${g})`, e.marginBottom = `var(--pittorica-space-${g})`), y && (e.padding = `var(--pittorica-space-${y})`), m && (e.paddingTop = `var(--pittorica-space-${m})`), w && (e.paddingRight = `var(--pittorica-space-${w})`), v && (e.paddingBottom = `var(--pittorica-space-${v})`), $ && (e.paddingLeft = `var(--pittorica-space-${$})`), h && (e.paddingLeft = `var(--pittorica-space-${h})`, e.paddingRight = `var(--pittorica-space-${h})`), f && (e.paddingTop = `var(--pittorica-space-${f})`, e.paddingBottom = `var(--pittorica-space-${f})`);
  const B = {
    ...q,
    ...e
  };
  return /* @__PURE__ */ p(
    k,
    {
      className: I("pittorica-box", R),
      style: B,
      disabled: A,
      required: V,
      ...j,
      children: r
    }
  );
};
b.displayName = "Box";
const T = 16, o = T / 2, N = {
  wave: { d: `M0 ${o} Q6 0 12 ${o} T24 ${o}`, w: 24 },
  zigzag: { d: `M0 ${o} L5 2 L15 14 L20 ${o}`, w: 20 },
  square: { d: `M0 ${o} V4 H12 V12 H24 V${o}`, w: 24 },
  scallop: { d: "M0 12 Q10 0 20 12", w: 20 },
  dashed: { d: `M0 ${o} H8`, w: 16 },
  double: { d: "M0 6 H100 M0 10 H100", w: 100 },
  cross: { d: "M4 4 L12 12 M12 4 L4 12", w: 16 },
  dots: { d: `M0 ${o} H0.1`, w: 16 }
}, _ = ({
  variant: i = "solid",
  color: r = "slate",
  thickness: a = "1px",
  children: t,
  className: s,
  style: n,
  as: d,
  ...c
}) => {
  const l = E(), g = i !== "solid" && i in N, m = r !== "inherit" && !(r != null && r.startsWith("#")) && !(r != null && r.startsWith("rgb")) ? `var(--pittorica-${r}-9)` : r;
  if (g) {
    const v = `pittorica-pattern-${i}-${l}`, { d: $, w: h } = N[i], f = ["dots", "scallop", "wave"].includes(i) ? "round" : "square", u = i === "dots" ? "3" : "1.5";
    return /* @__PURE__ */ p(
      b,
      {
        as: d || "div",
        role: "separator",
        "aria-orientation": "horizontal",
        className: M(
          "pittorica-divider",
          `pittorica-divider--${i}`,
          s
        ),
        style: { color: m, ...n },
        ...c,
        children: /* @__PURE__ */ C(
          "svg",
          {
            width: "100%",
            height: T,
            fill: "none",
            style: { display: "block", overflow: "visible" },
            children: [
              /* @__PURE__ */ p("defs", { children: /* @__PURE__ */ p(
                "pattern",
                {
                  id: v,
                  width: h,
                  height: T,
                  patternUnits: "userSpaceOnUse",
                  children: /* @__PURE__ */ p(
                    "path",
                    {
                      d: $,
                      stroke: "currentColor",
                      strokeWidth: u,
                      strokeLinecap: f,
                      fill: "none"
                    }
                  )
                }
              ) }),
              /* @__PURE__ */ p("rect", { width: "100%", height: "100%", fill: `url(#${v})` })
            ]
          }
        )
      }
    );
  }
  return /* @__PURE__ */ p(
    b,
    {
      as: d || (t ? "div" : "hr"),
      ...t ? { role: "separator" } : {},
      className: M(
        "pittorica-divider",
        "pittorica-divider--solid",
        { "pittorica-divider--with-children": !!t },
        s
      ),
      style: {
        "--divider-color": m,
        "--divider-thickness": typeof a == "number" ? `${a}px` : a,
        ...n
      },
      ...c,
      children: t && /* @__PURE__ */ p("span", { className: "pittorica-divider__content", children: t })
    }
  );
};
_.displayName = "Divider";
export {
  _ as Divider
};
