// @pittorica/badge-react@0.26.3/dist/index.js
import { jsx as $, Fragment as k, jsxs as W } from "react/jsx-runtime";
function N(i) {
  var r, t, a = "";
  if (typeof i == "string" || typeof i == "number") a += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var e = i.length;
    for (r = 0; r < e; r++) i[r] && (t = N(i[r])) && (a && (a += " "), a += t);
  } else for (t in i) i[t] && (a && (a += " "), a += t);
  return a;
}
function D() {
  for (var i, r, t = 0, a = "", e = arguments.length; t < e; t++) (i = arguments[t]) && (r = N(i)) && (a && (a += " "), a += r);
  return a;
}
function T(i) {
  var r, t, a = "";
  if (typeof i == "string" || typeof i == "number") a += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var e = i.length;
    for (r = 0; r < e; r++) i[r] && (t = T(i[r])) && (a && (a += " "), a += t);
  } else for (t in i) i[t] && (a && (a += " "), a += t);
  return a;
}
function F() {
  for (var i, r, t = 0, a = "", e = arguments.length; t < e; t++) (i = arguments[t]) && (r = T(i)) && (a && (a += " "), a += r);
  return a;
}
const B = ({
  as: i,
  children: r,
  display: t,
  m: a,
  mt: e,
  mr: s,
  mb: d,
  ml: g,
  mx: n,
  my: o,
  p: c,
  pt: m,
  pr: f,
  pb: v,
  pl: y,
  px: h,
  py: u,
  width: l,
  height: b,
  position: x,
  style: j,
  className: A,
  disabled: L,
  required: R,
  ...q
}) => {
  const w = i || "div", p = {};
  t && (p.display = t), l && (p.width = l), b && (p.height = b), x && (p.position = x), a && (p.margin = `var(--pittorica-space-${a})`), e && (p.marginTop = `var(--pittorica-space-${e})`), s && (p.marginRight = `var(--pittorica-space-${s})`), d && (p.marginBottom = `var(--pittorica-space-${d})`), g && (p.marginLeft = `var(--pittorica-space-${g})`), n && (p.marginLeft = `var(--pittorica-space-${n})`, p.marginRight = `var(--pittorica-space-${n})`), o && (p.marginTop = `var(--pittorica-space-${o})`, p.marginBottom = `var(--pittorica-space-${o})`), c && (p.padding = `var(--pittorica-space-${c})`), m && (p.paddingTop = `var(--pittorica-space-${m})`), f && (p.paddingRight = `var(--pittorica-space-${f})`), v && (p.paddingBottom = `var(--pittorica-space-${v})`), y && (p.paddingLeft = `var(--pittorica-space-${y})`), h && (p.paddingLeft = `var(--pittorica-space-${h})`, p.paddingRight = `var(--pittorica-space-${h})`), u && (p.paddingTop = `var(--pittorica-space-${u})`, p.paddingBottom = `var(--pittorica-space-${u})`);
  const S = {
    ...j,
    ...p
  };
  return /* @__PURE__ */ $(
    w,
    {
      className: F("pittorica-box", A),
      style: S,
      disabled: L,
      required: R,
      ...q,
      children: r
    }
  );
};
B.displayName = "Box";
const z = ({
  children: i,
  badgeContent: r,
  variant: t = "standard",
  color: a = "indigo",
  invisible: e = !1,
  max: s = 99,
  className: d,
  style: g,
  as: n,
  ...o
}) => {
  const c = t === "dot", m = !c && typeof r == "number" && r > s ? `${s}+` : r, f = a !== "inherit" && !(a != null && a.startsWith("#")) && !(a != null && a.startsWith("rgb")), v = {
    ...g,
    backgroundColor: f ? `var(--pittorica-${a}-9)` : a,
    color: "white"
  };
  return e ? /* @__PURE__ */ $(k, { children: i }) : /* @__PURE__ */ W("div", { className: "pittorica-badge-container", children: [
    i,
    /* @__PURE__ */ $(
      B,
      {
        as: n || "span",
        className: D(
          "pittorica-badge",
          `pittorica-badge--${t}`,
          d
        ),
        style: v,
        ...o,
        children: !c && m
      }
    )
  ] });
};
z.displayName = "Badge";
export {
  z as Badge
};
