// @pittorica/kbd-react@0.26.3/dist/index.js
import { jsx as y } from "react/jsx-runtime";
function z(i) {
  var a, t, r = "";
  if (typeof i == "string" || typeof i == "number") r += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (a = 0; a < o; a++) i[a] && (t = z(i[a])) && (r && (r += " "), r += t);
  } else for (t in i) i[t] && (r && (r += " "), r += t);
  return r;
}
function S() {
  for (var i, a, t = 0, r = "", o = arguments.length; t < o; t++) (i = arguments[t]) && (a = z(i)) && (r && (r += " "), r += a);
  return r;
}
function A(i) {
  var a, t, r = "";
  if (typeof i == "string" || typeof i == "number") r += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (a = 0; a < o; a++) i[a] && (t = A(i[a])) && (r && (r += " "), r += t);
  } else for (t in i) i[t] && (r && (r += " "), r += t);
  return r;
}
function N() {
  for (var i, a, t = 0, r = "", o = arguments.length; t < o; t++) (i = arguments[t]) && (a = A(i)) && (r && (r += " "), r += a);
  return r;
}
function w(i) {
  var a, t, r = "";
  if (typeof i == "string" || typeof i == "number") r += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (a = 0; a < o; a++) i[a] && (t = w(i[a])) && (r && (r += " "), r += t);
  } else for (t in i) i[t] && (r && (r += " "), r += t);
  return r;
}
function C() {
  for (var i, a, t = 0, r = "", o = arguments.length; t < o; t++) (i = arguments[t]) && (a = w(i)) && (r && (r += " "), r += a);
  return r;
}
const B = ({
  as: i,
  children: a,
  display: t,
  m: r,
  mt: o,
  mr: l,
  mb: s,
  ml: d,
  mx: n,
  my: p,
  p: g,
  pt: m,
  pr: f,
  pb: u,
  pl: v,
  px: c,
  py: h,
  width: $,
  height: x,
  position: b,
  style: j,
  className: L,
  disabled: R,
  required: W,
  ...k
}) => {
  const q = i || "div", e = {};
  t && (e.display = t), $ && (e.width = $), x && (e.height = x), b && (e.position = b), r && (e.margin = `var(--pittorica-space-${r})`), o && (e.marginTop = `var(--pittorica-space-${o})`), l && (e.marginRight = `var(--pittorica-space-${l})`), s && (e.marginBottom = `var(--pittorica-space-${s})`), d && (e.marginLeft = `var(--pittorica-space-${d})`), n && (e.marginLeft = `var(--pittorica-space-${n})`, e.marginRight = `var(--pittorica-space-${n})`), p && (e.marginTop = `var(--pittorica-space-${p})`, e.marginBottom = `var(--pittorica-space-${p})`), g && (e.padding = `var(--pittorica-space-${g})`), m && (e.paddingTop = `var(--pittorica-space-${m})`), f && (e.paddingRight = `var(--pittorica-space-${f})`), u && (e.paddingBottom = `var(--pittorica-space-${u})`), v && (e.paddingLeft = `var(--pittorica-space-${v})`), c && (e.paddingLeft = `var(--pittorica-space-${c})`, e.paddingRight = `var(--pittorica-space-${c})`), h && (e.paddingTop = `var(--pittorica-space-${h})`, e.paddingBottom = `var(--pittorica-space-${h})`);
  const K = {
    ...j,
    ...e
  };
  return /* @__PURE__ */ y(
    q,
    {
      className: C("pittorica-box", L),
      style: K,
      disabled: R,
      required: W,
      ...k,
      children: a
    }
  );
};
B.displayName = "Box";
const T = ({
  children: i,
  as: a,
  size: t,
  weight: r = "regular",
  align: o,
  truncate: l = !1,
  color: s,
  className: d,
  style: n,
  ...p
}) => {
  const g = a || "span", m = (s == null ? void 0 : s.startsWith("#")) || (s == null ? void 0 : s.startsWith("rgb")) || (s == null ? void 0 : s.startsWith("hsl")), f = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), u = (() => {
    if (s)
      return s === "inherit" ? "inherit" : m ? s : f.has(s) ? `var(--pittorica-${s}-9)` : s;
  })(), v = t ? typeof t == "string" ? `pittorica-text--size-${t}` : N({
    [`pittorica-text--size-${t.initial}`]: t.initial,
    [`pittorica-text--xs-size-${t.xs}`]: t.xs,
    [`pittorica-text--sm-size-${t.sm}`]: t.sm,
    [`pittorica-text--md-size-${t.md}`]: t.md,
    [`pittorica-text--lg-size-${t.lg}`]: t.lg,
    [`pittorica-text--xl-size-${t.xl}`]: t.xl
  }) : null, c = {
    ...n,
    textAlign: o,
    color: u
  };
  return /* @__PURE__ */ y(
    B,
    {
      as: g,
      className: N(
        "pittorica-text",
        v,
        { "pittorica-text--truncate": l },
        d
      ),
      "data-weight": r,
      style: c,
      ...p,
      children: i
    }
  );
};
T.displayName = "Text";
const D = ({ className: i, style: a, ...t }) => /* @__PURE__ */ y(
  T,
  {
    as: "kbd",
    className: S("pittorica-kbd", i),
    style: a,
    ...t
  }
);
D.displayName = "Kbd";
export {
  D as Kbd
};
