// @pittorica/flex-react@0.26.3/dist/index.js
import { jsx as j } from "react/jsx-runtime";
function R(i) {
  var a, r, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var p = i.length;
    for (a = 0; a < p; a++) i[a] && (r = R(i[a])) && (t && (t += " "), t += r);
  } else for (r in i) i[r] && (t && (t += " "), t += r);
  return t;
}
function k() {
  for (var i, a, r = 0, t = "", p = arguments.length; r < p; r++) (i = arguments[r]) && (a = R(i)) && (t && (t += " "), t += a);
  return t;
}
function A(i) {
  var a, r, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var p = i.length;
    for (a = 0; a < p; a++) i[a] && (r = A(i[a])) && (t && (t += " "), t += r);
  } else for (r in i) i[r] && (t && (t += " "), t += r);
  return t;
}
function O() {
  for (var i, a, r = 0, t = "", p = arguments.length; r < p; r++) (i = arguments[r]) && (a = A(i)) && (t && (t += " "), t += a);
  return t;
}
const B = ({
  as: i,
  children: a,
  display: r,
  m: t,
  mt: p,
  mr: g,
  mb: o,
  ml: d,
  mx: s,
  my: n,
  p: l,
  pt: m,
  pr: $,
  pb: c,
  pl: h,
  px: y,
  py: u,
  width: x,
  height: b,
  position: T,
  style: F,
  className: L,
  disabled: N,
  required: q,
  ...w
}) => {
  const C = i || "div", e = {};
  r && (e.display = r), x && (e.width = x), b && (e.height = b), T && (e.position = T), t && (e.margin = `var(--pittorica-space-${t})`), p && (e.marginTop = `var(--pittorica-space-${p})`), g && (e.marginRight = `var(--pittorica-space-${g})`), o && (e.marginBottom = `var(--pittorica-space-${o})`), d && (e.marginLeft = `var(--pittorica-space-${d})`), s && (e.marginLeft = `var(--pittorica-space-${s})`, e.marginRight = `var(--pittorica-space-${s})`), n && (e.marginTop = `var(--pittorica-space-${n})`, e.marginBottom = `var(--pittorica-space-${n})`), l && (e.padding = `var(--pittorica-space-${l})`), m && (e.paddingTop = `var(--pittorica-space-${m})`), $ && (e.paddingRight = `var(--pittorica-space-${$})`), c && (e.paddingBottom = `var(--pittorica-space-${c})`), h && (e.paddingLeft = `var(--pittorica-space-${h})`), y && (e.paddingLeft = `var(--pittorica-space-${y})`, e.paddingRight = `var(--pittorica-space-${y})`), u && (e.paddingTop = `var(--pittorica-space-${u})`, e.paddingBottom = `var(--pittorica-space-${u})`);
  const S = {
    ...F,
    ...e
  };
  return /* @__PURE__ */ j(
    C,
    {
      className: O("pittorica-box", L),
      style: S,
      disabled: N,
      required: q,
      ...w,
      children: a
    }
  );
};
B.displayName = "Box";
const v = (i) => typeof i == "string" && i.startsWith("auto-");
function f(i, a) {
  return !a || v(a) ? [] : typeof a == "string" ? [`pittorica-flex--${i}-${a}`] : Object.entries(a).filter(([, r]) => r !== void 0 && !v(r)).map(
    ([r, t]) => r === "initial" ? `pittorica-flex--${i}-${t}` : `pittorica-flex--${r}-${i}-${t}`
  );
}
const W = ({
  children: i,
  direction: a,
  align: r,
  justify: t,
  wrap: p,
  gap: g,
  basis: o,
  className: d,
  style: s,
  as: n,
  ...l
}) => {
  const m = n || "div", $ = [
    ...f("direction", a),
    ...f("align", r),
    ...f("justify", t),
    ...f("wrap", p),
    ...f("gap", g)
  ], c = { ...s };
  return v(o) && (c["--pittorica-flex-basis"] = o.split("-")[1]), /* @__PURE__ */ j(
    B,
    {
      as: m,
      className: k("pittorica-flex", $, d),
      style: c,
      ...l,
      children: i
    }
  );
};
W.displayName = "Flex";
export {
  W as Flex
};
