// eslint-plugin-package-json@0.89.1/lib/rules/no-empty-fields.mjs
import { createRule } from "../createRule.mjs";
import { fixRemoveArrayElement, fixRemoveObjectProperty } from "eslint-fix-utils";

//#region src/rules/no-empty-fields.ts
const getDataAndMessageId = (node) => {
	switch (node.type) {
		case "JSONArrayExpression": return {
			data: { expressionType: "array" },
			messageId: "emptyExpression"
		};
		case "JSONObjectExpression": return {
			data: { expressionType: "object" },
			messageId: "emptyExpression"
		};
		case "JSONProperty": return {
			data: { field: node.key.value },
			messageId: "emptyFields"
		};
	}
};
const report = (context, node) => {
	const { data, messageId } = getDataAndMessageId(node);
	context.report({
		data,
		messageId,
		node,
		suggest: [{
			fix: node.type === "JSONProperty" ? fixRemoveObjectProperty(context, node) : fixRemoveArrayElement(context, node, node.parent),
			messageId: "remove"
		}]
	});
};
const getNode = (node) => {
	return node.parent.type === "JSONProperty" ? node.parent : node;
};
const getTopLevelProperty = (node) => {
	let n = node;
	while (n.parent.parent?.parent?.type !== void 0 && n.parent.parent.parent.type !== "Program") n = n.parent;
	return n.type === "JSONProperty" ? n.key : void 0;
};
const rule = createRule({
	create(context) {
		const ignoreProperties = context.options[0]?.ignoreProperties ?? [];
		return {
			JSONArrayExpression(node) {
				const topLevelProperty = getTopLevelProperty(node);
				if (!topLevelProperty) return;
				if (!node.elements.length) {
					const topLevelPropertyName = topLevelProperty.value;
					if (!ignoreProperties.includes(topLevelPropertyName)) report(context, getNode(node));
				}
			},
			JSONObjectExpression(node) {
				const topLevelProperty = getTopLevelProperty(node);
				if (!topLevelProperty) return;
				if (!node.properties.length) {
					const topLevelPropertyName = topLevelProperty.value;
					if (!ignoreProperties.includes(topLevelPropertyName)) report(context, getNode(node));
				}
			}
		};
	},
	meta: {
		defaultOptions: [{ ignoreProperties: [] }],
		docs: {
			category: "Best Practices",
			description: "Reports on unnecessary empty arrays and objects.",
			recommended: true
		},
		hasSuggestions: true,
		messages: {
			emptyExpression: "This {{ expressionType }} does nothing and can be removed.",
			emptyFields: "The field '{{ field }}' does nothing and can be removed.",
			remove: "Remove this empty field."
		},
		schema: [{
			additionalProperties: false,
			properties: { ignoreProperties: {
				description: "Array of top-level properties to ignore.",
				items: { type: "string" },
				type: "array"
			} },
			type: "object"
		}],
		type: "suggestion"
	},
	name: "no-empty-fields"
});

//#endregion
export { rule };
