// eslint-plugin-package-json@0.89.1/lib/rules/require-attribution.mjs
import { createRule } from "../createRule.mjs";
import { fixRemoveObjectProperty } from "eslint-fix-utils";

//#region src/rules/require-attribution.ts
const rule = createRule({
	create(context) {
		const preferContributorsOnly = context.options[0]?.preferContributorsOnly ?? false;
		const ignorePrivate = context.options[0]?.ignorePrivate ?? true;
		let authorPropertyNode;
		let contributorsPropertyNode;
		let isPrivatePackage = false;
		return {
			"Program > JSONExpressionStatement > JSONObjectExpression > JSONProperty[key.value=author]"(node) {
				authorPropertyNode = node;
			},
			"Program > JSONExpressionStatement > JSONObjectExpression > JSONProperty[key.value=contributors]"(node) {
				contributorsPropertyNode = node;
				const contributorsValue = node.value;
				if (contributorsValue.type !== "JSONArrayExpression" || !contributorsValue.elements.some((element) => !!element)) context.report({
					messageId: "noContributors",
					node
				});
			},
			"Program > JSONExpressionStatement > JSONObjectExpression > JSONProperty[key.value=private]"(node) {
				if (node.value.type === "JSONLiteral" && node.value.value === true) isPrivatePackage = true;
			},
			"Program:exit"(node) {
				if (ignorePrivate && isPrivatePackage) return;
				if (preferContributorsOnly) {
					if (authorPropertyNode) context.report({
						messageId: "contributorsOnly",
						node: authorPropertyNode,
						suggest: [{
							fix: fixRemoveObjectProperty(context, authorPropertyNode),
							messageId: "removeAuthor"
						}]
					});
					if (!contributorsPropertyNode) context.report({
						messageId: "missingContributor",
						node
					});
				} else if (!authorPropertyNode && !contributorsPropertyNode) context.report({
					messageId: "missing",
					node
				});
			}
		};
	},
	meta: {
		defaultOptions: [{
			ignorePrivate: true,
			preferContributorsOnly: false
		}],
		docs: {
			category: "Best Practices",
			description: "Ensures that proper attribution is included, requiring that either `author` or `contributors` is defined, and that if `contributors` is present, it should include at least one contributor.",
			recommended: true
		},
		hasSuggestions: true,
		messages: {
			contributorsOnly: "Only `contributors` should be defined for attribution.",
			missing: "Property attribution is required. Either `author` or `contributors` should be defined.",
			missingContributor: "Property attribution is required. `contributors` should be defined.",
			noContributors: "At least one contributor should be defined.",
			removeAuthor: "Remove `author`."
		},
		schema: [{
			additionalProperties: false,
			properties: {
				ignorePrivate: {
					description: "Skip attribution requirements for packages with `\"private\": true`.",
					type: "boolean"
				},
				preferContributorsOnly: {
					description: "Require that only `contributors` is present, and `author` is not defined.",
					type: "boolean"
				}
			},
			type: "object"
		}],
		type: "suggestion"
	},
	name: "require-attribution"
});

//#endregion
export { rule };
