// eslint-plugin-package-json@0.89.1/lib/rules/restrict-private-properties.mjs
import { createRule } from "../createRule.mjs";
import { isJSONStringLiteral } from "../utils/predicates.mjs";
import { fixRemoveObjectProperty } from "eslint-fix-utils";

//#region src/rules/restrict-private-properties.ts
const defaultBlockedProperties = ["files", "publishConfig"];
const rule = createRule({
	create(context) {
		const blockedProperties = context.options[0]?.blockedProperties ?? defaultBlockedProperties;
		return { "Program > JSONExpressionStatement > JSONObjectExpression"(node) {
			if (!node.properties.some((property) => isJSONStringLiteral(property.key) && property.key.value === "private" && property.value.type === "JSONLiteral" && property.value.value === true)) return;
			for (const property of node.properties) if (isJSONStringLiteral(property.key) && blockedProperties.includes(property.key.value)) {
				const isEmpty = property.value.type === "JSONArrayExpression" && property.value.elements.length === 0 || property.value.type === "JSONObjectExpression" && property.value.properties.length === 0;
				context.report({
					data: { property: property.key.value },
					messageId: "unnecessaryProperty",
					node: property,
					...isEmpty ? { fix: fixRemoveObjectProperty(context, property) } : { suggest: [{
						data: { property: property.key.value },
						fix: fixRemoveObjectProperty(context, property),
						messageId: "removePropertySuggestion"
					}] }
				});
			}
		} };
	},
	meta: {
		defaultOptions: [{ blockedProperties: defaultBlockedProperties }],
		docs: {
			category: "Best Practices",
			description: "Disallows unnecessary properties in private packages.",
			recommended: false
		},
		fixable: "code",
		hasSuggestions: true,
		messages: {
			removePropertySuggestion: "Remove the '{{ property }}' field.",
			unnecessaryProperty: "The '{{ property }}' field is unnecessary in private packages and can be removed."
		},
		schema: [{
			additionalProperties: false,
			properties: { blockedProperties: {
				description: "Array of property names to disallow in private packages.",
				items: { type: "string" },
				type: "array"
			} },
			type: "object"
		}],
		type: "suggestion"
	},
	name: "restrict-private-properties"
});

//#endregion
export { rule };
