// eslint-plugin-package-json@0.89.1/lib/rules/scripts-name-casing.mjs
import { createRule } from "../createRule.mjs";
import { kebabCase } from "change-case";

//#region src/rules/scripts-name-casing.ts
const BUILT_IN_SCRIPTS_IN_CAMEL_CASE = new Set(["prepublishOnly"]);
const rule = createRule({
	create(context) {
		return { "Program > JSONExpressionStatement > JSONObjectExpression > JSONProperty[key.value=scripts]"(node) {
			if (node.value.type === "JSONObjectExpression") for (const property of node.value.properties) {
				const keyNode = property.key;
				const key = keyNode.value;
				if (BUILT_IN_SCRIPTS_IN_CAMEL_CASE.has(key)) continue;
				const kebabCaseKey = key.split(":").map((segment) => kebabCase(segment)).join(":");
				if (kebabCaseKey !== key) context.report({
					data: { property: key },
					messageId: "invalidCase",
					node: keyNode,
					suggest: [{
						data: { property: key },
						fix: (fixer) => {
							return fixer.replaceText(keyNode, JSON.stringify(kebabCaseKey));
						},
						messageId: "convertToKebabCase"
					}]
				});
			}
		} };
	},
	meta: {
		docs: {
			category: "Stylistic",
			description: "Enforce that names for `scripts` are in kebab case (optionally separated by colons)."
		},
		hasSuggestions: true,
		messages: {
			convertToKebabCase: "Convert {{ property }} to kebab case.",
			invalidCase: "Script name {{ property }} should be in kebab case."
		},
		schema: [],
		type: "suggestion"
	},
	name: "scripts-name-casing"
});

//#endregion
export { rule };
