// eslint-plugin-package-json@0.89.1/lib/rules/specify-peers-locally.mjs
import { createRule } from "../createRule.mjs";
import { isJSONStringLiteral } from "../utils/predicates.mjs";

//#region src/rules/specify-peers-locally.ts
const rule = createRule({
	create(context) {
		const devDependencyNames = /* @__PURE__ */ new Set();
		let devDependenciesObjectNode;
		const peerDependencyMap = {};
		return {
			"Program > JSONExpressionStatement > JSONObjectExpression > JSONProperty[key.type=JSONLiteral][value.type=JSONObjectExpression][key.value=devDependencies]"(node) {
				devDependenciesObjectNode = node.value;
				for (const devDependencyPropertyNode of node.value.properties) {
					const dependencyKey = devDependencyPropertyNode.key;
					if (isJSONStringLiteral(dependencyKey)) devDependencyNames.add(dependencyKey.value);
				}
			},
			"Program > JSONExpressionStatement > JSONObjectExpression > JSONProperty[key.type=JSONLiteral][value.type=JSONObjectExpression][key.value=peerDependencies]"(node) {
				for (const peerDependencyPropertyNode of node.value.properties) {
					const dependencyKey = peerDependencyPropertyNode.key;
					if (isJSONStringLiteral(dependencyKey)) peerDependencyMap[dependencyKey.value] = peerDependencyPropertyNode;
				}
			},
			"Program:exit"() {
				for (const [peerDependencyName, peerDependencyNode] of Object.entries(peerDependencyMap)) if (!devDependencyNames.has(peerDependencyName)) {
					const peerDependencyValue = peerDependencyNode.value;
					context.report({
						data: { name: peerDependencyName },
						messageId: "devDependencyNotDefined",
						node: peerDependencyNode,
						suggest: devDependenciesObjectNode && isJSONStringLiteral(peerDependencyValue) ? [{
							data: { name: peerDependencyName },
							fix: (fixer) => {
								const updatedDevDependencies = {
									...JSON.parse(context.sourceCode.getText(devDependenciesObjectNode)),
									[peerDependencyName]: peerDependencyValue.value
								};
								const sortedDevDependencies = Object.fromEntries(Object.entries(updatedDevDependencies).sort((a, b) => a[0] > b[0] ? 1 : -1));
								return fixer.replaceText(devDependenciesObjectNode, JSON.stringify(sortedDevDependencies, null, 2).split("\n").join("\n  "));
							},
							messageId: "addToDevDependencies"
						}] : []
					});
				}
			}
		};
	},
	meta: {
		docs: {
			description: "Requires that all peer dependencies are also declared as dev dependencies",
			recommended: true
		},
		hasSuggestions: true,
		messages: {
			addToDevDependencies: "Add \"{{ name }}\" to `devDependencies`",
			devDependencyNotDefined: "Peer dependency \"{{ name }}\" is not also declared in `devDependencies`."
		},
		schema: [],
		type: "problem"
	},
	name: "specify-peers-locally"
});

//#endregion
export { rule };
