// eslint-plugin-package-json@0.89.1/lib/rules/valid-package-definition.mjs
import { createRule } from "../createRule.mjs";
import { validate } from "package-json-validator";

//#region src/rules/valid-package-definition.ts
const ignoredErrors = [
	/^Url not valid/i,
	/^Invalid version range for .+?: file:/i,
	/^author field should have name/i
];
const isUsableError = (errorText) => ignoredErrors.every((pattern) => !pattern.test(errorText));
const rule = createRule({
	create(context) {
		const ignoreProperties = context.options[0]?.ignoreProperties ?? [];
		return { "Program:exit"() {
			const usableErrors = validate(context.sourceCode.text).errors?.filter((error) => {
				return isUsableError(error.message) && !ignoreProperties.includes(error.field);
			}) ?? [];
			for (const error of usableErrors) if (error.message) context.report({
				message: error.message,
				node: context.sourceCode.ast
			});
		} };
	},
	meta: {
		defaultOptions: [{ ignoreProperties: [] }],
		deprecated: true,
		docs: {
			category: "Best Practices",
			description: "Enforce that package.json has all properties required by the npm spec"
		},
		schema: [{
			additionalProperties: false,
			properties: { ignoreProperties: {
				description: "Array of top-level package properties to ignore.",
				items: { type: "string" },
				type: "array"
			} },
			type: "object"
		}],
		type: "problem"
	},
	name: "valid-package-definition"
});

//#endregion
export { rule };
