// @pittorica/grid-react@0.26.3/dist/index.js
import { jsx as R } from "react/jsx-runtime";
function j(i) {
  var a, r, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var e = i.length;
    for (a = 0; a < e; a++) i[a] && (r = j(i[a])) && (t && (t += " "), t += r);
  } else for (r in i) i[r] && (t && (t += " "), t += r);
  return t;
}
function k() {
  for (var i, a, r = 0, t = "", e = arguments.length; r < e; r++) (i = arguments[r]) && (a = j(i)) && (t && (t += " "), t += a);
  return t;
}
function A(i) {
  var a, r, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var e = i.length;
    for (a = 0; a < e; a++) i[a] && (r = A(i[a])) && (t && (t += " "), t += r);
  } else for (r in i) i[r] && (t && (t += " "), t += r);
  return t;
}
function F() {
  for (var i, a, r = 0, t = "", e = arguments.length; r < e; r++) (i = arguments[r]) && (a = A(i)) && (t && (t += " "), t += a);
  return t;
}
const B = ({
  as: i,
  children: a,
  display: r,
  m: t,
  mt: e,
  mr: g,
  mb: d,
  ml: f,
  mx: n,
  my: c,
  p: m,
  pt: l,
  pr: $,
  pb: u,
  pl: y,
  px: s,
  py: h,
  width: b,
  height: x,
  position: T,
  style: L,
  className: N,
  disabled: q,
  required: w,
  ...C
}) => {
  const G = i || "div", p = {};
  r && (p.display = r), b && (p.width = b), x && (p.height = x), T && (p.position = T), t && (p.margin = `var(--pittorica-space-${t})`), e && (p.marginTop = `var(--pittorica-space-${e})`), g && (p.marginRight = `var(--pittorica-space-${g})`), d && (p.marginBottom = `var(--pittorica-space-${d})`), f && (p.marginLeft = `var(--pittorica-space-${f})`), n && (p.marginLeft = `var(--pittorica-space-${n})`, p.marginRight = `var(--pittorica-space-${n})`), c && (p.marginTop = `var(--pittorica-space-${c})`, p.marginBottom = `var(--pittorica-space-${c})`), m && (p.padding = `var(--pittorica-space-${m})`), l && (p.paddingTop = `var(--pittorica-space-${l})`), $ && (p.paddingRight = `var(--pittorica-space-${$})`), u && (p.paddingBottom = `var(--pittorica-space-${u})`), y && (p.paddingLeft = `var(--pittorica-space-${y})`), s && (p.paddingLeft = `var(--pittorica-space-${s})`, p.paddingRight = `var(--pittorica-space-${s})`), h && (p.paddingTop = `var(--pittorica-space-${h})`, p.paddingBottom = `var(--pittorica-space-${h})`);
  const S = {
    ...L,
    ...p
  };
  return /* @__PURE__ */ R(
    G,
    {
      className: F("pittorica-box", N),
      style: S,
      disabled: q,
      required: w,
      ...C,
      children: a
    }
  );
};
B.displayName = "Box";
const v = (i) => typeof i == "string" && i.startsWith("auto-");
function o(i, a) {
  return !a || v(a) ? [] : typeof a == "string" ? [`pittorica-grid--${i}-${a}`] : Object.entries(a).filter(([, r]) => r !== void 0 && !v(r)).map(
    ([r, t]) => r === "initial" ? `pittorica-grid--${i}-${t}` : `pittorica-grid--${r}-${i}-${t}`
  );
}
const O = ({
  children: i,
  columns: a,
  rows: r,
  gap: t,
  gapX: e,
  gapY: g,
  flow: d,
  align: f,
  justify: n,
  className: c,
  style: m,
  as: l,
  ...$
}) => {
  const u = l || "div", y = [
    ...o("columns", a),
    ...o("rows", r),
    ...o("gap", t),
    ...o("gapX", e),
    ...o("gapY", g),
    ...o("flow", d),
    ...o("align", f),
    ...o("justify", n)
  ], s = {};
  return v(a) && (s.gridTemplateColumns = `repeat(auto-fit, minmax(${a.split("-")[1]}, 1fr))`), v(r) && (s.gridTemplateRows = `repeat(auto-fit, minmax(${r.split("-")[1]}, 1fr))`), /* @__PURE__ */ R(
    B,
    {
      as: u,
      className: k("pittorica-grid", y, c),
      style: { ...m, ...s },
      ...$,
      children: i
    }
  );
};
O.displayName = "Grid";
export {
  O as Grid
};
