// eslint-plugin-spellbookx@0.10.0/dist/index.js

import { fileURLToPath as _fileURLToPath } from 'node:url';
import { dirname } from 'node:path';
import { createRequire as _createRequire } from 'node:module';

const __filename = _fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const require = _createRequire(import.meta.url);

import { defineConfig, globalIgnores } from "eslint/config";
import prettierRecommended from "eslint-plugin-prettier/recommended";
import cspell from "@cspell/eslint-plugin/configs";
import importPlugin from "eslint-plugin-import";
import jsdoc from "eslint-plugin-jsdoc";
import nodeDependencies from "eslint-plugin-node-dependencies";
import prettierPlugin from "eslint-plugin-prettier";
import simpleImportSort from "eslint-plugin-simple-import-sort";
import unicorn from "eslint-plugin-unicorn";
import globals from "globals";
import tseslint from "typescript-eslint";
import js from "@eslint/js";
import eslintReact from "@eslint-react/eslint-plugin";
import packageJson from "eslint-plugin-package-json";
import * as jsoncParser from "jsonc-eslint-parser";
import json from "@eslint/json";
import markdown from "@eslint/markdown";
import vuoto from "eslint-plugin-vuoto";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

//#region src/configs/cspell.ts
const configCspell = defineConfig([cspell.recommended, prettierRecommended]);

//#endregion
//#region src/configs/ignores.ts
const configIgnores = defineConfig([globalIgnores([
	"**/node_modules/**",
	"**/dist/**",
	"**/build/**",
	"**/out/**",
	"**/out-tsc/**",
	"**/coverage/**",
	"**/.astro/**",
	"**/.next/**",
	"**/.nuxt/**",
	"**/.svelte-kit/**",
	"**/.turbo/**",
	"**/.pnpm/**",
	"**/.npm/**",
	"**/.yarn/**",
	"**/.pnp/**",
	"**/package-lock.json",
	"**/pnpm-lock.yaml",
	"**/yarn.lock",
	"**/bun.lockb",
	"**/__pycache__/**",
	"**/*.pyc",
	"**/*.pyo",
	"**/.mypy_cache/**",
	"**/.pytest_cache/**",
	"**/.tox/**",
	"**/.ruff_cache/**",
	"**/.venv/**",
	"**/venv/**",
	"**/env/**",
	"**/.ipynb_checkpoints/**",
	"**/target/**",
	"**/.cargo/**",
	"**/Cargo.lock",
	"**/go.sum",
	"**/go.work",
	"**/go.work.sum",
	"**/vendor/**",
	"**/.gopath/**",
	"**/.cache/go-build/**",
	"**/*LICENSE*",
	"**/*.log",
	"**/.cache/**",
	"**/.temp/**",
	"**/.tmp/**",
	"**/.DS_Store",
	"**/.idea/**",
	"**/.vscode/**",
	"**/.nx/**",
	"**/.cspell/**",
	"**/.cursor/**",
	"**/.history/**",
	"**/.terraform/**",
	"**/.devcontainer/**",
	"**/.direnv/**",
	"**/.editorconfig",
	"**/.eslintcache",
	"**/.babelrc",
	"**/.prettier*",
	"**/.sass-cache/**",
	"**/.gradle/**",
	"**/.docker/**",
	"**/.kube/**",
	"**/.git/**",
	"**/.svn/**",
	"**/.hg/**",
	"**/.gemini/**",
	"**/.codacy/**",
	".github/instructions/**",
	"**/*:Zone.Identifier",
	"**/Thumbs.db",
	"**/desktop.ini",
	"**/$RECYCLE.BIN/**",
	"**/System Volume Information/**",
	"**/pagefile.sys",
	"**/swapfile.sys",
	"**/hiberfil.sys",
	"**/*.hbs"
])]);

//#endregion
//#region src/configs/javascript.ts
const configJavascript = defineConfig([
	{
		files: ["**/*.{js,jsx,mjs,cjs,ts,tsx,mts,cts}"],
		plugins: {
			jsdoc,
			import: importPlugin,
			prettier: prettierPlugin,
			"simple-import-sort": simpleImportSort
		},
		extends: [
			js.configs.recommended,
			...tseslint.configs.recommended,
			unicorn.configs.recommended
		],
		rules: {
			"prettier/prettier": "error",
			"no-unused-vars": "off",
			"@typescript-eslint/no-unused-vars": ["error", {
				argsIgnorePattern: "^_",
				varsIgnorePattern: "^_"
			}],
			"sort-imports": "off",
			"simple-import-sort/exports": "error",
			"simple-import-sort/imports": ["error", { groups: [
				["^@dotenvx/dotenvx", "^dotenv"],
				[String.raw`^\u0000`],
				["^@env"],
				[
					"^assert",
					"^buffer",
					"^child_process",
					"^cluster",
					"^console",
					"^constants",
					"^crypto",
					"^dgram",
					"^dns",
					"^domain",
					"^events",
					"^fs",
					"^http",
					"^https",
					"^inspector",
					"^module",
					"^net",
					"^os",
					"^path",
					"^perf_hooks",
					"^process",
					"^punycode",
					"^querystring",
					"^readline",
					"^repl",
					"^stream",
					"^string_decoder",
					"^timers",
					"^tls",
					"^tty",
					"^url",
					"^util",
					"^v8",
					"^vm",
					"^zlib"
				],
				["^node:"],
				[
					"^@fastify",
					"^@hapi/hapi",
					"^@koa/koa",
					"^@nestjs",
					"^express",
					"^fastify",
					"^hapi",
					"^koa",
					"^loopback",
					"^nest",
					"^sails"
				],
				[
					"^body-parser",
					"^connect-redis",
					"^cookie-parser",
					"^cors",
					"^express-rate-limit",
					"^express-session",
					"^helmet",
					"^morgan",
					"^passport",
					String.raw`^pino-`,
					"^redis",
					"^winston"
				],
				[
					"^@angular",
					"^react",
					"^solid-js",
					"^svelte",
					"^vue"
				],
				[
					"^@tanstack/router",
					"^react-dom",
					"^react-helmet",
					"^react-intl",
					"^react-router",
					"^react-router-dom"
				],
				[
					"^@nuxt/kit",
					"^@remix-run",
					"^@sveltejs/kit",
					"^gatsby",
					"^next"
				],
				[
					"^@expo",
					String.raw`^@expo\/`,
					"^@react-native",
					"^expo",
					String.raw`^expo-`,
					"^react-native",
					"^react-navigation"
				],
				[
					"^@reduxjs/toolkit",
					"^jotai",
					"^mobx",
					"^recoil",
					"^redux",
					"^valtio",
					"^zustand"
				],
				[
					"^@apollo/client",
					"^@tanstack/react-query",
					"^axios",
					"^graphql",
					"^swr"
				],
				[
					"^@chakra-ui",
					"^@headlessui/react",
					"^@lottiefiles",
					"^@material-ui",
					"^@mui",
					"^@nextui-org/react",
					"^@radix-ui",
					"^@pittorica/pitto",
					"^radix-ui",
					"^antd",
					"^framer-motion",
					"^native-base",
					"^react-native-paper",
					"^shadcn-ui",
					"^tailwindcss"
				],
				[
					"^@emotion",
					"^class-variance-authority",
					"^clsx",
					"^lucide-react",
					"^styled-components",
					"^tailwind-merge",
					"^twin.macro",
					"^tw-animate-css"
				],
				[
					String.raw`^@expo\/vector-icons`,
					"^@fortawesome",
					"^@tabler/icons-react",
					"^lucide",
					"^react-feather",
					"^react-icons",
					"^react-native-feather",
					"^react-native-vector-icons"
				],
				[
					"^@testing-library",
					"^cypress",
					"^jest",
					"^playwright",
					"^vitest"
				],
				["^[a-z]", String.raw`^@\w`],
				[
					"@org/",
					"^@my-org/",
					"^@workspace/",
					"^@repo/"
				],
				[
					String.raw`^.+\.(avi|mkv|mov|mp4|webm)$`,
					String.raw`^.+\.(mp3|ogg|wav|weba)$`,
					String.raw`^.+\.(gif|jpe?g|png|svg|webp)$`,
					String.raw`^.+\.lottie$`,
					String.raw`^.+\.(eot|otf|ttf|woff|woff2)$`
				],
				[String.raw`^.+\.json$`],
				[String.raw`^.+\.less$`, String.raw`^.+\.s?css$`],
				[String.raw`^\.?\.\/`]
			] }],
			"unicorn/no-null": "off",
			"unicorn/prevent-abbreviations": "off"
		}
	},
	{
		files: ["**/*.{ts,js,mts,mjs,cts,cjs}"],
		ignores: ["**/*.{tsx,jsx}"],
		extends: [
			nodeDependencies.configs["flat/recommended"],
			jsdoc.configs["flat/contents-typescript-flavor"],
			jsdoc.configs["flat/logical-typescript-flavor"]
		],
		languageOptions: { globals: { ...globals.node } },
		rules: {
			"unicorn/filename-case": ["error", { case: "kebabCase" }],
			"jsdoc/require-jsdoc": ["warn", { require: {
				FunctionDeclaration: true,
				MethodDefinition: true,
				ClassDeclaration: true
			} }],
			"jsdoc/require-description": "warn"
		}
	},
	{
		files: ["**/*.{tsx,jsx}"],
		extends: [eslintReact.configs["recommended-typescript"]],
		languageOptions: {
			sourceType: "module",
			globals: {
				...globals.browser,
				...globals.es2022
			},
			parser: tseslint.parser,
			parserOptions: {
				projectService: true,
				tsconfigRootDir: import.meta.dirname
			}
		},
		rules: {
			"unicorn/filename-case": "off",
			"jsdoc/require-jsdoc": "off",
			"jsdoc/require-description": "off",
			"jsdoc/require-param": "off",
			"jsdoc/require-returns": "off",
			"jsdoc/check-tag-names": "off"
		}
	},
	{
		files: ["**/*.{cjs,cts}"],
		languageOptions: { globals: { ...globals.commonjs } },
		rules: {
			"@typescript-eslint/no-require-imports": "off",
			"unicorn/prefer-module": "off"
		}
	},
	prettierRecommended
]);

//#endregion
//#region src/configs/json.ts
const configJson = defineConfig([
	{
		files: ["**/*.json"],
		plugins: { json },
		language: "json/json",
		extends: ["json/recommended"],
		languageOptions: { parser: jsoncParser },
		rules: { "prettier/prettier": "off" }
	},
	{
		files: ["**/*.json5"],
		plugins: { json },
		language: "json/json5",
		extends: ["json/recommended"],
		languageOptions: { parser: jsoncParser }
	},
	{
		files: [
			"**/*.jsonc",
			"**/tsconfig*.json",
			"**/.vscode/**/*.json"
		],
		plugins: { json },
		language: "json/jsonc",
		extends: ["json/recommended"],
		languageOptions: { parser: jsoncParser }
	},
	{
		files: ["**/package.json"],
		plugins: {
			json,
			"package-json": packageJson
		},
		language: "json/json",
		extends: [packageJson.configs.recommended],
		languageOptions: { parser: jsoncParser },
		rules: {
			"package-json/order-properties": "error",
			"package-json/sort-collections": "error",
			"package-json/require-description": "error",
			"package-json/require-bugs": "error",
			"package-json/require-keywords": "error",
			"package-json/require-name": "error",
			"package-json/require-version": "error",
			"package-json/valid-description": "error",
			"package-json/valid-license": "error",
			"package-json/valid-name": "error",
			"package-json/valid-package-definition": "error",
			"package-json/valid-version": "error",
			"prettier/prettier": "off"
		}
	},
	prettierRecommended
]);

//#endregion
//#region src/configs/markdown.ts
const configMarkdown = defineConfig([{
	files: ["**/*.md"],
	plugins: { markdown },
	language: "markdown/gfm",
	extends: ["markdown/recommended"]
}, prettierRecommended]);

//#endregion
//#region src/configs/vuoto.ts
const configVuoto = defineConfig([{ plugins: { vuoto } }, vuoto.configs.all]);

//#endregion
//#region src/configs/recommended.ts
const configRecommended = defineConfig([
	configIgnores,
	configJavascript,
	configJson,
	configMarkdown,
	configVuoto,
	configCspell,
	prettierRecommended
]);

//#endregion
//#region src/helpers.ts
const packagePath = fileURLToPath(new URL("../package.json", import.meta.url));
const packageJson$1 = JSON.parse(readFileSync(packagePath, "utf8"));

//#endregion
//#region src/index.ts
const configs = {
	cspell: configCspell,
	ignores: configIgnores,
	javascript: configJavascript,
	json: configJson,
	markdown: configMarkdown,
	recommended: configRecommended,
	vuoto: configVuoto
};
const plugin = {
	meta: {
		name: packageJson$1.name,
		version: packageJson$1.version,
		namespace: "spellbookx"
	},
	configs,
	rules: {},
	processors: {}
};

//#endregion
export { plugin as default };
