// @pittorica/checkbox-card-react@0.26.3/dist/index.js
import { jsx as k, jsxs as _ } from "react/jsx-runtime";
import { createContext as D, useState as F, useMemo as G, use as H } from "react";
function w(t) {
  var e, i, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (e = 0; e < o; e++) t[e] && (i = w(t[e])) && (a && (a += " "), a += i);
  } else for (i in t) t[i] && (a && (a += " "), a += i);
  return a;
}
function I() {
  for (var t, e, i = 0, a = "", o = arguments.length; i < o; i++) (t = arguments[i]) && (e = w(t)) && (a && (a += " "), a += e);
  return a;
}
function L(t) {
  var e, i, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (e = 0; e < o; e++) t[e] && (i = L(t[e])) && (a && (a += " "), a += i);
  } else for (i in t) t[i] && (a && (a += " "), a += i);
  return a;
}
function J() {
  for (var t, e, i = 0, a = "", o = arguments.length; i < o; i++) (t = arguments[i]) && (e = L(t)) && (a && (a += " "), a += e);
  return a;
}
const j = ({
  as: t,
  children: e,
  display: i,
  m: a,
  mt: o,
  mr: n,
  mb: p,
  ml: c,
  mx: d,
  my: l,
  p: f,
  pt: h,
  pr: u,
  pb: m,
  pl: g,
  px: v,
  py: s,
  width: y,
  height: b,
  position: x,
  style: $,
  className: C,
  disabled: N,
  required: T,
  ...R
}) => {
  const A = t || "div", r = {};
  i && (r.display = i), y && (r.width = y), b && (r.height = b), x && (r.position = x), a && (r.margin = `var(--pittorica-space-${a})`), o && (r.marginTop = `var(--pittorica-space-${o})`), n && (r.marginRight = `var(--pittorica-space-${n})`), p && (r.marginBottom = `var(--pittorica-space-${p})`), c && (r.marginLeft = `var(--pittorica-space-${c})`), d && (r.marginLeft = `var(--pittorica-space-${d})`, r.marginRight = `var(--pittorica-space-${d})`), l && (r.marginTop = `var(--pittorica-space-${l})`, r.marginBottom = `var(--pittorica-space-${l})`), f && (r.padding = `var(--pittorica-space-${f})`), h && (r.paddingTop = `var(--pittorica-space-${h})`), u && (r.paddingRight = `var(--pittorica-space-${u})`), m && (r.paddingBottom = `var(--pittorica-space-${m})`), g && (r.paddingLeft = `var(--pittorica-space-${g})`), v && (r.paddingLeft = `var(--pittorica-space-${v})`, r.paddingRight = `var(--pittorica-space-${v})`), s && (r.paddingTop = `var(--pittorica-space-${s})`, r.paddingBottom = `var(--pittorica-space-${s})`);
  const B = {
    ...$,
    ...r
  };
  return /* @__PURE__ */ k(
    A,
    {
      className: J("pittorica-box", C),
      style: B,
      disabled: N,
      required: T,
      ...R,
      children: e
    }
  );
};
j.displayName = "Box";
function q(t) {
  var e, i, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (e = 0; e < o; e++) t[e] && (i = q(t[e])) && (a && (a += " "), a += i);
  } else for (i in t) t[i] && (a && (a += " "), a += i);
  return a;
}
function K() {
  for (var t, e, i = 0, a = "", o = arguments.length; i < o; i++) (t = arguments[i]) && (e = q(t)) && (a && (a += " "), a += e);
  return a;
}
function S(t) {
  var e, i, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (e = 0; e < o; e++) t[e] && (i = S(t[e])) && (a && (a += " "), a += i);
  } else for (i in t) t[i] && (a && (a += " "), a += i);
  return a;
}
function P() {
  for (var t, e, i = 0, a = "", o = arguments.length; i < o; i++) (t = arguments[i]) && (e = S(t)) && (a && (a += " "), a += e);
  return a;
}
const E = ({
  as: t,
  children: e,
  display: i,
  m: a,
  mt: o,
  mr: n,
  mb: p,
  ml: c,
  mx: d,
  my: l,
  p: f,
  pt: h,
  pr: u,
  pb: m,
  pl: g,
  px: v,
  py: s,
  width: y,
  height: b,
  position: x,
  style: $,
  className: C,
  disabled: N,
  required: T,
  ...R
}) => {
  const A = t || "div", r = {};
  i && (r.display = i), y && (r.width = y), b && (r.height = b), x && (r.position = x), a && (r.margin = `var(--pittorica-space-${a})`), o && (r.marginTop = `var(--pittorica-space-${o})`), n && (r.marginRight = `var(--pittorica-space-${n})`), p && (r.marginBottom = `var(--pittorica-space-${p})`), c && (r.marginLeft = `var(--pittorica-space-${c})`), d && (r.marginLeft = `var(--pittorica-space-${d})`, r.marginRight = `var(--pittorica-space-${d})`), l && (r.marginTop = `var(--pittorica-space-${l})`, r.marginBottom = `var(--pittorica-space-${l})`), f && (r.padding = `var(--pittorica-space-${f})`), h && (r.paddingTop = `var(--pittorica-space-${h})`), u && (r.paddingRight = `var(--pittorica-space-${u})`), m && (r.paddingBottom = `var(--pittorica-space-${m})`), g && (r.paddingLeft = `var(--pittorica-space-${g})`), v && (r.paddingLeft = `var(--pittorica-space-${v})`, r.paddingRight = `var(--pittorica-space-${v})`), s && (r.paddingTop = `var(--pittorica-space-${s})`, r.paddingBottom = `var(--pittorica-space-${s})`);
  const B = {
    ...$,
    ...r
  };
  return /* @__PURE__ */ k(
    A,
    {
      className: P("pittorica-box", C),
      style: B,
      disabled: N,
      required: T,
      ...R,
      children: e
    }
  );
};
E.displayName = "Box";
const W = ({
  variant: t = "surface",
  translucent: e = !1,
  children: i,
  className: a,
  style: o,
  as: n,
  ...p
}) => /* @__PURE__ */ k(
  E,
  {
    as: n || "div",
    className: K(
      "pittorica-card",
      `pittorica-card--variant-${t}`,
      { "pittorica-card--translucent": e },
      a
    ),
    style: o,
    ...p,
    children: i
  }
);
W.displayName = "Card";
const z = D(
  null
), M = ({
  value: t,
  defaultValue: e = [],
  onValueChange: i,
  orientation: a = "vertical",
  color: o = "indigo",
  disabled: n,
  translucent: p,
  name: c,
  required: d,
  children: l,
  className: f,
  as: h,
  ...u
}) => {
  const [m, g] = F(e), v = t !== void 0, s = v ? t : m, y = ($) => {
    const C = s.includes($) ? s.filter((N) => N !== $) : [...s, $];
    v || g(C), i == null || i(C);
  }, b = G(
    () => ({
      value: s,
      onItemChange: y,
      color: o,
      disabled: n,
      translucent: p,
      name: c,
      required: d
    }),
    [s, o, n, p, c, d]
  );
  return /* @__PURE__ */ k(z, { value: b, children: /* @__PURE__ */ k(
    j,
    {
      as: h || "div",
      className: I("pittorica-checkbox-card-root", f),
      "data-orientation": a,
      role: "group",
      ...u,
      children: l
    }
  ) });
}, O = ({
  value: t,
  children: e,
  className: i,
  style: a,
  translucent: o,
  as: n,
  ...p
}) => {
  var m, g;
  const c = H(z);
  if (!c)
    throw new Error("CheckboxCard.Item must be used within CheckboxCard.Root");
  const d = o ?? c.translucent, l = c.value.includes(t), h = c.color !== "inherit" && !((m = c.color) != null && m.startsWith("#")) && !((g = c.color) != null && g.startsWith("rgb")) ? `var(--pittorica-${c.color}-9)` : c.color;
  return /* @__PURE__ */ _(
    W,
    {
      as: n || "label",
      translucent: d,
      className: I("pittorica-checkbox-card-item", i),
      "data-state": l ? "checked" : "unchecked",
      "data-disabled": c.disabled,
      style: {
        "--_checkbox-card-color": h,
        ...a
      },
      ...p,
      children: [
        /* @__PURE__ */ k(
          "input",
          {
            type: "checkbox",
            className: "pittorica-checkbox-card-input",
            style: { position: "absolute", opacity: 0, pointerEvents: "none" },
            name: c.name,
            value: t,
            checked: l,
            disabled: c.disabled,
            required: c.required,
            onChange: () => c.onItemChange(t)
          }
        ),
        e
      ]
    }
  );
}, X = Object.assign(M, {
  Item: O
});
M.displayName = "CheckboxCard.Root";
O.displayName = "CheckboxCard.Item";
export {
  X as CheckboxCard
};
