// @pittorica/avatar-react@0.26.3/dist/index.js
import { jsx as b } from "react/jsx-runtime";
import { useState as q, useEffect as z } from "react";
function N(a) {
  var i, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var p = a.length;
    for (i = 0; i < p; i++) a[i] && (r = N(a[i])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function k() {
  for (var a, i, r = 0, t = "", p = arguments.length; r < p; r++) (a = arguments[r]) && (i = N(a)) && (t && (t += " "), t += i);
  return t;
}
function T(a) {
  var i, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var p = a.length;
    for (i = 0; i < p; i++) a[i] && (r = T(a[i])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function W() {
  for (var a, i, r = 0, t = "", p = arguments.length; r < p; r++) (a = arguments[r]) && (i = T(a)) && (t && (t += " "), t += i);
  return t;
}
const A = ({
  as: a,
  children: i,
  display: r,
  m: t,
  mt: p,
  mr: n,
  mb: l,
  ml: h,
  mx: c,
  my: d,
  p: u,
  pt: g,
  pr: $,
  pb: v,
  pl: x,
  px: m,
  py: o,
  width: L,
  height: s,
  position: f,
  style: y,
  className: E,
  disabled: S,
  required: B,
  ...R
}) => {
  const w = a || "div", e = {};
  r && (e.display = r), L && (e.width = L), s && (e.height = s), f && (e.position = f), t && (e.margin = `var(--pittorica-space-${t})`), p && (e.marginTop = `var(--pittorica-space-${p})`), n && (e.marginRight = `var(--pittorica-space-${n})`), l && (e.marginBottom = `var(--pittorica-space-${l})`), h && (e.marginLeft = `var(--pittorica-space-${h})`), c && (e.marginLeft = `var(--pittorica-space-${c})`, e.marginRight = `var(--pittorica-space-${c})`), d && (e.marginTop = `var(--pittorica-space-${d})`, e.marginBottom = `var(--pittorica-space-${d})`), u && (e.padding = `var(--pittorica-space-${u})`), g && (e.paddingTop = `var(--pittorica-space-${g})`), $ && (e.paddingRight = `var(--pittorica-space-${$})`), v && (e.paddingBottom = `var(--pittorica-space-${v})`), x && (e.paddingLeft = `var(--pittorica-space-${x})`), m && (e.paddingLeft = `var(--pittorica-space-${m})`, e.paddingRight = `var(--pittorica-space-${m})`), o && (e.paddingTop = `var(--pittorica-space-${o})`, e.paddingBottom = `var(--pittorica-space-${o})`);
  const j = {
    ...y,
    ...e
  };
  return /* @__PURE__ */ b(
    w,
    {
      className: W("pittorica-box", E),
      style: j,
      disabled: S,
      required: B,
      ...R,
      children: i
    }
  );
};
A.displayName = "Box";
const _ = ({
  src: a,
  alt: i,
  fallback: r,
  size: t = "3",
  radius: p = "full",
  color: n = "indigo",
  className: l,
  style: h,
  as: c,
  ...d
}) => {
  const [u, g] = q(
    a ? "loading" : "idle"
  );
  z(() => {
    if (!a) return;
    const s = new Image(), f = () => g("loaded"), y = () => g("error");
    return s.addEventListener("load", f), s.addEventListener("error", y), s.src = a, () => {
      s.removeEventListener("load", f), s.removeEventListener("error", y);
    };
  }, [a]);
  const $ = {
    1: "24px",
    2: "32px",
    3: "40px",
    4: "48px",
    5: "64px",
    6: "80px",
    7: "96px",
    8: "128px",
    9: "160px"
  }, v = n !== "inherit" && !(n != null && n.startsWith("#")) && !(n != null && n.startsWith("rgb")), x = v ? `var(--pittorica-${n}-3)` : n, m = v ? `var(--pittorica-${n}-11)` : "inherit", o = $[t];
  return /* @__PURE__ */ b(
    A,
    {
      as: c || "div",
      className: k("pittorica-avatar", l),
      style: {
        width: o,
        height: o,
        borderRadius: p === "full" ? "50%" : `var(--pittorica-radius-${p})`,
        fontSize: `calc(${o} / 2.5)`,
        "--_avatar-bg": x,
        "--_avatar-text": m,
        ...h
      },
      ...d,
      children: u === "loaded" && a ? /* @__PURE__ */ b("img", { src: a, alt: i, className: "pittorica-avatar-image" }) : /* @__PURE__ */ b("div", { className: "pittorica-avatar-fallback", "aria-label": i, children: r || (i == null ? void 0 : i.charAt(0)) || "?" })
    }
  );
};
_.displayName = "Avatar";
export {
  _ as Avatar
};
