// @pittorica/slider-react@0.26.3/dist/index.js
import { jsx as w, jsxs as E } from "react/jsx-runtime";
import { useState as D, useRef as U, useCallback as W } from "react";
function B(i) {
  var e, a, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var s = i.length;
    for (e = 0; e < s; e++) i[e] && (a = B(i[e])) && (t && (t += " "), t += a);
  } else for (a in i) i[a] && (t && (t += " "), t += a);
  return t;
}
function I() {
  for (var i, e, a = 0, t = "", s = arguments.length; a < s; a++) (i = arguments[a]) && (e = B(i)) && (t && (t += " "), t += e);
  return t;
}
function j(i) {
  var e, a, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var s = i.length;
    for (e = 0; e < s; e++) i[e] && (a = j(i[e])) && (t && (t += " "), t += a);
  } else for (a in i) i[a] && (t && (t += " "), t += a);
  return t;
}
function V() {
  for (var i, e, a = 0, t = "", s = arguments.length; a < s; a++) (i = arguments[a]) && (e = j(i)) && (t && (t += " "), t += e);
  return t;
}
const k = ({
  as: i,
  children: e,
  display: a,
  m: t,
  mt: s,
  mr: p,
  mb: n,
  ml: c,
  mx: u,
  my: f,
  p: $,
  pt: b,
  pr: x,
  pb: N,
  pl: T,
  px: m,
  py: d,
  width: g,
  height: h,
  position: y,
  style: L,
  className: P,
  disabled: R,
  required: S,
  ...l
}) => {
  const v = i || "div", r = {};
  a && (r.display = a), g && (r.width = g), h && (r.height = h), y && (r.position = y), t && (r.margin = `var(--pittorica-space-${t})`), s && (r.marginTop = `var(--pittorica-space-${s})`), p && (r.marginRight = `var(--pittorica-space-${p})`), n && (r.marginBottom = `var(--pittorica-space-${n})`), c && (r.marginLeft = `var(--pittorica-space-${c})`), u && (r.marginLeft = `var(--pittorica-space-${u})`, r.marginRight = `var(--pittorica-space-${u})`), f && (r.marginTop = `var(--pittorica-space-${f})`, r.marginBottom = `var(--pittorica-space-${f})`), $ && (r.padding = `var(--pittorica-space-${$})`), b && (r.paddingTop = `var(--pittorica-space-${b})`), x && (r.paddingRight = `var(--pittorica-space-${x})`), N && (r.paddingBottom = `var(--pittorica-space-${N})`), T && (r.paddingLeft = `var(--pittorica-space-${T})`), m && (r.paddingLeft = `var(--pittorica-space-${m})`, r.paddingRight = `var(--pittorica-space-${m})`), d && (r.paddingTop = `var(--pittorica-space-${d})`, r.paddingBottom = `var(--pittorica-space-${d})`);
  const o = {
    ...L,
    ...r
  };
  return /* @__PURE__ */ w(
    v,
    {
      className: V("pittorica-box", P),
      style: o,
      disabled: R,
      required: S,
      ...l,
      children: e
    }
  );
};
k.displayName = "Box";
const X = ({
  value: i,
  defaultValue: e = 0,
  min: a = 0,
  max: t = 100,
  step: s = 1,
  disabled: p = !1,
  color: n = "source",
  onValueChange: c,
  required: u = !1,
  className: f,
  style: $,
  as: b,
  ...x
}) => {
  const [N, T] = D(e), m = U(null), d = i !== void 0, g = d ? i : N, h = (g - a) / (t - a) * 100, y = W(
    (l, v) => {
      const r = m.current || v;
      if (!r) return;
      const o = r.getBoundingClientRect();
      if (!o || o.width <= 0) return;
      const q = Math.max(0, Math.min((typeof l == "number" ? l : 0) - (o.left || 0), o.width)) / o.width * (t - a) + a, A = Math.round(q / s) * s, M = Math.max(a, Math.min(t, A));
      Number.isNaN(M) || (d || T(M), c == null || c(M));
    },
    [t, a, s, d, c]
  ), L = (l) => {
    if (p) return;
    y(l.clientX, l.currentTarget);
    const v = (o) => y(o.clientX), r = () => {
      globalThis.removeEventListener("pointermove", v), globalThis.removeEventListener("pointerup", r);
    };
    globalThis.addEventListener("pointermove", v), globalThis.addEventListener("pointerup", r);
  }, R = n !== "inherit" && !(n != null && n.startsWith("#")) && !(n != null && n.startsWith("rgb")) ? `var(--pittorica-${n}-9)` : n;
  return /* @__PURE__ */ E(
    k,
    {
      as: b || "div",
      className: I("pittorica-slider-root", f),
      "data-disabled": p,
      onPointerDown: L,
      "aria-required": u,
      style: { "--pittorica-source-color": R, ...$ },
      ...x,
      children: [
        /* @__PURE__ */ w("div", { className: "pittorica-slider-track", ref: m, children: /* @__PURE__ */ w(
          "div",
          {
            className: "pittorica-slider-range",
            style: { width: `${h}%` }
          }
        ) }),
        /* @__PURE__ */ w(
          "div",
          {
            role: "slider",
            "aria-valuemin": a,
            "aria-valuemax": t,
            "aria-valuenow": g,
            "aria-disabled": p,
            "aria-required": u,
            tabIndex: p ? -1 : 0,
            className: "pittorica-slider-thumb",
            style: { position: "absolute", left: `calc(${h}% - 8px)` }
          }
        )
      ]
    }
  );
};
X.displayName = "Slider";
export {
  X as Slider
};
