// @pittorica/dropdown-menu-react@0.26.3/dist/index.js
import { jsxs as g, jsx as u } from "react/jsx-runtime";
import { createContext as M, useState as b, useRef as A, useCallback as m, use as N, useLayoutEffect as j, useEffect as P, useMemo as H } from "react";
import { createPortal as O } from "react-dom";
function E(t) {
  var o, n, e = "";
  if (typeof t == "string" || typeof t == "number") e += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var r = t.length;
    for (o = 0; o < r; o++) t[o] && (n = E(t[o])) && (e && (e += " "), e += n);
  } else for (n in t) t[n] && (e && (e += " "), e += n);
  return e;
}
function W() {
  for (var t, o, n = 0, e = "", r = arguments.length; n < r; n++) (t = arguments[n]) && (o = E(t)) && (e && (e += " "), e += o);
  return e;
}
const D = M(null), z = ({
  triggerRef: t,
  children: o,
  onClose: n,
  itemCount: e
}) => {
  const r = A(null), [l, d] = b(-1), [c, v] = b({ top: 0, left: 0 }), f = m(() => {
    const s = t.current, i = r.current;
    if (!i || !s) return;
    const a = s.getBoundingClientRect(), { innerWidth: p, innerHeight: L } = globalThis, { offsetWidth: y, offsetHeight: x } = i, k = a.left + y > p ? a.right - y : a.left, T = a.bottom + 4 + x > L ? a.top - x - 4 : a.bottom + 4;
    v({ left: k, top: T });
  }, [t]);
  j(() => {
    if (!t.current || !r.current) return;
    const s = new ResizeObserver(() => {
      f();
    });
    return s.observe(t.current), s.observe(r.current), () => s.disconnect();
  }, [t, f]);
  const w = m(
    (s) => {
      var a, p;
      const i = s.target;
      !((a = r.current) != null && a.contains(i)) && !((p = t.current) != null && p.contains(i)) && n();
    },
    [n, t]
  ), h = m(
    (s) => {
      e !== 0 && (s.key === "ArrowDown" && d((i) => (i + 1) % e), s.key === "ArrowUp" && d((i) => (i - 1 + e) % e), s.key === "Escape" && n());
    },
    [e, n]
  );
  P(() => (globalThis.addEventListener("keydown", h), globalThis.addEventListener("mousedown", w), () => {
    globalThis.removeEventListener("keydown", h), globalThis.removeEventListener("mousedown", w);
  }), [h, w]);
  const I = H(
    () => ({
      activeIndex: l,
      setActiveIndex: d,
      closeAll: n
    }),
    [l, n]
  );
  return O(
    /* @__PURE__ */ u(D, { value: I, children: /* @__PURE__ */ u(
      "div",
      {
        ref: r,
        className: "pittorica-dropdown-content",
        style: { top: c.top, left: c.left },
        children: o
      }
    ) }),
    document.body
  );
}, U = ({
  children: t,
  onClick: o,
  index: n,
  icon: e,
  shortcut: r,
  className: l,
  ...d
}) => {
  const c = N(D);
  if (!c) return null;
  const v = c.activeIndex === n;
  return /* @__PURE__ */ g(
    "div",
    {
      ...d,
      className: W("pittorica-dropdown-item", l),
      "data-active": v,
      onMouseEnter: () => c.setActiveIndex(n),
      onClick: (f) => {
        o == null || o(f), c.closeAll();
      },
      children: [
        e && /* @__PURE__ */ u("span", { className: "pittorica-dropdown-item-icon", children: e }),
        /* @__PURE__ */ u("span", { className: "pittorica-dropdown-item-content", children: t }),
        r && /* @__PURE__ */ u("kbd", { className: "pittorica-dropdown-item-shortcut", children: r })
      ]
    }
  );
}, V = ({
  renderTrigger: t,
  children: o,
  itemCount: n
}) => {
  const [e, r] = b(!1), l = A(null), d = m(() => r((c) => !c), []);
  return /* @__PURE__ */ g("div", { style: { display: "inline-block" }, children: [
    t({ ref: l, onClick: d }),
    e && /* @__PURE__ */ u(
      z,
      {
        triggerRef: l,
        onClose: () => r(!1),
        itemCount: n,
        children: o
      }
    )
  ] });
};
export {
  V as DropdownMenu,
  U as DropdownMenuItem
};
