// @pittorica/switch-react@0.26.3/dist/index.js
import { jsx as y } from "react/jsx-runtime";
import { forwardRef as W, createElement as b, useState as _ } from "react";
function x(t) {
  var e, r, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var a = t.length;
    for (e = 0; e < a; e++) t[e] && (r = x(t[e])) && (i && (i += " "), i += r);
  } else for (r in t) t[r] && (i && (i += " "), i += r);
  return i;
}
function z() {
  for (var t, e, r = 0, i = "", a = arguments.length; r < a; r++) (t = arguments[r]) && (e = x(t)) && (i && (i += " "), i += e);
  return i;
}
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var F = {
  outline: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  },
  filled: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "currentColor",
    stroke: "none"
  }
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const N = (t, e, r, i) => {
  const a = W(
    ({ color: p = "currentColor", size: l = 24, stroke: g = 2, title: c, className: d, children: s, ...f }, u) => b(
      "svg",
      {
        ref: u,
        ...F[t],
        width: l,
        height: l,
        className: ["tabler-icon", `tabler-icon-${e}`, d].join(" "),
        ...t === "filled" ? {
          fill: p
        } : {
          strokeWidth: g,
          stroke: p
        },
        ...f
      },
      [
        c && b("title", { key: "svg-title" }, c),
        ...i.map(([m, n]) => b(m, n)),
        ...Array.isArray(s) ? s : [s]
      ]
    )
  );
  return a.displayName = `${r}`, a;
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const I = [["path", { d: "M3 12a9 9 0 1 0 18 0a9 9 0 1 0 -18 0", key: "svg-0" }]], M = N("outline", "circle", "Circle", I);
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const E = [["path", { d: "M7 3.34a10 10 0 1 1 -4.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 4.995 -8.336z", key: "svg-0" }]], U = N("filled", "circle-filled", "CircleFilled", E);
function C(t) {
  var e, r, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var a = t.length;
    for (e = 0; e < a; e++) t[e] && (r = C(t[e])) && (i && (i += " "), i += r);
  } else for (r in t) t[r] && (i && (i += " "), i += r);
  return i;
}
function D() {
  for (var t, e, r = 0, i = "", a = arguments.length; r < a; r++) (t = arguments[r]) && (e = C(t)) && (i && (i += " "), i += e);
  return i;
}
const A = ({
  as: t,
  children: e,
  display: r,
  m: i,
  mt: a,
  mr: p,
  mb: l,
  ml: g,
  mx: c,
  my: d,
  p: s,
  pt: f,
  pr: u,
  pb: m,
  pl: n,
  px: h,
  py: w,
  width: $,
  height: v,
  position: k,
  style: T,
  className: B,
  disabled: L,
  required: R,
  ...S
}) => {
  const j = t || "div", o = {};
  r && (o.display = r), $ && (o.width = $), v && (o.height = v), k && (o.position = k), i && (o.margin = `var(--pittorica-space-${i})`), a && (o.marginTop = `var(--pittorica-space-${a})`), p && (o.marginRight = `var(--pittorica-space-${p})`), l && (o.marginBottom = `var(--pittorica-space-${l})`), g && (o.marginLeft = `var(--pittorica-space-${g})`), c && (o.marginLeft = `var(--pittorica-space-${c})`, o.marginRight = `var(--pittorica-space-${c})`), d && (o.marginTop = `var(--pittorica-space-${d})`, o.marginBottom = `var(--pittorica-space-${d})`), s && (o.padding = `var(--pittorica-space-${s})`), f && (o.paddingTop = `var(--pittorica-space-${f})`), u && (o.paddingRight = `var(--pittorica-space-${u})`), m && (o.paddingBottom = `var(--pittorica-space-${m})`), n && (o.paddingLeft = `var(--pittorica-space-${n})`), h && (o.paddingLeft = `var(--pittorica-space-${h})`, o.paddingRight = `var(--pittorica-space-${h})`), w && (o.paddingTop = `var(--pittorica-space-${w})`, o.paddingBottom = `var(--pittorica-space-${w})`);
  const q = {
    ...T,
    ...o
  };
  return /* @__PURE__ */ y(
    j,
    {
      className: D("pittorica-box", B),
      style: q,
      disabled: L,
      required: R,
      ...S,
      children: e
    }
  );
};
A.displayName = "Box";
const G = ({
  checked: t,
  defaultChecked: e,
  onCheckedChange: r,
  disabled: i,
  color: a = "source",
  name: p,
  className: l,
  style: g,
  required: c = !1,
  as: d,
  ...s
}) => {
  const [f, u] = _(
    e ?? !1
  ), m = t !== void 0, n = m ? t : f, h = (k) => {
    i || (m || u(!n), r == null || r(!n), typeof s.onClick == "function" && s.onClick(k));
  }, $ = a !== "inherit" && !(a != null && a.startsWith("#")) && !(a != null && a.startsWith("rgb")) ? `var(--pittorica-${a}-9)` : a, v = d || "button";
  return /* @__PURE__ */ y(
    A,
    {
      as: v,
      type: v === "button" ? "button" : void 0,
      role: "switch",
      name: p,
      "aria-checked": n,
      "aria-required": c,
      "data-state": n ? "checked" : "unchecked",
      disabled: i,
      onClick: h,
      className: z("pittorica-switch-root", l),
      style: {
        "--pittorica-source-color": $,
        ...g
      },
      ...s,
      children: /* @__PURE__ */ y("span", { className: "pittorica-switch-thumb", children: n ? /* @__PURE__ */ y(U, { size: 12, stroke: 2.5 }) : /* @__PURE__ */ y(M, { size: 12, stroke: 2.5 }) })
    }
  );
};
G.displayName = "Switch";
export {
  G as Switch
};
