// @pittorica/alert-dialog-react@0.26.3/dist/index.js
import { jsx as n, jsxs as w, Fragment as P } from "react/jsx-runtime";
import { useState as x, useId as W, useRef as C, useEffect as T, useLayoutEffect as F } from "react";
import { createPortal as K } from "react-dom";
function L(t) {
  var i, e, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (i = 0; i < o; i++) t[i] && (e = L(t[i])) && (a && (a += " "), a += e);
  } else for (e in t) t[e] && (a && (a += " "), a += e);
  return a;
}
function D() {
  for (var t, i, e = 0, a = "", o = arguments.length; e < o; e++) (t = arguments[e]) && (i = L(t)) && (a && (a += " "), a += i);
  return a;
}
function k(t) {
  var i, e, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (i = 0; i < o; i++) t[i] && (e = k(t[i])) && (a && (a += " "), a += e);
  } else for (e in t) t[e] && (a && (a += " "), a += e);
  return a;
}
function M() {
  for (var t, i, e = 0, a = "", o = arguments.length; e < o; e++) (t = arguments[e]) && (i = k(t)) && (a && (a += " "), a += i);
  return a;
}
const h = ({
  as: t,
  children: i,
  display: e,
  m: a,
  mt: o,
  mr: m,
  mb: l,
  ml: p,
  mx: d,
  my: u,
  p: g,
  pt: f,
  pr: v,
  pb: b,
  pl: c,
  px: s,
  py: y,
  width: $,
  height: A,
  position: N,
  style: R,
  className: j,
  disabled: B,
  required: E,
  ...q
}) => {
  const I = t || "div", r = {};
  e && (r.display = e), $ && (r.width = $), A && (r.height = A), N && (r.position = N), a && (r.margin = `var(--pittorica-space-${a})`), o && (r.marginTop = `var(--pittorica-space-${o})`), m && (r.marginRight = `var(--pittorica-space-${m})`), l && (r.marginBottom = `var(--pittorica-space-${l})`), p && (r.marginLeft = `var(--pittorica-space-${p})`), d && (r.marginLeft = `var(--pittorica-space-${d})`, r.marginRight = `var(--pittorica-space-${d})`), u && (r.marginTop = `var(--pittorica-space-${u})`, r.marginBottom = `var(--pittorica-space-${u})`), g && (r.padding = `var(--pittorica-space-${g})`), f && (r.paddingTop = `var(--pittorica-space-${f})`), v && (r.paddingRight = `var(--pittorica-space-${v})`), b && (r.paddingBottom = `var(--pittorica-space-${b})`), c && (r.paddingLeft = `var(--pittorica-space-${c})`), s && (r.paddingLeft = `var(--pittorica-space-${s})`, r.paddingRight = `var(--pittorica-space-${s})`), y && (r.paddingTop = `var(--pittorica-space-${y})`, r.paddingBottom = `var(--pittorica-space-${y})`);
  const S = {
    ...R,
    ...r
  };
  return /* @__PURE__ */ n(
    I,
    {
      className: M("pittorica-box", j),
      style: S,
      disabled: B,
      required: E,
      ...q,
      children: i
    }
  );
};
h.displayName = "Box";
const _ = ({
  children: t,
  sourceColor: i = "indigo",
  appearance: e = "light",
  radius: a = "medium",
  className: o,
  style: m,
  id: l
}) => {
  const p = i.startsWith("#") || i.startsWith("rgb"), d = {
    id: l,
    className: ["pittorica-theme", o].filter(Boolean).join(" "),
    "data-source-color": p ? "custom" : i,
    "data-appearance": e,
    "data-radius": a,
    style: {
      ...m,
      ...p ? { "--pittorica-source-color": i } : {}
    }
  };
  return /* @__PURE__ */ n("div", { ...d, children: t });
}, Q = ({
  children: t,
  color: i
}) => /* @__PURE__ */ n(h, { mb: "3", children: /* @__PURE__ */ n(
  "h2",
  {
    className: "pittorica-alert-dialog-title",
    style: {
      "--_alert-dialog-title-color": i ? `var(--pittorica-${i}-9)` : void 0
    },
    children: t
  }
) }), U = ({
  children: t
}) => /* @__PURE__ */ n(h, { mb: "6", children: /* @__PURE__ */ n("p", { className: "pittorica-alert-dialog-description", children: t }) }), V = ({ children: t }) => /* @__PURE__ */ n(h, { mt: "6", className: "pittorica-alert-dialog-actions", children: t }), z = ({
  open: t,
  onClose: i,
  children: e,
  className: a,
  appearance: o,
  closeOnOverlayClick: m = !1,
  closeOnEsc: l = !1
}) => {
  const [p, d] = x(!1), u = W(), g = C(null), [f, v] = x();
  if (T(() => {
    d(!0);
  }, []), F(() => {
    if (t && g.current) {
      const c = g.current.closest(
        ".pittorica-theme"
      );
      if (c) {
        const s = c.dataset.appearance;
        v(s || void 0);
      }
    }
  }, [t]), T(() => {
    if (typeof document > "u" || !document.body || !t) return;
    const c = document.body.style.overflow;
    document.body.style.overflow = "hidden";
    const s = (y) => {
      l && y.key === "Escape" && i();
    };
    return globalThis.addEventListener("keydown", s), () => {
      document.body.style.overflow = c, globalThis.removeEventListener("keydown", s);
    };
  }, [t, l, i]), !t || !p || typeof document > "u") return null;
  const b = o === "inherit" ? f : o ?? f;
  return /* @__PURE__ */ w(P, { children: [
    /* @__PURE__ */ n("div", { ref: g, style: { display: "none" }, "aria-hidden": "true" }),
    K(
      /* @__PURE__ */ w(
        "div",
        {
          className: D("pittorica-alert-dialog-portal"),
          onClick: m ? i : void 0,
          children: [
            /* @__PURE__ */ n("div", { className: "pittorica-alert-dialog-overlay", "aria-hidden": "true" }),
            /* @__PURE__ */ n(_, { appearance: b, children: /* @__PURE__ */ n(
              h,
              {
                className: D("pittorica-alert-dialog-content", a),
                role: "alertdialog",
                "aria-modal": "true",
                "aria-labelledby": u,
                onClick: (c) => c.stopPropagation(),
                children: e
              }
            ) })
          ]
        }
      ),
      document.body
    )
  ] });
};
z.displayName = "AlertDialog";
export {
  z as AlertDialog,
  V as AlertDialogActions,
  U as AlertDialogDescription,
  Q as AlertDialogTitle
};
