// @pittorica/progress-react@0.26.3/dist/index.js
import { jsx as f } from "react/jsx-runtime";
function N(a) {
  var e, i, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var p = a.length;
    for (e = 0; e < p; e++) a[e] && (i = N(a[e])) && (t && (t += " "), t += i);
  } else for (i in a) a[i] && (t && (t += " "), t += i);
  return t;
}
function M() {
  for (var a, e, i = 0, t = "", p = arguments.length; i < p; i++) (a = arguments[i]) && (e = N(a)) && (t && (t += " "), t += e);
  return t;
}
function T(a) {
  var e, i, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var p = a.length;
    for (e = 0; e < p; e++) a[e] && (i = T(a[e])) && (t && (t += " "), t += i);
  } else for (i in a) a[i] && (t && (t += " "), t += i);
  return t;
}
function P() {
  for (var a, e, i = 0, t = "", p = arguments.length; i < p; i++) (a = arguments[i]) && (e = T(a)) && (t && (t += " "), t += e);
  return t;
}
const w = ({
  as: a,
  children: e,
  display: i,
  m: t,
  mt: p,
  mr: n,
  mb: o,
  ml: c,
  mx: g,
  my: s,
  p: d,
  pt: h,
  pr: l,
  pb: u,
  pl: y,
  px: m,
  py: v,
  width: $,
  height: b,
  position: x,
  style: A,
  className: B,
  disabled: L,
  required: R,
  ...j
}) => {
  const q = a || "div", r = {};
  i && (r.display = i), $ && (r.width = $), b && (r.height = b), x && (r.position = x), t && (r.margin = `var(--pittorica-space-${t})`), p && (r.marginTop = `var(--pittorica-space-${p})`), n && (r.marginRight = `var(--pittorica-space-${n})`), o && (r.marginBottom = `var(--pittorica-space-${o})`), c && (r.marginLeft = `var(--pittorica-space-${c})`), g && (r.marginLeft = `var(--pittorica-space-${g})`, r.marginRight = `var(--pittorica-space-${g})`), s && (r.marginTop = `var(--pittorica-space-${s})`, r.marginBottom = `var(--pittorica-space-${s})`), d && (r.padding = `var(--pittorica-space-${d})`), h && (r.paddingTop = `var(--pittorica-space-${h})`), l && (r.paddingRight = `var(--pittorica-space-${l})`), u && (r.paddingBottom = `var(--pittorica-space-${u})`), y && (r.paddingLeft = `var(--pittorica-space-${y})`), m && (r.paddingLeft = `var(--pittorica-space-${m})`, r.paddingRight = `var(--pittorica-space-${m})`), v && (r.paddingTop = `var(--pittorica-space-${v})`, r.paddingBottom = `var(--pittorica-space-${v})`);
  const S = {
    ...A,
    ...r
  };
  return /* @__PURE__ */ f(
    q,
    {
      className: P("pittorica-box", B),
      style: S,
      disabled: L,
      required: R,
      ...j,
      children: e
    }
  );
};
w.displayName = "Box";
const W = ({
  value: a = 0,
  max: e = 100,
  color: i = "indigo",
  className: t,
  style: p,
  as: n,
  ...o
}) => {
  const c = Math.min(Math.max(0, a / e * 100), 100), s = i !== "inherit" && !(i != null && i.startsWith("#")) && !(i != null && i.startsWith("rgb")) ? `var(--pittorica-${i}-9)` : i;
  return /* @__PURE__ */ f(
    w,
    {
      as: n || "div",
      role: "progressbar",
      "aria-valuenow": a,
      "aria-valuemin": 0,
      "aria-valuemax": e,
      className: M("pittorica-progress-root", t),
      style: {
        "--pittorica-source-color": s,
        ...p
      },
      ...o,
      children: /* @__PURE__ */ f(
        "div",
        {
          className: "pittorica-progress-indicator",
          style: { width: `${c}%` }
        }
      )
    }
  );
};
W.displayName = "Progress";
export {
  W as Progress
};
