// @pittorica/radio-card-react@0.26.3/dist/index.js
import { jsx as C } from "react/jsx-runtime";
import { createContext as W, useState as E, useMemo as M, use as O } from "react";
function N(a) {
  var r, i, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var e = a.length;
    for (r = 0; r < e; r++) a[r] && (i = N(a[r])) && (t && (t += " "), t += i);
  } else for (i in a) a[i] && (t && (t += " "), t += i);
  return t;
}
function k() {
  for (var a, r, i = 0, t = "", e = arguments.length; i < e; i++) (a = arguments[i]) && (r = N(a)) && (t && (t += " "), t += r);
  return t;
}
function j(a) {
  var r, i, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var e = a.length;
    for (r = 0; r < e; r++) a[r] && (i = j(a[r])) && (t && (t += " "), t += i);
  } else for (i in a) a[i] && (t && (t += " "), t += i);
  return t;
}
function U() {
  for (var a, r, i = 0, t = "", e = arguments.length; i < e; i++) (a = arguments[i]) && (r = j(a)) && (t && (t += " "), t += r);
  return t;
}
const x = ({
  as: a,
  children: r,
  display: i,
  m: t,
  mt: e,
  mr: c,
  mb: p,
  ml: d,
  mx: l,
  my: s,
  p: f,
  pt: h,
  pr: y,
  pb: m,
  pl: g,
  px: u,
  py: n,
  width: v,
  height: b,
  position: R,
  style: $,
  className: A,
  disabled: B,
  required: L,
  ...S
}) => {
  const I = a || "div", o = {};
  i && (o.display = i), v && (o.width = v), b && (o.height = b), R && (o.position = R), t && (o.margin = `var(--pittorica-space-${t})`), e && (o.marginTop = `var(--pittorica-space-${e})`), c && (o.marginRight = `var(--pittorica-space-${c})`), p && (o.marginBottom = `var(--pittorica-space-${p})`), d && (o.marginLeft = `var(--pittorica-space-${d})`), l && (o.marginLeft = `var(--pittorica-space-${l})`, o.marginRight = `var(--pittorica-space-${l})`), s && (o.marginTop = `var(--pittorica-space-${s})`, o.marginBottom = `var(--pittorica-space-${s})`), f && (o.padding = `var(--pittorica-space-${f})`), h && (o.paddingTop = `var(--pittorica-space-${h})`), y && (o.paddingRight = `var(--pittorica-space-${y})`), m && (o.paddingBottom = `var(--pittorica-space-${m})`), g && (o.paddingLeft = `var(--pittorica-space-${g})`), u && (o.paddingLeft = `var(--pittorica-space-${u})`, o.paddingRight = `var(--pittorica-space-${u})`), n && (o.paddingTop = `var(--pittorica-space-${n})`, o.paddingBottom = `var(--pittorica-space-${n})`);
  const D = {
    ...$,
    ...o
  };
  return /* @__PURE__ */ C(
    I,
    {
      className: U("pittorica-box", A),
      style: D,
      disabled: B,
      required: L,
      ...S,
      children: r
    }
  );
};
x.displayName = "Box";
const q = W(null), z = () => {
  const a = O(q);
  if (!a)
    throw new Error(
      "RadioCard components must be wrapped in <RadioCard.Root />"
    );
  return a;
}, T = ({
  children: a,
  value: r,
  defaultValue: i,
  onValueChange: t,
  color: e = "indigo",
  disabled: c,
  name: p,
  required: d,
  columns: l = "1",
  className: s,
  style: f,
  as: h,
  ...y
}) => {
  const [m, g] = E(i), u = r !== void 0, n = u ? r : m, v = ($) => {
    u || g($), t == null || t($);
  }, b = M(
    () => ({
      value: n,
      onValueChange: v,
      color: e,
      disabled: c,
      name: p,
      required: d
    }),
    [n, e, c, p, d]
  );
  return /* @__PURE__ */ C(q, { value: b, children: /* @__PURE__ */ C(
    x,
    {
      as: h || "div",
      ...y,
      role: "radiogroup",
      className: k("pittorica-radio-card-root", s),
      style: {
        display: "grid",
        gridTemplateColumns: `repeat(${l}, minmax(0, 1fr))`,
        gap: "var(--pittorica-space-3)",
        ...f
      },
      children: a
    }
  ) });
}, w = ({
  children: a,
  value: r,
  disabled: i,
  className: t,
  style: e,
  as: c,
  ...p
}) => {
  const {
    value: d,
    onValueChange: l,
    color: s,
    disabled: f,
    name: h,
    required: y
  } = z(), m = d === r, g = f || i, n = s !== "inherit" && !(s != null && s.startsWith("#")) && !(s != null && s.startsWith("rgb")) ? `var(--pittorica-${s}-9)` : s, v = c || "button";
  return /* @__PURE__ */ C(
    x,
    {
      as: v,
      type: v === "button" ? "button" : void 0,
      role: "radio",
      name: h,
      "aria-checked": m,
      "aria-required": y,
      "data-state": m ? "checked" : "unchecked",
      disabled: g,
      className: k("pittorica-radio-card-item", t),
      onClick: () => !g && l(r),
      style: {
        "--pittorica-source-color": n,
        ...e
      },
      ...p,
      children: a
    }
  );
}, H = Object.assign(T, {
  Root: T,
  Item: w,
  displayName: "RadioCard"
});
w.displayName = "RadioCard.Item";
export {
  H as RadioCard
};
