// @pittorica/text-react@0.26.3/dist/index.js
import { jsx as T } from "react/jsx-runtime";
function N(a) {
  var e, t, i = "";
  if (typeof a == "string" || typeof a == "number") i += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var p = a.length;
    for (e = 0; e < p; e++) a[e] && (t = N(a[e])) && (i && (i += " "), i += t);
  } else for (t in a) a[t] && (i && (i += " "), i += t);
  return i;
}
function b() {
  for (var a, e, t = 0, i = "", p = arguments.length; t < p; t++) (a = arguments[t]) && (e = N(a)) && (i && (i += " "), i += e);
  return i;
}
function A(a) {
  var e, t, i = "";
  if (typeof a == "string" || typeof a == "number") i += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var p = a.length;
    for (e = 0; e < p; e++) a[e] && (t = A(a[e])) && (i && (i += " "), i += t);
  } else for (t in a) a[t] && (i && (i += " "), i += t);
  return i;
}
function W() {
  for (var a, e, t = 0, i = "", p = arguments.length; t < p; t++) (a = arguments[t]) && (e = A(a)) && (i && (i += " "), i += e);
  return i;
}
const w = ({
  as: a,
  children: e,
  display: t,
  m: i,
  mt: p,
  mr: g,
  mb: s,
  ml: m,
  mx: n,
  my: c,
  p: d,
  pt: f,
  pr: u,
  pb: l,
  pl: v,
  px: o,
  py: x,
  width: $,
  height: h,
  position: y,
  style: B,
  className: C,
  disabled: L,
  required: R,
  ...j
}) => {
  const q = a || "div", r = {};
  t && (r.display = t), $ && (r.width = $), h && (r.height = h), y && (r.position = y), i && (r.margin = `var(--pittorica-space-${i})`), p && (r.marginTop = `var(--pittorica-space-${p})`), g && (r.marginRight = `var(--pittorica-space-${g})`), s && (r.marginBottom = `var(--pittorica-space-${s})`), m && (r.marginLeft = `var(--pittorica-space-${m})`), n && (r.marginLeft = `var(--pittorica-space-${n})`, r.marginRight = `var(--pittorica-space-${n})`), c && (r.marginTop = `var(--pittorica-space-${c})`, r.marginBottom = `var(--pittorica-space-${c})`), d && (r.padding = `var(--pittorica-space-${d})`), f && (r.paddingTop = `var(--pittorica-space-${f})`), u && (r.paddingRight = `var(--pittorica-space-${u})`), l && (r.paddingBottom = `var(--pittorica-space-${l})`), v && (r.paddingLeft = `var(--pittorica-space-${v})`), o && (r.paddingLeft = `var(--pittorica-space-${o})`, r.paddingRight = `var(--pittorica-space-${o})`), x && (r.paddingTop = `var(--pittorica-space-${x})`, r.paddingBottom = `var(--pittorica-space-${x})`);
  const S = {
    ...B,
    ...r
  };
  return /* @__PURE__ */ T(
    q,
    {
      className: W("pittorica-box", C),
      style: S,
      disabled: L,
      required: R,
      ...j,
      children: e
    }
  );
};
w.displayName = "Box";
const k = ({
  children: a,
  as: e,
  size: t,
  weight: i = "regular",
  align: p,
  truncate: g = !1,
  color: s,
  className: m,
  style: n,
  ...c
}) => {
  const d = e || "span", f = (s == null ? void 0 : s.startsWith("#")) || (s == null ? void 0 : s.startsWith("rgb")) || (s == null ? void 0 : s.startsWith("hsl")), u = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), l = (() => {
    if (s)
      return s === "inherit" ? "inherit" : f ? s : u.has(s) ? `var(--pittorica-${s}-9)` : s;
  })(), v = t ? typeof t == "string" ? `pittorica-text--size-${t}` : b({
    [`pittorica-text--size-${t.initial}`]: t.initial,
    [`pittorica-text--xs-size-${t.xs}`]: t.xs,
    [`pittorica-text--sm-size-${t.sm}`]: t.sm,
    [`pittorica-text--md-size-${t.md}`]: t.md,
    [`pittorica-text--lg-size-${t.lg}`]: t.lg,
    [`pittorica-text--xl-size-${t.xl}`]: t.xl
  }) : null, o = {
    ...n,
    textAlign: p,
    color: l
  };
  return /* @__PURE__ */ T(
    w,
    {
      as: d,
      className: b(
        "pittorica-text",
        v,
        { "pittorica-text--truncate": g },
        m
      ),
      "data-weight": i,
      style: o,
      ...c,
      children: a
    }
  );
};
k.displayName = "Text";
export {
  k as Text
};
