// @pittorica/container-react@0.26.3/dist/index.js
import { jsx as x } from "react/jsx-runtime";
function N(i) {
  var p, r, a = "";
  if (typeof i == "string" || typeof i == "number") a += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (p = 0; p < o; p++) i[p] && (r = N(i[p])) && (a && (a += " "), a += r);
  } else for (r in i) i[r] && (a && (a += " "), a += r);
  return a;
}
function S() {
  for (var i, p, r = 0, a = "", o = arguments.length; r < o; r++) (i = arguments[r]) && (p = N(i)) && (a && (a += " "), a += p);
  return a;
}
function T(i) {
  var p, r, a = "";
  if (typeof i == "string" || typeof i == "number") a += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (p = 0; p < o; p++) i[p] && (r = T(i[p])) && (a && (a += " "), a += r);
  } else for (r in i) i[r] && (a && (a += " "), a += r);
  return a;
}
function k() {
  for (var i, p, r = 0, a = "", o = arguments.length; r < o; r++) (i = arguments[r]) && (p = T(i)) && (a && (a += " "), a += p);
  return a;
}
const A = ({
  as: i,
  children: p,
  display: r,
  m: a,
  mt: o,
  mr: s,
  mb: c,
  ml: d,
  mx: e,
  my: n,
  p: l,
  pt: f,
  pr: v,
  pb: y,
  pl: u,
  px: g,
  py: m,
  width: $,
  height: h,
  position: b,
  style: B,
  className: L,
  disabled: R,
  required: j,
  ...q
}) => {
  const w = i || "div", t = {};
  r && (t.display = r), $ && (t.width = $), h && (t.height = h), b && (t.position = b), a && (t.margin = `var(--pittorica-space-${a})`), o && (t.marginTop = `var(--pittorica-space-${o})`), s && (t.marginRight = `var(--pittorica-space-${s})`), c && (t.marginBottom = `var(--pittorica-space-${c})`), d && (t.marginLeft = `var(--pittorica-space-${d})`), e && (t.marginLeft = `var(--pittorica-space-${e})`, t.marginRight = `var(--pittorica-space-${e})`), n && (t.marginTop = `var(--pittorica-space-${n})`, t.marginBottom = `var(--pittorica-space-${n})`), l && (t.padding = `var(--pittorica-space-${l})`), f && (t.paddingTop = `var(--pittorica-space-${f})`), v && (t.paddingRight = `var(--pittorica-space-${v})`), y && (t.paddingBottom = `var(--pittorica-space-${y})`), u && (t.paddingLeft = `var(--pittorica-space-${u})`), g && (t.paddingLeft = `var(--pittorica-space-${g})`, t.paddingRight = `var(--pittorica-space-${g})`), m && (t.paddingTop = `var(--pittorica-space-${m})`, t.paddingBottom = `var(--pittorica-space-${m})`);
  const C = {
    ...B,
    ...t
  };
  return /* @__PURE__ */ x(
    w,
    {
      className: k("pittorica-box", L),
      style: C,
      disabled: R,
      required: j,
      ...q,
      children: p
    }
  );
};
A.displayName = "Box";
const z = ({
  children: i,
  maxWidth: p = "lg",
  fixed: r = !1,
  disableGutters: a = !1,
  className: o,
  style: s,
  as: c,
  ...d
}) => {
  const e = c || "div", n = {
    // Logic: Map maxWidth to breakpoint token if not in 'fixed' mode
    maxWidth: p && !r ? `var(--pittorica-bp-${p})` : void 0,
    ...s
  };
  return /* @__PURE__ */ x(
    A,
    {
      as: e,
      className: S(
        "pittorica-container",
        {
          "pittorica-container--disable-gutters": a
        },
        o
      ),
      "data-fixed": r,
      style: n,
      ...d,
      children: i
    }
  );
};
z.displayName = "Container";
export {
  z as Container
};
