// @pittorica/text-field-react@0.26.3/dist/index.js
import { jsx as x, jsxs as q } from "react/jsx-runtime";
import { createContext as K, useId as I, use as M } from "react";
function W(t) {
  var e, r, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var p = t.length;
    for (e = 0; e < p; e++) t[e] && (r = W(t[e])) && (i && (i += " "), i += r);
  } else for (r in t) t[r] && (i && (i += " "), i += r);
  return i;
}
function L() {
  for (var t, e, r = 0, i = "", p = arguments.length; r < p; r++) (t = arguments[r]) && (e = W(t)) && (i && (i += " "), i += e);
  return i;
}
function C(t) {
  var e, r, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var p = t.length;
    for (e = 0; e < p; e++) t[e] && (r = C(t[e])) && (i && (i += " "), i += r);
  } else for (r in t) t[r] && (i && (i += " "), i += r);
  return i;
}
function P() {
  for (var t, e, r = 0, i = "", p = arguments.length; r < p; r++) (t = arguments[r]) && (e = C(t)) && (i && (i += " "), i += e);
  return i;
}
const S = ({
  as: t,
  children: e,
  display: r,
  m: i,
  mt: p,
  mr: s,
  mb: o,
  ml: d,
  mx: n,
  my: c,
  p: u,
  pt: f,
  pr: v,
  pb: l,
  pl: m,
  px: g,
  py: h,
  width: $,
  height: y,
  position: b,
  style: T,
  className: N,
  disabled: w,
  required: R,
  ...F
}) => {
  const A = t || "div", a = {};
  r && (a.display = r), $ && (a.width = $), y && (a.height = y), b && (a.position = b), i && (a.margin = `var(--pittorica-space-${i})`), p && (a.marginTop = `var(--pittorica-space-${p})`), s && (a.marginRight = `var(--pittorica-space-${s})`), o && (a.marginBottom = `var(--pittorica-space-${o})`), d && (a.marginLeft = `var(--pittorica-space-${d})`), n && (a.marginLeft = `var(--pittorica-space-${n})`, a.marginRight = `var(--pittorica-space-${n})`), c && (a.marginTop = `var(--pittorica-space-${c})`, a.marginBottom = `var(--pittorica-space-${c})`), u && (a.padding = `var(--pittorica-space-${u})`), f && (a.paddingTop = `var(--pittorica-space-${f})`), v && (a.paddingRight = `var(--pittorica-space-${v})`), l && (a.paddingBottom = `var(--pittorica-space-${l})`), m && (a.paddingLeft = `var(--pittorica-space-${m})`), g && (a.paddingLeft = `var(--pittorica-space-${g})`, a.paddingRight = `var(--pittorica-space-${g})`), h && (a.paddingTop = `var(--pittorica-space-${h})`, a.paddingBottom = `var(--pittorica-space-${h})`);
  const B = {
    ...T,
    ...a
  };
  return /* @__PURE__ */ x(
    A,
    {
      className: P("pittorica-box", N),
      style: B,
      disabled: w,
      required: R,
      ...F,
      children: e
    }
  );
};
S.displayName = "Box";
function k(t) {
  var e, r, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var p = t.length;
    for (e = 0; e < p; e++) t[e] && (r = k(t[e])) && (i && (i += " "), i += r);
  } else for (r in t) t[r] && (i && (i += " "), i += r);
  return i;
}
function j() {
  for (var t, e, r = 0, i = "", p = arguments.length; r < p; r++) (t = arguments[r]) && (e = k(t)) && (i && (i += " "), i += e);
  return i;
}
function E(t) {
  var e, r, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var p = t.length;
    for (e = 0; e < p; e++) t[e] && (r = E(t[e])) && (i && (i += " "), i += r);
  } else for (r in t) t[r] && (i && (i += " "), i += r);
  return i;
}
function Q() {
  for (var t, e, r = 0, i = "", p = arguments.length; r < p; r++) (t = arguments[r]) && (e = E(t)) && (i && (i += " "), i += e);
  return i;
}
const O = ({
  as: t,
  children: e,
  display: r,
  m: i,
  mt: p,
  mr: s,
  mb: o,
  ml: d,
  mx: n,
  my: c,
  p: u,
  pt: f,
  pr: v,
  pb: l,
  pl: m,
  px: g,
  py: h,
  width: $,
  height: y,
  position: b,
  style: T,
  className: N,
  disabled: w,
  required: R,
  ...F
}) => {
  const A = t || "div", a = {};
  r && (a.display = r), $ && (a.width = $), y && (a.height = y), b && (a.position = b), i && (a.margin = `var(--pittorica-space-${i})`), p && (a.marginTop = `var(--pittorica-space-${p})`), s && (a.marginRight = `var(--pittorica-space-${s})`), o && (a.marginBottom = `var(--pittorica-space-${o})`), d && (a.marginLeft = `var(--pittorica-space-${d})`), n && (a.marginLeft = `var(--pittorica-space-${n})`, a.marginRight = `var(--pittorica-space-${n})`), c && (a.marginTop = `var(--pittorica-space-${c})`, a.marginBottom = `var(--pittorica-space-${c})`), u && (a.padding = `var(--pittorica-space-${u})`), f && (a.paddingTop = `var(--pittorica-space-${f})`), v && (a.paddingRight = `var(--pittorica-space-${v})`), l && (a.paddingBottom = `var(--pittorica-space-${l})`), m && (a.paddingLeft = `var(--pittorica-space-${m})`), g && (a.paddingLeft = `var(--pittorica-space-${g})`, a.paddingRight = `var(--pittorica-space-${g})`), h && (a.paddingTop = `var(--pittorica-space-${h})`, a.paddingBottom = `var(--pittorica-space-${h})`);
  const B = {
    ...T,
    ...a
  };
  return /* @__PURE__ */ x(
    A,
    {
      className: Q("pittorica-box", N),
      style: B,
      disabled: w,
      required: R,
      ...F,
      children: e
    }
  );
};
O.displayName = "Box";
const D = ({
  children: t,
  as: e,
  size: r,
  weight: i = "regular",
  align: p,
  truncate: s = !1,
  color: o,
  className: d,
  style: n,
  ...c
}) => {
  const u = e || "span", f = (o == null ? void 0 : o.startsWith("#")) || (o == null ? void 0 : o.startsWith("rgb")) || (o == null ? void 0 : o.startsWith("hsl")), v = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), l = (() => {
    if (o)
      return o === "inherit" ? "inherit" : f ? o : v.has(o) ? `var(--pittorica-${o}-9)` : o;
  })(), m = r ? typeof r == "string" ? `pittorica-text--size-${r}` : j({
    [`pittorica-text--size-${r.initial}`]: r.initial,
    [`pittorica-text--xs-size-${r.xs}`]: r.xs,
    [`pittorica-text--sm-size-${r.sm}`]: r.sm,
    [`pittorica-text--md-size-${r.md}`]: r.md,
    [`pittorica-text--lg-size-${r.lg}`]: r.lg,
    [`pittorica-text--xl-size-${r.xl}`]: r.xl
  }) : null, g = {
    ...n,
    textAlign: p,
    color: l
  };
  return /* @__PURE__ */ x(
    O,
    {
      as: u,
      className: j(
        "pittorica-text",
        m,
        { "pittorica-text--truncate": s },
        d
      ),
      "data-weight": i,
      style: g,
      ...c,
      children: t
    }
  );
};
D.displayName = "Text";
const G = K(null), U = () => {
  const t = M(G);
  if (!t)
    throw new Error(
      "TextField components must be wrapped in <TextField.Root />"
    );
  return t;
}, z = ({
  children: t,
  label: e,
  helperText: r,
  error: i,
  color: p = "source",
  disabled: s,
  name: o,
  required: d,
  size: n = "sm",
  className: c,
  style: u,
  as: f,
  ...v
}) => {
  const l = I(), m = I(), h = p !== "inherit" && !(p != null && p.startsWith("#")) && !(p != null && p.startsWith("rgb")) ? `var(--pittorica-${p}-9)` : p;
  return /* @__PURE__ */ x(
    G,
    {
      value: { inputId: l, helperId: m, disabled: s, size: n, name: o, required: d },
      children: /* @__PURE__ */ q(
        S,
        {
          as: f || "div",
          className: L(
            "pittorica-text-field-root",
            `pittorica-text-field--${n}`,
            c
          ),
          "data-error": i,
          ...v,
          children: [
            e && /* @__PURE__ */ q(
              D,
              {
                as: "label",
                htmlFor: l,
                weight: "medium",
                style: {
                  paddingLeft: "4px",
                  fontSize: "var(--pittorica-font-size-1)",
                  marginBottom: "4px",
                  display: "inline-block"
                },
                children: [
                  e,
                  " ",
                  d && /* @__PURE__ */ x("span", { "aria-hidden": "true", children: "*" })
                ]
              }
            ),
            /* @__PURE__ */ x(
              "div",
              {
                className: "pittorica-text-field-input-wrapper",
                "data-disabled": s,
                style: {
                  "--pittorica-source-color": h,
                  ...u
                },
                children: t
              }
            ),
            r && /* @__PURE__ */ x("div", { id: m, className: "pittorica-text-field-helper", children: r })
          ]
        }
      )
    }
  );
}, H = ({
  className: t,
  ...e
}) => {
  const {
    inputId: r,
    helperId: i,
    disabled: p,
    name: s,
    required: o
  } = U();
  return /* @__PURE__ */ x(
    "input",
    {
      name: s,
      ...e,
      id: r,
      "aria-describedby": i,
      disabled: p,
      required: o,
      className: L("pittorica-text-field-input", t)
    }
  );
}, J = ({
  children: t,
  className: e,
  as: r,
  ...i
}) => /* @__PURE__ */ x(
  S,
  {
    as: r || "div",
    className: L("pittorica-text-field-slot", e),
    ...i,
    children: t
  }
), Y = Object.assign(z, {
  Root: z,
  Input: H,
  Slot: J
});
z.displayName = "TextField.Root";
H.displayName = "TextField.Input";
J.displayName = "TextField.Slot";
export {
  Y as TextField,
  H as TextFieldInput,
  z as TextFieldRoot,
  J as TextFieldSlot
};
