// @pittorica/tabs-react@0.26.3/dist/index.js
import { jsx as f } from "react/jsx-runtime";
import { createContext as W, useState as H, useMemo as M, useRef as U, use as V } from "react";
function A(t) {
  var i, e, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var s = t.length;
    for (i = 0; i < s; i++) t[i] && (e = A(t[i])) && (a && (a += " "), a += e);
  } else for (e in t) t[e] && (a && (a += " "), a += e);
  return a;
}
function T() {
  for (var t, i, e = 0, a = "", s = arguments.length; e < s; e++) (t = arguments[e]) && (i = A(t)) && (a && (a += " "), a += i);
  return a;
}
function C(t) {
  var i, e, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var s = t.length;
    for (i = 0; i < s; i++) t[i] && (e = C(t[i])) && (a && (a += " "), a += e);
  } else for (e in t) t[e] && (a && (a += " "), a += e);
  return a;
}
function z() {
  for (var t, i, e = 0, a = "", s = arguments.length; e < s; e++) (t = arguments[e]) && (i = C(t)) && (a && (a += " "), a += i);
  return a;
}
const h = ({
  as: t,
  children: i,
  display: e,
  m: a,
  mt: s,
  mr: l,
  mb: d,
  ml: b,
  mx: o,
  my: p,
  p: c,
  pt: g,
  pr: n,
  pb: v,
  pl: m,
  px: u,
  py: y,
  width: N,
  height: R,
  position: w,
  style: q,
  className: B,
  disabled: D,
  required: I,
  ...K
}) => {
  const E = t || "div", r = {};
  e && (r.display = e), N && (r.width = N), R && (r.height = R), w && (r.position = w), a && (r.margin = `var(--pittorica-space-${a})`), s && (r.marginTop = `var(--pittorica-space-${s})`), l && (r.marginRight = `var(--pittorica-space-${l})`), d && (r.marginBottom = `var(--pittorica-space-${d})`), b && (r.marginLeft = `var(--pittorica-space-${b})`), o && (r.marginLeft = `var(--pittorica-space-${o})`, r.marginRight = `var(--pittorica-space-${o})`), p && (r.marginTop = `var(--pittorica-space-${p})`, r.marginBottom = `var(--pittorica-space-${p})`), c && (r.padding = `var(--pittorica-space-${c})`), g && (r.paddingTop = `var(--pittorica-space-${g})`), n && (r.paddingRight = `var(--pittorica-space-${n})`), v && (r.paddingBottom = `var(--pittorica-space-${v})`), m && (r.paddingLeft = `var(--pittorica-space-${m})`), u && (r.paddingLeft = `var(--pittorica-space-${u})`, r.paddingRight = `var(--pittorica-space-${u})`), y && (r.paddingTop = `var(--pittorica-space-${y})`, r.paddingBottom = `var(--pittorica-space-${y})`);
  const O = {
    ...q,
    ...r
  };
  return /* @__PURE__ */ f(
    E,
    {
      className: z("pittorica-box", B),
      style: O,
      disabled: D,
      required: I,
      ...K,
      children: i
    }
  );
};
h.displayName = "Box";
const L = W(null), $ = () => {
  const t = V(L);
  if (!t)
    throw new Error("Tabs components must be wrapped in <Tabs.Root />");
  return t;
}, x = ({
  children: t,
  defaultValue: i,
  value: e,
  onValueChange: a,
  color: s = "indigo",
  className: l,
  as: d,
  ...b
}) => {
  const [o, p] = H(i), c = e !== void 0, g = c ? e : o, n = (u) => {
    c || p(u), a == null || a(u);
  }, v = M(
    () => ({
      value: g,
      onValueChange: n,
      color: s
    }),
    [g, s]
  );
  return /* @__PURE__ */ f(L, { value: v, children: /* @__PURE__ */ f(
    h,
    {
      as: d || "div",
      className: T("pittorica-tabs-root", l),
      ...b,
      children: t
    }
  ) });
}, k = ({
  children: t,
  className: i,
  as: e,
  ...a
}) => {
  const s = U(null), { onValueChange: l } = $();
  return /* @__PURE__ */ f(
    h,
    {
      as: e || "div",
      ref: s,
      role: "tablist",
      onKeyDown: (o) => {
        var v;
        const p = o.target;
        if (p.getAttribute("role") !== "tab") return;
        const c = [
          ...((v = s.current) == null ? void 0 : v.querySelectorAll('[role="tab"]')) || []
        ], g = c.indexOf(p);
        let n = null;
        switch (o.key) {
          case "ArrowRight": {
            n = (g + 1) % c.length;
            break;
          }
          case "ArrowLeft": {
            n = (g - 1 + c.length) % c.length;
            break;
          }
          case "Home": {
            n = 0;
            break;
          }
          case "End": {
            n = c.length - 1;
            break;
          }
        }
        if (n !== null) {
          o.preventDefault();
          const m = c[n];
          m.focus();
          const u = m.dataset.value;
          u && l(u);
        }
      },
      className: T("pittorica-tabs-list", i),
      ...a,
      children: t
    }
  );
}, S = ({
  value: t,
  children: i,
  className: e,
  style: a,
  as: s,
  ...l
}) => {
  const { value: d, onValueChange: b, color: o } = $(), p = d === t, g = o !== "inherit" && !(o != null && o.startsWith("#")) && !(o != null && o.startsWith("rgb")) ? `var(--pittorica-${o}-9)` : o, n = s || "button";
  return /* @__PURE__ */ f(
    h,
    {
      as: n,
      role: "tab",
      type: n === "button" ? "button" : void 0,
      "aria-selected": p,
      tabIndex: p ? 0 : -1,
      "data-state": p ? "active" : "inactive",
      "data-value": t,
      className: T("pittorica-tabs-trigger", e),
      onClick: () => b(t),
      style: {
        "--pittorica-source-color": g,
        ...a
      },
      ...l,
      children: i
    }
  );
}, j = ({
  value: t,
  children: i,
  className: e,
  as: a,
  ...s
}) => {
  const { value: l } = $(), d = l === t;
  return /* @__PURE__ */ f(
    h,
    {
      as: a || "div",
      role: "tabpanel",
      tabIndex: 0,
      "data-state": d ? "active" : "inactive",
      className: T("pittorica-tabs-content", e),
      ...s,
      children: d ? i : null
    }
  );
}, J = Object.assign(x, {
  Root: x,
  List: k,
  Trigger: S,
  Content: j
});
x.displayName = "Tabs.Root";
k.displayName = "Tabs.List";
S.displayName = "Tabs.Trigger";
j.displayName = "Tabs.Content";
export {
  J as Tabs
};
