// eslint-fix-utils@0.4.1/lib/index.mjs
//#region src/addObjectProperty.ts
/**
* Given an ObjectProperty, and a rule fixer, this function yields the addition
* of a new property to the given ObjectExpression, along with any necessary commas.
* The new property is at the end of the object.
* @param context ESLint Rule Context
* @param fixer Rule fixer
* @param objectExpression The object expression to which the property will be added
* @param propertyKey The key for the new property
* @param propertyValue The value for the new property
* @yields fixer addition for the new property into the object expression, as well as any necessary commas
*/
function* addObjectProperty(context, fixer, objectExpression, propertyKey, propertyValue) {
	const sourceCode = context.sourceCode;
	let insertPosition;
	let needsComma = false;
	if (objectExpression.properties.length > 0) {
		const tokenAfter = sourceCode.getTokenAfter(objectExpression.properties[objectExpression.properties.length - 1]);
		if (tokenAfter?.value === ",") insertPosition = tokenAfter.range;
		else {
			needsComma = true;
			insertPosition = objectExpression.properties.at(-1)?.range;
		}
	}
	insertPosition ??= [0, 1];
	if (needsComma) {
		yield fixer.insertTextAfterRange(insertPosition, ",");
		insertPosition = [insertPosition[0] + 1, insertPosition[1] + 1];
	} else yield fixer.insertTextAfterRange(insertPosition, "\n");
	const textToInsert = `  ${`${JSON.stringify(propertyKey)}: ${JSON.stringify(propertyValue)}`}\n`;
	yield fixer.insertTextAfterRange(insertPosition, textToInsert);
}

//#endregion
//#region src/fixAddObjectProperty.ts
/**
* Given an ObjectExpression and the key and value of a property,
* this function returns a fixer function that you can provide to a report
* descriptor that will add the new property to it.
* @param context ESLint Rule Context
* @param objectExpression The object expression to which the property will be added
* @param propertyKey The key for the new property
* @param propertyValue The value for the new property
* @returns a fixer function that you can provide to a report descriptor that
* adds a new property to an object expression, along with any commas that
* are needed.
*/
const fixAddObjectProperty = (context, objectExpression, propertyKey, propertyValue) => {
	return (fixer) => addObjectProperty(context, fixer, objectExpression, propertyKey, propertyValue);
};

//#endregion
//#region src/removeArrayElement.ts
/**
* Given an ArrayExpression or the list of elements an ArrayExpression has,
* the index or node within that array that you want to remove, and a rule fixer,
* this function yields removals for the node itself, as well as any trailing
* commas that are no longer necessary.
* @param context ESLint Rule Context
* @param fixer Rule fixer
* @param elementOrIndex The child expression, spread element, or a numeric
* index of the child
* @param parentOrElements The array expression node, or its `.elements` array
* @yields fixer removals for the node itself, as well as any trailing commas
* that are no longer necessary.
*/
function* removeArrayElement(context, fixer, elementOrIndex, parentOrElements) {
	const [element, index] = getElementAndIndex(Array.isArray(parentOrElements) ? parentOrElements : parentOrElements.elements, elementOrIndex);
	if (!element) throw new Error("Cannot remove a null (blank) array element.");
	const tokenAfter = context.sourceCode.getTokenAfter(element);
	const tokenBefore = context.sourceCode.getTokenBefore(element);
	if (index > 0 && tokenAfter?.value !== "," && tokenBefore?.value === ",") yield fixer.remove(tokenBefore);
	yield fixer.remove(element);
	if (tokenAfter?.value === ",") yield fixer.remove(tokenAfter);
}
function getElementAndIndex(elements, elementOrIndex) {
	if (typeof elementOrIndex === "number") return [elements[elementOrIndex], elementOrIndex];
	const index = elements.indexOf(elementOrIndex);
	if (index === -1) throw new Error("Node is not a child of the parent array.");
	return [elements[index], index];
}

//#endregion
//#region src/fixRemoveArrayElement.ts
/**
* Given an ArrayExpression or the list of elements an ArrayExpression has, and
* the index or node within that array that you want to remove, this function
* returns a fixer function that you can provide to a report descriptor that
* will remove that node along with any trailing comma.
* @param context ESLint Rule Context
* @param elementOrIndex The child expression, spread element, or a numeric
* index of the child
* @param parentOrElements The array expression node, or its `.elements` array
* @returns a fixer function that you can provide to a report descriptor, that
* removes an element from an array expression, along with any commas that are
* no longer necessary.
*/
const fixRemoveArrayElement = (context, elementOrIndex, parentOrElements) => {
	return (fixer) => removeArrayElement(context, fixer, elementOrIndex, parentOrElements);
};

//#endregion
//#region src/removeObjectProperty.ts
/**
* Given an ObjectProperty, and a rule fixer, this function yields removals for
* the node itself, as well as any trailing commas that are no longer necessary.
* @param context ESLint Rule Context
* @param fixer Rule fixer
* @param property The property node
* @yields fixer removals for the node itself, as well as any trailing commas
* that are no longer necessary.
*/
function* removeObjectProperty(context, fixer, property) {
	const tokenAfter = context.sourceCode.getTokenAfter(property);
	const tokenBefore = context.sourceCode.getTokenBefore(property);
	if (tokenAfter?.value !== "," && tokenBefore?.value === ",") yield fixer.remove(tokenBefore);
	yield fixer.remove(property);
	if (tokenAfter?.value === ",") yield fixer.remove(tokenAfter);
}

//#endregion
//#region src/fixRemoveObjectProperty.ts
/**
* Given an ObjectProperty, this function returns a fixer function that you can
* provide to a report descriptor that will remove that node along with any
* trailing comma.
* @param context ESLint Rule Context
* @param property The property node
* @returns a fixer function that you can provide to a report descriptor, that
* removes a property from an object expression, along with any commas that
* are no longer necessary.
*/
const fixRemoveObjectProperty = (context, property) => {
	return (fixer) => removeObjectProperty(context, fixer, property);
};

//#endregion
export { addObjectProperty, fixAddObjectProperty, fixRemoveArrayElement, fixRemoveObjectProperty, removeArrayElement, removeObjectProperty };
