// @pittorica/tooltip-react@0.26.3/dist/index.js
import { jsx as w, jsxs as k } from "react/jsx-runtime";
import { useState as A, useRef as C, useLayoutEffect as H, useCallback as I } from "react";
import { createPortal as M } from "react-dom";
function E(o) {
  var a, i, t = "";
  if (typeof o == "string" || typeof o == "number") t += o;
  else if (typeof o == "object") if (Array.isArray(o)) {
    var r = o.length;
    for (a = 0; a < r; a++) o[a] && (i = E(o[a])) && (t && (t += " "), t += i);
  } else for (i in o) o[i] && (t && (t += " "), t += i);
  return t;
}
function P() {
  for (var o, a, i = 0, t = "", r = arguments.length; i < r; i++) (o = arguments[i]) && (a = E(o)) && (t && (t += " "), t += a);
  return t;
}
function q(o) {
  var a, i, t = "";
  if (typeof o == "string" || typeof o == "number") t += o;
  else if (typeof o == "object") if (Array.isArray(o)) {
    var r = o.length;
    for (a = 0; a < r; a++) o[a] && (i = q(o[a])) && (t && (t += " "), t += i);
  } else for (i in o) o[i] && (t && (t += " "), t += i);
  return t;
}
function W() {
  for (var o, a, i = 0, t = "", r = arguments.length; i < r; i++) (o = arguments[i]) && (a = q(o)) && (t && (t += " "), t += a);
  return t;
}
const O = ({
  as: o,
  children: a,
  display: i,
  m: t,
  mt: r,
  mr: d,
  mb: m,
  ml: s,
  mx: c,
  my: u,
  p: b,
  pt: $,
  pr: x,
  pb: g,
  pl: N,
  px: p,
  py: h,
  width: R,
  height: v,
  position: y,
  style: B,
  className: L,
  disabled: n,
  required: l,
  ...T
}) => {
  const f = o || "div", e = {};
  i && (e.display = i), R && (e.width = R), v && (e.height = v), y && (e.position = y), t && (e.margin = `var(--pittorica-space-${t})`), r && (e.marginTop = `var(--pittorica-space-${r})`), d && (e.marginRight = `var(--pittorica-space-${d})`), m && (e.marginBottom = `var(--pittorica-space-${m})`), s && (e.marginLeft = `var(--pittorica-space-${s})`), c && (e.marginLeft = `var(--pittorica-space-${c})`, e.marginRight = `var(--pittorica-space-${c})`), u && (e.marginTop = `var(--pittorica-space-${u})`, e.marginBottom = `var(--pittorica-space-${u})`), b && (e.padding = `var(--pittorica-space-${b})`), $ && (e.paddingTop = `var(--pittorica-space-${$})`), x && (e.paddingRight = `var(--pittorica-space-${x})`), g && (e.paddingBottom = `var(--pittorica-space-${g})`), N && (e.paddingLeft = `var(--pittorica-space-${N})`), p && (e.paddingLeft = `var(--pittorica-space-${p})`, e.paddingRight = `var(--pittorica-space-${p})`), h && (e.paddingTop = `var(--pittorica-space-${h})`, e.paddingBottom = `var(--pittorica-space-${h})`);
  const j = {
    ...B,
    ...e
  };
  return /* @__PURE__ */ w(
    f,
    {
      className: W("pittorica-box", L),
      style: j,
      disabled: n,
      required: l,
      ...T,
      children: a
    }
  );
};
O.displayName = "Box";
const F = ({
  children: o,
  sourceColor: a = "indigo",
  appearance: i = "light",
  radius: t = "medium",
  className: r,
  style: d,
  id: m
}) => {
  const s = a.startsWith("#") || a.startsWith("rgb"), c = {
    id: m,
    className: ["pittorica-theme", r].filter(Boolean).join(" "),
    "data-source-color": s ? "custom" : a,
    "data-appearance": i,
    "data-radius": t,
    style: {
      ...d,
      ...s ? { "--pittorica-source-color": a } : {}
    }
  };
  return /* @__PURE__ */ w("div", { ...c, children: o });
}, z = ({
  children: o,
  content: a,
  side: i = "top",
  appearance: t,
  className: r,
  as: d,
  ...m
}) => {
  const [s, c] = A(!1), [u, b] = A({ top: 0, left: 0 }), [$, x] = A(i), [g, N] = A(), p = C(null), h = C(null);
  H(() => {
    if (s && p.current) {
      const n = p.current.closest(
        ".pittorica-theme"
      );
      if (n) {
        const l = n.dataset.appearance;
        N(l || void 0);
      }
    }
  }, [s]);
  const R = I(() => {
    if (!p.current) return;
    const n = p.current.getBoundingClientRect(), l = 8, T = 32;
    let f = i;
    i === "top" && n.top - T - l < 0 ? f = "bottom" : i === "bottom" && globalThis.window !== void 0 && n.bottom + T + l > window.innerHeight && (f = "top"), x(f);
    const e = f === "top" ? n.top - l : n.bottom + l, j = n.left + n.width / 2;
    b({ top: e, left: j });
  }, [i]), v = () => {
    R(), c(!0);
  }, y = () => c(!1), B = d || "span", L = t === "inherit" ? g : t ?? g;
  return /* @__PURE__ */ k(
    O,
    {
      as: B,
      ref: p,
      className: P("pittorica-tooltip-root", r),
      onMouseEnter: v,
      onMouseLeave: y,
      onFocus: v,
      onBlur: y,
      ...m,
      children: [
        o,
        s && typeof document < "u" && M(
          /* @__PURE__ */ w(F, { appearance: L, children: /* @__PURE__ */ w(
            "div",
            {
              ref: h,
              className: "pittorica-tooltip-content",
              role: "tooltip",
              style: {
                top: u.top,
                left: u.left,
                transform: $ === "top" ? "translate(-50%, -100%)" : "translate(-50%, 0)",
                position: "fixed",
                pointerEvents: "none"
              },
              children: a
            }
          ) }),
          document.body
        )
      ]
    }
  );
};
z.displayName = "Tooltip";
export {
  z as Tooltip
};
