// @pittorica/sheet-react@0.26.3/dist/index.js
import { jsx as f, jsxs as j, Fragment as X } from "react/jsx-runtime";
import { forwardRef as _, createElement as C, useState as W, useRef as F, useEffect as I, useLayoutEffect as G } from "react";
import { createPortal as Y } from "react-dom";
function S(t) {
  var e, i, r = "";
  if (typeof t == "string" || typeof t == "number") r += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (e = 0; e < o; e++) t[e] && (i = S(t[e])) && (r && (r += " "), r += i);
  } else for (i in t) t[i] && (r && (r += " "), r += i);
  return r;
}
function D() {
  for (var t, e, i = 0, r = "", o = arguments.length; i < o; i++) (t = arguments[i]) && (e = S(t)) && (r && (r += " "), r += e);
  return r;
}
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var H = {
  outline: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  },
  filled: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "currentColor",
    stroke: "none"
  }
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const J = (t, e, i, r) => {
  const o = _(
    ({ color: s = "currentColor", size: p = 24, stroke: c = 2, title: n, className: d, children: m, ...g }, u) => C(
      "svg",
      {
        ref: u,
        ...H[t],
        width: p,
        height: p,
        className: ["tabler-icon", `tabler-icon-${e}`, d].join(" "),
        strokeWidth: c,
        stroke: s,
        ...g
      },
      [
        n && C("title", { key: "svg-title" }, n),
        ...r.map(([h, v]) => C(h, v)),
        ...Array.isArray(m) ? m : [m]
      ]
    )
  );
  return o.displayName = `${i}`, o;
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const K = [["path", { d: "M18 6l-12 12", key: "svg-0" }], ["path", { d: "M6 6l12 12", key: "svg-1" }]], Q = J("outline", "x", "X", K);
function q(t) {
  var e, i, r = "";
  if (typeof t == "string" || typeof t == "number") r += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (e = 0; e < o; e++) t[e] && (i = q(t[e])) && (r && (r += " "), r += i);
  } else for (i in t) t[i] && (r && (r += " "), r += i);
  return r;
}
function U() {
  for (var t, e, i = 0, r = "", o = arguments.length; i < o; i++) (t = arguments[i]) && (e = q(t)) && (r && (r += " "), r += e);
  return r;
}
const x = ({
  as: t,
  children: e,
  display: i,
  m: r,
  mt: o,
  mr: s,
  mb: p,
  ml: c,
  mx: n,
  my: d,
  p: m,
  pt: g,
  pr: u,
  pb: h,
  pl: v,
  px: y,
  py: l,
  width: $,
  height: b,
  position: w,
  style: N,
  className: k,
  disabled: A,
  required: B,
  ...R
}) => {
  const T = t || "div", a = {};
  i && (a.display = i), $ && (a.width = $), b && (a.height = b), w && (a.position = w), r && (a.margin = `var(--pittorica-space-${r})`), o && (a.marginTop = `var(--pittorica-space-${o})`), s && (a.marginRight = `var(--pittorica-space-${s})`), p && (a.marginBottom = `var(--pittorica-space-${p})`), c && (a.marginLeft = `var(--pittorica-space-${c})`), n && (a.marginLeft = `var(--pittorica-space-${n})`, a.marginRight = `var(--pittorica-space-${n})`), d && (a.marginTop = `var(--pittorica-space-${d})`, a.marginBottom = `var(--pittorica-space-${d})`), m && (a.padding = `var(--pittorica-space-${m})`), g && (a.paddingTop = `var(--pittorica-space-${g})`), u && (a.paddingRight = `var(--pittorica-space-${u})`), h && (a.paddingBottom = `var(--pittorica-space-${h})`), v && (a.paddingLeft = `var(--pittorica-space-${v})`), y && (a.paddingLeft = `var(--pittorica-space-${y})`, a.paddingRight = `var(--pittorica-space-${y})`), l && (a.paddingTop = `var(--pittorica-space-${l})`, a.paddingBottom = `var(--pittorica-space-${l})`);
  const L = {
    ...N,
    ...a
  };
  return /* @__PURE__ */ f(
    T,
    {
      className: U("pittorica-box", k),
      style: L,
      disabled: A,
      required: B,
      ...R,
      children: e
    }
  );
};
x.displayName = "Box";
function E(t) {
  var e, i, r = "";
  if (typeof t == "string" || typeof t == "number") r += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (e = 0; e < o; e++) t[e] && (i = E(t[e])) && (r && (r += " "), r += i);
  } else for (i in t) t[i] && (r && (r += " "), r += i);
  return r;
}
function V() {
  for (var t, e, i = 0, r = "", o = arguments.length; i < o; i++) (t = arguments[i]) && (e = E(t)) && (r && (r += " "), r += e);
  return r;
}
function M(t) {
  var e, i, r = "";
  if (typeof t == "string" || typeof t == "number") r += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (e = 0; e < o; e++) t[e] && (i = M(t[e])) && (r && (r += " "), r += i);
  } else for (i in t) t[i] && (r && (r += " "), r += i);
  return r;
}
function Z() {
  for (var t, e, i = 0, r = "", o = arguments.length; i < o; i++) (t = arguments[i]) && (e = M(t)) && (r && (r += " "), r += e);
  return r;
}
const z = ({
  as: t,
  children: e,
  display: i,
  m: r,
  mt: o,
  mr: s,
  mb: p,
  ml: c,
  mx: n,
  my: d,
  p: m,
  pt: g,
  pr: u,
  pb: h,
  pl: v,
  px: y,
  py: l,
  width: $,
  height: b,
  position: w,
  style: N,
  className: k,
  disabled: A,
  required: B,
  ...R
}) => {
  const T = t || "div", a = {};
  i && (a.display = i), $ && (a.width = $), b && (a.height = b), w && (a.position = w), r && (a.margin = `var(--pittorica-space-${r})`), o && (a.marginTop = `var(--pittorica-space-${o})`), s && (a.marginRight = `var(--pittorica-space-${s})`), p && (a.marginBottom = `var(--pittorica-space-${p})`), c && (a.marginLeft = `var(--pittorica-space-${c})`), n && (a.marginLeft = `var(--pittorica-space-${n})`, a.marginRight = `var(--pittorica-space-${n})`), d && (a.marginTop = `var(--pittorica-space-${d})`, a.marginBottom = `var(--pittorica-space-${d})`), m && (a.padding = `var(--pittorica-space-${m})`), g && (a.paddingTop = `var(--pittorica-space-${g})`), u && (a.paddingRight = `var(--pittorica-space-${u})`), h && (a.paddingBottom = `var(--pittorica-space-${h})`), v && (a.paddingLeft = `var(--pittorica-space-${v})`), y && (a.paddingLeft = `var(--pittorica-space-${y})`, a.paddingRight = `var(--pittorica-space-${y})`), l && (a.paddingTop = `var(--pittorica-space-${l})`, a.paddingBottom = `var(--pittorica-space-${l})`);
  const L = {
    ...N,
    ...a
  };
  return /* @__PURE__ */ f(
    T,
    {
      className: Z("pittorica-box", k),
      style: L,
      disabled: A,
      required: B,
      ...R,
      children: e
    }
  );
};
z.displayName = "Box";
const P = ({
  children: t,
  variant: e = "filled",
  color: i = "source",
  size: r = "3",
  className: o,
  style: s,
  as: p,
  ...c
}) => {
  const n = i !== "inherit" && !(i != null && i.startsWith("#")) && !(i != null && i.startsWith("rgb")), d = {
    ...s,
    "--pittorica-source-color": n ? `var(--pittorica-${i}-9)` : i,
    "--pittorica-on-source-color": "#ffffff"
  }, m = p || (c.href ? "a" : "button");
  return /* @__PURE__ */ f(
    z,
    {
      as: m,
      className: V(
        "pittorica-icon-button",
        `pittorica-icon-button--${e}`,
        `pittorica-icon-button--${r}`,
        o
      ),
      style: d,
      ...c,
      children: t
    }
  );
};
P.displayName = "IconButton";
const O = ({
  children: t,
  sourceColor: e = "indigo",
  appearance: i = "light",
  radius: r = "medium",
  className: o,
  style: s,
  id: p
}) => {
  const c = e.startsWith("#") || e.startsWith("rgb"), n = {
    id: p,
    className: ["pittorica-theme", o].filter(Boolean).join(" "),
    "data-source-color": c ? "custom" : e,
    "data-appearance": i,
    "data-radius": r,
    style: {
      ...s,
      ...c ? { "--pittorica-source-color": e } : {}
    }
  };
  return /* @__PURE__ */ f("div", { ...n, children: t });
}, tt = ({
  isOpen: t,
  onClose: e,
  children: i,
  side: r = "right",
  title: o,
  appearance: s,
  className: p,
  as: c,
  ...n
}) => {
  const [d, m] = W(!1), g = F(null), [u, h] = W();
  return I(() => {
    m(!0);
  }, []), G(() => {
    if (t && g.current) {
      const l = g.current.closest(
        ".pittorica-theme"
      );
      if (l) {
        const $ = l.dataset.appearance;
        h($ || void 0);
      }
    }
  }, [t]), I(() => {
    if (!(typeof document > "u"))
      return t && (document.body.style.overflow = "hidden"), () => {
        document.body.style.overflow = "";
      };
  }, [t]), !t || !d || typeof document > "u" ? null : /* @__PURE__ */ j(X, { children: [
    /* @__PURE__ */ f("div", { ref: g, style: { display: "none" }, "aria-hidden": "true" }),
    Y(
      /* @__PURE__ */ f("div", { className: "pittorica-sheet-overlay", onClick: e, children: /* @__PURE__ */ f(O, { appearance: s === "inherit" ? u : s ?? u, children: /* @__PURE__ */ j(
        x,
        {
          as: c || "div",
          className: D(
            "pittorica-sheet-content",
            `pittorica-sheet--${r}`,
            p
          ),
          role: "dialog",
          "aria-modal": "true",
          onClick: (l) => l.stopPropagation(),
          ...n,
          children: [
            (r === "bottom" || r === "top") && /* @__PURE__ */ f("div", { className: "pittorica-sheet-handle" }),
            /* @__PURE__ */ j(
              x,
              {
                p: "4",
                style: {
                  display: "flex",
                  alignItems: "center",
                  justifyContent: "space-between"
                },
                children: [
                  o && /* @__PURE__ */ f(
                    "h2",
                    {
                      style: {
                        margin: 0,
                        fontSize: "var(--pittorica-font-size-4)"
                      },
                      children: o
                    }
                  ),
                  /* @__PURE__ */ f(
                    P,
                    {
                      variant: "text",
                      color: "inherit",
                      onClick: (l) => {
                        l.stopPropagation(), e();
                      },
                      "aria-label": "Close",
                      as: "button",
                      children: /* @__PURE__ */ f(Q, { size: 20 })
                    }
                  ),
                  " "
                ]
              }
            ),
            /* @__PURE__ */ f(x, { p: "4", style: { flexGrow: 1, overflowY: "auto" }, children: i })
          ]
        }
      ) }) }),
      document.body
    )
  ] });
};
tt.displayName = "Sheet";
export {
  tt as Sheet
};
