// @pittorica/button-react@0.26.3/dist/index.js
import { jsx as x } from "react/jsx-runtime";
function B(i) {
  var p, a, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var e = i.length;
    for (p = 0; p < e; p++) i[p] && (a = B(i[p])) && (t && (t += " "), t += a);
  } else for (a in i) i[a] && (t && (t += " "), t += a);
  return t;
}
function S() {
  for (var i, p, a = 0, t = "", e = arguments.length; a < e; a++) (i = arguments[a]) && (p = B(i)) && (t && (t += " "), t += p);
  return t;
}
function N(i) {
  var p, a, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var e = i.length;
    for (p = 0; p < e; p++) i[p] && (a = N(i[p])) && (t && (t += " "), t += a);
  } else for (a in i) i[a] && (t && (t += " "), t += a);
  return t;
}
function W() {
  for (var i, p, a = 0, t = "", e = arguments.length; a < e; a++) (i = arguments[a]) && (p = N(i)) && (t && (t += " "), t += p);
  return t;
}
const T = ({
  as: i,
  children: p,
  display: a,
  m: t,
  mt: e,
  mr: c,
  mb: g,
  ml: m,
  mx: s,
  my: o,
  p: n,
  pt: d,
  pr: v,
  pb: $,
  pl: b,
  px: f,
  py: u,
  width: h,
  height: y,
  position: l,
  style: w,
  className: A,
  disabled: L,
  required: R,
  ...j
}) => {
  const q = i || "div", r = {};
  a && (r.display = a), h && (r.width = h), y && (r.height = y), l && (r.position = l), t && (r.margin = `var(--pittorica-space-${t})`), e && (r.marginTop = `var(--pittorica-space-${e})`), c && (r.marginRight = `var(--pittorica-space-${c})`), g && (r.marginBottom = `var(--pittorica-space-${g})`), m && (r.marginLeft = `var(--pittorica-space-${m})`), s && (r.marginLeft = `var(--pittorica-space-${s})`, r.marginRight = `var(--pittorica-space-${s})`), o && (r.marginTop = `var(--pittorica-space-${o})`, r.marginBottom = `var(--pittorica-space-${o})`), n && (r.padding = `var(--pittorica-space-${n})`), d && (r.paddingTop = `var(--pittorica-space-${d})`), v && (r.paddingRight = `var(--pittorica-space-${v})`), $ && (r.paddingBottom = `var(--pittorica-space-${$})`), b && (r.paddingLeft = `var(--pittorica-space-${b})`), f && (r.paddingLeft = `var(--pittorica-space-${f})`, r.paddingRight = `var(--pittorica-space-${f})`), u && (r.paddingTop = `var(--pittorica-space-${u})`, r.paddingBottom = `var(--pittorica-space-${u})`);
  const k = {
    ...w,
    ...r
  };
  return /* @__PURE__ */ x(
    q,
    {
      className: W("pittorica-box", A),
      style: k,
      disabled: L,
      required: R,
      ...j,
      children: p
    }
  );
};
T.displayName = "Box";
const V = ({
  children: i,
  variant: p = "filled",
  size: a = "sm",
  color: t = "source",
  disabled: e = !1,
  appearance: c,
  className: g,
  style: m,
  as: s,
  ...o
}) => {
  const n = t !== "inherit" && !(t != null && t.startsWith("#")) && !(t != null && t.startsWith("rgb")), d = {
    "--pittorica-button-color": n ? `var(--pittorica-${t}-9)` : t,
    "--pittorica-button-on-color": n ? `var(--pittorica-on-${t}-9)` : "var(--pittorica-white)",
    "--pittorica-button-soft-bg": n ? `var(--pittorica-${t}-3)` : `color-mix(in srgb, ${t} 15%, var(--pittorica-white))`,
    "--pittorica-button-soft-text": n ? `var(--pittorica-${t}-11)` : `color-mix(in srgb, ${t} 80%, var(--pittorica-black))`,
    ...m
  }, v = s || (o.href ? "a" : "button");
  return /* @__PURE__ */ x(
    T,
    {
      as: v,
      ...o,
      disabled: e,
      "aria-disabled": e,
      className: S(
        "pittorica-button",
        `pittorica-button--${p}`,
        `pittorica-button--${a}`,
        g
      ),
      style: d,
      "data-appearance": c,
      children: i
    }
  );
};
V.displayName = "Button";
export {
  V as Button
};
