// @pittorica/section-react@0.26.3/dist/index.js
import { jsx as x } from "react/jsx-runtime";
function T(i) {
  var p, r, a = "";
  if (typeof i == "string" || typeof i == "number") a += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (p = 0; p < o; p++) i[p] && (r = T(i[p])) && (a && (a += " "), a += r);
  } else for (r in i) i[r] && (a && (a += " "), a += r);
  return a;
}
function k() {
  for (var i, p, r = 0, a = "", o = arguments.length; r < o; r++) (i = arguments[r]) && (p = T(i)) && (a && (a += " "), a += p);
  return a;
}
function N(i) {
  var p, r, a = "";
  if (typeof i == "string" || typeof i == "number") a += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (p = 0; p < o; p++) i[p] && (r = N(i[p])) && (a && (a += " "), a += r);
  } else for (r in i) i[r] && (a && (a += " "), a += r);
  return a;
}
function z() {
  for (var i, p, r = 0, a = "", o = arguments.length; r < o; r++) (i = arguments[r]) && (p = N(i)) && (a && (a += " "), a += p);
  return a;
}
const A = ({
  as: i,
  children: p,
  display: r,
  m: a,
  mt: o,
  mr: e,
  mb: d,
  ml: m,
  mx: c,
  my: s,
  p: f,
  pt: v,
  pr: l,
  pb: u,
  pl: y,
  px: n,
  py: g,
  width: $,
  height: h,
  position: b,
  style: B,
  className: L,
  disabled: R,
  required: j,
  ...q
}) => {
  const S = i || "div", t = {};
  r && (t.display = r), $ && (t.width = $), h && (t.height = h), b && (t.position = b), a && (t.margin = `var(--pittorica-space-${a})`), o && (t.marginTop = `var(--pittorica-space-${o})`), e && (t.marginRight = `var(--pittorica-space-${e})`), d && (t.marginBottom = `var(--pittorica-space-${d})`), m && (t.marginLeft = `var(--pittorica-space-${m})`), c && (t.marginLeft = `var(--pittorica-space-${c})`, t.marginRight = `var(--pittorica-space-${c})`), s && (t.marginTop = `var(--pittorica-space-${s})`, t.marginBottom = `var(--pittorica-space-${s})`), f && (t.padding = `var(--pittorica-space-${f})`), v && (t.paddingTop = `var(--pittorica-space-${v})`), l && (t.paddingRight = `var(--pittorica-space-${l})`), u && (t.paddingBottom = `var(--pittorica-space-${u})`), y && (t.paddingLeft = `var(--pittorica-space-${y})`), n && (t.paddingLeft = `var(--pittorica-space-${n})`, t.paddingRight = `var(--pittorica-space-${n})`), g && (t.paddingTop = `var(--pittorica-space-${g})`, t.paddingBottom = `var(--pittorica-space-${g})`);
  const w = {
    ...B,
    ...t
  };
  return /* @__PURE__ */ x(
    S,
    {
      className: z("pittorica-box", L),
      style: w,
      disabled: R,
      required: j,
      ...q,
      children: p
    }
  );
};
A.displayName = "Box";
const C = ({
  children: i,
  as: p,
  size: r = "3",
  className: a,
  ...o
}) => /* @__PURE__ */ x(
  A,
  {
    as: p || "section",
    className: k("pittorica-section", a),
    "data-size": r,
    ...o,
    children: i
  }
);
C.displayName = "Section";
export {
  C as Section
};
