// @pittorica/input-otp-react@0.26.3/dist/index.js
import { jsx as i, jsxs as I } from "react/jsx-runtime";
import a, { forwardRef as x, createElement as d } from "react";
import { OTPInput as k, OTPInputContext as C } from "input-otp";
function h(t) {
  var r, e, o = "";
  if (typeof t == "string" || typeof t == "number") o += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var n = t.length;
    for (r = 0; r < n; r++) t[r] && (e = h(t[r])) && (o && (o += " "), o += e);
  } else for (e in t) t[e] && (o && (o += " "), o += e);
  return o;
}
function u() {
  for (var t, r, e = 0, o = "", n = arguments.length; e < n; e++) (t = arguments[e]) && (r = h(t)) && (o && (o += " "), o += r);
  return o;
}
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var R = {
  outline: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  },
  filled: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "currentColor",
    stroke: "none"
  }
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const b = (t, r, e, o) => {
  const n = x(
    ({ color: s = "currentColor", size: p = 24, stroke: l = 2, title: c, className: y, children: m, ...N }, O) => d(
      "svg",
      {
        ref: O,
        ...R[t],
        width: p,
        height: p,
        className: ["tabler-icon", `tabler-icon-${r}`, y].join(" "),
        strokeWidth: l,
        stroke: s,
        ...N
      },
      [
        c && d("title", { key: "svg-title" }, c),
        ...o.map(([P, T]) => d(P, T)),
        ...Array.isArray(m) ? m : [m]
      ]
    )
  );
  return n.displayName = `${e}`, n;
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const j = [["path", { d: "M5 12l14 0", key: "svg-0" }]], A = b("outline", "minus", "Minus", j), f = a.forwardRef(
  ({ className: t, containerClassName: r, ...e }, o) => /* @__PURE__ */ i(
    k,
    {
      ref: o,
      containerClassName: u(
        "pittorica-input-otp-container",
        r
      ),
      className: u("pittorica-input-otp", t),
      ...e
    }
  )
);
f.displayName = "InputOTP.Root";
const v = a.forwardRef(({ className: t, ...r }, e) => /* @__PURE__ */ i(
  "div",
  {
    ref: e,
    className: u("pittorica-input-otp-group", t),
    ...r
  }
));
v.displayName = "InputOTP.Group";
const w = a.forwardRef(({ index: t, className: r, ...e }, o) => {
  const s = a.use(C).slots[t];
  if (!s) return null;
  const { char: p, hasFakeCaret: l, isActive: c } = s;
  return /* @__PURE__ */ I(
    "div",
    {
      ref: o,
      className: u("pittorica-input-otp-slot", r),
      "data-active": c,
      ...e,
      children: [
        p,
        l && /* @__PURE__ */ i("div", { className: "pittorica-input-otp-caret", children: /* @__PURE__ */ i("div", { className: "pittorica-input-otp-caret-inner" }) })
      ]
    }
  );
});
w.displayName = "InputOTP.Slot";
const g = a.forwardRef(({ ...t }, r) => /* @__PURE__ */ i(
  "div",
  {
    ref: r,
    role: "separator",
    className: "pittorica-input-otp-separator",
    ...t,
    children: /* @__PURE__ */ i(A, { size: 16 })
  }
));
g.displayName = "InputOTP.Separator";
const B = Object.assign(f, {
  Root: f,
  Group: v,
  Slot: w,
  Separator: g
});
export {
  B as InputOTP
};
