// @pittorica/box-react@0.26.3/dist/index.js
import { jsx as b } from "react/jsx-runtime";
function T(t) {
  var o, r, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var p = t.length;
    for (o = 0; o < p; o++) t[o] && (r = T(t[o])) && (a && (a += " "), a += r);
  } else for (r in t) t[r] && (a && (a += " "), a += r);
  return a;
}
function k() {
  for (var t, o, r = 0, a = "", p = arguments.length; r < p; r++) (t = arguments[r]) && (o = T(t)) && (a && (a += " "), a += o);
  return a;
}
const q = ({
  as: t,
  children: o,
  display: r,
  m: a,
  mt: p,
  mr: f,
  mb: g,
  ml: v,
  mx: c,
  my: s,
  p: d,
  pt: $,
  pr: m,
  pb: l,
  pl: u,
  px: n,
  py: e,
  width: h,
  height: y,
  position: B,
  style: L,
  className: R,
  disabled: j,
  required: x,
  ...A
}) => {
  const N = t || "div", i = {};
  r && (i.display = r), h && (i.width = h), y && (i.height = y), B && (i.position = B), a && (i.margin = `var(--pittorica-space-${a})`), p && (i.marginTop = `var(--pittorica-space-${p})`), f && (i.marginRight = `var(--pittorica-space-${f})`), g && (i.marginBottom = `var(--pittorica-space-${g})`), v && (i.marginLeft = `var(--pittorica-space-${v})`), c && (i.marginLeft = `var(--pittorica-space-${c})`, i.marginRight = `var(--pittorica-space-${c})`), s && (i.marginTop = `var(--pittorica-space-${s})`, i.marginBottom = `var(--pittorica-space-${s})`), d && (i.padding = `var(--pittorica-space-${d})`), $ && (i.paddingTop = `var(--pittorica-space-${$})`), m && (i.paddingRight = `var(--pittorica-space-${m})`), l && (i.paddingBottom = `var(--pittorica-space-${l})`), u && (i.paddingLeft = `var(--pittorica-space-${u})`), n && (i.paddingLeft = `var(--pittorica-space-${n})`, i.paddingRight = `var(--pittorica-space-${n})`), e && (i.paddingTop = `var(--pittorica-space-${e})`, i.paddingBottom = `var(--pittorica-space-${e})`);
  const S = {
    ...L,
    ...i
  };
  return /* @__PURE__ */ b(
    N,
    {
      className: k("pittorica-box", R),
      style: S,
      disabled: j,
      required: x,
      ...A,
      children: o
    }
  );
};
q.displayName = "Box";
export {
  q as Box
};
