// @pittorica/inset-react@0.26.3/dist/index.js
import { jsx as x } from "react/jsx-runtime";
function T(a) {
  var p, r, i = "";
  if (typeof a == "string" || typeof a == "number") i += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var e = a.length;
    for (p = 0; p < e; p++) a[p] && (r = T(a[p])) && (i && (i += " "), i += r);
  } else for (r in a) a[r] && (i && (i += " "), i += r);
  return i;
}
function k() {
  for (var a, p, r = 0, i = "", e = arguments.length; r < e; r++) (a = arguments[r]) && (p = T(a)) && (i && (i += " "), i += p);
  return i;
}
function N(a) {
  var p, r, i = "";
  if (typeof a == "string" || typeof a == "number") i += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var e = a.length;
    for (p = 0; p < e; p++) a[p] && (r = N(a[p])) && (i && (i += " "), i += r);
  } else for (r in a) a[r] && (i && (i += " "), i += r);
  return i;
}
function S() {
  for (var a, p, r = 0, i = "", e = arguments.length; r < e; r++) (a = arguments[r]) && (p = N(a)) && (i && (i += " "), i += p);
  return i;
}
const A = ({
  as: a,
  children: p,
  display: r,
  m: i,
  mt: e,
  mr: o,
  mb: s,
  ml: n,
  mx: c,
  my: d,
  p: l,
  pt: v,
  pr: f,
  pb: u,
  pl: y,
  px: g,
  py: m,
  width: $,
  height: h,
  position: b,
  style: B,
  className: L,
  disabled: R,
  required: j,
  ...q
}) => {
  const w = a || "div", t = {};
  r && (t.display = r), $ && (t.width = $), h && (t.height = h), b && (t.position = b), i && (t.margin = `var(--pittorica-space-${i})`), e && (t.marginTop = `var(--pittorica-space-${e})`), o && (t.marginRight = `var(--pittorica-space-${o})`), s && (t.marginBottom = `var(--pittorica-space-${s})`), n && (t.marginLeft = `var(--pittorica-space-${n})`), c && (t.marginLeft = `var(--pittorica-space-${c})`, t.marginRight = `var(--pittorica-space-${c})`), d && (t.marginTop = `var(--pittorica-space-${d})`, t.marginBottom = `var(--pittorica-space-${d})`), l && (t.padding = `var(--pittorica-space-${l})`), v && (t.paddingTop = `var(--pittorica-space-${v})`), f && (t.paddingRight = `var(--pittorica-space-${f})`), u && (t.paddingBottom = `var(--pittorica-space-${u})`), y && (t.paddingLeft = `var(--pittorica-space-${y})`), g && (t.paddingLeft = `var(--pittorica-space-${g})`, t.paddingRight = `var(--pittorica-space-${g})`), m && (t.paddingTop = `var(--pittorica-space-${m})`, t.paddingBottom = `var(--pittorica-space-${m})`);
  const I = {
    ...B,
    ...t
  };
  return /* @__PURE__ */ x(
    w,
    {
      className: S("pittorica-box", L),
      style: I,
      disabled: R,
      required: j,
      ...q,
      children: p
    }
  );
};
A.displayName = "Box";
const z = ({
  children: a,
  side: p = "all",
  clip: r = !0,
  className: i,
  style: e,
  as: o,
  ...s
}) => /* @__PURE__ */ x(
  A,
  {
    as: o || "div",
    className: k("pittorica-inset", `pittorica-inset--${p}`, i),
    style: {
      overflow: r ? "hidden" : "visible",
      ...e
    },
    ...s,
    children: a
  }
);
z.displayName = "Inset";
export {
  z as Inset
};
