// @pittorica/code-react@0.26.3/dist/index.js
import { jsx as t, jsxs as l } from "react/jsx-runtime";
import { useState as d, useEffect as L } from "react";
import { PrismLight as e } from "react-syntax-highlighter";
import j from "react-syntax-highlighter/dist/esm/styles/prism/vsc-dark-plus.js";
import T from "react-syntax-highlighter/dist/esm/languages/prism/bash.js";
import k from "react-syntax-highlighter/dist/esm/languages/prism/css.js";
import w from "react-syntax-highlighter/dist/esm/languages/prism/javascript.js";
import N from "react-syntax-highlighter/dist/esm/languages/prism/json.js";
import I from "react-syntax-highlighter/dist/esm/languages/prism/markup.js";
import P from "react-syntax-highlighter/dist/esm/languages/prism/typescript.js";
import { clsx as g } from "clsx";
import { IconClipboardCheck as F, IconClipboard as D } from "@tabler/icons-react";
import { Box as f } from "@pittorica/box-react";
import { Text as H } from "@pittorica/text-react";
e.registerLanguage("typescript", P);
e.registerLanguage("javascript", w);
e.registerLanguage("bash", T);
e.registerLanguage("json", N);
e.registerLanguage("css", k);
e.registerLanguage("html", I);
const $ = ({
  children: r,
  language: i = "typescript",
  showLineNumbers: u = !0,
  filename: h,
  className: a,
  style: n,
  as: y,
  ...v
}) => {
  const [s, c] = d(!1), [b, C] = d(!1);
  L(() => {
    C(!0);
  }, []);
  const o = String(r), p = !o.includes(`
`), x = (m) => {
    m.preventDefault(), m.stopPropagation(), typeof navigator < "u" && navigator.clipboard && navigator.clipboard.writeText(o).then(() => {
      c(!0), setTimeout(() => c(!1), 2e3);
    }).catch((z) => {
      console.error("Failed to copy text:", z);
    });
  }, S = y || (p ? "code" : "pre");
  return p ? /* @__PURE__ */ t(
    S,
    {
      className: g("pittorica-code-inline", a),
      style: n,
      ...v,
      children: r
    }
  ) : /* @__PURE__ */ l(f, { className: g("pittorica-code-container", a), style: n, children: [
    /* @__PURE__ */ l("div", { className: "pittorica-code-header", children: [
      /* @__PURE__ */ t(
        H,
        {
          weight: "bold",
          size: "2",
          style: {
            textTransform: "lowercase",
            opacity: 0.8
          },
          children: h || i
        }
      ),
      /* @__PURE__ */ t(
        "button",
        {
          type: "button",
          onClick: x,
          className: "pittorica-code-copy-button",
          "aria-label": s ? "Copied" : "Copy to clipboard",
          children: s ? /* @__PURE__ */ t(F, { size: 18, color: "var(--pittorica-green-9)" }) : /* @__PURE__ */ t(D, { size: 18 })
        }
      )
    ] }),
    /* @__PURE__ */ t(f, { style: { position: "relative" }, children: b ? /* @__PURE__ */ t(
      e,
      {
        language: i,
        style: j,
        showLineNumbers: u,
        PreTag: "div",
        lineNumberStyle: {
          minWidth: "2.5em",
          paddingRight: "1.5em",
          opacity: 0.3,
          textAlign: "right",
          userSelect: "none"
        },
        customStyle: {
          margin: 0,
          padding: "1.5rem",
          fontSize: "var(--pittorica-font-size-2)",
          backgroundColor: "transparent",
          lineHeight: "1.7",
          fontFamily: "var(--pittorica-font-mono)"
        },
        children: o.replace(/\n$/, "")
      }
    ) : /* @__PURE__ */ t(
      "div",
      {
        style: {
          padding: "1.5rem",
          fontSize: "var(--pittorica-font-size-2)",
          lineHeight: "1.7",
          fontFamily: "var(--pittorica-font-mono)",
          whiteSpace: "pre-wrap",
          opacity: 0.5
        },
        children: o.replace(/\n$/, "")
      }
    ) })
  ] });
};
$.displayName = "Code";
export {
  $ as Code
};
