// @pittorica/table-react@0.26.3/dist/index.js
import { jsx as p } from "react/jsx-runtime";
function x(t) {
  var r, i, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (r = 0; r < o; r++) t[r] && (i = x(t[r])) && (a && (a += " "), a += i);
  } else for (i in t) t[i] && (a && (a += " "), a += i);
  return a;
}
function c() {
  for (var t, r, i = 0, a = "", o = arguments.length; i < o; i++) (t = arguments[i]) && (r = x(t)) && (a && (a += " "), a += r);
  return a;
}
function C(t) {
  var r, i, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (r = 0; r < o; r++) t[r] && (i = C(t[r])) && (a && (a += " "), a += i);
  } else for (i in t) t[i] && (a && (a += " "), a += i);
  return a;
}
function z() {
  for (var t, r, i = 0, a = "", o = arguments.length; i < o; i++) (t = arguments[i]) && (r = C(t)) && (a && (a += " "), a += r);
  return a;
}
const s = ({
  as: t,
  children: r,
  display: i,
  m: a,
  mt: o,
  mr: g,
  mb: b,
  ml: f,
  mx: n,
  my: l,
  p: v,
  pt: y,
  pr: h,
  pb: u,
  pl: $,
  px: d,
  py: m,
  width: T,
  height: N,
  position: R,
  style: A,
  className: H,
  disabled: L,
  required: j,
  ...q
}) => {
  const k = t || "div", e = {};
  i && (e.display = i), T && (e.width = T), N && (e.height = N), R && (e.position = R), a && (e.margin = `var(--pittorica-space-${a})`), o && (e.marginTop = `var(--pittorica-space-${o})`), g && (e.marginRight = `var(--pittorica-space-${g})`), b && (e.marginBottom = `var(--pittorica-space-${b})`), f && (e.marginLeft = `var(--pittorica-space-${f})`), n && (e.marginLeft = `var(--pittorica-space-${n})`, e.marginRight = `var(--pittorica-space-${n})`), l && (e.marginTop = `var(--pittorica-space-${l})`, e.marginBottom = `var(--pittorica-space-${l})`), v && (e.padding = `var(--pittorica-space-${v})`), y && (e.paddingTop = `var(--pittorica-space-${y})`), h && (e.paddingRight = `var(--pittorica-space-${h})`), u && (e.paddingBottom = `var(--pittorica-space-${u})`), $ && (e.paddingLeft = `var(--pittorica-space-${$})`), d && (e.paddingLeft = `var(--pittorica-space-${d})`, e.paddingRight = `var(--pittorica-space-${d})`), m && (e.paddingTop = `var(--pittorica-space-${m})`, e.paddingBottom = `var(--pittorica-space-${m})`);
  const S = {
    ...A,
    ...e
  };
  return /* @__PURE__ */ p(
    k,
    {
      className: z("pittorica-box", H),
      style: S,
      disabled: L,
      required: j,
      ...q,
      children: r
    }
  );
};
s.displayName = "Box";
const D = ({
  children: t,
  className: r,
  as: i,
  ...a
}) => /* @__PURE__ */ p(
  s,
  {
    as: i || "div",
    className: c("pittorica-table-container", r),
    ...a,
    children: t
  }
), B = ({
  children: t,
  className: r,
  as: i,
  ...a
}) => /* @__PURE__ */ p(
  s,
  {
    as: i || "table",
    className: c("pittorica-table-root", r),
    ...a,
    children: t
  }
), E = ({
  children: t,
  className: r,
  as: i,
  ...a
}) => /* @__PURE__ */ p(
  s,
  {
    as: i || "thead",
    className: c("pittorica-table-header", r),
    ...a,
    children: t
  }
), F = ({
  children: t,
  className: r,
  as: i,
  ...a
}) => /* @__PURE__ */ p(
  s,
  {
    as: i || "tbody",
    className: c("pittorica-table-body", r),
    ...a,
    children: t
  }
), G = ({
  children: t,
  className: r,
  as: i,
  ...a
}) => /* @__PURE__ */ p(
  s,
  {
    as: i || "tr",
    className: c("pittorica-table-row", r),
    ...a,
    children: t
  }
), w = ({
  children: t,
  align: r = "left",
  className: i,
  as: a,
  ...o
}) => /* @__PURE__ */ p(
  s,
  {
    as: a || "td",
    "data-align": r,
    className: c("pittorica-table-cell", i),
    ...o,
    children: t
  }
), I = ({
  children: t,
  align: r = "left",
  className: i,
  as: a,
  ...o
}) => /* @__PURE__ */ p(
  s,
  {
    as: a || "th",
    "data-align": r,
    className: c("pittorica-table-column-header", i),
    ...o,
    children: t
  }
), K = {
  Container: D,
  Root: B,
  Header: E,
  Body: F,
  Row: G,
  Cell: w,
  ColumnHeader: I
};
B.displayName = "Table.Root";
w.displayName = "Table.Cell";
export {
  K as Table
};
