// @pittorica/checkbox-react@0.26.3/dist/index.js
import { jsx as $, jsxs as C } from "react/jsx-runtime";
import E, { forwardRef as D, createElement as B } from "react";
function L(i) {
  var a, e, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (a = 0; a < o; a++) i[a] && (e = L(i[a])) && (t && (t += " "), t += e);
  } else for (e in i) i[e] && (t && (t += " "), t += e);
  return t;
}
function G() {
  for (var i, a, e = 0, t = "", o = arguments.length; e < o; e++) (i = arguments[e]) && (a = L(i)) && (t && (t += " "), t += a);
  return t;
}
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var H = {
  outline: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  },
  filled: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "currentColor",
    stroke: "none"
  }
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const S = (i, a, e, t) => {
  const o = D(
    ({ color: p = "currentColor", size: s = 24, stroke: g = 2, title: n, className: l, children: c, ...m }, h) => B(
      "svg",
      {
        ref: h,
        ...H[i],
        width: s,
        height: s,
        className: ["tabler-icon", `tabler-icon-${a}`, l].join(" "),
        ...i === "filled" ? {
          fill: p
        } : {
          strokeWidth: g,
          stroke: p
        },
        ...m
      },
      [
        n && B("title", { key: "svg-title" }, n),
        ...t.map(([f, v]) => B(f, v)),
        ...Array.isArray(c) ? c : [c]
      ]
    )
  );
  return o.displayName = `${e}`, o;
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const J = [["path", { d: "M3 5a2 2 0 0 1 2 -2h14a2 2 0 0 1 2 2v14a2 2 0 0 1 -2 2h-14a2 2 0 0 1 -2 -2v-14", key: "svg-0" }]], K = S("outline", "square", "Square", J);
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const O = [["path", { d: "M18.333 2c1.96 0 3.56 1.537 3.662 3.472l.005 .195v12.666c0 1.96 -1.537 3.56 -3.472 3.662l-.195 .005h-12.666a3.667 3.667 0 0 1 -3.662 -3.472l-.005 -.195v-12.666c0 -1.96 1.537 -3.56 3.472 -3.662l.195 -.005h12.666zm-2.626 7.293a1 1 0 0 0 -1.414 0l-3.293 3.292l-1.293 -1.292l-.094 -.083a1 1 0 0 0 -1.32 1.497l2 2l.094 .083a1 1 0 0 0 1.32 -.083l4 -4l.083 -.094a1 1 0 0 0 -.083 -1.32z", key: "svg-0" }]], P = S("filled", "square-check-filled", "SquareCheckFilled", O);
function j(i) {
  var a, e, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (a = 0; a < o; a++) i[a] && (e = j(i[a])) && (t && (t += " "), t += e);
  } else for (e in i) i[e] && (t && (t += " "), t += e);
  return t;
}
function Q() {
  for (var i, a, e = 0, t = "", o = arguments.length; e < o; e++) (i = arguments[e]) && (a = j(i)) && (t && (t += " "), t += a);
  return t;
}
const W = ({
  as: i,
  children: a,
  display: e,
  m: t,
  mt: o,
  mr: p,
  mb: s,
  ml: g,
  mx: n,
  my: l,
  p: c,
  pt: m,
  pr: h,
  pb: f,
  pl: v,
  px: d,
  py: u,
  width: y,
  height: x,
  position: b,
  style: w,
  className: k,
  disabled: N,
  required: A,
  ...T
}) => {
  const q = i || "div", r = {};
  e && (r.display = e), y && (r.width = y), x && (r.height = x), b && (r.position = b), t && (r.margin = `var(--pittorica-space-${t})`), o && (r.marginTop = `var(--pittorica-space-${o})`), p && (r.marginRight = `var(--pittorica-space-${p})`), s && (r.marginBottom = `var(--pittorica-space-${s})`), g && (r.marginLeft = `var(--pittorica-space-${g})`), n && (r.marginLeft = `var(--pittorica-space-${n})`, r.marginRight = `var(--pittorica-space-${n})`), l && (r.marginTop = `var(--pittorica-space-${l})`, r.marginBottom = `var(--pittorica-space-${l})`), c && (r.padding = `var(--pittorica-space-${c})`), m && (r.paddingTop = `var(--pittorica-space-${m})`), h && (r.paddingRight = `var(--pittorica-space-${h})`), f && (r.paddingBottom = `var(--pittorica-space-${f})`), v && (r.paddingLeft = `var(--pittorica-space-${v})`), d && (r.paddingLeft = `var(--pittorica-space-${d})`, r.paddingRight = `var(--pittorica-space-${d})`), u && (r.paddingTop = `var(--pittorica-space-${u})`, r.paddingBottom = `var(--pittorica-space-${u})`);
  const z = {
    ...w,
    ...r
  };
  return /* @__PURE__ */ $(
    q,
    {
      className: Q("pittorica-box", k),
      style: z,
      disabled: N,
      required: A,
      ...T,
      children: a
    }
  );
};
W.displayName = "Box";
function _(i) {
  var a, e, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (a = 0; a < o; a++) i[a] && (e = _(i[a])) && (t && (t += " "), t += e);
  } else for (e in i) i[e] && (t && (t += " "), t += e);
  return t;
}
function R() {
  for (var i, a, e = 0, t = "", o = arguments.length; e < o; e++) (i = arguments[e]) && (a = _(i)) && (t && (t += " "), t += a);
  return t;
}
function I(i) {
  var a, e, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (a = 0; a < o; a++) i[a] && (e = I(i[a])) && (t && (t += " "), t += e);
  } else for (e in i) i[e] && (t && (t += " "), t += e);
  return t;
}
function U() {
  for (var i, a, e = 0, t = "", o = arguments.length; e < o; e++) (i = arguments[e]) && (a = I(i)) && (t && (t += " "), t += a);
  return t;
}
const F = ({
  as: i,
  children: a,
  display: e,
  m: t,
  mt: o,
  mr: p,
  mb: s,
  ml: g,
  mx: n,
  my: l,
  p: c,
  pt: m,
  pr: h,
  pb: f,
  pl: v,
  px: d,
  py: u,
  width: y,
  height: x,
  position: b,
  style: w,
  className: k,
  disabled: N,
  required: A,
  ...T
}) => {
  const q = i || "div", r = {};
  e && (r.display = e), y && (r.width = y), x && (r.height = x), b && (r.position = b), t && (r.margin = `var(--pittorica-space-${t})`), o && (r.marginTop = `var(--pittorica-space-${o})`), p && (r.marginRight = `var(--pittorica-space-${p})`), s && (r.marginBottom = `var(--pittorica-space-${s})`), g && (r.marginLeft = `var(--pittorica-space-${g})`), n && (r.marginLeft = `var(--pittorica-space-${n})`, r.marginRight = `var(--pittorica-space-${n})`), l && (r.marginTop = `var(--pittorica-space-${l})`, r.marginBottom = `var(--pittorica-space-${l})`), c && (r.padding = `var(--pittorica-space-${c})`), m && (r.paddingTop = `var(--pittorica-space-${m})`), h && (r.paddingRight = `var(--pittorica-space-${h})`), f && (r.paddingBottom = `var(--pittorica-space-${f})`), v && (r.paddingLeft = `var(--pittorica-space-${v})`), d && (r.paddingLeft = `var(--pittorica-space-${d})`, r.paddingRight = `var(--pittorica-space-${d})`), u && (r.paddingTop = `var(--pittorica-space-${u})`, r.paddingBottom = `var(--pittorica-space-${u})`);
  const z = {
    ...w,
    ...r
  };
  return /* @__PURE__ */ $(
    q,
    {
      className: U("pittorica-box", k),
      style: z,
      disabled: N,
      required: A,
      ...T,
      children: a
    }
  );
};
F.displayName = "Box";
const M = ({
  children: i,
  as: a,
  size: e,
  weight: t = "regular",
  align: o,
  truncate: p = !1,
  color: s,
  className: g,
  style: n,
  ...l
}) => {
  const c = a || "span", m = (s == null ? void 0 : s.startsWith("#")) || (s == null ? void 0 : s.startsWith("rgb")) || (s == null ? void 0 : s.startsWith("hsl")), h = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), f = (() => {
    if (s)
      return s === "inherit" ? "inherit" : m ? s : h.has(s) ? `var(--pittorica-${s}-9)` : s;
  })(), v = e ? typeof e == "string" ? `pittorica-text--size-${e}` : R({
    [`pittorica-text--size-${e.initial}`]: e.initial,
    [`pittorica-text--xs-size-${e.xs}`]: e.xs,
    [`pittorica-text--sm-size-${e.sm}`]: e.sm,
    [`pittorica-text--md-size-${e.md}`]: e.md,
    [`pittorica-text--lg-size-${e.lg}`]: e.lg,
    [`pittorica-text--xl-size-${e.xl}`]: e.xl
  }) : null, d = {
    ...n,
    textAlign: o,
    color: f
  };
  return /* @__PURE__ */ $(
    F,
    {
      as: c,
      className: R(
        "pittorica-text",
        v,
        { "pittorica-text--truncate": p },
        g
      ),
      "data-weight": t,
      style: d,
      ...l,
      children: i
    }
  );
};
M.displayName = "Text";
const V = ({
  checked: i,
  defaultChecked: a,
  onChange: e,
  color: t = "source",
  disabled: o = !1,
  label: p,
  name: s,
  value: g,
  className: n,
  style: l,
  required: c = !1,
  as: m,
  ...h
}) => {
  const [f, v] = E.useState(
    a ?? !1
  ), d = i === void 0 ? f : i, y = t !== "inherit" && !(t != null && t.startsWith("#")) && !(t != null && t.startsWith("rgb")) ? `var(--pittorica-${t}-9)` : t, x = (w) => {
    if (o) return;
    const k = w.target.checked;
    v(k), e == null || e(k);
  };
  return /* @__PURE__ */ C(
    W,
    {
      as: m || "label",
      className: G("pittorica-checkbox-root", n),
      "data-disabled": o,
      style: { "--_checkbox-color": y, ...l },
      ...h,
      children: [
        /* @__PURE__ */ $(
          "input",
          {
            type: "checkbox",
            className: "pittorica-checkbox-input",
            name: s,
            value: g,
            checked: d,
            disabled: o,
            required: c,
            onChange: x
          }
        ),
        /* @__PURE__ */ $("div", { className: "pittorica-checkbox-control", "aria-hidden": "true", children: d ? /* @__PURE__ */ $(P, { size: 20 }) : /* @__PURE__ */ $(K, { size: 20, stroke: 1.5 }) }),
        p && /* @__PURE__ */ C(M, { weight: "medium", style: { color: "inherit" }, children: [
          p,
          " ",
          c && /* @__PURE__ */ $("span", { "aria-hidden": "true", children: "*" })
        ] })
      ]
    }
  );
};
V.displayName = "Checkbox";
export {
  V as Checkbox
};
