// @pittorica/flex-react@0.25.0/dist/index.js
import { jsx as j } from "react/jsx-runtime";
function A(i) {
  var a, r, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var e = i.length;
    for (a = 0; a < e; a++) i[a] && (r = A(i[a])) && (t && (t += " "), t += r);
  } else for (r in i) i[r] && (t && (t += " "), t += r);
  return t;
}
function O() {
  for (var i, a, r = 0, t = "", e = arguments.length; r < e; r++) (i = arguments[r]) && (a = A(i)) && (t && (t += " "), t += a);
  return t;
}
function R(i) {
  var a, r, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var e = i.length;
    for (a = 0; a < e; a++) i[a] && (r = R(i[a])) && (t && (t += " "), t += r);
  } else for (r in i) i[r] && (t && (t += " "), t += r);
  return t;
}
function S() {
  for (var i, a, r = 0, t = "", e = arguments.length; r < e; r++) (i = arguments[r]) && (a = R(i)) && (t && (t += " "), t += a);
  return t;
}
const B = ({
  as: i,
  children: a,
  display: r,
  m: t,
  mt: e,
  mr: g,
  mb: o,
  ml: d,
  mx: s,
  my: n,
  p: l,
  pt: m,
  pr: $,
  pb: c,
  pl: h,
  px: y,
  py: v,
  width: x,
  height: b,
  position: T,
  style: F,
  className: L,
  disabled: N,
  ...w
}) => {
  const C = i || "div", p = {};
  r && (p.display = r), x && (p.width = x), b && (p.height = b), T && (p.position = T), t && (p.margin = `var(--pittorica-space-${t})`), e && (p.marginTop = `var(--pittorica-space-${e})`), g && (p.marginRight = `var(--pittorica-space-${g})`), o && (p.marginBottom = `var(--pittorica-space-${o})`), d && (p.marginLeft = `var(--pittorica-space-${d})`), s && (p.marginLeft = `var(--pittorica-space-${s})`, p.marginRight = `var(--pittorica-space-${s})`), n && (p.marginTop = `var(--pittorica-space-${n})`, p.marginBottom = `var(--pittorica-space-${n})`), l && (p.padding = `var(--pittorica-space-${l})`), m && (p.paddingTop = `var(--pittorica-space-${m})`), $ && (p.paddingRight = `var(--pittorica-space-${$})`), c && (p.paddingBottom = `var(--pittorica-space-${c})`), h && (p.paddingLeft = `var(--pittorica-space-${h})`), y && (p.paddingLeft = `var(--pittorica-space-${y})`, p.paddingRight = `var(--pittorica-space-${y})`), v && (p.paddingTop = `var(--pittorica-space-${v})`, p.paddingBottom = `var(--pittorica-space-${v})`);
  const k = {
    ...F,
    ...p
  };
  return /* @__PURE__ */ j(
    C,
    {
      className: S("pittorica-box", L),
      style: k,
      disabled: N,
      ...w,
      children: a
    }
  );
};
B.displayName = "Box";
const u = (i) => typeof i == "string" && i.startsWith("auto-");
function f(i, a) {
  return !a || u(a) ? [] : typeof a == "string" ? [`pittorica-flex--${i}-${a}`] : Object.entries(a).filter(([, r]) => r !== void 0 && !u(r)).map(
    ([r, t]) => r === "initial" ? `pittorica-flex--${i}-${t}` : `pittorica-flex--${r}-${i}-${t}`
  );
}
const W = ({
  children: i,
  direction: a,
  align: r,
  justify: t,
  wrap: e,
  gap: g,
  basis: o,
  className: d,
  style: s,
  as: n,
  ...l
}) => {
  const m = n || "div", $ = [
    ...f("direction", a),
    ...f("align", r),
    ...f("justify", t),
    ...f("wrap", e),
    ...f("gap", g)
  ], c = { ...s };
  return u(o) && (c["--pittorica-flex-basis"] = o.split("-")[1]), /* @__PURE__ */ j(
    B,
    {
      as: m,
      className: O("pittorica-flex", $, d),
      style: c,
      ...l,
      children: i
    }
  );
};
W.displayName = "Flex";
export {
  W as Flex
};
