// @pittorica/dialog-react@0.26.3/dist/index.js
import { jsx as q, jsxs as ht, Fragment as gt } from "react/jsx-runtime";
import * as ir from "react";
import X, { useState as fe, createContext as _r, forwardRef as Rr, useRef as we, useCallback as pe, useEffect as Se, useMemo as Et, Fragment as Tt, PureComponent as xt, useId as ur, useLayoutEffect as Ot, use as wt } from "react";
import { createPortal as St } from "react-dom";
function Pt(e, r) {
  if (e == null) return {};
  var t = {};
  for (var n in e) if ({}.hasOwnProperty.call(e, n)) {
    if (r.indexOf(n) !== -1) continue;
    t[n] = e[n];
  }
  return t;
}
function me() {
  return me = Object.assign ? Object.assign.bind() : function(e) {
    for (var r = 1; r < arguments.length; r++) {
      var t = arguments[r];
      for (var n in t) ({}).hasOwnProperty.call(t, n) && (e[n] = t[n]);
    }
    return e;
  }, me.apply(null, arguments);
}
function _t(e) {
  return e && e.__esModule && Object.prototype.hasOwnProperty.call(e, "default") ? e.default : e;
}
var xe = { exports: {} }, Oe = { exports: {} }, N = {};
/** @license React v16.13.1
 * react-is.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var cr;
function Rt() {
  if (cr) return N;
  cr = 1;
  var e = typeof Symbol == "function" && Symbol.for, r = e ? Symbol.for("react.element") : 60103, t = e ? Symbol.for("react.portal") : 60106, n = e ? Symbol.for("react.fragment") : 60107, o = e ? Symbol.for("react.strict_mode") : 60108, a = e ? Symbol.for("react.profiler") : 60114, i = e ? Symbol.for("react.provider") : 60109, u = e ? Symbol.for("react.context") : 60110, m = e ? Symbol.for("react.async_mode") : 60111, f = e ? Symbol.for("react.concurrent_mode") : 60111, v = e ? Symbol.for("react.forward_ref") : 60112, d = e ? Symbol.for("react.suspense") : 60113, p = e ? Symbol.for("react.suspense_list") : 60120, g = e ? Symbol.for("react.memo") : 60115, E = e ? Symbol.for("react.lazy") : 60116, b = e ? Symbol.for("react.block") : 60121, w = e ? Symbol.for("react.fundamental") : 60117, $ = e ? Symbol.for("react.responder") : 60118, O = e ? Symbol.for("react.scope") : 60119;
  function T(s) {
    if (typeof s == "object" && s !== null) {
      var k = s.$$typeof;
      switch (k) {
        case r:
          switch (s = s.type, s) {
            case m:
            case f:
            case n:
            case a:
            case o:
            case d:
              return s;
            default:
              switch (s = s && s.$$typeof, s) {
                case u:
                case v:
                case E:
                case g:
                case i:
                  return s;
                default:
                  return k;
              }
          }
        case t:
          return k;
      }
    }
  }
  function F(s) {
    return T(s) === f;
  }
  return N.AsyncMode = m, N.ConcurrentMode = f, N.ContextConsumer = u, N.ContextProvider = i, N.Element = r, N.ForwardRef = v, N.Fragment = n, N.Lazy = E, N.Memo = g, N.Portal = t, N.Profiler = a, N.StrictMode = o, N.Suspense = d, N.isAsyncMode = function(s) {
    return F(s) || T(s) === m;
  }, N.isConcurrentMode = F, N.isContextConsumer = function(s) {
    return T(s) === u;
  }, N.isContextProvider = function(s) {
    return T(s) === i;
  }, N.isElement = function(s) {
    return typeof s == "object" && s !== null && s.$$typeof === r;
  }, N.isForwardRef = function(s) {
    return T(s) === v;
  }, N.isFragment = function(s) {
    return T(s) === n;
  }, N.isLazy = function(s) {
    return T(s) === E;
  }, N.isMemo = function(s) {
    return T(s) === g;
  }, N.isPortal = function(s) {
    return T(s) === t;
  }, N.isProfiler = function(s) {
    return T(s) === a;
  }, N.isStrictMode = function(s) {
    return T(s) === o;
  }, N.isSuspense = function(s) {
    return T(s) === d;
  }, N.isValidElementType = function(s) {
    return typeof s == "string" || typeof s == "function" || s === n || s === f || s === a || s === o || s === d || s === p || typeof s == "object" && s !== null && (s.$$typeof === E || s.$$typeof === g || s.$$typeof === i || s.$$typeof === u || s.$$typeof === v || s.$$typeof === w || s.$$typeof === $ || s.$$typeof === O || s.$$typeof === b);
  }, N.typeOf = T, N;
}
var D = {};
/** @license React v16.13.1
 * react-is.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var sr;
function Ft() {
  return sr || (sr = 1, process.env.NODE_ENV !== "production" && (function() {
    var e = typeof Symbol == "function" && Symbol.for, r = e ? Symbol.for("react.element") : 60103, t = e ? Symbol.for("react.portal") : 60106, n = e ? Symbol.for("react.fragment") : 60107, o = e ? Symbol.for("react.strict_mode") : 60108, a = e ? Symbol.for("react.profiler") : 60114, i = e ? Symbol.for("react.provider") : 60109, u = e ? Symbol.for("react.context") : 60110, m = e ? Symbol.for("react.async_mode") : 60111, f = e ? Symbol.for("react.concurrent_mode") : 60111, v = e ? Symbol.for("react.forward_ref") : 60112, d = e ? Symbol.for("react.suspense") : 60113, p = e ? Symbol.for("react.suspense_list") : 60120, g = e ? Symbol.for("react.memo") : 60115, E = e ? Symbol.for("react.lazy") : 60116, b = e ? Symbol.for("react.block") : 60121, w = e ? Symbol.for("react.fundamental") : 60117, $ = e ? Symbol.for("react.responder") : 60118, O = e ? Symbol.for("react.scope") : 60119;
    function T(l) {
      return typeof l == "string" || typeof l == "function" || // Note: its typeof might be other than 'symbol' or 'number' if it's a polyfill.
      l === n || l === f || l === a || l === o || l === d || l === p || typeof l == "object" && l !== null && (l.$$typeof === E || l.$$typeof === g || l.$$typeof === i || l.$$typeof === u || l.$$typeof === v || l.$$typeof === w || l.$$typeof === $ || l.$$typeof === O || l.$$typeof === b);
    }
    function F(l) {
      if (typeof l == "object" && l !== null) {
        var U = l.$$typeof;
        switch (U) {
          case r:
            var ue = l.type;
            switch (ue) {
              case m:
              case f:
              case n:
              case a:
              case o:
              case d:
                return ue;
              default:
                var Te = ue && ue.$$typeof;
                switch (Te) {
                  case u:
                  case v:
                  case E:
                  case g:
                  case i:
                    return Te;
                  default:
                    return U;
                }
            }
          case t:
            return U;
        }
      }
    }
    var s = m, k = f, Y = u, z = i, H = r, h = v, V = n, ne = E, Z = g, J = t, oe = a, B = o, K = d, ae = !1;
    function ie(l) {
      return ae || (ae = !0, console.warn("The ReactIs.isAsyncMode() alias has been deprecated, and will be removed in React 17+. Update your code to use ReactIs.isConcurrentMode() instead. It has the exact same API.")), c(l) || F(l) === m;
    }
    function c(l) {
      return F(l) === f;
    }
    function y(l) {
      return F(l) === u;
    }
    function P(l) {
      return F(l) === i;
    }
    function S(l) {
      return typeof l == "object" && l !== null && l.$$typeof === r;
    }
    function x(l) {
      return F(l) === v;
    }
    function A(l) {
      return F(l) === n;
    }
    function _(l) {
      return F(l) === E;
    }
    function R(l) {
      return F(l) === g;
    }
    function I(l) {
      return F(l) === t;
    }
    function M(l) {
      return F(l) === a;
    }
    function C(l) {
      return F(l) === o;
    }
    function L(l) {
      return F(l) === d;
    }
    D.AsyncMode = s, D.ConcurrentMode = k, D.ContextConsumer = Y, D.ContextProvider = z, D.Element = H, D.ForwardRef = h, D.Fragment = V, D.Lazy = ne, D.Memo = Z, D.Portal = J, D.Profiler = oe, D.StrictMode = B, D.Suspense = K, D.isAsyncMode = ie, D.isConcurrentMode = c, D.isContextConsumer = y, D.isContextProvider = P, D.isElement = S, D.isForwardRef = x, D.isFragment = A, D.isLazy = _, D.isMemo = R, D.isPortal = I, D.isProfiler = M, D.isStrictMode = C, D.isSuspense = L, D.isValidElementType = T, D.typeOf = F;
  })()), D;
}
var fr;
function Fr() {
  return fr || (fr = 1, process.env.NODE_ENV === "production" ? Oe.exports = Rt() : Oe.exports = Ft()), Oe.exports;
}
/*
object-assign
(c) Sindre Sorhus
@license MIT
*/
var Re, dr;
function At() {
  if (dr) return Re;
  dr = 1;
  var e = Object.getOwnPropertySymbols, r = Object.prototype.hasOwnProperty, t = Object.prototype.propertyIsEnumerable;
  function n(a) {
    if (a == null)
      throw new TypeError("Object.assign cannot be called with null or undefined");
    return Object(a);
  }
  function o() {
    try {
      if (!Object.assign)
        return !1;
      var a = new String("abc");
      if (a[5] = "de", Object.getOwnPropertyNames(a)[0] === "5")
        return !1;
      for (var i = {}, u = 0; u < 10; u++)
        i["_" + String.fromCharCode(u)] = u;
      var m = Object.getOwnPropertyNames(i).map(function(v) {
        return i[v];
      });
      if (m.join("") !== "0123456789")
        return !1;
      var f = {};
      return "abcdefghijklmnopqrst".split("").forEach(function(v) {
        f[v] = v;
      }), Object.keys(Object.assign({}, f)).join("") === "abcdefghijklmnopqrst";
    } catch {
      return !1;
    }
  }
  return Re = o() ? Object.assign : function(a, i) {
    for (var u, m = n(a), f, v = 1; v < arguments.length; v++) {
      u = Object(arguments[v]);
      for (var d in u)
        r.call(u, d) && (m[d] = u[d]);
      if (e) {
        f = e(u);
        for (var p = 0; p < f.length; p++)
          t.call(u, f[p]) && (m[f[p]] = u[f[p]]);
      }
    }
    return m;
  }, Re;
}
var Fe, lr;
function ze() {
  if (lr) return Fe;
  lr = 1;
  var e = "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED";
  return Fe = e, Fe;
}
var Ae, vr;
function Ar() {
  return vr || (vr = 1, Ae = Function.call.bind(Object.prototype.hasOwnProperty)), Ae;
}
var Ie, pr;
function It() {
  if (pr) return Ie;
  pr = 1;
  var e = function() {
  };
  if (process.env.NODE_ENV !== "production") {
    var r = /* @__PURE__ */ ze(), t = {}, n = /* @__PURE__ */ Ar();
    e = function(a) {
      var i = "Warning: " + a;
      typeof console < "u" && console.error(i);
      try {
        throw new Error(i);
      } catch {
      }
    };
  }
  function o(a, i, u, m, f) {
    if (process.env.NODE_ENV !== "production") {
      for (var v in a)
        if (n(a, v)) {
          var d;
          try {
            if (typeof a[v] != "function") {
              var p = Error(
                (m || "React class") + ": " + u + " type `" + v + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof a[v] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`."
              );
              throw p.name = "Invariant Violation", p;
            }
            d = a[v](i, v, m, u, null, r);
          } catch (E) {
            d = E;
          }
          if (d && !(d instanceof Error) && e(
            (m || "React class") + ": type specification of " + u + " `" + v + "` is invalid; the type checker function must return `null` or an `Error` but returned a " + typeof d + ". You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument)."
          ), d instanceof Error && !(d.message in t)) {
            t[d.message] = !0;
            var g = f ? f() : "";
            e(
              "Failed " + u + " type: " + d.message + (g ?? "")
            );
          }
        }
    }
  }
  return o.resetWarningCache = function() {
    process.env.NODE_ENV !== "production" && (t = {});
  }, Ie = o, Ie;
}
var Ce, yr;
function Ct() {
  if (yr) return Ce;
  yr = 1;
  var e = Fr(), r = At(), t = /* @__PURE__ */ ze(), n = /* @__PURE__ */ Ar(), o = /* @__PURE__ */ It(), a = function() {
  };
  process.env.NODE_ENV !== "production" && (a = function(u) {
    var m = "Warning: " + u;
    typeof console < "u" && console.error(m);
    try {
      throw new Error(m);
    } catch {
    }
  });
  function i() {
    return null;
  }
  return Ce = function(u, m) {
    var f = typeof Symbol == "function" && Symbol.iterator, v = "@@iterator";
    function d(c) {
      var y = c && (f && c[f] || c[v]);
      if (typeof y == "function")
        return y;
    }
    var p = "<<anonymous>>", g = {
      array: $("array"),
      bigint: $("bigint"),
      bool: $("boolean"),
      func: $("function"),
      number: $("number"),
      object: $("object"),
      string: $("string"),
      symbol: $("symbol"),
      any: O(),
      arrayOf: T,
      element: F(),
      elementType: s(),
      instanceOf: k,
      node: h(),
      objectOf: z,
      oneOf: Y,
      oneOfType: H,
      shape: ne,
      exact: Z
    };
    function E(c, y) {
      return c === y ? c !== 0 || 1 / c === 1 / y : c !== c && y !== y;
    }
    function b(c, y) {
      this.message = c, this.data = y && typeof y == "object" ? y : {}, this.stack = "";
    }
    b.prototype = Error.prototype;
    function w(c) {
      if (process.env.NODE_ENV !== "production")
        var y = {}, P = 0;
      function S(A, _, R, I, M, C, L) {
        if (I = I || p, C = C || R, L !== t) {
          if (m) {
            var l = new Error(
              "Calling PropTypes validators directly is not supported by the `prop-types` package. Use `PropTypes.checkPropTypes()` to call them. Read more at http://fb.me/use-check-prop-types"
            );
            throw l.name = "Invariant Violation", l;
          } else if (process.env.NODE_ENV !== "production" && typeof console < "u") {
            var U = I + ":" + R;
            !y[U] && // Avoid spamming the console because they are often not actionable except for lib authors
            P < 3 && (a(
              "You are manually calling a React.PropTypes validation function for the `" + C + "` prop on `" + I + "`. This is deprecated and will throw in the standalone `prop-types` package. You may be seeing this warning due to a third-party PropTypes library. See https://fb.me/react-warning-dont-call-proptypes for details."
            ), y[U] = !0, P++);
          }
        }
        return _[R] == null ? A ? _[R] === null ? new b("The " + M + " `" + C + "` is marked as required " + ("in `" + I + "`, but its value is `null`.")) : new b("The " + M + " `" + C + "` is marked as required in " + ("`" + I + "`, but its value is `undefined`.")) : null : c(_, R, I, M, C);
      }
      var x = S.bind(null, !1);
      return x.isRequired = S.bind(null, !0), x;
    }
    function $(c) {
      function y(P, S, x, A, _, R) {
        var I = P[S], M = B(I);
        if (M !== c) {
          var C = K(I);
          return new b(
            "Invalid " + A + " `" + _ + "` of type " + ("`" + C + "` supplied to `" + x + "`, expected ") + ("`" + c + "`."),
            { expectedType: c }
          );
        }
        return null;
      }
      return w(y);
    }
    function O() {
      return w(i);
    }
    function T(c) {
      function y(P, S, x, A, _) {
        if (typeof c != "function")
          return new b("Property `" + _ + "` of component `" + x + "` has invalid PropType notation inside arrayOf.");
        var R = P[S];
        if (!Array.isArray(R)) {
          var I = B(R);
          return new b("Invalid " + A + " `" + _ + "` of type " + ("`" + I + "` supplied to `" + x + "`, expected an array."));
        }
        for (var M = 0; M < R.length; M++) {
          var C = c(R, M, x, A, _ + "[" + M + "]", t);
          if (C instanceof Error)
            return C;
        }
        return null;
      }
      return w(y);
    }
    function F() {
      function c(y, P, S, x, A) {
        var _ = y[P];
        if (!u(_)) {
          var R = B(_);
          return new b("Invalid " + x + " `" + A + "` of type " + ("`" + R + "` supplied to `" + S + "`, expected a single ReactElement."));
        }
        return null;
      }
      return w(c);
    }
    function s() {
      function c(y, P, S, x, A) {
        var _ = y[P];
        if (!e.isValidElementType(_)) {
          var R = B(_);
          return new b("Invalid " + x + " `" + A + "` of type " + ("`" + R + "` supplied to `" + S + "`, expected a single ReactElement type."));
        }
        return null;
      }
      return w(c);
    }
    function k(c) {
      function y(P, S, x, A, _) {
        if (!(P[S] instanceof c)) {
          var R = c.name || p, I = ie(P[S]);
          return new b("Invalid " + A + " `" + _ + "` of type " + ("`" + I + "` supplied to `" + x + "`, expected ") + ("instance of `" + R + "`."));
        }
        return null;
      }
      return w(y);
    }
    function Y(c) {
      if (!Array.isArray(c))
        return process.env.NODE_ENV !== "production" && (arguments.length > 1 ? a(
          "Invalid arguments supplied to oneOf, expected an array, got " + arguments.length + " arguments. A common mistake is to write oneOf(x, y, z) instead of oneOf([x, y, z])."
        ) : a("Invalid argument supplied to oneOf, expected an array.")), i;
      function y(P, S, x, A, _) {
        for (var R = P[S], I = 0; I < c.length; I++)
          if (E(R, c[I]))
            return null;
        var M = JSON.stringify(c, function(L, l) {
          var U = K(l);
          return U === "symbol" ? String(l) : l;
        });
        return new b("Invalid " + A + " `" + _ + "` of value `" + String(R) + "` " + ("supplied to `" + x + "`, expected one of " + M + "."));
      }
      return w(y);
    }
    function z(c) {
      function y(P, S, x, A, _) {
        if (typeof c != "function")
          return new b("Property `" + _ + "` of component `" + x + "` has invalid PropType notation inside objectOf.");
        var R = P[S], I = B(R);
        if (I !== "object")
          return new b("Invalid " + A + " `" + _ + "` of type " + ("`" + I + "` supplied to `" + x + "`, expected an object."));
        for (var M in R)
          if (n(R, M)) {
            var C = c(R, M, x, A, _ + "." + M, t);
            if (C instanceof Error)
              return C;
          }
        return null;
      }
      return w(y);
    }
    function H(c) {
      if (!Array.isArray(c))
        return process.env.NODE_ENV !== "production" && a("Invalid argument supplied to oneOfType, expected an instance of array."), i;
      for (var y = 0; y < c.length; y++) {
        var P = c[y];
        if (typeof P != "function")
          return a(
            "Invalid argument supplied to oneOfType. Expected an array of check functions, but received " + ae(P) + " at index " + y + "."
          ), i;
      }
      function S(x, A, _, R, I) {
        for (var M = [], C = 0; C < c.length; C++) {
          var L = c[C], l = L(x, A, _, R, I, t);
          if (l == null)
            return null;
          l.data && n(l.data, "expectedType") && M.push(l.data.expectedType);
        }
        var U = M.length > 0 ? ", expected one of type [" + M.join(", ") + "]" : "";
        return new b("Invalid " + R + " `" + I + "` supplied to " + ("`" + _ + "`" + U + "."));
      }
      return w(S);
    }
    function h() {
      function c(y, P, S, x, A) {
        return J(y[P]) ? null : new b("Invalid " + x + " `" + A + "` supplied to " + ("`" + S + "`, expected a ReactNode."));
      }
      return w(c);
    }
    function V(c, y, P, S, x) {
      return new b(
        (c || "React class") + ": " + y + " type `" + P + "." + S + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + x + "`."
      );
    }
    function ne(c) {
      function y(P, S, x, A, _) {
        var R = P[S], I = B(R);
        if (I !== "object")
          return new b("Invalid " + A + " `" + _ + "` of type `" + I + "` " + ("supplied to `" + x + "`, expected `object`."));
        for (var M in c) {
          var C = c[M];
          if (typeof C != "function")
            return V(x, A, _, M, K(C));
          var L = C(R, M, x, A, _ + "." + M, t);
          if (L)
            return L;
        }
        return null;
      }
      return w(y);
    }
    function Z(c) {
      function y(P, S, x, A, _) {
        var R = P[S], I = B(R);
        if (I !== "object")
          return new b("Invalid " + A + " `" + _ + "` of type `" + I + "` " + ("supplied to `" + x + "`, expected `object`."));
        var M = r({}, P[S], c);
        for (var C in M) {
          var L = c[C];
          if (n(c, C) && typeof L != "function")
            return V(x, A, _, C, K(L));
          if (!L)
            return new b(
              "Invalid " + A + " `" + _ + "` key `" + C + "` supplied to `" + x + "`.\nBad object: " + JSON.stringify(P[S], null, "  ") + `
Valid keys: ` + JSON.stringify(Object.keys(c), null, "  ")
            );
          var l = L(R, C, x, A, _ + "." + C, t);
          if (l)
            return l;
        }
        return null;
      }
      return w(y);
    }
    function J(c) {
      switch (typeof c) {
        case "number":
        case "string":
        case "undefined":
          return !0;
        case "boolean":
          return !c;
        case "object":
          if (Array.isArray(c))
            return c.every(J);
          if (c === null || u(c))
            return !0;
          var y = d(c);
          if (y) {
            var P = y.call(c), S;
            if (y !== c.entries) {
              for (; !(S = P.next()).done; )
                if (!J(S.value))
                  return !1;
            } else
              for (; !(S = P.next()).done; ) {
                var x = S.value;
                if (x && !J(x[1]))
                  return !1;
              }
          } else
            return !1;
          return !0;
        default:
          return !1;
      }
    }
    function oe(c, y) {
      return c === "symbol" ? !0 : y ? y["@@toStringTag"] === "Symbol" || typeof Symbol == "function" && y instanceof Symbol : !1;
    }
    function B(c) {
      var y = typeof c;
      return Array.isArray(c) ? "array" : c instanceof RegExp ? "object" : oe(y, c) ? "symbol" : y;
    }
    function K(c) {
      if (typeof c > "u" || c === null)
        return "" + c;
      var y = B(c);
      if (y === "object") {
        if (c instanceof Date)
          return "date";
        if (c instanceof RegExp)
          return "regexp";
      }
      return y;
    }
    function ae(c) {
      var y = K(c);
      switch (y) {
        case "array":
        case "object":
          return "an " + y;
        case "boolean":
        case "date":
        case "regexp":
          return "a " + y;
        default:
          return y;
      }
    }
    function ie(c) {
      return !c.constructor || !c.constructor.name ? p : c.constructor.name;
    }
    return g.checkPropTypes = o, g.resetWarningCache = o.resetWarningCache, g.PropTypes = g, g;
  }, Ce;
}
var $e, mr;
function $t() {
  if (mr) return $e;
  mr = 1;
  var e = /* @__PURE__ */ ze();
  function r() {
  }
  function t() {
  }
  return t.resetWarningCache = r, $e = function() {
    function n(i, u, m, f, v, d) {
      if (d !== e) {
        var p = new Error(
          "Calling PropTypes validators directly is not supported by the `prop-types` package. Use PropTypes.checkPropTypes() to call them. Read more at http://fb.me/use-check-prop-types"
        );
        throw p.name = "Invariant Violation", p;
      }
    }
    n.isRequired = n;
    function o() {
      return n;
    }
    var a = {
      array: n,
      bigint: n,
      bool: n,
      func: n,
      number: n,
      object: n,
      string: n,
      symbol: n,
      any: n,
      arrayOf: o,
      element: n,
      elementType: n,
      instanceOf: o,
      node: n,
      objectOf: o,
      oneOf: o,
      oneOfType: o,
      shape: o,
      exact: o,
      checkPropTypes: t,
      resetWarningCache: r
    };
    return a.PropTypes = a, a;
  }, $e;
}
var br;
function Mt() {
  if (br) return xe.exports;
  if (br = 1, process.env.NODE_ENV !== "production") {
    var e = Fr(), r = !0;
    xe.exports = /* @__PURE__ */ Ct()(e.isElement, r);
  } else
    xe.exports = /* @__PURE__ */ $t()();
  return xe.exports;
}
var j = /* @__PURE__ */ Mt();
const Ir = /* @__PURE__ */ _t(j);
var Le = "data-focus-lock", Cr = "data-focus-lock-disabled", Nt = "data-no-focus-lock", Dt = "data-autofocus-inside", jt = "data-no-autofocus";
function Me(e, r) {
  return typeof e == "function" ? e(r) : e && (e.current = r), e;
}
function kt(e, r) {
  var t = fe(function() {
    return {
      // value
      value: e,
      // last callback
      callback: r,
      // "memoized" public interface
      facade: {
        get current() {
          return t.value;
        },
        set current(n) {
          var o = t.value;
          o !== n && (t.value = n, t.callback(n, o));
        }
      }
    };
  })[0];
  return t.callback = r, t.facade;
}
var Lt = typeof window < "u" ? ir.useLayoutEffect : ir.useEffect, hr = /* @__PURE__ */ new WeakMap();
function Wt(e, r) {
  var t = kt(null, function(n) {
    return e.forEach(function(o) {
      return Me(o, n);
    });
  });
  return Lt(function() {
    var n = hr.get(t);
    if (n) {
      var o = new Set(n), a = new Set(e), i = t.current;
      o.forEach(function(u) {
        a.has(u) || Me(u, null);
      }), a.forEach(function(u) {
        o.has(u) || Me(u, i);
      });
    }
    hr.set(t, e);
  }, [e]), t;
}
var Ne = {
  width: "1px",
  height: "0px",
  padding: 0,
  overflow: "hidden",
  position: "fixed",
  top: "1px",
  left: "1px"
};
process.env.NODE_ENV !== "production" && Ir.node;
var We = function() {
  return We = Object.assign || function(r) {
    for (var t, n = 1, o = arguments.length; n < o; n++) {
      t = arguments[n];
      for (var a in t) Object.prototype.hasOwnProperty.call(t, a) && (r[a] = t[a]);
    }
    return r;
  }, We.apply(this, arguments);
};
function $r(e) {
  return e;
}
function Mr(e, r) {
  r === void 0 && (r = $r);
  var t = [], n = !1, o = {
    read: function() {
      if (n)
        throw new Error("Sidecar: could not `read` from an `assigned` medium. `read` could be used only with `useMedium`.");
      return t.length ? t[t.length - 1] : e;
    },
    useMedium: function(a) {
      var i = r(a, n);
      return t.push(i), function() {
        t = t.filter(function(u) {
          return u !== i;
        });
      };
    },
    assignSyncMedium: function(a) {
      for (n = !0; t.length; ) {
        var i = t;
        t = [], i.forEach(a);
      }
      t = {
        push: function(u) {
          return a(u);
        },
        filter: function() {
          return t;
        }
      };
    },
    assignMedium: function(a) {
      n = !0;
      var i = [];
      if (t.length) {
        var u = t;
        t = [], u.forEach(a), i = t;
      }
      var m = function() {
        var v = i;
        i = [], v.forEach(a);
      }, f = function() {
        return Promise.resolve().then(m);
      };
      f(), t = {
        push: function(v) {
          i.push(v), f();
        },
        filter: function(v) {
          return i = i.filter(v), t;
        }
      };
    }
  };
  return o;
}
function Ve(e, r) {
  return r === void 0 && (r = $r), Mr(e, r);
}
function qt(e) {
  e === void 0 && (e = {});
  var r = Mr(null);
  return r.options = We({ async: !0, ssr: !1 }, e), r;
}
var Nr = Ve({}, function(e) {
  var r = e.target, t = e.currentTarget;
  return {
    target: r,
    currentTarget: t
  };
}), Dr = Ve(), Bt = Ve(), Ut = qt({
  async: !0,
  ssr: typeof document < "u"
}), Yt = /* @__PURE__ */ _r(void 0), zt = [], Ge = /* @__PURE__ */ Rr(function(r, t) {
  var n, o = fe(), a = o[0], i = o[1], u = we(), m = we(!1), f = we(null), v = fe({}), d = v[1], p = r.children, g = r.disabled, E = g === void 0 ? !1 : g, b = r.noFocusGuards, w = b === void 0 ? !1 : b, $ = r.persistentFocus, O = $ === void 0 ? !1 : $, T = r.crossFrame, F = T === void 0 ? !0 : T, s = r.autoFocus, k = s === void 0 ? !0 : s, Y = r.allowTextSelection, z = r.group, H = r.className, h = r.whiteList, V = r.hasPositiveIndices, ne = r.shards, Z = ne === void 0 ? zt : ne, J = r.as, oe = J === void 0 ? "div" : J, B = r.lockProps, K = B === void 0 ? {} : B, ae = r.sideCar, ie = r.returnFocus, c = ie === void 0 ? !1 : ie, y = r.focusOptions, P = r.onActivation, S = r.onDeactivation, x = fe({}), A = x[0], _ = pe(function(Q) {
    var ce = Q.captureFocusRestore;
    if (!f.current) {
      var se, ee = (se = document) == null ? void 0 : se.activeElement;
      f.current = ee, ee !== document.body && (f.current = ce(ee));
    }
    u.current && P && P(u.current), m.current = !0, d();
  }, [P]), R = pe(function() {
    m.current = !1, S && S(u.current), d();
  }, [S]), I = pe(function(Q) {
    var ce = f.current;
    if (ce) {
      var se = (typeof ce == "function" ? ce() : ce) || document.body, ee = typeof c == "function" ? c(se) : c;
      if (ee) {
        var ar = typeof ee == "object" ? ee : void 0;
        f.current = null, Q ? Promise.resolve().then(function() {
          return se.focus(ar);
        }) : se.focus(ar);
      }
    }
  }, [c]), M = pe(function(Q) {
    m.current && Nr.useMedium(Q);
  }, []), C = Dr.useMedium, L = pe(function(Q) {
    u.current !== Q && (u.current = Q, i(Q));
  }, []);
  process.env.NODE_ENV !== "production" && (typeof Y < "u" && console.warn("React-Focus-Lock: allowTextSelection is deprecated and enabled by default"), Se(function() {
    !u.current && typeof oe != "string" && console.error("FocusLock: could not obtain ref to internal node");
  }, []));
  var l = me((n = {}, n[Cr] = E && "disabled", n[Le] = z, n), K), U = w !== !0, ue = U && w !== "tail", Te = Wt([t, L]), bt = Et(function() {
    return {
      observed: u,
      shards: Z,
      enabled: !E,
      get active() {
        return m.current;
      }
    };
  }, [E, m, Z, u]);
  return /* @__PURE__ */ X.createElement(Tt, null, U && [
    /* @__PURE__ */ X.createElement("div", {
      key: "guard-first",
      "data-focus-guard": !0,
      tabIndex: E ? -1 : 0,
      style: Ne
    }),
    V ? /* @__PURE__ */ X.createElement("div", {
      key: "guard-nearest",
      "data-focus-guard": !0,
      tabIndex: E ? -1 : 1,
      style: Ne
    }) : null
  ], !E && /* @__PURE__ */ X.createElement(ae, {
    id: A,
    sideCar: Ut,
    observed: a,
    disabled: E,
    persistentFocus: O,
    crossFrame: F,
    autoFocus: k,
    whiteList: h,
    shards: Z,
    onActivation: _,
    onDeactivation: R,
    returnFocus: I,
    focusOptions: y,
    noFocusGuards: w
  }), /* @__PURE__ */ X.createElement(oe, me({
    ref: Te
  }, l, {
    className: H,
    onBlur: C,
    onFocus: M
  }), /* @__PURE__ */ X.createElement(Yt.Provider, {
    value: bt
  }, p)), ue && /* @__PURE__ */ X.createElement("div", {
    "data-focus-guard": !0,
    tabIndex: E ? -1 : 0,
    style: Ne
  }));
});
Ge.propTypes = process.env.NODE_ENV !== "production" ? {
  children: j.node,
  disabled: j.bool,
  returnFocus: j.oneOfType([j.bool, j.object, j.func]),
  focusOptions: j.object,
  noFocusGuards: j.bool,
  hasPositiveIndices: j.bool,
  allowTextSelection: j.bool,
  autoFocus: j.bool,
  persistentFocus: j.bool,
  crossFrame: j.bool,
  group: j.string,
  className: j.string,
  whiteList: j.func,
  shards: j.arrayOf(j.any),
  as: j.oneOfType([j.string, j.func, j.object]),
  lockProps: j.object,
  onActivation: j.func,
  onDeactivation: j.func,
  sideCar: j.any.isRequired
} : {};
function qe(e, r) {
  return qe = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(t, n) {
    return t.__proto__ = n, t;
  }, qe(e, r);
}
function Vt(e, r) {
  e.prototype = Object.create(r.prototype), e.prototype.constructor = e, qe(e, r);
}
function be(e) {
  "@babel/helpers - typeof";
  return be = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(r) {
    return typeof r;
  } : function(r) {
    return r && typeof Symbol == "function" && r.constructor === Symbol && r !== Symbol.prototype ? "symbol" : typeof r;
  }, be(e);
}
function Gt(e, r) {
  if (be(e) != "object" || !e) return e;
  var t = e[Symbol.toPrimitive];
  if (t !== void 0) {
    var n = t.call(e, r);
    if (be(n) != "object") return n;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (r === "string" ? String : Number)(e);
}
function Ht(e) {
  var r = Gt(e, "string");
  return be(r) == "symbol" ? r : r + "";
}
function Jt(e, r, t) {
  return (r = Ht(r)) in e ? Object.defineProperty(e, r, {
    value: t,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : e[r] = t, e;
}
function Kt(e, r) {
  if (process.env.NODE_ENV !== "production") {
    if (typeof e != "function")
      throw new Error("Expected reducePropsToState to be a function.");
    if (typeof r != "function")
      throw new Error("Expected handleStateChangeOnClient to be a function.");
  }
  function t(n) {
    return n.displayName || n.name || "Component";
  }
  return function(o) {
    if (process.env.NODE_ENV !== "production" && typeof o != "function")
      throw new Error("Expected WrappedComponent to be a React component.");
    var a = [], i;
    function u() {
      i = e(a.map(function(f) {
        return f.props;
      })), r(i);
    }
    var m = /* @__PURE__ */ (function(f) {
      Vt(v, f);
      function v() {
        return f.apply(this, arguments) || this;
      }
      v.peek = function() {
        return i;
      };
      var d = v.prototype;
      return d.componentDidMount = function() {
        a.push(this), u();
      }, d.componentDidUpdate = function() {
        u();
      }, d.componentWillUnmount = function() {
        var g = a.indexOf(this);
        a.splice(g, 1), u();
      }, d.render = function() {
        return /* @__PURE__ */ X.createElement(o, this.props);
      }, v;
    })(xt);
    return Jt(m, "displayName", "SideEffect(" + t(o) + ")"), m;
  };
}
var G = function(e) {
  for (var r = Array(e.length), t = 0; t < e.length; ++t)
    r[t] = e[t];
  return r;
}, te = function(e) {
  return Array.isArray(e) ? e : [e];
}, jr = function(e) {
  return Array.isArray(e) ? e[0] : e;
}, Qt = function(e) {
  if (e.nodeType !== Node.ELEMENT_NODE)
    return !1;
  var r = window.getComputedStyle(e, null);
  return !r || !r.getPropertyValue ? !1 : r.getPropertyValue("display") === "none" || r.getPropertyValue("visibility") === "hidden";
}, kr = function(e) {
  return e.parentNode && e.parentNode.nodeType === Node.DOCUMENT_FRAGMENT_NODE ? (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    e.parentNode.host
  ) : e.parentNode;
}, Lr = function(e) {
  return e === document || e && e.nodeType === Node.DOCUMENT_NODE;
}, Xt = function(e) {
  return e.hasAttribute("inert");
}, Zt = function(e, r) {
  return !e || Lr(e) || !Qt(e) && !Xt(e) && r(kr(e));
}, Wr = function(e, r) {
  var t = e.get(r);
  if (t !== void 0)
    return t;
  var n = Zt(r, Wr.bind(void 0, e));
  return e.set(r, n), n;
}, en = function(e, r) {
  return e && !Lr(e) ? nn(e) ? r(kr(e)) : !1 : !0;
}, qr = function(e, r) {
  var t = e.get(r);
  if (t !== void 0)
    return t;
  var n = en(r, qr.bind(void 0, e));
  return e.set(r, n), n;
}, Br = function(e) {
  return e.dataset;
}, rn = function(e) {
  return e.tagName === "BUTTON";
}, Ur = function(e) {
  return e.tagName === "INPUT";
}, Yr = function(e) {
  return Ur(e) && e.type === "radio";
}, tn = function(e) {
  return !((Ur(e) || rn(e)) && (e.type === "hidden" || e.disabled));
}, nn = function(e) {
  var r = e.getAttribute(jt);
  return ![!0, "true", ""].includes(r);
}, He = function(e) {
  var r;
  return !!(e && (!((r = Br(e)) === null || r === void 0) && r.focusGuard));
}, Be = function(e) {
  return !He(e);
}, on = function(e) {
  return !!e;
}, an = function(e, r) {
  var t = Math.max(0, e.tabIndex), n = Math.max(0, r.tabIndex), o = t - n, a = e.index - r.index;
  if (o) {
    if (!t)
      return 1;
    if (!n)
      return -1;
  }
  return o || a;
}, un = function(e) {
  return e.tabIndex < 0 && !e.hasAttribute("tabindex") ? 0 : e.tabIndex;
}, Je = function(e, r, t) {
  return G(e).map(function(n, o) {
    var a = un(n);
    return {
      node: n,
      index: o,
      tabIndex: t && a === -1 ? (n.dataset || {}).focusGuard ? 0 : -1 : a
    };
  }).filter(function(n) {
    return !r || n.tabIndex >= 0;
  }).sort(an);
}, cn = [
  "button:enabled",
  "select:enabled",
  "textarea:enabled",
  "input:enabled",
  // elements with explicit roles will also use explicit tabindex
  // '[role="button"]',
  "a[href]",
  "area[href]",
  "summary",
  "iframe",
  "object",
  "embed",
  "audio[controls]",
  "video[controls]",
  "[tabindex]",
  "[contenteditable]",
  "[autofocus]"
], Ke = cn.join(","), sn = "".concat(Ke, ", [data-focus-guard]"), zr = function(e, r) {
  return G((e.shadowRoot || e).children).reduce(function(t, n) {
    return t.concat(n.matches(r ? sn : Ke) ? [n] : [], zr(n));
  }, []);
}, fn = function(e, r) {
  var t;
  return e instanceof HTMLIFrameElement && (!((t = e.contentDocument) === null || t === void 0) && t.body) ? ve([e.contentDocument.body], r) : [e];
}, ve = function(e, r) {
  return e.reduce(function(t, n) {
    var o, a = zr(n, r), i = (o = []).concat.apply(o, a.map(function(u) {
      return fn(u, r);
    }));
    return t.concat(
      // add all tabbables inside and within shadow DOMs in DOM order
      i,
      // add if node is tabbable itself
      n.parentNode ? G(n.parentNode.querySelectorAll(Ke)).filter(function(u) {
        return u === n;
      }) : []
    );
  }, []);
}, dn = function(e) {
  var r = e.querySelectorAll("[".concat(Dt, "]"));
  return G(r).map(function(t) {
    return ve([t]);
  }).reduce(function(t, n) {
    return t.concat(n);
  }, []);
}, Qe = function(e, r) {
  return G(e).filter(function(t) {
    return Wr(r, t);
  }).filter(function(t) {
    return tn(t);
  });
}, gr = function(e, r) {
  return r === void 0 && (r = /* @__PURE__ */ new Map()), G(e).filter(function(t) {
    return qr(r, t);
  });
}, Xe = function(e, r, t) {
  return Je(Qe(ve(e, t), r), !0, t);
}, he = function(e, r) {
  return Je(Qe(ve(e), r), !1);
}, ln = function(e, r) {
  return Qe(dn(e), r);
}, re = function(e, r) {
  return e.shadowRoot ? re(e.shadowRoot, r) : Object.getPrototypeOf(e).contains !== void 0 && Object.getPrototypeOf(e).contains.call(e, r) ? !0 : G(e.children).some(function(t) {
    var n;
    if (t instanceof HTMLIFrameElement) {
      var o = (n = t.contentDocument) === null || n === void 0 ? void 0 : n.body;
      return o ? re(o, r) : !1;
    }
    return re(t, r);
  });
}, vn = function(e) {
  for (var r = /* @__PURE__ */ new Set(), t = e.length, n = 0; n < t; n += 1)
    for (var o = n + 1; o < t; o += 1) {
      var a = e[n].compareDocumentPosition(e[o]);
      (a & Node.DOCUMENT_POSITION_CONTAINED_BY) > 0 && r.add(o), (a & Node.DOCUMENT_POSITION_CONTAINS) > 0 && r.add(n);
    }
  return e.filter(function(i, u) {
    return !r.has(u);
  });
}, Vr = function(e) {
  return e.parentNode ? Vr(e.parentNode) : e;
}, Ze = function(e) {
  var r = te(e);
  return r.filter(Boolean).reduce(function(t, n) {
    var o = n.getAttribute(Le);
    return t.push.apply(t, o ? vn(G(Vr(n).querySelectorAll("[".concat(Le, '="').concat(o, '"]:not([').concat(Cr, '="disabled"])')))) : [n]), t;
  }, []);
}, pn = function(e) {
  try {
    return e();
  } catch {
    return;
  }
}, ge = function(e) {
  if (e === void 0 && (e = document), !(!e || !e.activeElement)) {
    var r = e.activeElement;
    return r.shadowRoot ? ge(r.shadowRoot) : r instanceof HTMLIFrameElement && pn(function() {
      return r.contentWindow.document;
    }) ? ge(r.contentWindow.document) : r;
  }
}, yn = function(e, r) {
  return e === r;
}, mn = function(e, r) {
  return !!G(e.querySelectorAll("iframe")).some(function(t) {
    return yn(t, r);
  });
}, Gr = function(e, r) {
  return r === void 0 && (r = ge(jr(e).ownerDocument)), !r || r.dataset && r.dataset.focusGuard ? !1 : Ze(e).some(function(t) {
    return re(t, r) || mn(t, r);
  });
}, bn = function(e) {
  e === void 0 && (e = document);
  var r = ge(e);
  return r ? G(e.querySelectorAll("[".concat(Nt, "]"))).some(function(t) {
    return re(t, r);
  }) : !1;
}, hn = function(e, r) {
  return r.filter(Yr).filter(function(t) {
    return t.name === e.name;
  }).filter(function(t) {
    return t.checked;
  })[0] || e;
}, er = function(e, r) {
  return Yr(e) && e.name ? hn(e, r) : e;
}, gn = function(e) {
  var r = /* @__PURE__ */ new Set();
  return e.forEach(function(t) {
    return r.add(er(t, e));
  }), e.filter(function(t) {
    return r.has(t);
  });
}, Er = function(e) {
  return e[0] && e.length > 1 ? er(e[0], e) : e[0];
}, Tr = function(e, r) {
  return e.indexOf(er(r, e));
}, Ue = "NEW_FOCUS", En = function(e, r, t, n, o) {
  var a = e.length, i = e[0], u = e[a - 1], m = He(n);
  if (!(n && e.indexOf(n) >= 0)) {
    var f = n !== void 0 ? t.indexOf(n) : -1, v = o ? t.indexOf(o) : f, d = o ? e.indexOf(o) : -1;
    if (f === -1)
      return d !== -1 ? d : Ue;
    if (d === -1)
      return Ue;
    var p = f - v, g = t.indexOf(i), E = t.indexOf(u), b = gn(t), w = n !== void 0 ? b.indexOf(n) : -1, $ = o ? b.indexOf(o) : w, O = b.filter(function(z) {
      return z.tabIndex >= 0;
    }), T = n !== void 0 ? O.indexOf(n) : -1, F = o ? O.indexOf(o) : T, s = T >= 0 && F >= 0 ? (
      // old/new are tabbables, measure distance in tabbable space
      F - T
    ) : (
      // or else measure in focusable space
      $ - w
    );
    if (!p && d >= 0 || r.length === 0)
      return d;
    var k = Tr(e, r[0]), Y = Tr(e, r[r.length - 1]);
    if (f <= g && m && Math.abs(p) > 1)
      return Y;
    if (f >= E && m && Math.abs(p) > 1)
      return k;
    if (p && Math.abs(s) > 1)
      return d;
    if (f <= g)
      return Y;
    if (f > E)
      return k;
    if (p)
      return Math.abs(p) > 1 ? d : (a + d + p) % a;
  }
}, Tn = function(e) {
  return function(r) {
    var t, n = (t = Br(r)) === null || t === void 0 ? void 0 : t.autofocus;
    return (
      // @ts-expect-error
      r.autofocus || //
      n !== void 0 && n !== "false" || //
      e.indexOf(r) >= 0
    );
  };
}, xr = function(e, r, t) {
  var n = e.map(function(a) {
    var i = a.node;
    return i;
  }), o = gr(n.filter(Tn(t)));
  return o && o.length ? Er(o) : Er(gr(r));
}, Ye = function(e, r) {
  return r === void 0 && (r = []), r.push(e), e.parentNode && Ye(e.parentNode.host || e.parentNode, r), r;
}, De = function(e, r) {
  for (var t = Ye(e), n = Ye(r), o = 0; o < t.length; o += 1) {
    var a = t[o];
    if (n.indexOf(a) >= 0)
      return a;
  }
  return !1;
}, Hr = function(e, r, t) {
  var n = te(e), o = te(r), a = n[0], i = !1;
  return o.filter(Boolean).forEach(function(u) {
    i = De(i || u, u) || i, t.filter(Boolean).forEach(function(m) {
      var f = De(a, m);
      f && (!i || re(f, i) ? i = f : i = De(f, i));
    });
  }), i;
}, Or = function(e, r) {
  return e.reduce(function(t, n) {
    return t.concat(ln(n, r));
  }, []);
}, xn = function(e, r) {
  var t = /* @__PURE__ */ new Map();
  return r.forEach(function(n) {
    return t.set(n.node, n);
  }), e.map(function(n) {
    return t.get(n);
  }).filter(on);
}, On = function(e, r) {
  var t = ge(te(e).length > 0 ? document : jr(e).ownerDocument), n = Ze(e).filter(Be), o = Hr(t || e, e, n), a = /* @__PURE__ */ new Map(), i = he(n, a), u = i.filter(function(E) {
    var b = E.node;
    return Be(b);
  });
  if (u[0]) {
    var m = he([o], a).map(function(E) {
      var b = E.node;
      return b;
    }), f = xn(m, u), v = f.map(function(E) {
      var b = E.node;
      return b;
    }), d = f.filter(function(E) {
      var b = E.tabIndex;
      return b >= 0;
    }).map(function(E) {
      var b = E.node;
      return b;
    }), p = En(v, d, m, t, r);
    if (p === Ue) {
      var g = (
        // first try only tabbable, and the fallback to all focusable, as long as at least one element should be picked for focus
        xr(i, d, Or(n, a)) || xr(i, v, Or(n, a))
      );
      if (g)
        return { node: g };
      console.warn("focus-lock: cannot find any node to move focus into");
      return;
    }
    return p === void 0 ? p : f[p];
  }
}, wn = function(e) {
  var r = Ze(e).filter(Be), t = Hr(e, e, r), n = Je(ve([t], !0), !0, !0), o = ve(r, !1);
  return n.map(function(a) {
    var i = a.node, u = a.index;
    return {
      node: i,
      index: u,
      lockItem: o.indexOf(i) >= 0,
      guard: He(i)
    };
  });
}, rr = function(e, r) {
  e && ("focus" in e && e.focus(r), "contentWindow" in e && e.contentWindow && e.contentWindow.focus());
}, je = 0, ke = !1, Jr = function(e, r, t) {
  t === void 0 && (t = {});
  var n = On(e, r);
  if (!ke && n) {
    if (je > 2) {
      console.error("FocusLock: focus-fighting detected. Only one focus management system could be active. See https://github.com/theKashey/focus-lock/#focus-fighting"), ke = !0, setTimeout(function() {
        ke = !1;
      }, 1);
      return;
    }
    je++, rr(n.node, t.focusOptions), je--;
  }
};
function ye(e) {
  if (!e)
    return null;
  if (typeof WeakRef > "u")
    return function() {
      return e || null;
    };
  var r = e ? new WeakRef(e) : null;
  return function() {
    return (r == null ? void 0 : r.deref()) || null;
  };
}
var Sn = function(e) {
  if (!e)
    return null;
  for (var r = [], t = e; t && t !== document.body; )
    r.push({
      current: ye(t),
      parent: ye(t.parentElement),
      left: ye(t.previousElementSibling),
      right: ye(t.nextElementSibling)
    }), t = t.parentElement;
  return {
    element: ye(e),
    stack: r,
    ownerDocument: e.ownerDocument
  };
}, Pn = function(e) {
  var r, t, n, o, a;
  if (e)
    for (var i = e.stack, u = e.ownerDocument, m = /* @__PURE__ */ new Map(), f = 0, v = i; f < v.length; f++) {
      var d = v[f], p = (r = d.parent) === null || r === void 0 ? void 0 : r.call(d);
      if (p && u.contains(p)) {
        for (var g = (t = d.left) === null || t === void 0 ? void 0 : t.call(d), E = d.current(), b = p.contains(E) ? E : void 0, w = (n = d.right) === null || n === void 0 ? void 0 : n.call(d), $ = Xe([p], m), O = (
          // that is element itself
          (a = (o = b ?? // or something in it's place
          (g == null ? void 0 : g.nextElementSibling)) !== null && o !== void 0 ? o : (
            // or somebody to the right, still close enough
            w
          )) !== null && a !== void 0 ? a : (
            // or somebody to the left, something?
            g
          )
        ); O; ) {
          for (var T = 0, F = $; T < F.length; T++) {
            var s = F[T];
            if (O != null && O.contains(s.node))
              return s.node;
          }
          O = O.nextElementSibling;
        }
        if ($.length)
          return $[0].node;
      }
    }
}, Kr = function(e) {
  var r = Sn(e);
  return function() {
    return Pn(r);
  };
}, _n = function(e, r, t) {
  if (!e || !r)
    return console.error("no element or scope given"), {};
  var n = te(r);
  if (n.every(function(i) {
    return !re(i, e);
  }))
    return console.error("Active element is not contained in the scope"), {};
  var o = t ? Xe(n, /* @__PURE__ */ new Map()) : he(n, /* @__PURE__ */ new Map()), a = o.findIndex(function(i) {
    var u = i.node;
    return u === e;
  });
  if (a !== -1)
    return {
      prev: o[a - 1],
      next: o[a + 1],
      first: o[0],
      last: o[o.length - 1]
    };
}, Rn = function(e, r) {
  var t = r ? Xe(te(e), /* @__PURE__ */ new Map()) : he(te(e), /* @__PURE__ */ new Map());
  return {
    first: t[0],
    last: t[t.length - 1]
  };
}, Fn = function(e) {
  return Object.assign({
    scope: document.body,
    cycle: !0,
    onlyTabbable: !0
  }, e);
}, Qr = function(e, r, t) {
  r === void 0 && (r = {});
  var n = Fn(r), o = _n(e, n.scope, n.onlyTabbable);
  if (o) {
    var a = t(o, n.cycle);
    a && rr(a.node, n.focusOptions);
  }
}, An = function(e, r) {
  r === void 0 && (r = {}), Qr(e, r, function(t, n) {
    var o = t.next, a = t.first;
    return o || n && a;
  });
}, In = function(e, r) {
  r === void 0 && (r = {}), Qr(e, r, function(t, n) {
    var o = t.prev, a = t.last;
    return o || n && a;
  });
}, Xr = function(e, r, t) {
  var n, o = Rn(e, (n = r.onlyTabbable) !== null && n !== void 0 ? n : !0), a = o[t];
  a && rr(a.node, r.focusOptions);
}, Cn = function(e, r) {
  r === void 0 && (r = {}), Xr(e, r, "first");
}, $n = function(e, r) {
  r === void 0 && (r = {}), Xr(e, r, "last");
};
function tr(e) {
  setTimeout(e, 1);
}
var Mn = function(r) {
  return r && "current" in r ? r.current : r;
}, Zr = function() {
  return document && document.activeElement === document.body;
}, Nn = function() {
  return Zr() || bn();
}, de = null, W = null, wr = function() {
  return null;
}, le = null, Ee = !1, nr = !1, Dn = function() {
  return !0;
}, jn = function(r) {
  return (de.whiteList || Dn)(r);
}, kn = function(r, t) {
  le = {
    observerNode: r,
    portaledElement: t
  };
}, Ln = function(r) {
  return le && le.portaledElement === r;
};
function Sr(e, r, t, n) {
  var o = null, a = e;
  do {
    var i = n[a];
    if (i.guard)
      i.node.dataset.focusAutoGuard && (o = i);
    else if (i.lockItem) {
      if (a !== e)
        return;
      o = null;
    } else
      break;
  } while ((a += t) !== r);
  o && (o.node.tabIndex = 0);
}
var Wn = function(r) {
  return r ? !!Ee : Ee === "meanwhile";
}, qn = function e(r, t, n) {
  return t && (t.host === r && (!t.activeElement || n.contains(t.activeElement)) || t.parentNode && e(r, t.parentNode, n));
}, Bn = function(r, t) {
  return t.some(function(n) {
    return qn(r, n, n);
  });
}, et = function(r) {
  return he(r, /* @__PURE__ */ new Map());
}, Un = function(r) {
  return !et([r.parentNode]).some(function(t) {
    return t.node === r;
  });
}, Pe = function() {
  var r = !1;
  if (de) {
    var t = de, n = t.observed, o = t.persistentFocus, a = t.autoFocus, i = t.shards, u = t.crossFrame, m = t.focusOptions, f = t.noFocusGuards, v = n || le && le.portaledElement;
    if (Zr() && W && W !== document.body && (!document.body.contains(W) || Un(W))) {
      var d = wr();
      d && d.focus();
    }
    var p = document && document.activeElement;
    if (v) {
      var g = [v].concat(i.map(Mn).filter(Boolean)), E = function() {
        if (!Wn(u) || !f || !W || nr)
          return !1;
        var T = et(g), F = T.findIndex(function(s) {
          var k = s.node;
          return k === W;
        });
        return F === 0 || F === T.length - 1;
      };
      if ((!p || jn(p)) && (o || E() || !Nn() || !W && a) && (v && !(Gr(g) || p && Bn(p, g) || Ln(p)) && (document && !W && p && !a ? (p.blur && p.blur(), document.body.focus()) : (r = Jr(g, W, {
        focusOptions: m
      }), le = {})), W = document && document.activeElement, W !== document.body && (wr = Kr(W)), Ee = !1), document && p !== document.activeElement && document.querySelector("[data-focus-auto-guard]")) {
        var b = document && document.activeElement, w = wn(g), $ = w.map(function(O) {
          var T = O.node;
          return T;
        }).indexOf(b);
        $ > -1 && (w.filter(function(O) {
          var T = O.guard, F = O.node;
          return T && F.dataset.focusAutoGuard;
        }).forEach(function(O) {
          var T = O.node;
          return T.removeAttribute("tabIndex");
        }), Sr($, w.length, 1, w), Sr($, -1, -1, w));
      }
    }
  }
  return r;
}, rt = function(r) {
  Pe() && r && (r.stopPropagation(), r.preventDefault());
}, or = function() {
  return tr(Pe);
}, Yn = function(r) {
  var t = r.target, n = r.currentTarget;
  n.contains(t) || kn(n, t);
}, zn = function() {
  return null;
};
process.env.NODE_ENV !== "production" && Ir.node.isRequired;
var tt = function() {
  nr = !0;
}, nt = function() {
  nr = !1, Ee = "just", tr(function() {
    Ee = "meanwhile";
  });
}, Vn = function() {
  document.addEventListener("focusin", rt), document.addEventListener("focusout", or), window.addEventListener("focus", tt), window.addEventListener("blur", nt);
}, Gn = function() {
  document.removeEventListener("focusin", rt), document.removeEventListener("focusout", or), window.removeEventListener("focus", tt), window.removeEventListener("blur", nt);
};
function Hn(e) {
  return e.filter(function(r) {
    var t = r.disabled;
    return !t;
  });
}
var ot = {
  moveFocusInside: Jr,
  focusInside: Gr,
  focusNextElement: An,
  focusPrevElement: In,
  focusFirstElement: Cn,
  focusLastElement: $n,
  captureFocusRestore: Kr
};
function Jn(e) {
  var r = e.slice(-1)[0];
  r && !de && Vn();
  var t = de, n = t && r && r.id === t.id;
  de = r, t && !n && (t.onDeactivation(), e.filter(function(o) {
    var a = o.id;
    return a === t.id;
  }).length || t.returnFocus(!r)), r ? (W = null, (!n || t.observed !== r.observed) && r.onActivation(ot), Pe(), tr(Pe)) : (Gn(), W = null);
}
Nr.assignSyncMedium(Yn);
Dr.assignMedium(or);
Bt.assignMedium(function(e) {
  return e(ot);
});
const Kn = Kt(Hn, Jn)(zn);
var at = /* @__PURE__ */ Rr(function(r, t) {
  return /* @__PURE__ */ X.createElement(Ge, me({
    sideCar: Kn,
    ref: t
  }, r));
}), it = Ge.propTypes || {};
it.sideCar;
var Qn = Pt(it, ["sideCar"]);
at.propTypes = process.env.NODE_ENV !== "production" ? Qn : {};
function ut(e) {
  var r, t, n = "";
  if (typeof e == "string" || typeof e == "number") n += e;
  else if (typeof e == "object") if (Array.isArray(e)) {
    var o = e.length;
    for (r = 0; r < o; r++) e[r] && (t = ut(e[r])) && (n && (n += " "), n += t);
  } else for (t in e) e[t] && (n && (n += " "), n += t);
  return n;
}
function Xn() {
  for (var e, r, t = 0, n = "", o = arguments.length; t < o; t++) (e = arguments[t]) && (r = ut(e)) && (n && (n += " "), n += r);
  return n;
}
function ct(e) {
  var r, t, n = "";
  if (typeof e == "string" || typeof e == "number") n += e;
  else if (typeof e == "object") if (Array.isArray(e)) {
    var o = e.length;
    for (r = 0; r < o; r++) e[r] && (t = ct(e[r])) && (n && (n += " "), n += t);
  } else for (t in e) e[t] && (n && (n += " "), n += t);
  return n;
}
function Zn() {
  for (var e, r, t = 0, n = "", o = arguments.length; t < o; t++) (e = arguments[t]) && (r = ct(e)) && (n && (n += " "), n += r);
  return n;
}
const _e = ({
  as: e,
  children: r,
  display: t,
  m: n,
  mt: o,
  mr: a,
  mb: i,
  ml: u,
  mx: m,
  my: f,
  p: v,
  pt: d,
  pr: p,
  pb: g,
  pl: E,
  px: b,
  py: w,
  width: $,
  height: O,
  position: T,
  style: F,
  className: s,
  disabled: k,
  required: Y,
  ...z
}) => {
  const H = e || "div", h = {};
  t && (h.display = t), $ && (h.width = $), O && (h.height = O), T && (h.position = T), n && (h.margin = `var(--pittorica-space-${n})`), o && (h.marginTop = `var(--pittorica-space-${o})`), a && (h.marginRight = `var(--pittorica-space-${a})`), i && (h.marginBottom = `var(--pittorica-space-${i})`), u && (h.marginLeft = `var(--pittorica-space-${u})`), m && (h.marginLeft = `var(--pittorica-space-${m})`, h.marginRight = `var(--pittorica-space-${m})`), f && (h.marginTop = `var(--pittorica-space-${f})`, h.marginBottom = `var(--pittorica-space-${f})`), v && (h.padding = `var(--pittorica-space-${v})`), d && (h.paddingTop = `var(--pittorica-space-${d})`), p && (h.paddingRight = `var(--pittorica-space-${p})`), g && (h.paddingBottom = `var(--pittorica-space-${g})`), E && (h.paddingLeft = `var(--pittorica-space-${E})`), b && (h.paddingLeft = `var(--pittorica-space-${b})`, h.paddingRight = `var(--pittorica-space-${b})`), w && (h.paddingTop = `var(--pittorica-space-${w})`, h.paddingBottom = `var(--pittorica-space-${w})`);
  const V = {
    ...F,
    ...h
  };
  return /* @__PURE__ */ q(
    H,
    {
      className: Zn("pittorica-box", s),
      style: V,
      disabled: k,
      required: Y,
      ...z,
      children: r
    }
  );
};
_e.displayName = "Box";
function st(e) {
  var r, t, n = "";
  if (typeof e == "string" || typeof e == "number") n += e;
  else if (typeof e == "object") if (Array.isArray(e)) {
    var o = e.length;
    for (r = 0; r < o; r++) e[r] && (t = st(e[r])) && (n && (n += " "), n += t);
  } else for (t in e) e[t] && (n && (n += " "), n += t);
  return n;
}
function eo() {
  for (var e, r, t = 0, n = "", o = arguments.length; t < o; t++) (e = arguments[t]) && (r = st(e)) && (n && (n += " "), n += r);
  return n;
}
function ft(e) {
  var r, t, n = "";
  if (typeof e == "string" || typeof e == "number") n += e;
  else if (typeof e == "object") if (Array.isArray(e)) {
    var o = e.length;
    for (r = 0; r < o; r++) e[r] && (t = ft(e[r])) && (n && (n += " "), n += t);
  } else for (t in e) e[t] && (n && (n += " "), n += t);
  return n;
}
function Pr() {
  for (var e, r, t = 0, n = "", o = arguments.length; t < o; t++) (e = arguments[t]) && (r = ft(e)) && (n && (n += " "), n += r);
  return n;
}
function dt(e) {
  var r, t, n = "";
  if (typeof e == "string" || typeof e == "number") n += e;
  else if (typeof e == "object") if (Array.isArray(e)) {
    var o = e.length;
    for (r = 0; r < o; r++) e[r] && (t = dt(e[r])) && (n && (n += " "), n += t);
  } else for (t in e) e[t] && (n && (n += " "), n += t);
  return n;
}
function ro() {
  for (var e, r, t = 0, n = "", o = arguments.length; t < o; t++) (e = arguments[t]) && (r = dt(e)) && (n && (n += " "), n += r);
  return n;
}
const lt = ({
  as: e,
  children: r,
  display: t,
  m: n,
  mt: o,
  mr: a,
  mb: i,
  ml: u,
  mx: m,
  my: f,
  p: v,
  pt: d,
  pr: p,
  pb: g,
  pl: E,
  px: b,
  py: w,
  width: $,
  height: O,
  position: T,
  style: F,
  className: s,
  disabled: k,
  required: Y,
  ...z
}) => {
  const H = e || "div", h = {};
  t && (h.display = t), $ && (h.width = $), O && (h.height = O), T && (h.position = T), n && (h.margin = `var(--pittorica-space-${n})`), o && (h.marginTop = `var(--pittorica-space-${o})`), a && (h.marginRight = `var(--pittorica-space-${a})`), i && (h.marginBottom = `var(--pittorica-space-${i})`), u && (h.marginLeft = `var(--pittorica-space-${u})`), m && (h.marginLeft = `var(--pittorica-space-${m})`, h.marginRight = `var(--pittorica-space-${m})`), f && (h.marginTop = `var(--pittorica-space-${f})`, h.marginBottom = `var(--pittorica-space-${f})`), v && (h.padding = `var(--pittorica-space-${v})`), d && (h.paddingTop = `var(--pittorica-space-${d})`), p && (h.paddingRight = `var(--pittorica-space-${p})`), g && (h.paddingBottom = `var(--pittorica-space-${g})`), E && (h.paddingLeft = `var(--pittorica-space-${E})`), b && (h.paddingLeft = `var(--pittorica-space-${b})`, h.paddingRight = `var(--pittorica-space-${b})`), w && (h.paddingTop = `var(--pittorica-space-${w})`, h.paddingBottom = `var(--pittorica-space-${w})`);
  const V = {
    ...F,
    ...h
  };
  return /* @__PURE__ */ q(
    H,
    {
      className: ro("pittorica-box", s),
      style: V,
      disabled: k,
      required: Y,
      ...z,
      children: r
    }
  );
};
lt.displayName = "Box";
const vt = ({
  children: e,
  as: r,
  size: t,
  weight: n = "regular",
  align: o,
  truncate: a = !1,
  color: i,
  className: u,
  style: m,
  ...f
}) => {
  const v = r || "span", d = (i == null ? void 0 : i.startsWith("#")) || (i == null ? void 0 : i.startsWith("rgb")) || (i == null ? void 0 : i.startsWith("hsl")), p = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), g = (() => {
    if (i)
      return i === "inherit" ? "inherit" : d ? i : p.has(i) ? `var(--pittorica-${i}-9)` : i;
  })(), E = t ? typeof t == "string" ? `pittorica-text--size-${t}` : Pr({
    [`pittorica-text--size-${t.initial}`]: t.initial,
    [`pittorica-text--xs-size-${t.xs}`]: t.xs,
    [`pittorica-text--sm-size-${t.sm}`]: t.sm,
    [`pittorica-text--md-size-${t.md}`]: t.md,
    [`pittorica-text--lg-size-${t.lg}`]: t.lg,
    [`pittorica-text--xl-size-${t.xl}`]: t.xl
  }) : null, b = {
    ...m,
    textAlign: o,
    color: g
  };
  return /* @__PURE__ */ q(
    lt,
    {
      as: v,
      className: Pr(
        "pittorica-text",
        E,
        { "pittorica-text--truncate": a },
        u
      ),
      "data-weight": n,
      style: b,
      ...f,
      children: e
    }
  );
};
vt.displayName = "Text";
const to = (e, r) => r ? typeof r == "string" ? [`pittorica-heading--${e}-${r}`] : Object.entries(r).filter(([t, n]) => n !== void 0).map(
  ([t, n]) => t === "initial" ? `pittorica-heading--${e}-${n}` : `pittorica-heading--${t}-${e}-${n}`
) : [], pt = ({
  children: e,
  as: r,
  size: t = "6",
  weight: n = "bold",
  variant: o = "classic",
  color: a,
  className: i,
  style: u,
  ...m
}) => {
  const f = to("size", t), v = r || "h1", d = a && a !== "inherit" && !a.startsWith("#") && !a.startsWith("rgb"), p = {
    ...o === "soft" && d && {
      "--pittorica-source-3": `var(--pittorica-${a}-3)`,
      "--pittorica-source-11": `var(--pittorica-${a}-11)`
    },
    ...o === "outline" && d && {
      "--pittorica-source-9": `var(--pittorica-${a}-9)`
    },
    ...u
  };
  return /* @__PURE__ */ q(
    vt,
    {
      as: v,
      weight: n,
      className: eo("pittorica-heading", f, i),
      "data-variant": o,
      style: p,
      color: a,
      ...m,
      children: e
    }
  );
};
pt.displayName = "Heading";
const no = ({
  children: e,
  sourceColor: r = "indigo",
  appearance: t = "light",
  radius: n = "medium",
  className: o,
  style: a,
  id: i
}) => {
  const u = r.startsWith("#") || r.startsWith("rgb"), m = {
    id: i,
    className: ["pittorica-theme", o].filter(Boolean).join(" "),
    "data-source-color": u ? "custom" : r,
    "data-appearance": t,
    "data-radius": n,
    style: {
      ...a,
      ...u ? { "--pittorica-source-color": r } : {}
    }
  };
  return /* @__PURE__ */ q("div", { ...m, children: e });
}, yt = _r(null), mt = () => {
  const e = wt(yt);
  if (!e)
    throw new Error("Dialog compound components must be used within a Dialog");
  return e;
}, oo = ({
  open: e,
  onClose: r,
  children: t,
  className: n,
  appearance: o,
  closeOnOverlayClick: a = !0,
  closeOnEsc: i = !0,
  as: u,
  ...m
}) => {
  const f = ur(), v = ur(), [d, p] = fe(), g = we(null), [E, b] = fe(!1);
  return Se(() => {
    b(!0);
  }, []), Se(() => {
    if (!e) return;
    const O = (T) => {
      T.key === "Escape" && i && r();
    };
    return globalThis.addEventListener("keydown", O), () => globalThis.removeEventListener("keydown", O);
  }, [e, r, i]), Se(() => {
    if (typeof document > "u" || !document.body || !e) return;
    const O = globalThis.getComputedStyle(document.body).overflow;
    return document.body.style.overflow = "hidden", () => {
      document.body.style.overflow = O;
    };
  }, [e]), Ot(() => {
    if (e && g.current) {
      const O = g.current.closest(
        ".pittorica-theme"
      );
      if (O) {
        const T = O.dataset.appearance;
        p(T || void 0);
      }
    }
  }, [e]), !e || !E ? null : /* @__PURE__ */ ht(gt, { children: [
    /* @__PURE__ */ q("div", { ref: g, style: { display: "none" }, "aria-hidden": "true" }),
    St(
      /* @__PURE__ */ q(
        "div",
        {
          className: "pittorica-dialog-overlay",
          onClick: a ? r : void 0,
          "aria-hidden": "true",
          children: /* @__PURE__ */ q(no, { appearance: o ?? d, children: /* @__PURE__ */ q(at, { returnFocus: !0, children: /* @__PURE__ */ q(
            _e,
            {
              as: u || "div",
              className: Xn("pittorica-dialog-content", n),
              onClick: (O) => O.stopPropagation(),
              role: "dialog",
              "aria-modal": "true",
              "aria-labelledby": f,
              "aria-describedby": v,
              ...m,
              children: /* @__PURE__ */ q(yt, { value: { titleId: f, descriptionId: v }, children: t })
            }
          ) }) })
        }
      ),
      document.body
    )
  ] });
}, ao = ({
  children: e,
  id: r,
  ...t
}) => {
  const { titleId: n } = mt();
  return /* @__PURE__ */ q(pt, { as: "h2", id: n, size: "6", mb: "3", ...t, children: e });
}, io = ({
  children: e,
  id: r,
  ...t
}) => {
  const { descriptionId: n } = mt();
  return /* @__PURE__ */ q(
    _e,
    {
      id: n,
      mb: "4",
      style: {
        color: "inherit",
        fontSize: "var(--pittorica-font-size-3)"
      },
      ...t,
      children: e
    }
  );
}, uo = ({
  children: e,
  ...r
}) => /* @__PURE__ */ q(
  _e,
  {
    style: {
      display: "flex",
      justifyContent: "flex-end",
      gap: "var(--pittorica-space-3)",
      marginTop: "var(--pittorica-space-6)"
    },
    ...r,
    children: e
  }
);
oo.displayName = "Dialog";
ao.displayName = "Dialog.Title";
io.displayName = "Dialog.Description";
uo.displayName = "Dialog.Actions";
export {
  oo as Dialog,
  uo as DialogActions,
  io as DialogDescription,
  ao as DialogTitle
};
