// @pittorica/stack-react@0.26.3/dist/index.js
import { jsx as h } from "react/jsx-runtime";
function B(t) {
  var i, r, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var p = t.length;
    for (i = 0; i < p; i++) t[i] && (r = B(t[i])) && (a && (a += " "), a += r);
  } else for (r in t) t[r] && (a && (a += " "), a += r);
  return a;
}
function C() {
  for (var t, i, r = 0, a = "", p = arguments.length; r < p; r++) (t = arguments[r]) && (i = B(t)) && (a && (a += " "), a += i);
  return a;
}
function A(t) {
  var i, r, a = "";
  if (typeof t == "string" || typeof t == "number") a += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var p = t.length;
    for (i = 0; i < p; i++) t[i] && (r = A(t[i])) && (a && (a += " "), a += r);
  } else for (r in t) t[r] && (a && (a += " "), a += r);
  return a;
}
function F() {
  for (var t, i, r = 0, a = "", p = arguments.length; r < p; r++) (t = arguments[r]) && (i = A(t)) && (a && (a += " "), a += i);
  return a;
}
const R = ({
  as: t,
  children: i,
  display: r,
  m: a,
  mt: p,
  mr: d,
  mb: o,
  ml: f,
  mx: s,
  my: c,
  p: g,
  pt: m,
  pr: v,
  pb: n,
  pl: b,
  px: $,
  py: u,
  width: x,
  height: N,
  position: j,
  style: L,
  className: k,
  disabled: S,
  required: T,
  ...q
}) => {
  const O = t || "div", e = {};
  r && (e.display = r), x && (e.width = x), N && (e.height = N), j && (e.position = j), a && (e.margin = `var(--pittorica-space-${a})`), p && (e.marginTop = `var(--pittorica-space-${p})`), d && (e.marginRight = `var(--pittorica-space-${d})`), o && (e.marginBottom = `var(--pittorica-space-${o})`), f && (e.marginLeft = `var(--pittorica-space-${f})`), s && (e.marginLeft = `var(--pittorica-space-${s})`, e.marginRight = `var(--pittorica-space-${s})`), c && (e.marginTop = `var(--pittorica-space-${c})`, e.marginBottom = `var(--pittorica-space-${c})`), g && (e.padding = `var(--pittorica-space-${g})`), m && (e.paddingTop = `var(--pittorica-space-${m})`), v && (e.paddingRight = `var(--pittorica-space-${v})`), n && (e.paddingBottom = `var(--pittorica-space-${n})`), b && (e.paddingLeft = `var(--pittorica-space-${b})`), $ && (e.paddingLeft = `var(--pittorica-space-${$})`, e.paddingRight = `var(--pittorica-space-${$})`), u && (e.paddingTop = `var(--pittorica-space-${u})`, e.paddingBottom = `var(--pittorica-space-${u})`);
  const W = {
    ...L,
    ...e
  };
  return /* @__PURE__ */ h(
    O,
    {
      className: F("pittorica-box", k),
      style: W,
      disabled: S,
      required: T,
      ...q,
      children: i
    }
  );
};
R.displayName = "Box";
const y = (t) => typeof t == "string" && t.startsWith("auto-");
function l(t, i) {
  return !i || y(i) ? [] : typeof i == "string" ? [`pittorica-flex--${t}-${i}`] : Object.entries(i).filter(([, r]) => r !== void 0 && !y(r)).map(
    ([r, a]) => r === "initial" ? `pittorica-flex--${t}-${a}` : `pittorica-flex--${r}-${t}-${a}`
  );
}
const w = ({
  children: t,
  direction: i,
  align: r,
  justify: a,
  wrap: p,
  gap: d,
  basis: o,
  className: f,
  style: s,
  as: c,
  ...g
}) => {
  const m = c || "div", v = [
    ...l("direction", i),
    ...l("align", r),
    ...l("justify", a),
    ...l("wrap", p),
    ...l("gap", d)
  ], n = { ...s };
  return y(o) && (n["--pittorica-flex-basis"] = o.split("-")[1]), /* @__PURE__ */ h(
    R,
    {
      as: m,
      className: C("pittorica-flex", v, f),
      style: n,
      ...g,
      children: t
    }
  );
};
w.displayName = "Flex";
const z = ({
  direction: t = "column",
  ...i
}) => /* @__PURE__ */ h(w, { direction: t, ...i });
z.displayName = "Stack";
export {
  z as Stack
};
