// eslint-plugin-package-json@0.89.1/lib/utils/createSimpleValidPropertyRule.mjs
import { createRule } from "../createRule.mjs";

//#region src/utils/createSimpleValidPropertyRule.ts
/**
* Given a top-level property name, and a validation function, create a rule that validates the property using the validation function.
* These rules will always be included in the recommended config.
* Note: this will only create a basic validation rule, with no options.  If you need
* to create a more complex rule, create it in its own file.
*/
const createSimpleValidPropertyRule = (propertyName, validationFunction, aliases = []) => {
	const ruleName = `valid-${propertyName}`;
	const propertyNames = [propertyName, ...aliases];
	return {
		rule: createRule({
			create(context) {
				const reportIssues = (result, node) => {
					if (result.errorMessages.length === 0) return;
					if (result.issues.length) for (const issue of result.issues) context.report({
						data: { error: issue.message },
						messageId: "validationError",
						node
					});
					const childrenWithIssues = result.childResults.filter((childResult) => childResult.errorMessages.length);
					if (node.type === "JSONObjectExpression" && childrenWithIssues.length) for (const childResult of childrenWithIssues) {
						const childNode = node.properties[childResult.index];
						reportIssues(childResult, childNode);
					}
					else if (node.type === "JSONArrayExpression" && childrenWithIssues.length) for (const childResult of childrenWithIssues) {
						const childNode = node.elements[childResult.index];
						if (childNode) reportIssues(childResult, childNode);
					}
				};
				return propertyNames.reduce((acc, name) => {
					acc[`Program > JSONExpressionStatement > JSONObjectExpression > JSONProperty[key.value=${name}]`] = (node) => {
						const valueNode = node.value;
						reportIssues(validationFunction(JSON.parse(context.sourceCode.getText(valueNode))), valueNode);
					};
					return acc;
				}, {});
			},
			meta: {
				docs: {
					category: "Best Practices",
					description: `Enforce that the \`${propertyName}\`${aliases.length ? ` (also: ${aliases.map((alias) => `\`${alias}\``).join(", ")})` : ""} property is valid.`,
					recommended: true
				},
				messages: { validationError: `Invalid ${propertyName}: {{ error }}` },
				schema: [],
				type: "problem"
			},
			name: ruleName
		}),
		ruleName
	};
};

//#endregion
export { createSimpleValidPropertyRule };
