// eslint-plugin-package-json@0.89.1/lib/utils/createSimpleRequirePropertyRule.mjs
import { createRule } from "../createRule.mjs";
import { isJSONStringLiteral } from "./predicates.mjs";

//#region src/utils/createSimpleRequirePropertyRule.ts
/**
* Given a top-level property name, create a rule that requires that property to be present.
* Optionally, include it in the recommended config.
* Note: this will only create a basic require rule, with no options.  If you need
* to create a more complex rule, create it in its own file.
*/
const createSimpleRequirePropertyRule = (propertyName, { category, fixValue, ignorePrivateDefault = false, isRecommended } = {}) => {
	const ruleName = `require-${propertyName}`;
	return {
		rule: createRule({
			create(context) {
				const enforceForPrivate = context.settings.packageJson?.enforceForPrivate;
				const ignorePrivate = context.options[0]?.ignorePrivate ?? (typeof enforceForPrivate === "boolean" ? !enforceForPrivate : ignorePrivateDefault);
				return { "Program > JSONExpressionStatement > JSONObjectExpression"(node) {
					if (ignorePrivate && node.properties.some((property) => isJSONStringLiteral(property.key) && property.key.value === "private" && property.value.type === "JSONLiteral" && property.value.value === true)) return;
					if (!node.properties.some((property) => isJSONStringLiteral(property.key) && property.key.value === propertyName)) context.report({
						data: { property: propertyName },
						fix: fixValue === void 0 ? void 0 : function* (fixer) {
							yield fixer.insertTextAfterRange([0, 1], `\n  "${propertyName}": ${JSON.stringify(fixValue)}`);
							yield node.properties.length > 0 ? fixer.insertTextAfterRange([0, 1], ",") : fixer.insertTextAfterRange([0, 1], "\n");
						},
						messageId: "missing",
						node: context.sourceCode.ast
					});
				} };
			},
			meta: {
				docs: {
					category,
					description: `Requires the \`${propertyName}\` property to be present.`,
					recommended: isRecommended
				},
				fixable: fixValue === void 0 ? void 0 : "code",
				messages: { missing: "Property '{{property}}' is required." },
				schema: [{
					additionalProperties: false,
					properties: { ignorePrivate: {
						default: ignorePrivateDefault,
						description: "Determines if this rule should be enforced when the package's `private` property is `true`.",
						type: "boolean"
					} },
					type: "object"
				}],
				type: "suggestion"
			},
			name: ruleName
		}),
		ruleName
	};
};

//#endregion
export { createSimpleRequirePropertyRule };
