// @pittorica/segmented-control-react@0.26.3/dist/index.js
import { jsx as y } from "react/jsx-runtime";
import { createContext as W, useState as E, useMemo as M, use as U } from "react";
function x(t) {
  var o, r, e = "";
  if (typeof t == "string" || typeof t == "number") e += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var i = t.length;
    for (o = 0; o < i; o++) t[o] && (r = x(t[o])) && (e && (e += " "), e += r);
  } else for (r in t) t[r] && (e && (e += " "), e += r);
  return e;
}
function N() {
  for (var t, o, r = 0, e = "", i = arguments.length; r < i; r++) (t = arguments[r]) && (o = x(t)) && (e && (e += " "), e += o);
  return e;
}
function R(t) {
  var o, r, e = "";
  if (typeof t == "string" || typeof t == "number") e += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var i = t.length;
    for (o = 0; o < i; o++) t[o] && (r = R(t[o])) && (e && (e += " "), e += r);
  } else for (r in t) t[r] && (e && (e += " "), e += r);
  return e;
}
function z() {
  for (var t, o, r = 0, e = "", i = arguments.length; r < i; r++) (t = arguments[r]) && (o = R(t)) && (e && (e += " "), e += o);
  return e;
}
const b = ({
  as: t,
  children: o,
  display: r,
  m: e,
  mt: i,
  mr: p,
  mb: m,
  ml: g,
  mx: d,
  my: n,
  p: u,
  pt: s,
  pr: c,
  pb: h,
  pl: v,
  px: l,
  py: f,
  width: $,
  height: C,
  position: S,
  style: A,
  className: B,
  disabled: L,
  required: j,
  ...q
}) => {
  const I = t || "div", a = {};
  r && (a.display = r), $ && (a.width = $), C && (a.height = C), S && (a.position = S), e && (a.margin = `var(--pittorica-space-${e})`), i && (a.marginTop = `var(--pittorica-space-${i})`), p && (a.marginRight = `var(--pittorica-space-${p})`), m && (a.marginBottom = `var(--pittorica-space-${m})`), g && (a.marginLeft = `var(--pittorica-space-${g})`), d && (a.marginLeft = `var(--pittorica-space-${d})`, a.marginRight = `var(--pittorica-space-${d})`), n && (a.marginTop = `var(--pittorica-space-${n})`, a.marginBottom = `var(--pittorica-space-${n})`), u && (a.padding = `var(--pittorica-space-${u})`), s && (a.paddingTop = `var(--pittorica-space-${s})`), c && (a.paddingRight = `var(--pittorica-space-${c})`), h && (a.paddingBottom = `var(--pittorica-space-${h})`), v && (a.paddingLeft = `var(--pittorica-space-${v})`), l && (a.paddingLeft = `var(--pittorica-space-${l})`, a.paddingRight = `var(--pittorica-space-${l})`), f && (a.paddingTop = `var(--pittorica-space-${f})`, a.paddingBottom = `var(--pittorica-space-${f})`);
  const D = {
    ...A,
    ...a
  };
  return /* @__PURE__ */ y(
    I,
    {
      className: z("pittorica-box", B),
      style: D,
      disabled: L,
      required: j,
      ...q,
      children: o
    }
  );
};
b.displayName = "Box";
const T = W(null), F = () => {
  const t = U(T);
  if (!t)
    throw new Error(
      "SegmentedControl components must be wrapped in <SegmentedControl.Root />"
    );
  return t;
}, k = ({
  children: t,
  value: o,
  defaultValue: r,
  onValueChange: e,
  color: i = "indigo",
  disabled: p,
  className: m,
  as: g,
  ...d
}) => {
  const [n, u] = E(r), s = o !== void 0, c = s ? o : n, h = (f) => {
    s || u(f), e == null || e(f);
  }, v = M(
    () => ({
      value: c,
      onValueChange: h,
      color: i,
      disabled: p
    }),
    [c, i, p]
  );
  return /* @__PURE__ */ y(T, { value: v, children: /* @__PURE__ */ y(
    b,
    {
      as: g || "div",
      ...d,
      role: "radiogroup",
      className: N("pittorica-segmented-control-root", m),
      children: t
    }
  ) });
}, w = ({
  children: t,
  value: o,
  disabled: r,
  className: e,
  style: i,
  as: p,
  ...m
}) => {
  const {
    value: g,
    onValueChange: d,
    color: n,
    disabled: u
  } = F(), s = g === o, c = u || r, v = n !== "inherit" && !(n != null && n.startsWith("#")) && !(n != null && n.startsWith("rgb")) ? `var(--pittorica-${n}-9)` : n, l = p || "button";
  return /* @__PURE__ */ y(
    b,
    {
      as: l,
      type: l === "button" ? "button" : void 0,
      role: "radio",
      "aria-checked": s,
      "data-state": s ? "checked" : "unchecked",
      disabled: c,
      className: N("pittorica-segmented-control-item", e),
      onClick: () => !c && d(o),
      style: {
        "--pittorica-source-color": v,
        ...i
      },
      ...m,
      children: t
    }
  );
}, J = {
  Root: k,
  Item: w
};
k.displayName = "SegmentedControl.Root";
w.displayName = "SegmentedControl.Item";
export {
  J as SegmentedControl
};
