// @pittorica/context-menu-react@0.26.3/dist/index.js
import { jsx as v, jsxs as k } from "react/jsx-runtime";
import { createContext as W, useState as y, useCallback as K, use as C, useRef as I, useLayoutEffect as N, useEffect as M, useMemo as U } from "react";
import { createPortal as V } from "react-dom";
function L(e) {
  var n, a, t = "";
  if (typeof e == "string" || typeof e == "number") t += e;
  else if (typeof e == "object") if (Array.isArray(e)) {
    var i = e.length;
    for (n = 0; n < i; n++) e[n] && (a = L(e[n])) && (t && (t += " "), t += a);
  } else for (a in e) e[a] && (t && (t += " "), t += a);
  return t;
}
function E() {
  for (var e, n, a = 0, t = "", i = arguments.length; a < i; a++) (e = arguments[a]) && (n = L(e)) && (t && (t += " "), t += n);
  return t;
}
function F(e) {
  var n, a, t = "";
  if (typeof e == "string" || typeof e == "number") t += e;
  else if (typeof e == "object") if (Array.isArray(e)) {
    var i = e.length;
    for (n = 0; n < i; n++) e[n] && (a = F(e[n])) && (t && (t += " "), t += a);
  } else for (a in e) e[a] && (t && (t += " "), t += a);
  return t;
}
function X() {
  for (var e, n, a = 0, t = "", i = arguments.length; a < i; a++) (e = arguments[a]) && (n = F(e)) && (t && (t += " "), t += n);
  return t;
}
const A = ({
  as: e,
  children: n,
  display: a,
  m: t,
  mt: i,
  mr: c,
  mb: o,
  ml: s,
  mx: l,
  my: d,
  p,
  pt: f,
  pr: x,
  pb: u,
  pl: g,
  px: m,
  py: h,
  width: b,
  height: T,
  position: $,
  style: R,
  className: j,
  disabled: P,
  required: B,
  ...S
}) => {
  const D = e || "div", r = {};
  a && (r.display = a), b && (r.width = b), T && (r.height = T), $ && (r.position = $), t && (r.margin = `var(--pittorica-space-${t})`), i && (r.marginTop = `var(--pittorica-space-${i})`), c && (r.marginRight = `var(--pittorica-space-${c})`), o && (r.marginBottom = `var(--pittorica-space-${o})`), s && (r.marginLeft = `var(--pittorica-space-${s})`), l && (r.marginLeft = `var(--pittorica-space-${l})`, r.marginRight = `var(--pittorica-space-${l})`), d && (r.marginTop = `var(--pittorica-space-${d})`, r.marginBottom = `var(--pittorica-space-${d})`), p && (r.padding = `var(--pittorica-space-${p})`), f && (r.paddingTop = `var(--pittorica-space-${f})`), x && (r.paddingRight = `var(--pittorica-space-${x})`), u && (r.paddingBottom = `var(--pittorica-space-${u})`), g && (r.paddingLeft = `var(--pittorica-space-${g})`), m && (r.paddingLeft = `var(--pittorica-space-${m})`, r.paddingRight = `var(--pittorica-space-${m})`), h && (r.paddingTop = `var(--pittorica-space-${h})`, r.paddingBottom = `var(--pittorica-space-${h})`);
  const H = {
    ...R,
    ...r
  };
  return /* @__PURE__ */ v(
    D,
    {
      className: X("pittorica-box", j),
      style: H,
      disabled: P,
      required: B,
      ...S,
      children: n
    }
  );
};
A.displayName = "Box";
const w = W(null), q = ({
  x: e,
  y: n,
  children: a,
  onClose: t,
  itemCount: i
}) => {
  const c = I(null), [o, s] = y(-1), [l, d] = y({ top: n, left: e }), [p, f] = y(!1);
  M(() => {
    const u = requestAnimationFrame(() => f(!0));
    return () => cancelAnimationFrame(u);
  }, []), N(() => {
    if (globalThis.window === void 0 || !c.current) return;
    const { innerWidth: u, innerHeight: g } = globalThis, { offsetWidth: m, offsetHeight: h } = c.current, b = requestAnimationFrame(() => {
      d({
        left: e + m > u ? e - m : e,
        top: n + h > g ? n - h : n
      });
    });
    return () => cancelAnimationFrame(b);
  }, [e, n]), M(() => {
    if (globalThis.window === void 0) return;
    const u = (g) => {
      if (i !== 0)
        switch (g.key) {
          case "ArrowDown": {
            s((m) => (m + 1) % i);
            break;
          }
          case "ArrowUp": {
            s((m) => (m - 1 + i) % i);
            break;
          }
          case "Escape": {
            t();
            break;
          }
        }
    };
    return globalThis.addEventListener("keydown", u), globalThis.addEventListener("click", t), () => {
      globalThis.removeEventListener("keydown", u), globalThis.removeEventListener("click", t);
    };
  }, [i, t]);
  const x = U(
    () => ({
      activeIndex: o,
      setActiveIndex: s,
      closeAll: t
    }),
    [o, t]
  );
  return p ? V(
    /* @__PURE__ */ v(w, { value: x, children: /* @__PURE__ */ v(
      "div",
      {
        ref: c,
        className: "pittorica-context-menu-content",
        style: { top: l.top, left: l.left },
        onClick: (u) => u.stopPropagation(),
        role: "menu",
        children: a
      }
    ) }),
    document.body
  ) : null;
}, Y = ({
  trigger: e,
  children: n,
  itemCount: a,
  as: t,
  ...i
}) => {
  const [c, o] = y(null), s = K(() => {
    o(null);
  }, []);
  return /* @__PURE__ */ k(
    A,
    {
      as: t || "div",
      onContextMenu: (p) => {
        p.preventDefault(), o({ x: p.clientX, y: p.clientY });
      },
      style: { display: "inline-block" },
      ...i,
      children: [
        e,
        c && /* @__PURE__ */ v(
          q,
          {
            x: c.x,
            y: c.y,
            onClose: s,
            itemCount: a,
            children: n
          }
        )
      ]
    }
  );
}, z = ({
  children: e,
  onClick: n,
  index: a,
  className: t,
  as: i,
  ...c
}) => {
  const o = C(w);
  if (!o) return null;
  const s = o.activeIndex === a, l = i || "div", d = (p) => {
    n && n(p), o.closeAll();
  };
  return /* @__PURE__ */ v(
    A,
    {
      as: l,
      className: E("pittorica-context-menu-item", t),
      "data-active": s,
      onMouseEnter: () => o.setActiveIndex(a),
      onClick: d,
      role: "menuitem",
      ...c,
      children: e
    }
  );
}, G = ({
  label: e,
  children: n,
  index: a,
  itemCount: t,
  className: i,
  as: c,
  ...o
}) => {
  const s = C(w), l = I(null), [d, p] = y(
    null
  );
  if (!s) return null;
  const f = s.activeIndex === a;
  return N(() => {
    if (f && l.current) {
      const g = l.current.getBoundingClientRect(), m = requestAnimationFrame(() => {
        p({ x: g.right, y: g.top });
      });
      return () => cancelAnimationFrame(m);
    }
    const u = requestAnimationFrame(() => p(null));
    return () => cancelAnimationFrame(u);
  }, [f]), /* @__PURE__ */ k(
    A,
    {
      as: c || "div",
      ref: l,
      className: E("pittorica-context-menu-item", i),
      "data-active": f,
      onMouseEnter: () => s.setActiveIndex(a),
      role: "menuitem",
      "aria-haspopup": "true",
      ...o,
      children: [
        e,
        /* @__PURE__ */ v("span", { className: "pittorica-context-menu-chevron", "aria-hidden": "true", children: "▶" }),
        f && d && /* @__PURE__ */ v(
          q,
          {
            x: d.x,
            y: d.y,
            onClose: s.closeAll,
            itemCount: t,
            children: n
          }
        )
      ]
    }
  );
};
Y.displayName = "ContextMenu";
z.displayName = "ContextMenu.Item";
G.displayName = "ContextMenu.Sub";
export {
  Y as ContextMenu,
  z as ContextMenuItem,
  G as ContextMenuSub
};
