// @pittorica/radio-react@0.26.3/dist/index.js
import { jsx as $ } from "react/jsx-runtime";
import { forwardRef as _, createElement as w, useState as z } from "react";
function C(i) {
  var a, r, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (a = 0; a < o; a++) i[a] && (r = C(i[a])) && (t && (t += " "), t += r);
  } else for (r in i) i[r] && (t && (t += " "), t += r);
  return t;
}
function I() {
  for (var i, a, r = 0, t = "", o = arguments.length; r < o; r++) (i = arguments[r]) && (a = C(i)) && (t && (t += " "), t += a);
  return t;
}
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var F = {
  outline: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  },
  filled: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "currentColor",
    stroke: "none"
  }
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const x = (i, a, r, t) => {
  const o = _(
    ({ color: l = "currentColor", size: c = 24, stroke: m = 2, title: p, className: d, children: n, ...g }, u) => w(
      "svg",
      {
        ref: u,
        ...F[i],
        width: c,
        height: c,
        className: ["tabler-icon", `tabler-icon-${a}`, d].join(" "),
        ...i === "filled" ? {
          fill: l
        } : {
          strokeWidth: m,
          stroke: l
        },
        ...g
      },
      [
        p && w("title", { key: "svg-title" }, p),
        ...t.map(([f, s]) => w(f, s)),
        ...Array.isArray(n) ? n : [n]
      ]
    )
  );
  return o.displayName = `${r}`, o;
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const M = [["path", { d: "M3 12a9 9 0 1 0 18 0a9 9 0 1 0 -18 0", key: "svg-0" }]], S = x("outline", "circle", "Circle", M);
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const E = [["path", { d: "M17 3.34a10 10 0 1 1 -14.995 8.984l-.005 -.324l.005 -.324a10 10 0 0 1 14.995 -8.336zm-1.293 5.953a1 1 0 0 0 -1.32 -.083l-.094 .083l-3.293 3.292l-1.293 -1.292l-.094 -.083a1 1 0 0 0 -1.403 1.403l.083 .094l2 2l.094 .083a1 1 0 0 0 1.226 0l.094 -.083l4 -4l.083 -.094a1 1 0 0 0 -.083 -1.32z", key: "svg-0" }]], D = x("filled", "circle-check-filled", "CircleCheckFilled", E);
function N(i) {
  var a, r, t = "";
  if (typeof i == "string" || typeof i == "number") t += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (a = 0; a < o; a++) i[a] && (r = N(i[a])) && (t && (t += " "), t += r);
  } else for (r in i) i[r] && (t && (t += " "), t += r);
  return t;
}
function G() {
  for (var i, a, r = 0, t = "", o = arguments.length; r < o; r++) (i = arguments[r]) && (a = N(i)) && (t && (t += " "), t += a);
  return t;
}
const R = ({
  as: i,
  children: a,
  display: r,
  m: t,
  mt: o,
  mr: l,
  mb: c,
  ml: m,
  mx: p,
  my: d,
  p: n,
  pt: g,
  pr: u,
  pb: f,
  pl: s,
  px: h,
  py: v,
  width: k,
  height: y,
  position: b,
  style: A,
  className: B,
  disabled: L,
  required: T,
  ...j
}) => {
  const q = i || "div", e = {};
  r && (e.display = r), k && (e.width = k), y && (e.height = y), b && (e.position = b), t && (e.margin = `var(--pittorica-space-${t})`), o && (e.marginTop = `var(--pittorica-space-${o})`), l && (e.marginRight = `var(--pittorica-space-${l})`), c && (e.marginBottom = `var(--pittorica-space-${c})`), m && (e.marginLeft = `var(--pittorica-space-${m})`), p && (e.marginLeft = `var(--pittorica-space-${p})`, e.marginRight = `var(--pittorica-space-${p})`), d && (e.marginTop = `var(--pittorica-space-${d})`, e.marginBottom = `var(--pittorica-space-${d})`), n && (e.padding = `var(--pittorica-space-${n})`), g && (e.paddingTop = `var(--pittorica-space-${g})`), u && (e.paddingRight = `var(--pittorica-space-${u})`), f && (e.paddingBottom = `var(--pittorica-space-${f})`), s && (e.paddingLeft = `var(--pittorica-space-${s})`), h && (e.paddingLeft = `var(--pittorica-space-${h})`, e.paddingRight = `var(--pittorica-space-${h})`), v && (e.paddingTop = `var(--pittorica-space-${v})`, e.paddingBottom = `var(--pittorica-space-${v})`);
  const W = {
    ...A,
    ...e
  };
  return /* @__PURE__ */ $(
    q,
    {
      className: G("pittorica-box", B),
      style: W,
      disabled: L,
      required: T,
      ...j,
      children: a
    }
  );
};
R.displayName = "Box";
const H = ({
  checked: i,
  defaultChecked: a,
  disabled: r,
  color: t = "source",
  className: o,
  style: l,
  onCheckedChange: c,
  name: m,
  required: p = !1,
  as: d,
  ...n
}) => {
  const [g, u] = z(
    a ?? !1
  ), f = i !== void 0, s = f ? i : g, h = t !== "inherit" && !(t != null && t.startsWith("#")) && !(t != null && t.startsWith("rgb")) ? `var(--pittorica-${t}-9)` : t, v = d || "button";
  return /* @__PURE__ */ $(
    R,
    {
      as: v,
      type: v === "button" ? "button" : void 0,
      role: "radio",
      name: m,
      "aria-checked": s,
      "aria-required": p,
      "data-state": s ? "checked" : "unchecked",
      disabled: r,
      onClick: (y) => {
        r || (f || u(!s), c == null || c(!s), typeof n.onClick == "function" && n.onClick(y));
      },
      className: I("pittorica-radio-root", o),
      style: {
        "--pittorica-source-color": h,
        ...l
      },
      ...n,
      children: s ? /* @__PURE__ */ $(D, { size: 14 }) : /* @__PURE__ */ $(S, { size: 14, opacity: 0.5 })
    }
  );
};
H.displayName = "Radio";
export {
  H as Radio
};
