// @pittorica/checkbox-group-react@0.26.3/dist/index.js
import { jsx as b, jsxs as z } from "react/jsx-runtime";
import Q, { forwardRef as U, createElement as B, createContext as D, useState as X, useMemo as Y, use as Z } from "react";
function j(a) {
  var i, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var o = a.length;
    for (i = 0; i < o; i++) a[i] && (r = j(a[i])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function V() {
  for (var a, i, r = 0, t = "", o = arguments.length; r < o; r++) (a = arguments[r]) && (i = j(a)) && (t && (t += " "), t += i);
  return t;
}
function G(a) {
  var i, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var o = a.length;
    for (i = 0; i < o; i++) a[i] && (r = G(a[i])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function tt() {
  for (var a, i, r = 0, t = "", o = arguments.length; r < o; r++) (a = arguments[r]) && (i = G(a)) && (t && (t += " "), t += i);
  return t;
}
const I = ({
  as: a,
  children: i,
  display: r,
  m: t,
  mt: o,
  mr: p,
  mb: c,
  ml: g,
  mx: n,
  my: l,
  p: d,
  pt: h,
  pr: u,
  pb: v,
  pl: m,
  px: s,
  py: f,
  width: $,
  height: x,
  position: y,
  style: k,
  className: w,
  disabled: N,
  required: C,
  ...T
}) => {
  const q = a || "div", e = {};
  r && (e.display = r), $ && (e.width = $), x && (e.height = x), y && (e.position = y), t && (e.margin = `var(--pittorica-space-${t})`), o && (e.marginTop = `var(--pittorica-space-${o})`), p && (e.marginRight = `var(--pittorica-space-${p})`), c && (e.marginBottom = `var(--pittorica-space-${c})`), g && (e.marginLeft = `var(--pittorica-space-${g})`), n && (e.marginLeft = `var(--pittorica-space-${n})`, e.marginRight = `var(--pittorica-space-${n})`), l && (e.marginTop = `var(--pittorica-space-${l})`, e.marginBottom = `var(--pittorica-space-${l})`), d && (e.padding = `var(--pittorica-space-${d})`), h && (e.paddingTop = `var(--pittorica-space-${h})`), u && (e.paddingRight = `var(--pittorica-space-${u})`), v && (e.paddingBottom = `var(--pittorica-space-${v})`), m && (e.paddingLeft = `var(--pittorica-space-${m})`), s && (e.paddingLeft = `var(--pittorica-space-${s})`, e.paddingRight = `var(--pittorica-space-${s})`), f && (e.paddingTop = `var(--pittorica-space-${f})`, e.paddingBottom = `var(--pittorica-space-${f})`);
  const R = {
    ...k,
    ...e
  };
  return /* @__PURE__ */ b(
    q,
    {
      className: tt("pittorica-box", w),
      style: R,
      disabled: N,
      required: C,
      ...T,
      children: i
    }
  );
};
I.displayName = "Box";
function W(a) {
  var i, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var o = a.length;
    for (i = 0; i < o; i++) a[i] && (r = W(a[i])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function at() {
  for (var a, i, r = 0, t = "", o = arguments.length; r < o; r++) (a = arguments[r]) && (i = W(a)) && (t && (t += " "), t += i);
  return t;
}
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var rt = {
  outline: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  },
  filled: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "currentColor",
    stroke: "none"
  }
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const S = (a, i, r, t) => {
  const o = U(
    ({ color: p = "currentColor", size: c = 24, stroke: g = 2, title: n, className: l, children: d, ...h }, u) => B(
      "svg",
      {
        ref: u,
        ...rt[a],
        width: c,
        height: c,
        className: ["tabler-icon", `tabler-icon-${i}`, l].join(" "),
        ...a === "filled" ? {
          fill: p
        } : {
          strokeWidth: g,
          stroke: p
        },
        ...h
      },
      [
        n && B("title", { key: "svg-title" }, n),
        ...t.map(([v, m]) => B(v, m)),
        ...Array.isArray(d) ? d : [d]
      ]
    )
  );
  return o.displayName = `${r}`, o;
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const et = [["path", { d: "M3 5a2 2 0 0 1 2 -2h14a2 2 0 0 1 2 2v14a2 2 0 0 1 -2 2h-14a2 2 0 0 1 -2 -2v-14", key: "svg-0" }]], it = S("outline", "square", "Square", et);
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const ot = [["path", { d: "M18.333 2c1.96 0 3.56 1.537 3.662 3.472l.005 .195v12.666c0 1.96 -1.537 3.56 -3.472 3.662l-.195 .005h-12.666a3.667 3.667 0 0 1 -3.662 -3.472l-.005 -.195v-12.666c0 -1.96 1.537 -3.56 3.472 -3.662l.195 -.005h12.666zm-2.626 7.293a1 1 0 0 0 -1.414 0l-3.293 3.292l-1.293 -1.292l-.094 -.083a1 1 0 0 0 -1.32 1.497l2 2l.094 .083a1 1 0 0 0 1.32 -.083l4 -4l.083 -.094a1 1 0 0 0 -.083 -1.32z", key: "svg-0" }]], ct = S("filled", "square-check-filled", "SquareCheckFilled", ot);
function M(a) {
  var i, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var o = a.length;
    for (i = 0; i < o; i++) a[i] && (r = M(a[i])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function st() {
  for (var a, i, r = 0, t = "", o = arguments.length; r < o; r++) (a = arguments[r]) && (i = M(a)) && (t && (t += " "), t += i);
  return t;
}
const E = ({
  as: a,
  children: i,
  display: r,
  m: t,
  mt: o,
  mr: p,
  mb: c,
  ml: g,
  mx: n,
  my: l,
  p: d,
  pt: h,
  pr: u,
  pb: v,
  pl: m,
  px: s,
  py: f,
  width: $,
  height: x,
  position: y,
  style: k,
  className: w,
  disabled: N,
  required: C,
  ...T
}) => {
  const q = a || "div", e = {};
  r && (e.display = r), $ && (e.width = $), x && (e.height = x), y && (e.position = y), t && (e.margin = `var(--pittorica-space-${t})`), o && (e.marginTop = `var(--pittorica-space-${o})`), p && (e.marginRight = `var(--pittorica-space-${p})`), c && (e.marginBottom = `var(--pittorica-space-${c})`), g && (e.marginLeft = `var(--pittorica-space-${g})`), n && (e.marginLeft = `var(--pittorica-space-${n})`, e.marginRight = `var(--pittorica-space-${n})`), l && (e.marginTop = `var(--pittorica-space-${l})`, e.marginBottom = `var(--pittorica-space-${l})`), d && (e.padding = `var(--pittorica-space-${d})`), h && (e.paddingTop = `var(--pittorica-space-${h})`), u && (e.paddingRight = `var(--pittorica-space-${u})`), v && (e.paddingBottom = `var(--pittorica-space-${v})`), m && (e.paddingLeft = `var(--pittorica-space-${m})`), s && (e.paddingLeft = `var(--pittorica-space-${s})`, e.paddingRight = `var(--pittorica-space-${s})`), f && (e.paddingTop = `var(--pittorica-space-${f})`, e.paddingBottom = `var(--pittorica-space-${f})`);
  const R = {
    ...k,
    ...e
  };
  return /* @__PURE__ */ b(
    q,
    {
      className: st("pittorica-box", w),
      style: R,
      disabled: N,
      required: C,
      ...T,
      children: i
    }
  );
};
E.displayName = "Box";
function F(a) {
  var i, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var o = a.length;
    for (i = 0; i < o; i++) a[i] && (r = F(a[i])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function A() {
  for (var a, i, r = 0, t = "", o = arguments.length; r < o; r++) (a = arguments[r]) && (i = F(a)) && (t && (t += " "), t += i);
  return t;
}
function O(a) {
  var i, r, t = "";
  if (typeof a == "string" || typeof a == "number") t += a;
  else if (typeof a == "object") if (Array.isArray(a)) {
    var o = a.length;
    for (i = 0; i < o; i++) a[i] && (r = O(a[i])) && (t && (t += " "), t += r);
  } else for (r in a) a[r] && (t && (t += " "), t += r);
  return t;
}
function nt() {
  for (var a, i, r = 0, t = "", o = arguments.length; r < o; r++) (a = arguments[r]) && (i = O(a)) && (t && (t += " "), t += i);
  return t;
}
const _ = ({
  as: a,
  children: i,
  display: r,
  m: t,
  mt: o,
  mr: p,
  mb: c,
  ml: g,
  mx: n,
  my: l,
  p: d,
  pt: h,
  pr: u,
  pb: v,
  pl: m,
  px: s,
  py: f,
  width: $,
  height: x,
  position: y,
  style: k,
  className: w,
  disabled: N,
  required: C,
  ...T
}) => {
  const q = a || "div", e = {};
  r && (e.display = r), $ && (e.width = $), x && (e.height = x), y && (e.position = y), t && (e.margin = `var(--pittorica-space-${t})`), o && (e.marginTop = `var(--pittorica-space-${o})`), p && (e.marginRight = `var(--pittorica-space-${p})`), c && (e.marginBottom = `var(--pittorica-space-${c})`), g && (e.marginLeft = `var(--pittorica-space-${g})`), n && (e.marginLeft = `var(--pittorica-space-${n})`, e.marginRight = `var(--pittorica-space-${n})`), l && (e.marginTop = `var(--pittorica-space-${l})`, e.marginBottom = `var(--pittorica-space-${l})`), d && (e.padding = `var(--pittorica-space-${d})`), h && (e.paddingTop = `var(--pittorica-space-${h})`), u && (e.paddingRight = `var(--pittorica-space-${u})`), v && (e.paddingBottom = `var(--pittorica-space-${v})`), m && (e.paddingLeft = `var(--pittorica-space-${m})`), s && (e.paddingLeft = `var(--pittorica-space-${s})`, e.paddingRight = `var(--pittorica-space-${s})`), f && (e.paddingTop = `var(--pittorica-space-${f})`, e.paddingBottom = `var(--pittorica-space-${f})`);
  const R = {
    ...k,
    ...e
  };
  return /* @__PURE__ */ b(
    q,
    {
      className: nt("pittorica-box", w),
      style: R,
      disabled: N,
      required: C,
      ...T,
      children: i
    }
  );
};
_.displayName = "Box";
const H = ({
  children: a,
  as: i,
  size: r,
  weight: t = "regular",
  align: o,
  truncate: p = !1,
  color: c,
  className: g,
  style: n,
  ...l
}) => {
  const d = i || "span", h = (c == null ? void 0 : c.startsWith("#")) || (c == null ? void 0 : c.startsWith("rgb")) || (c == null ? void 0 : c.startsWith("hsl")), u = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), v = (() => {
    if (c)
      return c === "inherit" ? "inherit" : h ? c : u.has(c) ? `var(--pittorica-${c}-9)` : c;
  })(), m = r ? typeof r == "string" ? `pittorica-text--size-${r}` : A({
    [`pittorica-text--size-${r.initial}`]: r.initial,
    [`pittorica-text--xs-size-${r.xs}`]: r.xs,
    [`pittorica-text--sm-size-${r.sm}`]: r.sm,
    [`pittorica-text--md-size-${r.md}`]: r.md,
    [`pittorica-text--lg-size-${r.lg}`]: r.lg,
    [`pittorica-text--xl-size-${r.xl}`]: r.xl
  }) : null, s = {
    ...n,
    textAlign: o,
    color: v
  };
  return /* @__PURE__ */ b(
    _,
    {
      as: d,
      className: A(
        "pittorica-text",
        m,
        { "pittorica-text--truncate": p },
        g
      ),
      "data-weight": t,
      style: s,
      ...l,
      children: a
    }
  );
};
H.displayName = "Text";
const J = ({
  checked: a,
  defaultChecked: i,
  onChange: r,
  color: t = "source",
  disabled: o = !1,
  label: p,
  name: c,
  value: g,
  className: n,
  style: l,
  required: d = !1,
  as: h,
  ...u
}) => {
  const [v, m] = Q.useState(
    i ?? !1
  ), s = a === void 0 ? v : a, f = t !== "inherit" && !(t != null && t.startsWith("#")) && !(t != null && t.startsWith("rgb")) ? `var(--pittorica-${t}-9)` : t, $ = (x) => {
    if (o) return;
    const y = x.target.checked;
    m(y), r == null || r(y);
  };
  return /* @__PURE__ */ z(
    E,
    {
      as: h || "label",
      className: at("pittorica-checkbox-root", n),
      "data-disabled": o,
      style: { "--_checkbox-color": f, ...l },
      ...u,
      children: [
        /* @__PURE__ */ b(
          "input",
          {
            type: "checkbox",
            className: "pittorica-checkbox-input",
            name: c,
            value: g,
            checked: s,
            disabled: o,
            required: d,
            onChange: $
          }
        ),
        /* @__PURE__ */ b("div", { className: "pittorica-checkbox-control", "aria-hidden": "true", children: s ? /* @__PURE__ */ b(ct, { size: 20 }) : /* @__PURE__ */ b(it, { size: 20, stroke: 1.5 }) }),
        p && /* @__PURE__ */ z(H, { weight: "medium", style: { color: "inherit" }, children: [
          p,
          " ",
          d && /* @__PURE__ */ b("span", { "aria-hidden": "true", children: "*" })
        ] })
      ]
    }
  );
};
J.displayName = "Checkbox";
const K = D(
  null
), L = ({
  value: a,
  defaultValue: i = [],
  onValueChange: r,
  orientation: t = "vertical",
  color: o,
  disabled: p,
  name: c,
  required: g,
  children: n,
  className: l,
  as: d,
  ...h
}) => {
  const [u, v] = X(i), m = a !== void 0, s = m ? a : u, f = (y) => {
    const k = s.includes(y) ? s.filter((w) => w !== y) : [...s, y];
    m || v(k), r == null || r(k);
  }, $ = Y(
    () => ({
      value: s,
      onItemChange: f,
      color: o,
      disabled: p,
      name: c,
      required: g
    }),
    [s, o, p, c, g]
  );
  return /* @__PURE__ */ b(K, { value: $, children: /* @__PURE__ */ b(
    I,
    {
      as: d || "div",
      className: V("pittorica-checkbox-group-root", l),
      "data-orientation": t,
      role: "group",
      ...h,
      children: n
    }
  ) });
}, P = ({
  value: a,
  as: i,
  ...r
}) => {
  const t = Z(K);
  if (!t)
    throw new Error(
      "CheckboxGroup.Item must be used within CheckboxGroup.Root"
    );
  return /* @__PURE__ */ b(
    J,
    {
      as: i || "label",
      ...r,
      name: t.name,
      value: a,
      checked: t.value.includes(a),
      onChange: () => t.onItemChange(a),
      color: r.color || t.color,
      disabled: r.disabled || t.disabled,
      required: r.required || t.required
    }
  );
}, dt = Object.assign(L, {
  Root: L,
  Item: P,
  displayName: "CheckboxGroup"
});
P.displayName = "CheckboxGroup.Item";
export {
  dt as CheckboxGroup
};
