// prettier-config-spellbookx@0.10.0/dist/index.js
//#region src/configs/astro.ts
const astro = {
	plugins: [
		"prettier-plugin-astro",
		"prettier-plugin-sh",
		"prettier-plugin-toml",
		"prettier-plugin-ini",
		"prettier-plugin-packagejson",
		"prettier-plugin-properties",
		"@prettier/plugin-xml"
	],
	trailingComma: "es5",
	tabWidth: 2,
	semi: true,
	singleQuote: true,
	printWidth: 80,
	endOfLine: "lf",
	bracketSpacing: true,
	overrides: [
		{
			files: "**/.czrc",
			options: { parser: "json" }
		},
		{
			files: ["*.toml"],
			options: { printWidth: 100 }
		},
		{
			files: ["*.astro"],
			options: { parser: "astro" }
		},
		{
			files: [
				"**/*.xml",
				"**/*.svg",
				"**/*.xhtml",
				"**/*.xsd",
				"**/*.xsl",
				"**/*.xslt",
				"**/*.plist",
				"**/*.axml",
				"**/*.xaml",
				"**/*.config",
				"**/*.resx",
				"**/*.csproj",
				"**/*.vbproj"
			],
			options: { parser: "xml" }
		}
	]
};

//#endregion
//#region src/configs/astro-prisma.ts
const astroPrisma = {
	plugins: [
		"prettier-plugin-astro",
		"prettier-plugin-sh",
		"prettier-plugin-toml",
		"prettier-plugin-ini",
		"prettier-plugin-packagejson",
		"prettier-plugin-properties",
		"prettier-plugin-prisma",
		"@prettier/plugin-xml"
	],
	trailingComma: "es5",
	tabWidth: 2,
	semi: true,
	singleQuote: true,
	printWidth: 80,
	endOfLine: "lf",
	bracketSpacing: true,
	overrides: [
		{
			files: "**/.czrc",
			options: { parser: "json" }
		},
		{
			files: ["*.toml"],
			options: { printWidth: 100 }
		},
		{
			files: ["*.astro"],
			options: { parser: "astro" }
		},
		{
			files: [
				"**/*.xml",
				"**/*.svg",
				"**/*.xhtml",
				"**/*.xsd",
				"**/*.xsl",
				"**/*.xslt",
				"**/*.plist",
				"**/*.axml",
				"**/*.xaml",
				"**/*.config",
				"**/*.resx",
				"**/*.csproj",
				"**/*.vbproj"
			],
			options: { parser: "xml" }
		}
	]
};

//#endregion
//#region src/configs/astro-prisma-tailwind.ts
const astroPrismaTailwind = {
	plugins: [
		"prettier-plugin-astro",
		"prettier-plugin-sh",
		"prettier-plugin-toml",
		"prettier-plugin-ini",
		"prettier-plugin-packagejson",
		"prettier-plugin-properties",
		"prettier-plugin-prisma",
		"@prettier/plugin-xml",
		"prettier-plugin-tailwindcss"
	],
	trailingComma: "es5",
	tabWidth: 2,
	semi: true,
	singleQuote: true,
	printWidth: 80,
	endOfLine: "lf",
	bracketSpacing: true,
	overrides: [
		{
			files: "**/.czrc",
			options: { parser: "json" }
		},
		{
			files: ["*.toml"],
			options: { printWidth: 100 }
		},
		{
			files: ["*.astro"],
			options: { parser: "astro" }
		},
		{
			files: [
				"**/*.xml",
				"**/*.svg",
				"**/*.xhtml",
				"**/*.xsd",
				"**/*.xsl",
				"**/*.xslt",
				"**/*.plist",
				"**/*.axml",
				"**/*.xaml",
				"**/*.config",
				"**/*.resx",
				"**/*.csproj",
				"**/*.vbproj"
			],
			options: { parser: "xml" }
		}
	]
};

//#endregion
//#region src/configs/astro-tailwind.ts
const astroTailwind = {
	plugins: [
		"prettier-plugin-astro",
		"prettier-plugin-sh",
		"prettier-plugin-toml",
		"prettier-plugin-ini",
		"prettier-plugin-packagejson",
		"prettier-plugin-properties",
		"@prettier/plugin-xml",
		"prettier-plugin-tailwindcss"
	],
	trailingComma: "es5",
	tabWidth: 2,
	semi: true,
	singleQuote: true,
	printWidth: 80,
	endOfLine: "lf",
	bracketSpacing: true,
	overrides: [
		{
			files: "**/.czrc",
			options: { parser: "json" }
		},
		{
			files: ["*.toml"],
			options: { printWidth: 100 }
		},
		{
			files: ["*.astro"],
			options: { parser: "astro" }
		},
		{
			files: [
				"**/*.xml",
				"**/*.svg",
				"**/*.xhtml",
				"**/*.xsd",
				"**/*.xsl",
				"**/*.xslt",
				"**/*.plist",
				"**/*.axml",
				"**/*.xaml",
				"**/*.config",
				"**/*.resx",
				"**/*.csproj",
				"**/*.vbproj"
			],
			options: { parser: "xml" }
		}
	]
};

//#endregion
//#region src/configs/base.ts
const base = {
	plugins: [
		"prettier-plugin-sh",
		"prettier-plugin-toml",
		"prettier-plugin-ini",
		"prettier-plugin-packagejson",
		"prettier-plugin-properties",
		"@prettier/plugin-xml"
	],
	trailingComma: "es5",
	tabWidth: 2,
	semi: true,
	singleQuote: true,
	printWidth: 80,
	endOfLine: "lf",
	bracketSpacing: true,
	overrides: [
		{
			files: "**/.czrc",
			options: { parser: "json" }
		},
		{
			files: ["*.toml"],
			options: { printWidth: 100 }
		},
		{
			files: ["*.astro"],
			options: { parser: "astro" }
		},
		{
			files: [
				"**/*.xml",
				"**/*.svg",
				"**/*.xhtml",
				"**/*.xsd",
				"**/*.xsl",
				"**/*.xslt",
				"**/*.plist",
				"**/*.axml",
				"**/*.xaml",
				"**/*.config",
				"**/*.resx",
				"**/*.csproj",
				"**/*.vbproj"
			],
			options: { parser: "xml" }
		}
	]
};

//#endregion
//#region src/configs/prisma.ts
const prisma = {
	plugins: [
		"prettier-plugin-sh",
		"prettier-plugin-toml",
		"prettier-plugin-ini",
		"prettier-plugin-packagejson",
		"prettier-plugin-properties",
		"prettier-plugin-prisma",
		"@prettier/plugin-xml"
	],
	trailingComma: "es5",
	tabWidth: 2,
	semi: true,
	singleQuote: true,
	printWidth: 80,
	endOfLine: "lf",
	bracketSpacing: true,
	overrides: [
		{
			files: "**/.czrc",
			options: { parser: "json" }
		},
		{
			files: ["*.toml"],
			options: { printWidth: 100 }
		},
		{
			files: ["*.astro"],
			options: { parser: "astro" }
		},
		{
			files: [
				"**/*.xml",
				"**/*.svg",
				"**/*.xhtml",
				"**/*.xsd",
				"**/*.xsl",
				"**/*.xslt",
				"**/*.plist",
				"**/*.axml",
				"**/*.xaml",
				"**/*.config",
				"**/*.resx",
				"**/*.csproj",
				"**/*.vbproj"
			],
			options: { parser: "xml" }
		}
	]
};

//#endregion
//#region src/configs/prisma-tailwind.ts
const prismaTailwind = {
	plugins: [
		"prettier-plugin-sh",
		"prettier-plugin-toml",
		"prettier-plugin-ini",
		"prettier-plugin-packagejson",
		"prettier-plugin-properties",
		"prettier-plugin-prisma",
		"@prettier/plugin-xml",
		"prettier-plugin-tailwindcss"
	],
	trailingComma: "es5",
	tabWidth: 2,
	semi: true,
	singleQuote: true,
	printWidth: 80,
	endOfLine: "lf",
	bracketSpacing: true,
	overrides: [
		{
			files: "**/.czrc",
			options: { parser: "json" }
		},
		{
			files: ["*.toml"],
			options: { printWidth: 100 }
		},
		{
			files: ["*.astro"],
			options: { parser: "astro" }
		},
		{
			files: [
				"**/*.xml",
				"**/*.svg",
				"**/*.xhtml",
				"**/*.xsd",
				"**/*.xsl",
				"**/*.xslt",
				"**/*.plist",
				"**/*.axml",
				"**/*.xaml",
				"**/*.config",
				"**/*.resx",
				"**/*.csproj",
				"**/*.vbproj"
			],
			options: { parser: "xml" }
		}
	]
};

//#endregion
//#region src/configs/tailwind.ts
const tailwind = {
	plugins: [
		"prettier-plugin-sh",
		"prettier-plugin-toml",
		"prettier-plugin-ini",
		"prettier-plugin-packagejson",
		"prettier-plugin-properties",
		"@prettier/plugin-xml",
		"prettier-plugin-tailwindcss"
	],
	trailingComma: "es5",
	tabWidth: 2,
	semi: true,
	singleQuote: true,
	printWidth: 80,
	endOfLine: "lf",
	bracketSpacing: true,
	overrides: [
		{
			files: "**/.czrc",
			options: { parser: "json" }
		},
		{
			files: ["*.toml"],
			options: { printWidth: 100 }
		},
		{
			files: ["*.astro"],
			options: { parser: "astro" }
		},
		{
			files: [
				"**/*.xml",
				"**/*.svg",
				"**/*.xhtml",
				"**/*.xsd",
				"**/*.xsl",
				"**/*.xslt",
				"**/*.plist",
				"**/*.axml",
				"**/*.xaml",
				"**/*.config",
				"**/*.resx",
				"**/*.csproj",
				"**/*.vbproj"
			],
			options: { parser: "xml" }
		}
	]
};

//#endregion
//#region src/index.ts
const configs = {
	base,
	astro,
	"astro-prisma": astroPrisma,
	"astro-tailwind": astroTailwind,
	"astro-prisma-tailwind": astroPrismaTailwind,
	tailwind,
	prisma,
	"prisma-tailwind": prismaTailwind
};

//#endregion
export { astro, astroPrisma, astroPrismaTailwind, astroTailwind, base, configs as default };
