// @pittorica/toast-react@0.26.3/dist/index.js
import { jsx as g, jsxs as j } from "react/jsx-runtime";
import { forwardRef as F, createElement as W, useState as C, useRef as G, useEffect as S, useCallback as H } from "react";
import { createPortal as J } from "react-dom";
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var K = {
  outline: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  },
  filled: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "currentColor",
    stroke: "none"
  }
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const Q = (i, e, t, r) => {
  const o = F(
    ({ color: p = "currentColor", size: s = 24, stroke: n = 2, title: c, className: d, children: l, ...m }, v) => W(
      "svg",
      {
        ref: v,
        ...K[i],
        width: s,
        height: s,
        className: ["tabler-icon", `tabler-icon-${e}`, d].join(" "),
        strokeWidth: n,
        stroke: p,
        ...m
      },
      [
        c && W("title", { key: "svg-title" }, c),
        ...r.map(([f, h]) => W(f, h)),
        ...Array.isArray(l) ? l : [l]
      ]
    )
  );
  return o.displayName = `${t}`, o;
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const U = [["path", { d: "M18 6l-12 12", key: "svg-0" }], ["path", { d: "M6 6l12 12", key: "svg-1" }]], Y = Q("outline", "x", "X", U);
function _(i) {
  var e, t, r = "";
  if (typeof i == "string" || typeof i == "number") r += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (e = 0; e < o; e++) i[e] && (t = _(i[e])) && (r && (r += " "), r += t);
  } else for (t in i) i[t] && (r && (r += " "), r += t);
  return r;
}
function Z() {
  for (var i, e, t = 0, r = "", o = arguments.length; t < o; t++) (i = arguments[t]) && (e = _(i)) && (r && (r += " "), r += e);
  return r;
}
const L = ({
  as: i,
  children: e,
  display: t,
  m: r,
  mt: o,
  mr: p,
  mb: s,
  ml: n,
  mx: c,
  my: d,
  p: l,
  pt: m,
  pr: v,
  pb: f,
  pl: h,
  px: u,
  py: $,
  width: y,
  height: b,
  position: x,
  style: w,
  className: T,
  disabled: N,
  required: B,
  ...k
}) => {
  const A = i || "div", a = {};
  t && (a.display = t), y && (a.width = y), b && (a.height = b), x && (a.position = x), r && (a.margin = `var(--pittorica-space-${r})`), o && (a.marginTop = `var(--pittorica-space-${o})`), p && (a.marginRight = `var(--pittorica-space-${p})`), s && (a.marginBottom = `var(--pittorica-space-${s})`), n && (a.marginLeft = `var(--pittorica-space-${n})`), c && (a.marginLeft = `var(--pittorica-space-${c})`, a.marginRight = `var(--pittorica-space-${c})`), d && (a.marginTop = `var(--pittorica-space-${d})`, a.marginBottom = `var(--pittorica-space-${d})`), l && (a.padding = `var(--pittorica-space-${l})`), m && (a.paddingTop = `var(--pittorica-space-${m})`), v && (a.paddingRight = `var(--pittorica-space-${v})`), f && (a.paddingBottom = `var(--pittorica-space-${f})`), h && (a.paddingLeft = `var(--pittorica-space-${h})`), u && (a.paddingLeft = `var(--pittorica-space-${u})`, a.paddingRight = `var(--pittorica-space-${u})`), $ && (a.paddingTop = `var(--pittorica-space-${$})`, a.paddingBottom = `var(--pittorica-space-${$})`);
  const R = {
    ...w,
    ...a
  };
  return /* @__PURE__ */ g(
    A,
    {
      className: Z("pittorica-box", T),
      style: R,
      disabled: N,
      required: B,
      ...k,
      children: e
    }
  );
};
L.displayName = "Box";
function I(i) {
  var e, t, r = "";
  if (typeof i == "string" || typeof i == "number") r += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (e = 0; e < o; e++) i[e] && (t = I(i[e])) && (r && (r += " "), r += t);
  } else for (t in i) i[t] && (r && (r += " "), r += t);
  return r;
}
function tt() {
  for (var i, e, t = 0, r = "", o = arguments.length; t < o; t++) (i = arguments[t]) && (e = I(i)) && (r && (r += " "), r += e);
  return r;
}
function M(i) {
  var e, t, r = "";
  if (typeof i == "string" || typeof i == "number") r += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (e = 0; e < o; e++) i[e] && (t = M(i[e])) && (r && (r += " "), r += t);
  } else for (t in i) i[t] && (r && (r += " "), r += t);
  return r;
}
function it() {
  for (var i, e, t = 0, r = "", o = arguments.length; t < o; t++) (i = arguments[t]) && (e = M(i)) && (r && (r += " "), r += e);
  return r;
}
const P = ({
  as: i,
  children: e,
  display: t,
  m: r,
  mt: o,
  mr: p,
  mb: s,
  ml: n,
  mx: c,
  my: d,
  p: l,
  pt: m,
  pr: v,
  pb: f,
  pl: h,
  px: u,
  py: $,
  width: y,
  height: b,
  position: x,
  style: w,
  className: T,
  disabled: N,
  required: B,
  ...k
}) => {
  const A = i || "div", a = {};
  t && (a.display = t), y && (a.width = y), b && (a.height = b), x && (a.position = x), r && (a.margin = `var(--pittorica-space-${r})`), o && (a.marginTop = `var(--pittorica-space-${o})`), p && (a.marginRight = `var(--pittorica-space-${p})`), s && (a.marginBottom = `var(--pittorica-space-${s})`), n && (a.marginLeft = `var(--pittorica-space-${n})`), c && (a.marginLeft = `var(--pittorica-space-${c})`, a.marginRight = `var(--pittorica-space-${c})`), d && (a.marginTop = `var(--pittorica-space-${d})`, a.marginBottom = `var(--pittorica-space-${d})`), l && (a.padding = `var(--pittorica-space-${l})`), m && (a.paddingTop = `var(--pittorica-space-${m})`), v && (a.paddingRight = `var(--pittorica-space-${v})`), f && (a.paddingBottom = `var(--pittorica-space-${f})`), h && (a.paddingLeft = `var(--pittorica-space-${h})`), u && (a.paddingLeft = `var(--pittorica-space-${u})`, a.paddingRight = `var(--pittorica-space-${u})`), $ && (a.paddingTop = `var(--pittorica-space-${$})`, a.paddingBottom = `var(--pittorica-space-${$})`);
  const R = {
    ...w,
    ...a
  };
  return /* @__PURE__ */ g(
    A,
    {
      className: it("pittorica-box", T),
      style: R,
      disabled: N,
      required: B,
      ...k,
      children: e
    }
  );
};
P.displayName = "Box";
const X = ({
  children: i,
  variant: e = "filled",
  color: t = "source",
  size: r = "3",
  className: o,
  style: p,
  as: s,
  ...n
}) => {
  const c = t !== "inherit" && !(t != null && t.startsWith("#")) && !(t != null && t.startsWith("rgb")), d = {
    ...p,
    "--pittorica-source-color": c ? `var(--pittorica-${t}-9)` : t,
    "--pittorica-on-source-color": "#ffffff"
  }, l = s || (n.href ? "a" : "button");
  return /* @__PURE__ */ g(
    P,
    {
      as: l,
      className: tt(
        "pittorica-icon-button",
        `pittorica-icon-button--${e}`,
        `pittorica-icon-button--${r}`,
        o
      ),
      style: d,
      ...n,
      children: i
    }
  );
};
X.displayName = "IconButton";
function O(i) {
  var e, t, r = "";
  if (typeof i == "string" || typeof i == "number") r += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (e = 0; e < o; e++) i[e] && (t = O(i[e])) && (r && (r += " "), r += t);
  } else for (t in i) i[t] && (r && (r += " "), r += t);
  return r;
}
function q() {
  for (var i, e, t = 0, r = "", o = arguments.length; t < o; t++) (i = arguments[t]) && (e = O(i)) && (r && (r += " "), r += e);
  return r;
}
function V(i) {
  var e, t, r = "";
  if (typeof i == "string" || typeof i == "number") r += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var o = i.length;
    for (e = 0; e < o; e++) i[e] && (t = V(i[e])) && (r && (r += " "), r += t);
  } else for (t in i) i[t] && (r && (r += " "), r += t);
  return r;
}
function rt() {
  for (var i, e, t = 0, r = "", o = arguments.length; t < o; t++) (i = arguments[t]) && (e = V(i)) && (r && (r += " "), r += e);
  return r;
}
const D = ({
  as: i,
  children: e,
  display: t,
  m: r,
  mt: o,
  mr: p,
  mb: s,
  ml: n,
  mx: c,
  my: d,
  p: l,
  pt: m,
  pr: v,
  pb: f,
  pl: h,
  px: u,
  py: $,
  width: y,
  height: b,
  position: x,
  style: w,
  className: T,
  disabled: N,
  required: B,
  ...k
}) => {
  const A = i || "div", a = {};
  t && (a.display = t), y && (a.width = y), b && (a.height = b), x && (a.position = x), r && (a.margin = `var(--pittorica-space-${r})`), o && (a.marginTop = `var(--pittorica-space-${o})`), p && (a.marginRight = `var(--pittorica-space-${p})`), s && (a.marginBottom = `var(--pittorica-space-${s})`), n && (a.marginLeft = `var(--pittorica-space-${n})`), c && (a.marginLeft = `var(--pittorica-space-${c})`, a.marginRight = `var(--pittorica-space-${c})`), d && (a.marginTop = `var(--pittorica-space-${d})`, a.marginBottom = `var(--pittorica-space-${d})`), l && (a.padding = `var(--pittorica-space-${l})`), m && (a.paddingTop = `var(--pittorica-space-${m})`), v && (a.paddingRight = `var(--pittorica-space-${v})`), f && (a.paddingBottom = `var(--pittorica-space-${f})`), h && (a.paddingLeft = `var(--pittorica-space-${h})`), u && (a.paddingLeft = `var(--pittorica-space-${u})`, a.paddingRight = `var(--pittorica-space-${u})`), $ && (a.paddingTop = `var(--pittorica-space-${$})`, a.paddingBottom = `var(--pittorica-space-${$})`);
  const R = {
    ...w,
    ...a
  };
  return /* @__PURE__ */ g(
    A,
    {
      className: rt("pittorica-box", T),
      style: R,
      disabled: N,
      required: B,
      ...k,
      children: e
    }
  );
};
D.displayName = "Box";
const z = ({
  children: i,
  as: e,
  size: t,
  weight: r = "regular",
  align: o,
  truncate: p = !1,
  color: s,
  className: n,
  style: c,
  ...d
}) => {
  const l = e || "span", m = (s == null ? void 0 : s.startsWith("#")) || (s == null ? void 0 : s.startsWith("rgb")) || (s == null ? void 0 : s.startsWith("hsl")), v = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), f = (() => {
    if (s)
      return s === "inherit" ? "inherit" : m ? s : v.has(s) ? `var(--pittorica-${s}-9)` : s;
  })(), h = t ? typeof t == "string" ? `pittorica-text--size-${t}` : q({
    [`pittorica-text--size-${t.initial}`]: t.initial,
    [`pittorica-text--xs-size-${t.xs}`]: t.xs,
    [`pittorica-text--sm-size-${t.sm}`]: t.sm,
    [`pittorica-text--md-size-${t.md}`]: t.md,
    [`pittorica-text--lg-size-${t.lg}`]: t.lg,
    [`pittorica-text--xl-size-${t.xl}`]: t.xl
  }) : null, u = {
    ...c,
    textAlign: o,
    color: f
  };
  return /* @__PURE__ */ g(
    D,
    {
      as: l,
      className: q(
        "pittorica-text",
        h,
        { "pittorica-text--truncate": p },
        n
      ),
      "data-weight": r,
      style: u,
      ...d,
      children: i
    }
  );
};
z.displayName = "Text";
const at = ({
  children: i,
  sourceColor: e = "indigo",
  appearance: t = "light",
  radius: r = "medium",
  className: o,
  style: p,
  id: s
}) => {
  const n = e.startsWith("#") || e.startsWith("rgb"), c = {
    id: s,
    className: ["pittorica-theme", o].filter(Boolean).join(" "),
    "data-source-color": n ? "custom" : e,
    "data-appearance": t,
    "data-radius": r,
    style: {
      ...p,
      ...n ? { "--pittorica-source-color": e } : {}
    }
  };
  return /* @__PURE__ */ g("div", { ...c, children: i });
}, E = "pittorica-toast", ct = (i) => {
  const e = Math.random().toString(36).slice(2, 9), t = new CustomEvent(E, { detail: { ...i, id: e } });
  globalThis.dispatchEvent(t);
}, et = ({
  t: i,
  onRemove: e
}) => {
  const t = i.color || "slate", r = t !== "inherit" && !(t != null && t.startsWith("#")) && !(t != null && t.startsWith("rgb")), o = r ? `var(--pittorica-${t}-3)` : `color-mix(in srgb, ${t} 12%, var(--pittorica-white))`, p = r ? `var(--pittorica-${t}-11)` : `color-mix(in srgb, ${t} 80%, var(--pittorica-black))`, s = r ? `var(--pittorica-${t}-6)` : `color-mix(in srgb, ${t} 20%, transparent)`, n = i.duration || 3e3;
  return /* @__PURE__ */ j(
    L,
    {
      className: "pittorica-toast-root",
      role: "status",
      "aria-live": "polite",
      style: {
        "--_toast-bg": o,
        "--_toast-text": p,
        "--_toast-border": s
      },
      children: [
        n > 0 && /* @__PURE__ */ g(
          "div",
          {
            className: "pittorica-toast-progress",
            style: {
              animation: `pittorica-toast-progress-shrink ${n}ms linear forwards`
            }
          }
        ),
        /* @__PURE__ */ j(L, { className: "pittorica-toast-content", children: [
          /* @__PURE__ */ g(z, { weight: "bold", style: { display: "block" }, children: i.title }),
          i.description && /* @__PURE__ */ g(
            z,
            {
              color: "inherit",
              size: "2",
              style: { display: "block", opacity: 0.8 },
              children: i.description
            }
          )
        ] }),
        /* @__PURE__ */ g(
          X,
          {
            variant: "text",
            color: "inherit",
            size: "1",
            onClick: () => e(i.id),
            "aria-label": "Close",
            children: /* @__PURE__ */ g(Y, { size: 18 })
          }
        )
      ]
    }
  );
}, pt = () => {
  const [i, e] = C([]), [t, r] = C(!1), o = G(/* @__PURE__ */ new Set());
  S(() => {
    r(!0);
  }, []);
  const p = H((n) => {
    e((c) => c.filter((d) => d.id !== n));
  }, []);
  if (S(() => {
    if (!t) return;
    const n = (c) => {
      const l = c.detail;
      if (e((m) => [...m, l]), l.duration !== 0) {
        const m = setTimeout(() => {
          p(l.id), o.current.delete(m);
        }, l.duration || 3e3);
        o.current.add(m);
      }
    };
    return globalThis.addEventListener(E, n), () => {
      globalThis.removeEventListener(E, n);
      for (const c of o.current) clearTimeout(c);
      o.current.clear();
    };
  }, [p, t]), !t || typeof document > "u") return null;
  const s = document.documentElement.dataset.appearance || document.body.dataset.appearance || "light";
  return J(
    /* @__PURE__ */ g(
      L,
      {
        className: "pittorica-toast-viewport",
        role: "region",
        "aria-label": "Notifications",
        children: /* @__PURE__ */ g(at, { appearance: s, children: i.map((n) => /* @__PURE__ */ g(et, { t: n, onRemove: p }, n.id)) })
      }
    ),
    document.body
  );
};
export {
  pt as ToastProvider,
  ct as toast
};
