// eslint-plugin-package-json@0.89.1/lib/plugin.mjs
import { rule } from "./rules/bin-name-casing.mjs";
import { rule as rule$1 } from "./rules/exports-subpaths-style.mjs";
import { rule as rule$2 } from "./rules/no-empty-fields.mjs";
import { rule as rule$3 } from "./rules/no-redundant-files.mjs";
import { rule as rule$4 } from "./rules/no-redundant-publishConfig.mjs";
import { rule as rule$5 } from "./rules/order-properties.mjs";
import { rule as rule$6 } from "./rules/repository-shorthand.mjs";
import { rule as rule$7 } from "./rules/require-attribution.mjs";
import { rules as rules$1 } from "./rules/require-properties.mjs";
import { rule as rule$8 } from "./rules/restrict-dependency-ranges.mjs";
import { rule as rule$9 } from "./rules/restrict-private-properties.mjs";
import { rule as rule$10 } from "./rules/scripts-name-casing.mjs";
import { rule as rule$11 } from "./rules/sort-collections.mjs";
import { rule as rule$12 } from "./rules/specify-peers-locally.mjs";
import { rule as rule$13 } from "./rules/unique-dependencies.mjs";
import { rule as rule$14 } from "./rules/valid-name.mjs";
import { rule as rule$15 } from "./rules/valid-package-definition.mjs";
import { rules as rules$2 } from "./rules/valid-properties.mjs";
import { rule as rule$16 } from "./rules/valid-repository-directory.mjs";
import { rule as rule$17 } from "./rules/valid-version.mjs";
import { createRequire } from "node:module";
import * as parserJsonc from "jsonc-eslint-parser";

//#region src/plugin.ts
const require = createRequire(import.meta.url);
const rules = {
	"bin-name-casing": rule,
	"exports-subpaths-style": rule$1,
	"no-empty-fields": rule$2,
	"no-redundant-files": rule$3,
	"no-redundant-publishConfig": rule$4,
	"order-properties": rule$5,
	"require-attribution": rule$7,
	...rules$1,
	"repository-shorthand": rule$6,
	"restrict-dependency-ranges": rule$8,
	"restrict-private-properties": rule$9,
	"scripts-name-casing": rule$10,
	"sort-collections": rule$11,
	"specify-peers-locally": rule$12,
	"unique-dependencies": rule$13,
	...rules$2,
	"valid-name": rule$14,
	"valid-package-definition": rule$15,
	"valid-repository-directory": rule$16,
	"valid-version": rule$17
};
const recommendedRules = { ...Object.fromEntries(Object.entries(rules).filter(([, rule]) => rule.meta.docs?.recommended).map(([name]) => ["package-json/" + name, "error"])) };
const recommendedPublishableRules = {
	...recommendedRules,
	...Object.fromEntries(Object.entries(rules).filter(([, rule]) => rule.meta.docs?.category === "Publishable").map(([name]) => ["package-json/" + name, "error"]))
};
const stylisticRules = { ...Object.fromEntries(Object.entries(rules).filter(([, rule]) => rule.meta.docs?.category === "Stylistic").map(([name]) => ["package-json/" + name, "error"])) };
const { name, version } = require("../package.json");
const plugin = {
	configs: {
		"legacy-recommended": {
			plugins: ["package-json"],
			rules: recommendedRules
		},
		recommended: {
			files: ["**/package.json"],
			languageOptions: { parser: parserJsonc },
			name: "package-json/recommended",
			plugins: { get "package-json"() {
				return plugin;
			} },
			rules: recommendedRules
		},
		"recommended-publishable": {
			files: ["**/package.json"],
			languageOptions: { parser: parserJsonc },
			name: "package-json/recommended-publishable",
			plugins: { get "package-json"() {
				return plugin;
			} },
			rules: recommendedPublishableRules
		},
		stylistic: {
			files: ["**/package.json"],
			languageOptions: { parser: parserJsonc },
			name: "package-json/stylistic",
			plugins: { get "package-json"() {
				return plugin;
			} },
			rules: stylisticRules
		}
	},
	meta: {
		name,
		version
	},
	rules
};

//#endregion
export { plugin };
