// @pittorica/select-react@0.26.3/dist/index.js
import { jsx as f, jsxs as B } from "react/jsx-runtime";
import { forwardRef as J, createElement as L, createContext as K, useId as W, use as P } from "react";
function I(t) {
  var r, e, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (r = 0; r < o; r++) t[r] && (e = I(t[r])) && (i && (i += " "), i += e);
  } else for (e in t) t[e] && (i && (i += " "), i += e);
  return i;
}
function z() {
  for (var t, r, e = 0, i = "", o = arguments.length; e < o; e++) (t = arguments[e]) && (r = I(t)) && (i && (i += " "), i += r);
  return i;
}
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
var Q = {
  outline: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: 2,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  },
  filled: {
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "currentColor",
    stroke: "none"
  }
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const U = (t, r, e, i) => {
  const o = J(
    ({ color: c = "currentColor", size: n = 24, stroke: p = 2, title: s, className: l, children: d, ...v }, h) => L(
      "svg",
      {
        ref: h,
        ...Q[t],
        width: n,
        height: n,
        className: ["tabler-icon", `tabler-icon-${r}`, l].join(" "),
        strokeWidth: p,
        stroke: c,
        ...v
      },
      [
        s && L("title", { key: "svg-title" }, s),
        ...i.map(([g, m]) => L(g, m)),
        ...Array.isArray(d) ? d : [d]
      ]
    )
  );
  return o.displayName = `${e}`, o;
};
/**
 * @license @tabler/icons-react v3.36.1 - MIT
 *
 * This source code is licensed under the MIT license.
 * See the LICENSE file in the root directory of this source tree.
 */
const V = [["path", { d: "M6 9l6 6l6 -6", key: "svg-0" }]], X = U("outline", "chevron-down", "ChevronDown", V);
function D(t) {
  var r, e, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (r = 0; r < o; r++) t[r] && (e = D(t[r])) && (i && (i += " "), i += e);
  } else for (e in t) t[e] && (i && (i += " "), i += e);
  return i;
}
function Y() {
  for (var t, r, e = 0, i = "", o = arguments.length; e < o; e++) (t = arguments[e]) && (r = D(t)) && (i && (i += " "), i += r);
  return i;
}
const w = ({
  as: t,
  children: r,
  display: e,
  m: i,
  mt: o,
  mr: c,
  mb: n,
  ml: p,
  mx: s,
  my: l,
  p: d,
  pt: v,
  pr: h,
  pb: g,
  pl: m,
  px: u,
  py: y,
  width: $,
  height: x,
  position: b,
  style: N,
  className: S,
  disabled: T,
  required: C,
  ...R
}) => {
  const A = t || "div", a = {};
  e && (a.display = e), $ && (a.width = $), x && (a.height = x), b && (a.position = b), i && (a.margin = `var(--pittorica-space-${i})`), o && (a.marginTop = `var(--pittorica-space-${o})`), c && (a.marginRight = `var(--pittorica-space-${c})`), n && (a.marginBottom = `var(--pittorica-space-${n})`), p && (a.marginLeft = `var(--pittorica-space-${p})`), s && (a.marginLeft = `var(--pittorica-space-${s})`, a.marginRight = `var(--pittorica-space-${s})`), l && (a.marginTop = `var(--pittorica-space-${l})`, a.marginBottom = `var(--pittorica-space-${l})`), d && (a.padding = `var(--pittorica-space-${d})`), v && (a.paddingTop = `var(--pittorica-space-${v})`), h && (a.paddingRight = `var(--pittorica-space-${h})`), g && (a.paddingBottom = `var(--pittorica-space-${g})`), m && (a.paddingLeft = `var(--pittorica-space-${m})`), u && (a.paddingLeft = `var(--pittorica-space-${u})`, a.paddingRight = `var(--pittorica-space-${u})`), y && (a.paddingTop = `var(--pittorica-space-${y})`, a.paddingBottom = `var(--pittorica-space-${y})`);
  const k = {
    ...N,
    ...a
  };
  return /* @__PURE__ */ f(
    A,
    {
      className: Y("pittorica-box", S),
      style: k,
      disabled: T,
      required: C,
      ...R,
      children: r
    }
  );
};
w.displayName = "Box";
function E(t) {
  var r, e, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (r = 0; r < o; r++) t[r] && (e = E(t[r])) && (i && (i += " "), i += e);
  } else for (e in t) t[e] && (i && (i += " "), i += e);
  return i;
}
function q() {
  for (var t, r, e = 0, i = "", o = arguments.length; e < o; e++) (t = arguments[e]) && (r = E(t)) && (i && (i += " "), i += r);
  return i;
}
function _(t) {
  var r, e, i = "";
  if (typeof t == "string" || typeof t == "number") i += t;
  else if (typeof t == "object") if (Array.isArray(t)) {
    var o = t.length;
    for (r = 0; r < o; r++) t[r] && (e = _(t[r])) && (i && (i += " "), i += e);
  } else for (e in t) t[e] && (i && (i += " "), i += e);
  return i;
}
function Z() {
  for (var t, r, e = 0, i = "", o = arguments.length; e < o; e++) (t = arguments[e]) && (r = _(t)) && (i && (i += " "), i += r);
  return i;
}
const F = ({
  as: t,
  children: r,
  display: e,
  m: i,
  mt: o,
  mr: c,
  mb: n,
  ml: p,
  mx: s,
  my: l,
  p: d,
  pt: v,
  pr: h,
  pb: g,
  pl: m,
  px: u,
  py: y,
  width: $,
  height: x,
  position: b,
  style: N,
  className: S,
  disabled: T,
  required: C,
  ...R
}) => {
  const A = t || "div", a = {};
  e && (a.display = e), $ && (a.width = $), x && (a.height = x), b && (a.position = b), i && (a.margin = `var(--pittorica-space-${i})`), o && (a.marginTop = `var(--pittorica-space-${o})`), c && (a.marginRight = `var(--pittorica-space-${c})`), n && (a.marginBottom = `var(--pittorica-space-${n})`), p && (a.marginLeft = `var(--pittorica-space-${p})`), s && (a.marginLeft = `var(--pittorica-space-${s})`, a.marginRight = `var(--pittorica-space-${s})`), l && (a.marginTop = `var(--pittorica-space-${l})`, a.marginBottom = `var(--pittorica-space-${l})`), d && (a.padding = `var(--pittorica-space-${d})`), v && (a.paddingTop = `var(--pittorica-space-${v})`), h && (a.paddingRight = `var(--pittorica-space-${h})`), g && (a.paddingBottom = `var(--pittorica-space-${g})`), m && (a.paddingLeft = `var(--pittorica-space-${m})`), u && (a.paddingLeft = `var(--pittorica-space-${u})`, a.paddingRight = `var(--pittorica-space-${u})`), y && (a.paddingTop = `var(--pittorica-space-${y})`, a.paddingBottom = `var(--pittorica-space-${y})`);
  const k = {
    ...N,
    ...a
  };
  return /* @__PURE__ */ f(
    A,
    {
      className: Z("pittorica-box", S),
      style: k,
      disabled: T,
      required: C,
      ...R,
      children: r
    }
  );
};
F.displayName = "Box";
const M = ({
  children: t,
  as: r,
  size: e,
  weight: i = "regular",
  align: o,
  truncate: c = !1,
  color: n,
  className: p,
  style: s,
  ...l
}) => {
  const d = r || "span", v = (n == null ? void 0 : n.startsWith("#")) || (n == null ? void 0 : n.startsWith("rgb")) || (n == null ? void 0 : n.startsWith("hsl")), h = /* @__PURE__ */ new Set([
    "danger",
    "success",
    "error",
    "info",
    "indigo",
    "crimson",
    "teal",
    "amber",
    "slate",
    "blue",
    "red"
  ]), g = (() => {
    if (n)
      return n === "inherit" ? "inherit" : v ? n : h.has(n) ? `var(--pittorica-${n}-9)` : n;
  })(), m = e ? typeof e == "string" ? `pittorica-text--size-${e}` : q({
    [`pittorica-text--size-${e.initial}`]: e.initial,
    [`pittorica-text--xs-size-${e.xs}`]: e.xs,
    [`pittorica-text--sm-size-${e.sm}`]: e.sm,
    [`pittorica-text--md-size-${e.md}`]: e.md,
    [`pittorica-text--lg-size-${e.lg}`]: e.lg,
    [`pittorica-text--xl-size-${e.xl}`]: e.xl
  }) : null, u = {
    ...s,
    textAlign: o,
    color: g
  };
  return /* @__PURE__ */ f(
    F,
    {
      as: d,
      className: q(
        "pittorica-text",
        m,
        { "pittorica-text--truncate": c },
        p
      ),
      "data-weight": i,
      style: u,
      ...l,
      children: t
    }
  );
};
M.displayName = "Text";
const O = K(null), tt = () => {
  const t = P(O);
  if (!t)
    throw new Error("Select components must be wrapped in <Select.Root />");
  return t;
}, j = ({
  children: t,
  label: r,
  helperText: e,
  error: i,
  color: o = "source",
  disabled: c,
  name: n,
  required: p,
  size: s = "sm",
  className: l,
  style: d,
  as: v,
  ...h
}) => {
  const g = W(), m = W(), y = o !== "inherit" && !(o != null && o.startsWith("#")) && !(o != null && o.startsWith("rgb")) ? `var(--pittorica-${o}-9)` : o;
  return /* @__PURE__ */ f(
    O,
    {
      value: { inputId: g, helperId: m, disabled: c, size: s, name: n, required: p },
      children: /* @__PURE__ */ B(
        w,
        {
          as: v || "div",
          className: z(
            "pittorica-select-root",
            `pittorica-select--${s}`,
            l
          ),
          "data-error": i,
          ...h,
          children: [
            r && /* @__PURE__ */ B(
              M,
              {
                as: "label",
                htmlFor: g,
                weight: "medium",
                style: {
                  paddingLeft: "4px",
                  fontSize: "var(--pittorica-font-size-1)",
                  marginBottom: "4px",
                  display: "inline-block"
                },
                children: [
                  r,
                  " ",
                  p && /* @__PURE__ */ f("span", { "aria-hidden": "true", children: "*" })
                ]
              }
            ),
            /* @__PURE__ */ B(
              "div",
              {
                className: "pittorica-select-wrapper",
                "data-disabled": c,
                style: {
                  "--pittorica-source-color": y,
                  ...d
                },
                children: [
                  t,
                  /* @__PURE__ */ f("div", { className: "pittorica-select-chevron", children: /* @__PURE__ */ f(X, { size: s === "xs" ? 14 : 18 }) })
                ]
              }
            ),
            e && /* @__PURE__ */ f("div", { id: m, className: "pittorica-select-helper", children: e })
          ]
        }
      )
    }
  );
}, G = ({
  children: t,
  className: r,
  as: e,
  ...i
}) => {
  const { inputId: o, helperId: c, disabled: n, name: p, required: s } = tt();
  return /* @__PURE__ */ f(
    w,
    {
      as: e || "select",
      name: p,
      id: o,
      disabled: n,
      "aria-describedby": c,
      "aria-required": s,
      className: z("pittorica-select-input", r),
      ...i,
      children: t
    }
  );
}, H = ({
  children: t,
  className: r,
  as: e,
  ...i
}) => /* @__PURE__ */ f(
  w,
  {
    as: e || "div",
    className: z("pittorica-select-slot", r),
    ...i,
    children: t
  }
), rt = Object.assign(j, {
  Root: j,
  Content: G,
  Slot: H
});
j.displayName = "Select.Root";
G.displayName = "Select.Content";
H.displayName = "Select.Slot";
export {
  rt as Select,
  G as SelectContent,
  j as SelectRoot,
  H as SelectSlot
};
